
```{r setup, echo=FALSE}
#opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, figalign="center", echo=FALSE, warning=FALSE, error=FALSE, message=FALSE)
## setup report details
clientname="Serge Gregoire"
clientemail="SGREGOIRE2@PARTNERS.ORG"
labPI="Wu"
lablocation="MGH/HSCI"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```


---

Expression Array and Chip-seq integration analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

---

## Methods summary
All MoGene 1.0 ST arrays were processed using the 'oligo' BioConductor package[^1],  Robust Multichip Average (RMA) normalized[^2] and differentially expressed genes identified using limma[^3]. ChIP-seq peaks postions relative to gene transcription start sites were determined with GREAT [^4] (using the  mm9 genome). 

---

## Workflow
The sections below provide descriptions of the included results and plots. 
There are two main sections: a) an analysis of differentially expressed genes from the microarray data and b) integration with the ChIP-seq results with Rcade.

---

##Bioconductor and R libraries used to process and visualize the data

```{r libraries, echo=TRUE}
library(oligo)
library(limma)
library(xtable)
library(Biobase)
library(pd.mogene.1.0.st.v1)
library("mogene10sttranscriptcluster.db")
library(plyr)
library(ggplot2)
library(biomaRt)
ensembl = useMart("ensembl",dataset="mmusculus_gene_ensembl")
filters=listFilters(ensembl)
attributes=listAttributes(ensembl)
```

---

##Variables
```{r VARIABLES, echo=TRUE}

if(file.exists("/n/home08/jhutchin/")){
	baseDir="/n/hsphS10/hsphfs1/chb/projects/sw_cardiomyocyte_differentiation"
} else if (file.exists("/Volumes/ody/")){
	baseDir="/Volumes/ody/projects/sw_cardiomyocyte_differentiation"
}
metaDir=file.path(baseDir, "meta")
mic.dataDir=file.path(baseDir, "data/microarray")
bed.dataDir=file.path(baseDir, "results/integrated/GREAT")
mic.resultsDir=file.path(baseDir, "results/microarray")
int.resultsDir=file.path(baseDir, "results/integrated/GREAT")
mic.covarsfilename="covars.desc" # do not use full path

mic.grouplabel="treatment"
mic.samplelabel="sample"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
chrs= c(seq(1, 22, 1), "X", "Y")
up.dist=1500
down.dist=1500
```

--
	
# MICROARRAY ANALYSIS
	
## Load the data
	- load phenotypes and array names from metadata file
- this file contains the names and descriptions of CEL files contained in same directory 

```{r dataload, results='hide', cache=TRUE}
covars <- read.delim(file.path(metaDir, mic.covarsfilename), row.names=1) ## simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns
celFiles <- list.celfiles(mic.dataDir, full.names=TRUE)
affyRaw <- read.celfiles(celFiles, pkgname="pd.mogene.1.0.st.v1")
pData(affyRaw) <- covars 
validObject(affyRaw) ## sanity check
````

## Sample metadata

```{r covars, results='asis'}
## Sample information table
pDataTable <- xtable(pData(affyRaw))
print(pDataTable, type='html')
```

## Background correct and normalize
- data was normalized for differential gene expression analysis using RMA in the oligo BioConductor package which normalizes the intensity values at the probe level, and  collapses probes into "core" transcripts based on annotations provided by Affymetrix

```{r normalize, results='hide'}
affyNorm.core <- rma(affyRaw, target="core", background=TRUE, normalize=TRUE)
rm(affyRaw) # cleanup
```

## Annotations
- loaded in feature data, i.e. the gene symbols that correspond to each probeset group on the array

```{r features, results='hide'}
# retrieve NetAffx Biological Annotation
featureData(affyNorm.core) <- getNetAffx(affyNorm.core, "transcript")
symbols <-  unlist(mget(as.character(pData(featureData(affyNorm.core))$transcriptclusterid), mogene10sttranscriptclusterSYMBOL, ifnotfound=NA))
entrezids <- unlist(mget(as.character(pData(featureData(affyNorm.core))$transcriptclusterid), mogene10sttranscriptclusterENTREZID, ifnotfound=NA))

# check to make sure data is correct
identical(length(featureData(affyNorm.core)$probesetid), length(symbols)) # sanity check, sane=TRUE
identical(length(featureData(affyNorm.core)$probesetid), length(entrezids)) # sanity check, sane=TRUE
gene.annots <- as.data.frame(cbind(symbols, entrezids))
head(gene.annots$symbols[!is.na(gene.annots$symbols)]) # sanity check, sane=>see gene ids
```

## Analyses
### Limma
A linear model for microarray data analysis (Limma) was performed on the samples to identify differentially expressed genes between all pair-wise comparisons of the sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we constructed two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.

#### Design matrix 
- make a matrix with arrays as rows, treatment groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group
- in this case it looks like this, where 1-4 are the 4 microarrays named in the covars.desc file in the same order as in the file

```{r design, results="asis"}
design <- model.matrix(~ -1+factor(pData(affyNorm.core)[,mic.grouplabel]))
# make sure the headings match
colnames(design) <- sub("factor.pData.affyNorm.core... mic.grouplabel..", "", colnames(design))
designTable <- xtable(design)
print(designTable, type='html')
```

#### Contrast matrix
- to perform specified pairwise comparisons
- in this table, columns are contrasts/comparisons and rows are sample groups
-  a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts(control_GFP-minus_dox,control_GFP-plus_dox, plus_dox-minus_dox, levels=c("control_GFP", "minus_dox", "plus_dox"))
contrastmatrixTable <- xtable(contrast.matrix)
print(contrastmatrixTable, type='html')
```

####Linear model
These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix 
- for each gene based on the given series of arrays

```{r linearmodel}
eset.core <- exprs(affyNorm.core) 
fit.core <- lmFit(eset.core, design) 
```

- using the contrast matrix, compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2.core <- contrasts.fit(fit.core, contrast.matrix) 
```


- compute moderated t-statistics and log-odds of differential expression 
- by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.core <- eBayes(fit2.core) 
```

## Results

```{r allstats}
all.stats <- llply(seq(1,3,1), function(n) {
	contrast <- gsub(" ", "", dimnames(fit2.core$contrasts)$Contrasts[n])
	stats.core <- topTable(fit2.core, coef=n, sort.by="B",number=length(symbols), genelist=cbind(gene.annots[,c("symbols", "entrezids")], fit2.core$genes))
	return(list(stats.core=stats.core, contrast=contrast))
})

```

---

# DATA INTEGRATION

## Load Chip-Seq peak postions relative to genes

```{r load_chipseq}
peaks <- lapply(as.list(list.files(pattern="tsv",path=bed.dataDir)), function(n) {
	peak.genes <- read.delim(file.path(bed.dataDir,n), header=T, fill=TRUE)
	names(peak.genes) <- c("gene", paste0("P", seq(1:(ncol(peak.genes)-1))))
	return(peak.genes)
})
```

### find genes with peaks near TSSs, and how many peaks are within the region

```{r min_dist}
peaks.in.region <- llply(peaks, function(n) {
	adply(n, 1, function(row) {
		num.row=as.numeric(row[2:length(row)])
		num.up <- length(which(abs(num.row[num.row<0])<up.dist))
		num.down <- length(which(num.row[num.row>0]<=down.dist))
		num.TSS.peaks <- num.up+num.down
			return(num.TSS.peaks=num.TSS.peaks)
		})
})

num.peaks.in.region <- llply(peaks.in.region, function(n) {
	just.num.peaks=n[, c("gene", "V1")]
	return(just.num.peaks)
})
```

### find genes with differential expression and ChIP-seq peaks

```{r combo_DE_peaks}

num.peaks.in.region.all <- merge(num.peaks.in.region[[1]], num.peaks.in.region[[2]], by="gene")
num.peaks.in.region.all <- num.peaks.in.region.all[!duplicated(num.peaks.in.region.all),]
num.peaks.in.region.all$sum <- apply(num.peaks.in.region.all[,2:3], 1 , function(n) sum(n))
DE.peaks <- merge(all.stats[[3]]$stats.core, num.peaks.in.region.all, by.x="symbols", by.y="gene", all=TRUE)



DE.peak.genes <- unique(DE.peaks[which(DE.peaks$sum>0 & DE.peaks$adj.P.Val<0.01 & abs(DE.peaks$logFC)>1),]$symbols)

DE.peak.genes.data <- DE.peaks[which(DE.peaks$sum>0 & DE.peaks$adj.P.Val<0.01 & abs(DE.peaks$logFC)>1),]

names(DE.peak.genes.data)[10:11] <- c("sample1.TSS.chipseq.peaks", "sample2.TSS.chipseq.peaks")
DE.peak.genes.data <- DE.peak.genes.data[!duplicated(DE.peak.genes.data),]
DE.peak.genes.data <- DE.peak.genes.data[order(DE.peak.genes.data$sum, decreasing=T),]
write.table(DE.peak.genes.data, file=file.path(int.resultsDir, "DE.peak.genes.data.xls"), quote=T, sep="\t", col.names=T, row.names=F)
```



[^1]: [BioC](http://www.bioconductor.org)
[^2]: [RMA](http://bioinformatics.oxfordjournals.org/content/19/2/185)
[^3]: [limma](http://www.bioconductor.org/packages/release/bioc/html/limma.html)
[^4]: [GREAT](http://bejerano.stanford.edu/great/public/html/)

