
```{r setup, echo=FALSE}
#opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, figalign="center", echo=FALSE, warning=FALSE, error=FALSE, message=FALSE)
## setup report details
clientname="Serge Gregoire"
clientemail="SGREGOIRE2@PARTNERS.ORG"
labPI="Wu"
lablocation="MGH/HSCI"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```


---

Expression Array and Chip-seq integration analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

---

## Methods summary
All MoGene 1.0 ST arrays were processed using the 'oligo' BioConductor package[^1],  Robust Multichip Average (RMA) normalized[^2] and differentially expressed genes  identified using limma[^3]. ChIP-seq reads were aligned with Bowtie [^4] to the mm9 genome and converted to BAM format [^5] before import. Differential expression data was integrated with ChIP-seq data with Rcade[^6].

---

## Workflow
The sections below provide descriptions of the included results and plots. 
There are two main sections: a) an analysis of differentially expressed genes from the microarray data and b) integration with the ChIP-seq results with Rcade.

---

##Bioconductor and R libraries used to process and visualize the data

```{r libraries, echo=TRUE}
library(oligo)
library(limma)
library(xtable)
library(Biobase)
library(pd.mogene.1.0.st.v1)
library("mogene10sttranscriptcluster.db")
library(plyr)
library(ggplot2)
library(biomaRt)
ensembl = useMart("ensembl",dataset="mmusculus_gene_ensembl")
filters=listFilters(ensembl)
attributes=listAttributes(ensembl)
library(Rcade)
library(parallel)
```

---

##Variables

```{r VARIABLES, echo=TRUE}
## Setup Data and Results directory variables
if(file.exists("/n/home08/jhutchin/")){
  baseDir="/n/hsphS10/hsphfs1/chb/projects/sw_cardiomyocyte_differentiation"
  } else if (file.exists("/Volumes/ody/")){
  baseDir="/Volumes/ody/projects/sw_cardiomyocyte_differentiation"
}
metaDir=file.path(baseDir, "meta")
mic.dataDir=file.path(baseDir, "data/microarray")
cs.bamDir=file.path(baseDir, "results/chipseq/bowtie") # directory with the ChIP-seq bam files 
mic.resultsDir=file.path(baseDir, "results/microarray")
int.resultsDir=file.path(baseDir, "results/integrated/Rcade_out")
mic.covarsfilename="covars.desc" # do not use full path
cs.covarsfilename="chipseq.covars.desc"

mic.grouplabel="treatment"
mic.samplelabel="sample"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
chrs= c(seq(1, 22, 1), "X", "Y")
```

---

# MICROARRAY ANALYSIS

## Load the data
- load phenotypes and array names from metadata file
  - this file contains the names and descriptions of CEL files contained in same directory 

```{r dataload, results='hide', cache=TRUE}
covars <- read.delim(file.path(metaDir, mic.covarsfilename), row.names=1) ## simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns
celFiles <- list.celfiles(mic.dataDir, full.names=TRUE)
affyRaw <- read.celfiles(celFiles, pkgname="pd.mogene.1.0.st.v1")
pData(affyRaw) <- covars 
validObject(affyRaw) ## sanity check
````

## Sample metadata

```{r covars, results='asis'}
## Sample information table
pDataTable <- xtable(pData(affyRaw))
print(pDataTable, type='html')
```

## Background correct and normalize
- data was normalized for differential gene expression analysis using RMA in the oligo BioConductor package which normalizes the intensity values at the probe level, and  collapses probes into "core" transcripts based on annotations provided by Affymetrix

```{r normalize, results='hide'}
affyNorm.core <- rma(affyRaw, target="core", background=TRUE, normalize=TRUE)
rm(affyRaw) # cleanup
```

## Annotations
- loaded in feature data, i.e. the gene symbols that correspond to each probeset group on the array

```{r features, results='hide'}
# retrieve NetAffx Biological Annotation
featureData(affyNorm.core) <- getNetAffx(affyNorm.core, "transcript")
symbols <-  unlist(mget(as.character(pData(featureData(affyNorm.core))$transcriptclusterid), mogene10sttranscriptclusterSYMBOL, ifnotfound=NA))
entrezids <- unlist(mget(as.character(pData(featureData(affyNorm.core))$transcriptclusterid), mogene10sttranscriptclusterENTREZID, ifnotfound=NA))

# check to make sure data is correct
identical(length(featureData(affyNorm.core)$probesetid), length(symbols)) # sanity check, sane=TRUE
identical(length(featureData(affyNorm.core)$probesetid), length(entrezids)) # sanity check, sane=TRUE
gene.annots <- as.data.frame(cbind(symbols, entrezids))
head(gene.annots$symbols[!is.na(gene.annots$symbols)]) # sanity check, sane=>see gene ids
```

## Analyses
### Limma
A linear model for microarray data analysis (Limma) was performed on the samples to identify differentially expressed genes between all pair-wise comparisons of the sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we constructed two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.

#### Design matrix 
- make a matrix with arrays as rows, treatment groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group
- in this case it looks like this, where 1-4 are the 4 microarrays named in the covars.desc file in the same order as in the file

```{r design, results="asis"}
design <- model.matrix(~ -1+factor(pData(affyNorm.core)[,mic.grouplabel]))
# make sure the headings match
colnames(design) <- sub("factor.pData.affyNorm.core... mic.grouplabel..", "", colnames(design))
designTable <- xtable(design)
print(designTable, type='html')
```

#### Contrast matrix
- to perform specified pairwise comparisons
- in this table, columns are contrasts/comparisons and rows are sample groups
-  a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts(control_GFP-minus_dox,control_GFP-plus_dox, plus_dox-minus_dox, levels=c("control_GFP", "minus_dox", "plus_dox"))
contrastmatrixTable <- xtable(contrast.matrix)
print(contrastmatrixTable, type='html')
```

####Linear model
These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix 
  - for each gene based on the given series of arrays

```{r linearmodel}
eset.core <- exprs(affyNorm.core) 
fit.core <- lmFit(eset.core, design) 
```

- using the contrast matrix, compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2.core <- contrasts.fit(fit.core, contrast.matrix) 
```


- compute moderated t-statistics and log-odds of differential expression 
  - by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.core <- eBayes(fit2.core) 
```

## Results
### Volcanoplot to visualize relationship between fold change and signifcance of changes in expression between groups
- genes outside the two vertical orange lines show more than 2-fold changes in expression between groups
- genes which pass an FDR  (false discovery rate) multiple testing correction of 0.01 (1% FDR) and show at least a 2-fold change in expression are colored blue
- the larger the pointsize, the more likely there is significant expression change (B = log-odds of differential expression) between the sample groups

```{r allstats}
all.stats <- llply(seq(1,3,1), function(n) {
    contrast <- gsub(" ", "", dimnames(fit2.core$contrasts)$Contrasts[n])
    stats.core <- topTable(fit2.core, coef=n, sort.by="B",number=length(symbols), genelist=cbind(gene.annots[,c("symbols", "entrezids")], fit2.core$genes))
    return(list(stats.core=stats.core, contrast=contrast))
    })

```

---

# DATA INTEGRATION

## Load Chip-Seq alignments and metadata

```{r load_chipseq}
cs.targets <- read.delim(file.path(metaDir, cs.covarsfilename), as.is = TRUE)
```

## Find regions where you will look for overlap
- upstream and downstream of transcription start sites

### get coordinates of transcription start sites for all genes in the microarray set
```{r gene_annotations}
mic.entrezids=na.omit(as.vector(unlist(gene.annots$entrezids)))
tss.anno <- getBM(attributes= c("entrezgene", "chromosome_name","transcript_start", "transcript_end", "strand"), filters="entrezgene", values=mic.entrezids, mart=ensembl)
tss.anno <- tss.anno[order(tss.anno$chromosome_name),]
colnames(tss.anno) <- c("ENTREZ","chr","start","end","str")
tss.anno=tss.anno[tss.anno$chr %in% chrs,]
```


### Get regions comprising the 1500bp upstream and 1500bp downstream  of transcription start sites of all genes (distances chosen based on GREAT[^8] analysis of "egfp_positive_cells_sample1_aligned.bed" file)
[relative positions of ChIP-seq peaks to TSS](../results/chipseq/GREAT.genome.annotations.distance.to.TSS.sample1.converted.bed.pdf)

```{r tss_bins}
ChIPannoZones <- defineBins(tss.anno, zone=c(-1500, 1500), geneID="ENTREZ") # these are the regions where you are going to look for changes in ChIP-seq peaks
```

## Rcade analysis

### loop through and run for all  comparison, then the other two comparison
### drop genes with ajdusted pvalues of more than 0.25 and log2 fold changes of less than 1

```{r Rcade}
cl <- makeCluster(8, "SOCK") ## for multicore processing
for (n in 1:3){ ## 3 different limma comparisons were made and stored in the stats.core dataframe within the "all.stats" list above
  DE <- all.stats[[n]]$stats.core
  ## remove genes with low fold change and non-significant DE
  DE.sig=DE[which(abs(DE$logFC)>1 & DE$adj.P.Val<0.25),]

  DElookup <- list(GeneID="entrezids", logFC="logFC", B="B", "symbols" ) # tell Rcade the headers of the appropriate columns
  Rcade <- RcadeAnalysis(DE.sig, ChIPannoZones, ChIPtargets=cs.targets, ChIPfileDir = cs.bamDir, cl=cl, shift = 0,  DElookup=DElookup)
  
  ## make directory for results output if necessary
  if (file.exists(file.path(int.resultsDir, all.stats[[n]]$contrast))) {
    next()
    } else {
      dir.create(file.path(int.resultsDir, all.stats[[n]]$contrast))
      } 
  
  pdf(file.path(file.path(int.resultsDir, all.stats[[n]]$contrast), "Rcade_plots.pdf"))
  plotPCA(Rcade)
  plotMM(Rcade)
  dev.off()
  
  exportRcade(Rcade, directory=file.path(int.resultsDir, all.stats[[n]]$contrast))
}
```


[^1]: [BioC](http://www.bioconductor.org)
[^3]: [RMA](http://bioinformatics.oxfordjournals.org/content/19/2/185)
[^4]: [limma](http://www.bioconductor.org/packages/release/bioc/html/limma.html)
[^5]: [Bowtie](http://bowtie-bio.sourceforge.net/index.shtml)
[^6]: [BAM format](http://genome.ucsc.edu/FAQ/FAQformat.html#format5.1)
[^7]: [Rcade](http://www.bioconductor.org/packages/2.12/bioc/html/Rcade.html)
,