=
```{r setup, echo=FALSE}
#opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, figalign="center", echo=FALSE, warning=FALSE, error=FALSE, message=FALSE)
## setup report details
clientname="Serge Gregoire"
clientemail="SGREGOIRE2@PARTNERS.ORG"
labPI="Wu"
lablocation="MGH/HSCI"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```


---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

---

## Methods summary
All MoGene 1.0 ST arrays were processed using the 'oligo' BioConductor package[^1], quality-controlled with arrayQualityMetrics[^2] and RMA normalized[^3]. Differentially expressed genes were identified using limma[^4].

---

## Workflow
The sections below provide descriptions of the included  results and plots. 

---

##Bioconductor and R libraries used to process and visualize the data

```{r libraries, echo=TRUE}
library(oligo)
library(arrayQualityMetrics)
library(limma)
library(pheatmap)
library(RColorBrewer)
library(xtable)
library(Biobase)
library(pd.mogene.1.0.st.v1)
library(plyr)
library(ggplot2)
library(biomaRt)
ensembl = useMart("ensembl",dataset="mmusculus_gene_ensembl")
filters=listFilters(ensembl)
attributes=listAttributes(ensembl)
```

---

##Variables

```{r VARIABLES, echo=TRUE}
## Setup Data and Results directory variables
metaDir="/n/hsphS10/hsphfs1/chb/projects/sw_cardiomyocyte_differentiation/meta"
dataDir="/n/hsphS10/hsphfs1/chb/projects/sw_cardiomyocyte_differentiation/data/microarray"
resultsDir="/n/hsphS10/hsphfs1/chb/projects/sw_cardiomyocyte_differentiation/results/microarray"
covarsfilename="covars.desc" # do not use full path
grouplabel="treatment"
samplelabel="sample"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
```

---

## Load the data
- load phenotypes and array names from metadata file
  - this file contains the names and descriptions of CEL files contained in same directory 

```{r dataload, results='hide', cache=TRUE}
covars <- read.delim(file.path(metaDir, covarsfilename), row.names=1) ## simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns
celFiles <- list.celfiles(dataDir, full.names=TRUE)
affyRaw <- read.celfiles(celFiles)
pData(affyRaw) <- covars 
validObject(affyRaw) ## sanity check
````

## Sample metadata

```{r covars, results='asis'}
## Sample information table
pDataTable <- xtable(pData(affyRaw))
print(pDataTable, type='html')
```

## QA/QC - 1 
- using arrayQualityMetrics library
- raw data

```{r rawQC, cache=TRUE}
arrayQualityMetrics(expressionset=affyRaw, outdir=file.path(resultsDir, 'report_raw'), force=TRUE, do.logtransform=TRUE)
```

[raw data QC report](./results/report_raw/index.html)

- some outliers when using the raw data before background correction and normalization 

## Background correct and normalize
- data was normalized for differential gene expression analysis using Robust Multichip Average (RMA)[^5] in the oligo BioConductor package which normalizes the intensity values at the probe level, and  collapses probes into "core" transcripts based on annotations provided by Affymetrix

```{r normalize, results='hide'}
affyNorm.core <- rma(affyRaw, target="core", background=TRUE, normalize=TRUE)
affyNorm.probeset<-rma(affyRaw, target="probeset", background=TRUE, normalize=TRUE)
rm(affyRaw) # cleanup
```

## QA/QC - 2 
- using arrayQualityMetrics library
- normalized data

```{r normQC, cache=TRUE}
arrayQualityMetrics(expressionset=affyNorm.core, outdir=file.path(resultsDir, 'report_rma.core'), force=TRUE, do.logtransform=FALSE)
```

[normalized data QC report](./results/report_rma.core/index.html)

- found no significant quality issues with any of the chips, as determined by (among other methods) visual inspection, intensity distributions or RNA degradation plots

- given there were no outliers with the normalized  corrected data, all samples were retained for the analysis 


## QA/QC - 3
### Primary Componenet Analysis Clustering
- the goal of this analysis is to blindly evaluate the variability within the raw data and determine whether this variability could predict the different sample categories by itself
- in basic terms, we are checking whether the samples cluster by sample type
  - its a good sign if they do

```{r pca, fig.cap="Primary Component Analysis of samples - all combinations of the 5 first primary components", fig.width=8.5, fig.height=8.5}
eset.core <- exprs(affyNorm.core) 
colnames(eset.core) <- pData(affyNorm.core)[,samplelabel]
myPca.core <- prcomp(t(eset.core))
tmpPCAData.core <- as.data.frame(myPca.core$x[,1:4])
plot(tmpPCAData.core, col=cbPalette[factor(pData(affyNorm.core)[,grouplabel])], pch=as.character(pData(affyNorm.core)[,grouplabel]), cex=2)
rm(tmpPCAData.core, myPca.core) ## cleanup
```

**Principal Component Analysis Figure**
The figure shows a scatterplot of the arrays along all dual combinations of the first four principal components. Each sample group is represented by a separate color, the symbol letter used corresponds to the first letter of the "treatment" column in the Sample metadata above. 
You can use this plot to explore if the arrays cluster, and whether this is according to an intended experimental factor or according to unintended causes such as batch effects. Principal component analysis is a dimension reduction and visualisation technique that is here used to project the multivariate data vector of each array into a two-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the arrays.

- from this plot it appears that the samples cluster fairly well, with the control_GFP samples clustering closer to the plus_dox samples than the minus_dox samples

## Annotations
- loaded in feature data, i.e. the gene symbols that correspond to each probeset group on the array

```{r features, results='hide'}
# retrieve NetAffx Biological Annotation
featureData(affyNorm.core) <- getNetAffx(affyNorm.core, "transcript")
featureData(affyNorm.probeset) <- getNetAffx(affyNorm.probeset, "probeset")
                        
# Extract the 'gene assignment' annotation
annot <- pData(featureData(affyNorm.core)[, c('geneassignment')])
categories <- pData(featureData(affyNorm.core)[, c('category')])
annot.probeset<-pData(featureData(affyNorm.probeset)[, c('geneassignment')])
categories.probeset <- pData(featureData(affyNorm.probeset)[, c('probesettype')])

# Generate a list of gene symbols and ensembl transcript ids from the gene assignment
desc <- annot[, 1]
desc.probeset<-annot.probeset[,1]

symbols <- unlist(lapply(desc, function(x) strsplit(x, ' // ')[[1]][2]))
symbols.probeset<-unlist(lapply(desc.probeset, function(x) strsplit(strsplit(x, " /// ")[[1]][1], " // ")[[1]][2]))

ensembl_transcripts <- unlist(lapply(desc, function(x) strsplit(strsplit(x, " // ")[[1]][5], " /// ")[[1]][2]))
ensembl_transcripts.probeset <- unlist(lapply(desc.probeset, function(x) strsplit(strsplit(x, " /// ")[[1]][2], " // ")[[1]][1]))

probeids=row.names(annot)
probeids.probeset=row.names(annot.probeset)

gene.annots<-cbind(probeids, cbind(symbols, ensembl_transcripts))
gene.annots.probeset=cbind(probeids.probeset, cbind(symbols.probeset, ensembl_transcripts.probeset))

# convert ensembl transcript ids to ensembl gene ids
ensembl_genes<- getBM(attributes=c("ensembl_gene_id", "ensembl_transcript_id"), filters="ensembl_transcript_id", value=ensembl_transcripts, mart=ensembl, uniqueRows=FALSE)
ensembl_genes.probeset<- getBM(attributes=c("ensembl_gene_id", "ensembl_transcript_id"), filters="ensembl_transcript_id", value=ensembl_transcripts.probeset, mart=ensembl, uniqueRows=FALSE)

gene.annots<-merge(gene.annots, ensembl_genes, by.x="ensembl_transcripts", by.y="ensembl_transcript_id", sort=FALSE, all.x=TRUE)
gene.annots.probeset<-merge(gene.annots.probeset, ensembl_genes.probeset, by.x="ensembl_transcripts.probeset", by.y="ensembl_transcript_id", sort=FALSE, all.x=TRUE)

gene.annots=gene.annots[order(gene.annots[, "probeids"]),]
gene.annots=as.data.frame(gene.annots)
gene.annots.probeset=gene.annots.probeset[order(gene.annots.probeset[, "probeids.probeset"]),]  
gene.annots.probeset=as.data.frame(gene.annots.probeset)

# check to make sure data is correct
identical(length(featureData(affyNorm.core)$probesetid), length(symbols)) # sanity check, sane=TRU E
identical(as.character(gene.annots$probeids), probeids) # sanity check, sane=TRUE
head(gene.annots$symbols[!is.na(gene.annots$symbols)]) # sanity check, sane=>see gene ids

identical(length(featureData(affyNorm.probeset)$probesetid), length(symbols.probeset)) # sanity check, sane=TRUE
identical(as.character(gene.annots.probeset$probeids), probeids.probeset) # sanity check, sane=TRUE
head(gene.annots.probeset$symbols[!is.na(gene.annots.probeset$symbols)]) # sanity check, sane=>see gene ids
```

## Analyses
### Limma
A linear model for microarray data analysis (Limma) was performed on the samples to identify differentially expressed genes between all pair-wise comparisons of the sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we constructed two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.

#### Design matrix 
- make a matrix with arrays as rows, treatment groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group
- in this case it looks like this, where 1-4 are the 4 microarrays named in the covars.desc file in the same order as in the file

```{r design, results="asis"}
design <- model.matrix(~ -1+factor(pData(affyNorm.core)[,grouplabel]))
# make sure the headings match
colnames(design) <- sub("factor.paste.pd.celltype, pd.treatment, sep . .....", "", colnames(design))
designTable <- xtable(design)
print(designTable, type='html')
```

#### Contrast matrix
- to perform specified pairwise comparisons
- in this table, columns are contrasts/comparisons and rows are sample groups
-  a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts(control_GFP-minus_dox,control_GFP-plus_dox, plus_dox-minus_dox, levels=c("control_GFP", "minus_dox", "plus_dox"))
contrastmatrixTable <- xtable(contrast.matrix)
print(contrastmatrixTable, type='html')
```

####Linear model
These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix 
  - for each gene based on the given series of arrays

```{r linearmodel}
fit.core <- lmFit(eset.core, design) 
eset.probeset <- exprs(affyNorm.probeset)
fit.probeset <- lmFit(eset.probeset, design)
```

- using the contrast matrix, compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2.core <- contrasts.fit(fit.core, contrast.matrix) 
fit2.probeset <- contrasts.fit(fit.probeset, contrast.matrix)
```


- compute moderated t-statistics and log-odds of differential expression 
  - by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.core <- eBayes(fit2.core) 
fit2.probeset <- eBayes(fit2.probeset)
```

## Results
### Volcanoplot to visualize relationship between fold change and signifcance of changes in expression between groups
- genes outside the two vertical orange lines show more than 2-fold changes in expression between groups
- genes which pass an FDR  (false discovery rate) multiple testing correction of 0.01 (1% FDR) and show at least a 2-fold change in expression are colored blue
- the larger the pointsize, the more likely there is significant expression change (B = log-odds of differential expression) between the sample groups

```{r volcanoplot, fig.cap="", fig.width=8.5, fig.height=8.5, fig.show='hold', results='hide'}
all.stats <- llply(seq(1,3,1), function(n) {
  contrast <- gsub(" ", "", dimnames(fit2.core$contrasts)$Contrasts[n])
  stats.core <- topTable(fit2.core, coef=n, sort.by="B",number=length(symbols), genelist=cbind(gene.annots[,c("symbols", "ensembl_transcripts")], fit2.core$genes))
  stats.core$Passes.0.01.FDR.threshold  <-  as.factor(stats.core$adj.P.Val<0.01 & abs(stats.core$logFC)>1)
  g <- ggplot(data=stats.core, aes(x=logFC, y=-log10(P.Value), color=Passes.0.01.FDR.threshold, size=B)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept=c(-1,1), color="orange", alpha=0.7, lwd=1.25) +
    xlab("log2 fold change") + ylab("-log10 p-value") +
    labs(title=contrast)
  
  g.simp=ggplot(data=stats.core, aes(x=abs(logFC), y=-log10(adj.P.Val))) +
    geom_point(alpha=0.2) +
    xlab("absolute log2 fold change") + ylab("-log10 adjusted p-value") +
    labs(title=contrast)+
    scale_x_continuous(limits = c(0,5))+
    scale_y_continuous(limits = c(0,9))
    
  
  return(list(g=g, g.simp=g.simp, stats.core=stats.core, contrast=contrast))
  })
lapply(all.stats, function(x) {
  x$g
  x$g.simp
  })


all.stats.probeset <- llply(seq(1,3,1), function(n) {
    contrast <- gsub(" ", "", dimnames(fit2.probeset$contrasts)$Contrasts[n])
    stats.probeset <- topTable(fit2.probeset, coef=n, sort.by="B",number=length(symbols.probeset), genelist=cbind(gene.annots.probeset[,c("symbols.probeset", "ensembl_transcripts.probeset")], fit2.probeset$genes))
    stats.probeset$Passes.0.01.FDR.threshold  <-  as.factor(stats.probeset$adj.P.Val<0.01 & abs(stats.probeset$logFC)>1)
    
    
    g <- ggplot(data=stats.probeset, aes(x=logFC, y=-log10(P.Value), color=Passes.0.01.FDR.threshold, size=B)) +
      geom_point(alpha=0.5) +
      geom_vline(xintercept=c(-1,1), color="orange", alpha=0.7, lwd=1.25) +
      xlab("log2 fold change") + ylab("-log10 p-value") +
      labs(title=contrast)
    return(list(g=g, stats.probeset=stats.probeset, contrast=contrast))
    })
lapply(all.stats.probeset, function(x) x$g)
```

### Obtain top results
- generate list of top 100 differentially expressed genes with at least 2 fold changes in expression and an adjusted pvalue <0.01,  as sorted by "B", the log-odds that a gene is differentially expressed

#### The summary table contains the following information: 
- logFC is the log2-fold change
- the AveExpr is the average expression value accross all arrays
- the moderated t-statistic (t) is the logFC to its standard error, the P.Value is the associated p-value
- the adj.P.Value is the p-value adjusted for multiple testing (by FDR) 
- the B-value (B) is the log-odds that a gene is differentially expressed (the-higher-the-better)
- the last 9 columns contain the log-transformed normalized expression levels for these genes in each sample

```{r toptable}
top.stats <- lapply(seq(1,3,1), function(n) {
  contrast <- gsub(" ", "", dimnames(fit2.core$contrasts)$Contrasts[n])
  stats.top.core <- topTable(fit2.core, coef=n, adjust="fdr", sort.by="p", p.value=0.01, lfc=1, number=200, genelist=cbind(gene.annots[,c("symbols", "ensembl_gene_id")], fit2.core$genes)) 
  # expression levels for these genes in all samples
  if(nrow(stats.top.core)>0) {
    eset.top.core  <-  exprs(affyNorm.core[stats.top.core[, "ID"], ]) 
    colnames(eset.top.core)  <-  pData(affyNorm.core)[,samplelabel]
    eset.stats.top.core <- cbind(stats.top.core, eset.top.core)
    } else { eset.stats.top.core<-"NA"}
  return(list(eset.stats.top.core=eset.stats.top.core, contrast=contrast))
  })
```



```{r allexpression_values}
all.exprs=as.data.frame(eset.core)
all.exprs$ID=row.names(all.exprs)
all.exprs<-merge(all.stats[[1]]$stats.core, all.exprs, by="ID")
all.exprs=all.exprs[,-(c(4,5,6,7,8,9,10))]
write.table(all.exprs, file=file.path(resultsDir, "all.genes.exprs.xls"),  sep ="\t",, row.names=F, col.names=T)
```

```{r overall_heatmap}
row.names(all.exprs)=all.exprs$ID
all.exprs=all.exprs[,-c(1:3)]
## filter for probes with minimum mean intensity in at least one group
min.int=6
all.exprs.min=all.exprs[apply(all.exprs, 1, function(n)  {
  any(unlist(aggregate(n, by=list(pData(affyNorm.core)$treatment), mean))>min.int)
  }),]
all.exprs.min.mad=all.exprs.min[apply(all.exprs.min, 1, function(n)  mad(n))>0.5,]
                                  
pheatmap(all.exprs.min.mad, main="Probesets above base expression and variance", color=brewer.pal(11, "RdBu"), fontsize_row=8, show_rownames=F, sub="sub")
```


```{r cardiac_genes_heatmap}
### pulled cardiac sp. genbes from Embl-gene Atlast (ArrayExpress?)
cardiac.genes.tab=read.delim(file.path(dataDir, "Embl.tissue.atlas.mouse.cardiac.muscle.tab"), header=T, skip=4)

cardiac.genes=unique(subset(cardiac.genes.tab, cardiac.genes.tab$P.value<1e-3)$Gene.name)
cardiac.gene.IDs=gene.annots[which(gene.annots$symbols %in% cardiac.genes),"probeids"]
pheatmap(all.exprs[cardiac.gene.IDs,], main="Cardiac Genes - pvalue<1e-3", color=brewer.pal(11, "RdBu"), fontsize_row=8, show_rownames=F, sub="sub")

cardiac.genes=unique(subset(cardiac.genes.tab, cardiac.genes.tab$P.value<1e-5)$Gene.name)
cardiac.gene.IDs=gene.annots[which(gene.annots$symbols %in% cardiac.genes),"probeids"]
pheatmap(all.exprs[cardiac.gene.IDs,], main="Cardiac Genes - pvalue<1e-5", color=brewer.pal(11, "RdBu"), fontsize_row=8, show_rownames=F, sub="sub")

cardiac.genes=unique(subset(cardiac.genes.tab, cardiac.genes.tab$P.value<1e-7)$Gene.name)
cardiac.gene.IDs=gene.annots[which(gene.annots$symbols %in% cardiac.genes),"probeids"]
exprs.cardiac.gene.IDs=all.exprs[cardiac.gene.IDs,]
row.names(exprs.cardiac.gene.IDs)=paste(row.names(exprs.cardiac.gene.IDs), gene.annots[which(gene.annots$probeids %in% row.names(exprs.cardiac.gene.IDs)),]$symbol, sep="-")
pheatmap(exprs.cardiac.gene.IDs, main="Cardiac Genes - pvalue<1e-7", color=brewer.pal(11, "RdBu"), fontsize_row=10, show_rownames=T, sub="sub")

```
### Output pvalues and expression levels to files

```{r outputs, results='hide'}
lapply(all.stats, function(x) {
  contrast=x$contrast
  write.table(x$stats.core, file=file.path(resultsDir, paste("all.genes.stats", contrast, "xls",sep=".")), sep="\t", row.names=F, col.names=T)
  })
lapply(top.stats, function(x){
  contrast=x$contrast
  write.table(x$eset.stats.top.core, file=file.path(resultsDir, paste("top100.stats", contrast, "xls",sep=".")), row.names=F, col.names=T, sep='\t')
  })
```


##########
[Links to Excel files with results for all genes in control_GFP vs minus_dox comparison](./results/all.genes.stats.control_GFP-minus_dox.xls)

[Links to Excel files with results for all genes in control_GFP vs plus_dox comparison](./results/all.genes.stats.control_GFP-plus_dox.xls)

[Links to Excel files with results for all genes in plus_dox vs minus_dox comparison](./results/all.genes.stats.plus_dox-minus_dox.xls)


[Links to Excel files with results for genes in top 200 log-odds of differential expression in control_GFP vs minus_dox comparison](./results/top100.stats.control_GFP-minus_dox.xls)

[Links to Excel files with results for genes in top 200 log-odds of differential expression in control_GFP vs plus_dox comparison](./results/top100.stats.control_GFP-plus_dox.xls)

[Links to Excel files with results for genes in top 200 log-odds of differential expression in plus_dox vs minus_dox comparison](./results/top100.stats.plus_dox-minus_dox.xls)


### Heatmaps of top 100 results
- the row names have the probeset id and the gene symbol separated by a dash
- the columns are the individual samples, even though each heatmap represents the expression levels of differentially expressed genes in a single pair-wise comparison between two groups, I included the expression levels for all 3 sample groups in each heatmap
- rows with "NA" for the gene symbol are mainly intronic/exonic control probes, seeing alot of these typically means there is very little difference between the groups 
- here, red denotes lower expression while and blue denotes higher

```{r heatmaps1, fig.cap="control_GFP-vs-minus_dox.top100pvalues.annotated.heatmap", fig.width=8.5, fig.height=15, fig.show='hold', results='hide'}
lapply(top.stats, function(n){
  eset.temp=n$eset.stats.top.core[1:100,]
  eset.temp=eset.temp[which(!is.na(eset.temp$ID)),]
  contrast=n$contrast
  ## add in gene symbols for each Affy probe
  row.names(eset.temp)=  paste(eset.temp$ID, eset.temp$symbols, sep= "-")
  #write out heatmaps with Affy probeset IDs
  pheatmap(eset.temp[,9:ncol(eset.temp)], main=contrast, color=brewer.pal(11, "RdBu"), fontsize_row=8)
})
```

### Heatmaps of top 25 results

pdf(file.path(resultsDir, "top25.heatmap.pdf"))
```{r heatmaps2, fig.cap="control_GFP-vs-minus_dox.top25pvalues.annotated.heatmap", fig.width=8.5, fig.height=15, fig.show='hold', results='hide'}
lapply(top.stats, function(n){
  eset.temp=n$eset.stats.top.core[1:25,]
  eset.temp=eset.temp[which(!is.na(eset.temp$ID)),]
  contrast=n$contrast
  ## add in gene symbols for each Affy probe
  row.names(eset.temp)=  paste(eset.temp$ID, eset.temp$symbols, sep= "-")
  #write out heatmaps with Affy probeset IDs
  pheatmap(eset.temp[,9:ncol(eset.temp)], main=contrast, color=brewer.pal(11, "RdBu"), fontsize_row=12)
})
dev.off()
```



---

  **How do the samples differ in terms of pathways?**
  
## PATHPRINT
  - an unpublished method that compares pathway activation in a set of samples to those of samples within the GEO dataset database that have been analysed on the same microarray platform[^6] 
- in this case, we are comparing the data to ~2500 datasets from the GEO DataSets database that were analysed on the Mouse Gene 1.0 ST array)


```{r pathprintsetup, results='hide'}
mic.data.platform <- "GPL6246"
mic.data.species <- "mouse"
mic.data.names <- pData(affyNorm.core)[,samplelabel]
                        
mic.data.fingerprint <- exprs2fingerprint(exprs = eset.core, platform = mic.data.platform, species = mic.data.species, progressBar = TRUE)
dimnames(mic.data.fingerprint)[[2]] <- mic.data.names
```

### Most variable pathways in data 
- allowing disagreement between samples within a group (i.e. plus_dox or control_GFP)
- a positive value indicates higher expression of genes in this pathway than typical
- a negative value indicates lower expression of genes in this pathway than typical
- a zero value indicates typical expression levels of genes in this pathway

```{r pathprintresults, fig.width=8.5, fig.height=11, fig.cap=""}
pheatmap(mic.data.fingerprint[apply(mic.data.fingerprint, 1, sd) > 0.6, ], fontsize_row=8,col = c("blue", "white", "red"), cluster_cols=FALSE)
```

### Variable pathways that are consistent within individual groups
- samples within a group must agree for call, must differ between plus and minus doxycycline samples

```{r pathprintresults2,fig.width=8.5, fig.height=11, fig.cap=""}
m=mic.data.fingerprint
pheatmap(m[which(m[,1]==m[,2] & m[,1]==m[,3] & m[,4]==m[,5] & m[,4]==m[,6] & m[,7]==m[,8] & m[,7]==m[,9] & m[,4]!=m[,7]),], labCol = mic.data.names, mar = c(10,25), col = c("blue", "white", "red"), cluster_cols=TRUE)
```

[List of genes within pathways](http://dl.dropbox.com/u/4253254/Resources/pathprint.Mm.gs.txt)

[^1]: [BioC](http://www.bioconductor.org)
[^2]: [arrayQualityMetrics](http://www.bioconductor.org/packages/2.10/bioc/html/arrayQualityMetrics.html)
[^3]: [RMA](http://bioinformatics.oxfordjournals.org/content/19/2/185)
[^4]: [limma](http://www.bioconductor.org/packages/release/bioc/html/limma.html)
[^5]: [RMA](http://bmbolstad.com/misc/ComputeRMAFAQ/ComputeRMAFAQ.html)
[^6]: [Geo_DataSets](http://www.ncbi.nlm.nih.gov/gds/)
)