---
title: "DESeq Analysis"
output:
  html_document:
    toc: true
    theme: spacelab
    fig_width: 12
    fig_height: 12
    fig_caption: false
    highlight: kate
    smart: false
    css: ~/Dropbox/Public/CSS/table.css
---


```{r setup, echo=FALSE}
options(width=200)
library(knitr)
opts_chunk$set(tidy=TRUE, cache=TRUE, echo=FALSE, highlight=TRUE, figalign="center", warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120))

clientname="Andrew Gagne"
clientemail="agagne@genome.med.harvard.edu"
lablocation="Biopolymers Facility"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```
---

RNAseq (DESeq) analysis for `r clientname` (`r clientemail`), at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

---

# METHODS SUMMARY

All reads were trimmed to remove adapter sequences and low quality calls.  Trimmed reads were aligned with [Star][1] and gene specific read counts for [Ensembl gene annotations][2] determined with the featureCounts program from the [subread][3] suite. Read counts were then normalized and differential gene expression calculated with the [DESeq2][4] Bioconductor package.

The sections below provide code to reproduce the post-counting results and plots. 

---

# SETUP

## Libraries

Bioconductor and R libraries used to process and visualize the data

```{r libraries}
library(knitr)
library(DESeq)
library(plyr)
library(reshape2)
library(ggplot2)
library(biomaRt)
library(scales)
library(devtools)
install_git("git://github.com/hbc/CHBUtils.git") # misc personal utilities
library(CHBUtils)
library(knitr)
devtools::install_github("wesanderson","karthik")
require(wesanderson)
# colorblind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
# Wes Anderson palette
pal <-  wes.palette(5,name = "Zissou")
library(RColorBrewer)
bigpalette <- c(brewer.pal(8, "Set2"), brewer.pal(8, "Set1"))
```

## Variables

Working directories, cutoffs and other variables necessary to the analysis.

```{r variables}
if (file.exists("/n/hsphS10/hsphfs1/chb/projects/seidman_rnaseq")) {
  baseDir <- "/n/hsphS10/hsphfs1/chb/projects/seidnman_rnaseq"
  } else if (file.exists("/Users/johnhutchinson/projects/seidman_rnaseq")){
    baseDir <- "/Users/johnhutchinson/projects/seidman_rnaseq"
    } else {
      baseDir <- "/Volumes/ody/projects/seidman_rnaseq"
      }
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")
dataDir <- file.path(baseDir, "project/final/2014-04-24_project")
count.file <- file.path(dataDir, "combined.counts")

pvalcutoff=0.05
numsig=10
rowcount.cutoff=10
logfoldcutoff=1
```


For this analysis, the cutoff for significant differential expression was an FDR (False Discovery Rate) of `r pvalcutoff` and a log2(foldchange) of expression of at least `r logfoldcutoff`.  

--- 

# DATA IMPORT

The data consists of counts generated for individual genes (using the Ensembl GTF annotation) from the Star aligments. The metadata (or covariates) that describes the experiment consists of the  information about sample ID, heart sublocation and sample age.

These data and metadata were and loaded into a DESeq *CountDataSet*.

```{r dataload_and_reshape, results='asis'}
metadata <- read.csv(file.path(metaDir,"project.csv" ))
combined.counts <- read.table(count.file, header=T, row.names=1)
# rename samples to match previous analyses naming scheme
names(combined.counts) <- sub("E14_5", "E14.5", names(combined.counts))
names(combined.counts) <- unlist(lapply(strsplit(names(combined.counts), "_"), function(x) paste(x[2], x[1], sep="_")))
metadata$description <- sub("E14_5", "E14.5", metadata$description)
metadata$description  <- unlist(lapply(strsplit(metadata$description, "_"), function(x) paste(x[2], x[1], sep="_")))

# only one replicate per sample, so just use samplenames get covars from the filenames of the counted samples
covars <- names(combined.counts)
## load up new count dataset
cds <- newCountDataSet(combined.counts, conditions=covars)
kable(as.matrix(metadata))
```

---

# DATA PREPROCESSING

## Normalization

Here is a sample of `r numsig` of the non-normalized counts of reads mapping to
genes in the annotation. These are just chosen at random for display purposes:

```{r rawcounts, results='asis'}
annotated_counts = head(as.data.frame(counts(cds)), numsig)
annotated_counts <- row2colnames(annotated_counts, "id")
annotated_counts = annotate_df(annotated_counts,
                               df_ensemblid_header="id",
                               biomart_ensembl_dataset="mmusculus_gene_ensembl",
                               biomart_ensemblid_filter="ensembl_gene_id",
                               biomart_genesymbol_attribute="mgi_symbol")
kable(annotated_counts)
```

To normalize for different depth of sequencing, size factors are calculated for each sample. 

Call $k_{ij}$ the count for gene $i$ in sample $j$. 

For all samples $j$, assume most of the counts for genes will have a constant ratio comparing any two samples in $j$, reflecting the assumption that most genes are not differentially expressed between samples. 

A reference set of counts is created by taking $f_i$, geometric mean of the counts for each gene $i$, and the median of of $\frac{k_i}{f_i}$ for sample $j$ is called the size factor for sample $j$. 

The size factors can be thought of as an attempt to normalize for the total number of reads mapping to transcripts, depending on the assumption that most of the transcripts are not vastly different between samples.

Here are the size factors for each sample in this experiment:

```{r sizefactors, results='asis'}
cds = estimateSizeFactors(cds)
kable(as.data.frame(sizeFactors(cds)))
```

These are the normalized counts, which are the raw counts divided by the size factor, for the same `r numsig` genes:

```{r normalized_counts, results='asis'}
annotated_normalized = head(as.data.frame(counts(cds, normalized=TRUE)), numsig)
annotated_normalized  <- row2colnames(annotated_normalized, 'id')
annotated_normalized = annotate_df(annotated_normalized,
                               df_ensemblid_header="id",
                               biomart_ensembl_dataset="mmusculus_gene_ensembl",
                               biomart_ensemblid_filter="ensembl_gene_id",
                               biomart_genesymbol_attribute="mgi_symbol")
kable(annotated_normalized)
```

We can see the assumption holds true by plotting the ratio of counts for any two randomly drawn samples and noting the distribution has a large central peak.

In blue are the raw counts and in red are the normalized counts. The effect of normalization is to shift the mean ratio towards 1.

Here are a couple examples:

```{r ratio_hist, out.width='50%', results='hide', fig.show='hold'}
raw.counts = counts(cds, normalized=FALSE)
cols <- sample(ncol(raw.counts),4, replace=F)
norm.counts = counts(cds, normalized=TRUE)

for(x in c(1,3)){
  rawdata = data.frame(ratio=raw.counts[,cols[x]] / raw.counts[,cols[x+1]])
  rawdata$set <- "raw"
  
  normdata = data.frame(ratio=norm.counts[,x] / norm.counts[,x+1])
  normdata$set <- "normalized"
  raw.norm.data <- rbind(rawdata, normdata)
  
  n = ggplot(raw.norm.data, aes(x=ratio, fill=set)) + geom_density(alpha=0.25) +
    scale_x_log10(breaks=c(0.01, 0.1, 1, 10, 100), labels=math_format(format=log10)) +
    labs(title="Normalized counts")
  print(n)
}
rm(raw.counts, norm.counts)
```

## Variance estimation

With replicates you can get a more accurate repesentation of the biological variability or dispersion. Dispersion describes the fluctuation of the observed counts for a gene around the mean counts for that gene and can be the square of the coefficient of biological variation. For example, if a gene's expression level fluctuates by 20% the calculated dispersion is $0.2^2 = 0.04$.

Typically you would calculate the dispersion from your replicates, but here we don't have any. Our assumption then becomes that the mean is a good predictor for the dispersion. Accepting this assumption, we can argue that 
> given two samples from different conditions and a number of genes with comparable expression levels, of which we expect only a minority to be influenced by the condition, we may take the dispersion estimated from comparing across their counts *across* conditions as ersatz for a proper extimate fo the variance across replicates. 

i.e. we assume that the estimated variance should not be too affected by the influence of the few differentially expressed genes

After the empirical dispersion values have been calculated for each gene, a dispersion-mean relationship is fitted to share information across genes to reduce the variability of the dispersion estimates. So we end up with two values for each gene: the empirical value (derived only from this gene's data), and the fitted value (i.e., the dispersion value typical for genes with an average expression similar to those of this gene). Because we have no replicates, and very few samples, we can only use the fitted value. Other methods or choosing between the empirical and fitted gene dispersion values aredesigned to detect false positives from dispersion outliers (i.e. genes with unusually high variability), but without replicates we can't catch these outliers.

Below is a plot of the estimated dispersion for genes with given counts; the dispersion increases as the mean count decreases meaning it is more difficult to accurately measure the expression level of
low-count genes. The red line represents the fitted value we will use.

```{r estimate_sizefactors, results='hide', out.width='50%'}
# sharingMode = maximum, most conservative approach to sharing information across genes to reduce variability of the dispersion estimates
cds <- estimateDispersions(cds, method="blind", sharingMode="fit-only")
ggplotDispEsts(cds)
```

---

# EXPLORATORY ANALYSES

## Unsupervised Clustering
- the goal of this analysis is to naiively evaluate the variability within the raw data and determine whether this variability can predict the different sample categories
- Principal Component Analysis  (PCA) is a dimension reduction and visualisation technique that is used to project the multivariate data vector of each sample into a two-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the samples.
- in basic terms, we are checking whether the samples cluster by sample type; its a good sign if they do

The plots show a scatterplot of the samples along all dual combinations of the first four principal components. Each sample group is represented by a separate color and shape. You can use these plots to explore if the samples cluster, and whether this is according to an intended experimental factor or according to unintended causes such as batch effects. 


```{r pca, out.width='50%', fig.show='hold'}
cds.annot <- cds
pData(cds.annot)$region <- sub("^.*_", "", pData(cds.annot)$condition)
pData(cds.annot)$stage <-  sub("_.*$", "", pData(cds.annot)$condition)

# this is a hack because I can't figure our how DESeq is using the pData, more than the minimal columns interfere with the nbinomTest function
PCAplot.cds(cds.annot, categories="region",categories2="stage", normalize.counts=FALSE, title="Raw counts", colorpalette=cbPalette, alpha=0.8, numcomponents=4)
PCAplot.cds(cds.annot, categories="region",categories2="stage", normalize.counts=TRUE, title="Normalized counts", colorpalette=cbPalette, alpha=0.8, numcomponents=4)
```

These plots only describe  qualitative differences, we explore them more quantitatively in the differential expression analyses below.

---

# DIRECTED ANALYSIS 

- quantitative assessment of sample differences (aka detecting differential expression)

## Filtering

Removing genes with no variance in the samples. 

Using this data, we perform independent filtering to eliminate genes that have no, or little chance of showing signicant evidence. Filtering should result in increased detection power, in terms of false discovery rate. Here, genes were filtered  based on the sum of counts from all samples, as below a certain number of counts, it is not possible to get a pvalue below the desired cutoff. We also filter out any genes without and Ensembl ID annotation.

```{r filter}
## get sum of counts for all samples for each gene
rowcounts <- rowSums(counts(cds, normalized=TRUE))
## filter the data based on the minimal row sum 
use <- (rowcounts > rowcount.cutoff)
cds.filt <- cds[use,]
## annotated?
use <- grepl("ENS", row.names(counts(cds.filt)))
cds.filt <- cds.filt[use,]
```

Here we discarded the genes with less than `r rowcount.cutoff` counts total in all samples, i.e. the bottom `r ecdf(rowcounts)(rowcount.cutoff)*100`% of genes by total counts.

## Statistical tests

Generate data for selected pairwise comparisons, comparing each sample to each other.

```{r pairwise_comparisons,  results='hide'}
## first construct the actual combinations
selected.combos <- matrix(c("RA_P10", "LV_E14.5",
                            "RV_P10", "LA_E14.5",
                            "RA_E14.5", "LV_P10",
                            "RV_E14.5", "LA_P10"), nrow=2)

setnames = apply(selected.combos, 2, function(x) paste(x[1], x[2], sep="-"))

# then run the comparisons
statistical.tests <- alply(selected.combos, 2, function(combo) {
  contrast <- paste(combo[1], combo[2], sep="-vs-")
  print(contrast)
  ## perform significance testing
  nbinomtests <- nbinomTest(cds.filt, combo[1], combo[2])
  nbinomtests <- annotate_df(nbinomtests,
                             df_ensemblid_header="id",
                            biomart_ensembl_dataset="mmusculus_gene_ensembl",
                            biomart_ensemblid_filter="ensembl_gene_id",
                            biomart_genesymbol_attribute="mgi_symbol")
  
  
  allcounts <- row2colnames(counts(cds.filt, normalize=T), "id")
  allresults <- merge(nbinomtests, allcounts, by="id")

  allresults <- allresults[order(allresults$padj),]
  DEresults <-  subset(allresults, padj<pvalcutoff, log2FoldChange>logfoldcutoff)
  return(list(contrast=contrast,allresults=allresults, DEresults=DEresults))
  })
```

Here are links to the count data and statistics for all genes and samples of those pairwise comparisons.

```{r output_all, cache=TRUE}
for (n in 1:length(setnames)){
  output <- statistical.tests[[n]]$allresults
  output.filename <- file.path(resultsDir, paste("counts.stats", setnames[n], "xls", sep="."))
  write.table(output, file=output.filename, quote=F, sep="\t", row.names=F, col.names=T)
  }
```

`r num=1`
[Counts and stats for filtered genes and samples in `r selected.combos[1,num]` vs. `r selected.combos[2,num]` comparison](../results/`r  paste("all.exprs.stats", setnames[num], "xls", sep=".")`)
`r num=num+1`
[Counts and stats for filtered genes and samples in `r selected.combos[1,num]` vs. `r selected.combos[2,num]` comparison](../results/`r paste("all.exprs.stats", setnames[num], "xls", sep=".")`)
`r num=num+1`
[Counts and stats for filtered genes and samples in `r selected.combos[1,num]` vs. `r selected.combos[2,num]` comparison](../results/`r paste("all.exprs.stats", setnames[num], "xls", sep=".")`)
`r num=num+1`
[Counts and stats for filtered genes and samples in `r selected.combos[1,num]` vs. `r selected.combos[2,num]` comparison](../results/`r paste("all.exprs.stats", setnames[num], "xls", sep=".")`)

## Selected comparisons

Here we take a closer look at our selected subset of the pairwise comparisons.

For each comparison, significant genes which passed a `r pvalcutoff*100`% false discovery rate are reported or highlighted.

There are 3 ways to visualize any significant results:  
  - MA-plot, with significantly varying transcripts shaded in red  
  - Volcano plot, with significantly varying transcripts shaded in red  
  - table of transcripts with significantly different expression levels  

`r num=1`

### `r selected.combos[1,num]` vs. `r selected.combos[2,num]`

There are `r nrow(statistical.tests[[num]]$DEresults)` genes with significant differential expression (FDR<`r pvalcutoff` and a log2 fold change in expression of at least `r logfoldcutoff`)

**M-A and volcano plots highlighting genes with significant differential expression** 

```{r plots1, out.width='50%', fig.show='hold', results='asis'}
allresults <- statistical.tests[[num]]$allresults
contrast <- statistical.tests[[num]]$contrast
DEresults <- statistical.tests[[num]]$DEresults
  
## get normalized counts for significant hits, relabel samples with condition rather than sampleID
DEplot <- plotDE(allresults, basemean_colid="baseMean", log2foldchange_colid="log2FoldChange", adj_pval_colid="padj", adj_pval_cutoff=pvalcutoff, plottitle=paste("MAplot for", contrast, sep=" "))
allresults$DE <- ifelse(abs(allresults$log2FoldChange)>logfoldcutoff & allresults$padj<pvalcutoff, "YES", "NO")
volcplot <-   ggplot(allresults, aes(x=log2FoldChange, y=-log10(padj), fill=DE) )+
  geom_point(size=5, alpha=0.25,  pch=21)+
  theme_bw()+
  geom_vline(xintercept=c(-1*logfoldcutoff,logfoldcutoff), color="darkgrey")+
  geom_hline(yintercept=-log10(pvalcutoff), color="darkgrey")+
  scale_y_continuous(name="-log10(adjusted pvalue)")+
  scale_x_continuous(limits=c(-10,10))+
  scale_fill_manual(values=c("#000000", "#FF0000"))+
  ggtitle(paste("Volcano plot for", contrast))

  
DEplot
volcplot
```

**Table of top `r numsig` transcripts showing significant differential expression** 

```{r DEtable1, results='asis'}
if(nrow(DEresults)>0){
kable(head(DEresults, numsig), row.names=FALSE)
} else{
  print(paste("No differential expressed genes to report"))
  }
```


`r num=2`

### `r selected.combos[1,num]` vs. `r selected.combos[2,num]`

There are `r nrow(statistical.tests[[num]]$DEresults)` genes with significant differential expression (FDR<`r pvalcutoff` and a log2 fold change in expression of at least `r logfoldcutoff`)

**M-A and Volcano plots highlighting genes with significant differential expression** 

```{r plots2, ref.label='plots1',out.width='50%', fig.show='hold', results='asis'}
```

**Table of top `r numsig` transcripts showing significant differential expression** 

```{r DEtable2, ref.label='DEtable1',results='asis'}
```


`r num=3`

### `r selected.combos[1,num]` vs. `r selected.combos[2,num]`

There are `r nrow(statistical.tests[[num]]$DEresults)` genes with significant differential expression (FDR<`r pvalcutoff` and a log2 fold change in expression of at least `r logfoldcutoff`)

**M-A and Volcano plots highlighting genes with significant differential expression** 

```{r plots3, ref.label='plots1',out.width='50%', fig.show='hold', results='asis'}
```

**Table of top `r numsig` transcripts showing significant differential expression** 

```{r DEtable3, ref.label='DEtable1',results='asis'}
```


`r num=4`

### `r selected.combos[1,num]` vs. `r selected.combos[2,num]`

There are `r nrow(statistical.tests[[num]]$DEresults)` genes with significant differential expression (FDR<`r pvalcutoff` and a log2 fold change in expression of at least `r logfoldcutoff`)

**M-A and Volcano plots highlighting genes with significant differential expression** 

```{r plots4, ref.label='plots1',out.width='50%', fig.show='hold', results='asis'}
```

**Table of top `r numsig` transcripts showing significant differential expression** 

```{r DEtable4, ref.label='DEtable1',results='asis'}
```


---

# R Session Information

```{r save_image, echo=FALSE}
save(statistical.tests, file=file.path(resultsDir, "RDATA-DESeq"))
sessionInfo()
```


[1]: http://bioinformatics.oxfordjournals.org/content/early/2012/10/25/bioinformatics.bts635 (Star)
[2]: http://tophat.cbcb.umd.edu/igenomes.html (iGenomes)
[3]: http://subread.sourceforge.net/ (subread)
[4]: http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html (DESeq2)
