
```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, fig.height=8, fig.width=8, figalign="center", echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev='svg')
options(width=200)

labPI="Vaidya"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```
---

Network analysis for the `r labPI` group.  Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

---

## Methods summary

Downloaded protein-protein interaction (PPI) network from [iRefIndex](http://irefindex.uio.no/wiki/iRefIndex), and derived significantly modified subnetworks of this PPI based on RNA-seq differential expression.

---

## General Setup

### Working directories, files and other variables necessary to the analysis.

```{r variables }
if (file.exists("/n/hsphS10/hsphfs1/chb/projects/vv_kidney_RNASeq/functional_project")) {
  baseDir <- "/n/hsphS10/hsphfs1/chb/projects/vv_kidney_RNASeq/functional_project"
} else {
  baseDir <- "/Volumes/ody/projects/vv_kidney_RNASeq/functional_project"
}
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
ensemblids <- as.character(read.delim(file.path(dataDir, "pvals", "all.exprs.stats.unfiltered.genes.Normal-vs-FA1day.xls"), header=T, sep="\t")$id)
```

### Libraries

Various R and Bioconductor libraries were used for the PPI network analysis.  
[iRefR](http://www.biomedcentral.com/1471-2105/12/455): A Bioconductor library for interfacing with the iRefIndex PPI database
[igraph](http://cran.r-project.org/web/packages/igraph/index.html): An R library for manipulation of networks (aka graphs).  
[BioNet](http://www.bioconductor.org/packages/2.12/bioc/html/BioNet.html): A Bioconductor library for the integrated analysis of gene expression and networks for the detection of functional subnetworks.  
[RCytoscape](http://www.bioconductor.org/packages/2.13/bioc/html/RCytoscape.html): A Biocondcutor library for interacting with  [Cytoscape](http://www.cytoscape.org/)  

```{r libraries}
library(BioNet)
library(iRefR)
library(igraph)
library(RCytoscape)
library(RColorBrewer) # color palettes
library(org.Mm.eg.db) # for annotation ID conversion
library(reshape2) # data format utility
```

---

## Data Setup

### Download database
Imported the most current mouse iRefIndex protein-protein interaction database. This database was subsetted to only those interactions that were mouse to mouse (iRefIndex catalogs any interaction where at least one protein is mouse derived (eg. mouse-human interactions) as mouse, and we don't want those)


```{r dldatabase}
iref <- get_irefindex("10090", "current", metaDir)
# subset to mouse-mouse interactions
iref <- iref[which(grepl("Mus", iref$taxa) & grepl("Mus", iref$taxb)),]
```

This database contains `r length(unique(iref$irigid))` unique interactions.

### Annotation Conversions
Our data has different annotations, which need to be unified. The RNA-seq (DESeq) results are described by ensembl ID and the PPI used unique 'icrogid' identifiers for proteins. The iRefIndex comes with an annotation table matching each icrogid to an entrez ID, so we can generate a table to interconvert between IDs as so:
rcogid <-> entrezID <-> ensemblID
For some icrogids, there were multiple matched ensemblids, for these I randomly selected the ensemblid.

```{r annotxform,cache=TRUE}
id_conversion_table <- create_id_conversion_table(iref, metaDir, output_filename="id_conversion_table", IDs_to_include="all")
id_conversion_table <- as.data.frame(id_conversion_table)

# only use PPI proteins with an entrez annotation
id_conversion_table <- subset(id_conversion_table, id_type=="entrezgene/locuslink")
id_conversion_table <- droplevels.data.frame(id_conversion_table)

# extract ensemblids for the PPI proteins adn remove dupes
ensemblids.iref <- melt(mget(as.character(id_conversion_table$id_value), org.Mm.egENSEMBL, ifnotfound=NA))
names(ensemblids.iref) <- c("ensemblid", "entrez")
ensemblids.iref <- ensemblids.iref[!duplicated(ensemblids.iref),]
# extract mgi_symbols for the PPI proteins adn remove dupes
symbols <- melt(mget(as.character(id_conversion_table$id_value), org.Mm.egSYMBOL, ifnotfound=NA))
names(symbols) <- c("symbol", "entrez")
symbols <- symbols[!duplicated(symbols),]
# merge ensemblids, symbols and entrezids
newids <- merge(ensemblids.iref, symbols)
# merge ensembldids, symbols and entrezids to the icrogids
id_conversion_table <- merge(newids, id_conversion_table, by.x="entrez", by.y="id_value")
# dump columns you won't need anymore
id_conversion_table$id_type <- NULL
id_conversion_table$irogid <- NULL
# CLEANUP the conversion table
# drop duplicate rows
id_conversion_table <- id_conversion_table[!duplicated(id_conversion_table),]
# for icrogids and entrez ids that have multiple ensemblid/mgi_symbol/entrez matches, just take the first one
id_conversion_table <- aggregate(cbind(cbind(as.character(ensemblid),as.character(symbol), entrez))~icrogid, function(n) n[1], data=id_conversion_table)
names(id_conversion_table) <- c("icrogid", "ensemblid", "symbol", "entrez")
# output for posterity
write.table(id_conversion_table, file.path(metaDir, "id_conversion_table.tab"), sep="\t", quote=F, row.names=F, col.names=T)
```

### Format Conversions
BioNet can only work with particular network/graph formats, the PPI is currently in MITAB format but needs to be converted to GraphNEL (via iGraph). Here, rather than explicitly convert the graph using R functions, I extract the relevant data from the MITAB graph and use it to make a new network/graph.

```{r converttograph, cache=TRUE}
# extract the ids (icrogids) of the protein partners in all interactions from the MITAB file
PPIs <- iref[,c("icrogida", "icrogidb")]
# we want to label the vertices/nodes (which are the icrogids) of the graph with the ensemblids, entrezids, symbols
# the format for these node labels has all the unique identifiers matching the protein partners in all interactions from the MITAB file in the first column
# further annotation are placed in subsequent columns
nodeannots <- id_conversion_table
nodeannots <- nodeannots[!duplicated(nodeannots),]
# the icrogids in the PPIs MUST match the icrogids in annotations
# remove any interactions that have unannotated icrogids
PPIs <- PPIs[apply(PPIs, 1, function(n) all(as.character(n) %in% as.character(nodeannots[,1]))),]

# put the interactions and annotations into a new undirected graph
igraph_trans <- graph.data.frame(PPIs, directed=F, vertices=nodeannots)
# remove self loops and multiple edges
igraph_trans <- simplify(igraph_trans, remove.loops=TRUE, remove.multiple=TRUE)
# convert to graphNEL format for use in BioNet
interactome <- igraph.to.graphNEL(igraph_trans)
```

### Pvalue and logFC import   
For each day of the time course:  
1) Import the expression statistics for the filtered gene set (pvalue and logfoldchange)  
2) Relate the RNAseq ensemblids to the PPI icrogids using the id interconversion table from above.  
3) Summarize stats for each icrogid, taking the mean fold change of expression and the lowest pvalue (if id present multiple times).   
4) Convert the gene expression fold changes to log2(foldchanges).  
5) Subset the PPI network to the genes examined in teh RNAseq experiment (i.e. the filtered set)
6) Assign scores and run the BioNet maximum scoring network identification algorithm. In simple terms, BioNet uses an algorithm to process these gene scores and find a connected subnetwork that maximizes these scores (ie. the sum of all node scores) while minimizing the number of connections needed to establish the subnetwork. In even simpler terms, BioNet will expand a network where the connected genes have positive scores, and stop expanding once enough genes have negative scores. To run this algorirhtm, the gene scores must be on a scale that flanks zero. There are alot of different ways you could derive gene scores with these properties but here I chose to simply pick a pvalue where the positive to negative transition occurs.    In this case, I chose a value of 0.1 and linearly shifted the pvalues so that for pvalues of >0.1 gene scores were negative, and for pvalues of <0.1 gene scores were positive.
7) Plot the subnetwork. Squares indicate genes with negative scores (i.e. non-significant pvalues), red coloring indicates a negative logfoldchange as compared to day 0 and green a postive log fold change.
 
```{r pvalueimport, out.width='75%', cache=TRUE}
statfiles <- list.files(file.path(dataDir, "pvals"), pattern="stats.Normal")
# reorder by day
statfiles <- statfiles[order(as.numeric(sub("day.xls", "", sub("all.exprs.stats.Normal-vs-FA", "", statfiles))))]
for (statfile in statfiles){
  daylabel <- sub(".xls", "", sub("all.exprs.stats.Normal-vs-", "", statfile))
  print(daylabel)
  # import filtered gene stats
  stats <- read.delim(file.path(dataDir, "pvals", statfile), header=T, sep="\t")
  # subset to columns of interest
  stats <- stats[,c("id", "foldChange", "pval")]
  # associate ensembl ids with icrogids
  stats <- merge(stats, id_conversion_table, by.x="id", by.y="ensemblid")
  
  # remove genes with no information about foldchange and/or pvalu
  stats <- subset(stats, !is.na(pval))
  stats <- subset(stats, !is.na(foldChange))
  
  # take median of fold change values for each icrogid
  # fuss with formats as BioNet will only take named vectors as input
  foldchange.med <- aggregate(foldChange ~ icrogid, median, data=stats)
  rownames(foldchange.med) <- foldchange.med$icrogid
  foldchange.med$icrogid <- NULL
  foldchange.med.v <- unlist(foldchange.med)
  names(foldchange.med.v) <- row.names(foldchange.med)
  # convert to log2folfoldchanges
  logfoldchange.v <- log2(foldchange.med.v)
  # restrict data to finite log2foldchanges
  logfoldchange.v <- logfoldchange.v[(is.finite(logfoldchange.v))]
  
  # take minimum pvalue for each icrogid
  # fuss with formats as BioNet is annoying and will only take named vectors as input
  pval.min <- aggregate(pval ~ icrogid, min, data=stats)
  rownames(pval.min) <- pval.min$icrogid
  pval.min$icrogid <- NULL
  pval.min.v <- unlist(pval.min)
  names(pval.min.v) <- row.names(pval.min)
  
  # subset pvalues and logfcs to common genes
  logfoldchange.v <- logfoldchange.v[intersect(names(pval.min.v), names(logfoldchange.v))]
  pval.min.v <- pval.min.v[intersect(names(pval.min.v), names(logfoldchange.v))]
  
  ## Subset PPI interactome to those genes that were examined in the RNAseq experiment
  subnet <- subNetwork(names(pval.min.v), interactome, neighbors="none")
  ## derive positive and negative scores for BioNet algorithm based on a -log10 pvalue cutoff of 1.3 (>1.3==positive, <1.3==negative)
  ## this is equivalent to making a pvalue of 0.05 the cutoff, but transformed onto a log10 scale
  scores <- -log10(pval.min.v)-(-log10(0.05))
  # find enriched subnetworks
  module <- runFastHeinz(subnet, scores)
  # plot the results
  #plotModule(module, scores = scores, diff.expr=-logfoldchange.v, labels=as.character(stats$symbol[match(nodes(module), stats$icrogid)]), main=daylabel)
  
  # convert to igraph and save for Rcytoscape
  module.igraph <- igraph.from.graphNEL(module)
  module.igraph <- set.vertex.attribute(module.igraph, "logfc", value=logfoldchange.v[get.vertex.attribute(module.igraph, "name")])
  module.igraph <- set.vertex.attribute(module.igraph, "scoreclassifier", value=ifelse(as.vector(scores[get.vertex.attribute(module.igraph, "name")])>0, "POS", "NEG"))
  assign(paste(daylabel, "igraph", sep="."), module.igraph)
  # cleanup
  rm(scores, module, module.igraph, subnet, stats, logfoldchange.v,foldchange.med, foldchange.med.v, pval.min, pval.min.v)
    }
```

## Export to Cytoscape

Networks were exported to Cytoscape and a consistent layout applied.

```{r rcytoscape, eval=FALSE}
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

for (daygraph in ls()[grep("igraph$", ls())]){
  daylabel <- sub(".igraph","", daygraph)
  subnet <- get(daygraph)
  subnet <- igraph.to.graphNEL(subnet)
  subnet <- initNodeAttribute(subnet, attribute.name="ensemblid", attribute.type="char", default.value="")
  subnet <- initNodeAttribute(subnet, attribute.name="symbol", attribute.type="char", default.value="")
  subnet <- initNodeAttribute(subnet, attribute.name="entrez", attribute.type="char", default.value="")
  subnet <- initNodeAttribute(subnet, attribute.name="score", attribute.type="numeric", default.value="")
  subnet <- initNodeAttribute(subnet, attribute.name="logfc", attribute.type="numeric", default.value="")
  subnet <- initNodeAttribute(subnet, attribute.name="scoreclassifier", attribute.type="char", default.value="")
  subnet <- initEdgeAttribute(subnet, attribute.name="weight", attribute.type="numeric", default.value="")
  
  cw <- new.CytoscapeWindow(daylabel, subnet)
  displayGraph(cw)
  layoutNetwork(cw, layout.name='force-directed')
  setDefaultNodeShape(cw, "ellipse")
  setDefaultNodeBorderWidth(cw, 20)
  setDefaultNodeSize(cw, 60)
  setDefaultNodeColor(cw, "#FFFFFF")
  setDefaultEdgeColor(cw, "#545454")
  setDefaultBackgroundColor(cw, "#FFFFFF")
  setDefaultEdgeLineWidth(cw,5)
  setDefaultNodeFontSize(cw, 24)
  
  setNodeColorRule (cw, node.attribute.name='logfc', c(-3,0,3), brewer.pal(3, "RdYlBu"), mode='interpolate')

  setNodeBorderColorRule(cw, node.attribute.name='scoreclassifier', control.points=c("POS", "NEG"), c("#000000", "#777777") , mode="lookup")
  setNodeBorderWidthRule(cw, node.attribute.name='scoreclassifier', attribute.values=c("POS", "NEG"), line.widths=c(0, 20) , default.width=20)

  setNodeLabelRule(cw, "symbol")
  
  hidePanel (cw, 'Data Panel')
  hidePanel (cw, 'Control Panel')
  hidePanel(cw, "Results Panel")
  redraw(cw)
  }
```

The cytoscape generated networks can be found [here](../results/Daily.Networks.pdf)

In these plots, each cirle represents one gne in the protein-protein interaction network. The interior color of the circle denotes expression change of that gene relative to controls (Day 0) with colors ranging from blue (lower expression at Day 0) through yellow (no expression change) to red (higher expression at Day 0). The presence of a gray border color denotes that the gene did not show significant expression change (p>0.05). Its interesting to note that many of the network hubs show no expression change.

[Here](../results/Daily.Networks.cys) are the same networks in Cytoscape format.