```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, echo=FALSE, cache=FALSE, highlight=TRUE, figalign="center", warning=FALSE,  error=FALSE, message=FALSE, fig.height=11, fig.width=11)
```

```{r libraries}
library(biomaRt)
ensembl = useMart("ensembl",dataset="mmusculus_gene_ensembl")
library(ggplot2)
library(xtable)
```


```{r variables}
dataDir <- "~/Work/Consults/Scadden-HIF1-secreted.factors-009/data/"
```

## Data Sources
- used two independent sources to get genes with predicted Hif1 binding sites 

### Hif1 associated genes from [cisred](http://www.cisred.org/)
"In promoter-based cisRED databases, sequence search regions for motif discovery extend from 1.5 Kb upstream to 200b downstream of a transcription start site, net of most types of repeats and of coding exons. Many transcription factor binding sites are located in such regions. For each target gene's search region, we use a base set of probabilistic ab initio discovery tools, in parallel, to find over-represented atomic motifs. Discovery methods use comparative genomics with over 40 vertebrate input genomes."

- cisred can be more stringent as 1) it requires consensus between a number of algorithms and 2) it doesn't simply scan sequences for a binding site but instead first "discovers" likely binding sites statistically overrepresented in multiple sequences and then matches them to known motifs and 3) it searches in a fairly small region upstream of the gene (1.5kb)

[Publication](http://www.ncbi.nlm.nih.gov/pubmed/16381958)

```{r datload_cisred_HIF1_targets}
cisred <- read.delim(paste(dataDir, "results.tsv", sep=""))
names(cisred) <- c("Index", "Atomic.motif.ID.crmMmus", "Annotation.pvalue", "Discovery.pvalue", "Motif.location", "Annotated.strand", "In.patterns.crmMmus", "Genes.in.region")
```

### Hif1 associated genes from [oPOSSUM](http://www.cisreg.ca/cgi-bin/oPOSSUM/opossum)  
- this database was created by Dr.Shannan Ho Sui from the HSPH bioinformatics core
- oPOSSUM is a web-based system for the detection of over-represented transcription factor binding sites in the promoters of sets of genes
- oPOSSUM combines a pre-computed database of conserved TFBSs in human and mouse promoters with statistical methods for identification of sites over-represented in a set of co-expressed genes
- oPOSSUM is less conservative than cisred in calling binding sites as 1) it does not require "discovery"" of the motif, but used precomputed (and biologically verified) binding motifs to scan the promotes of genes and 2) here it is searching in a larger region upstream of the genes (5kb)

[Publication](http://www.ncbi.nlm.nih.gov/pubmed/15933209?dopt=Abstract)


```{r dataload_oppossum_targets}
possum.data <- read.delim(paste(dataDir, "HIF1A.txt", sep=""))
possum.data$symbol <- toupper(possum.data$symbol)
```

### Mouse secretory protein data from [Secreted Protein Database](http://spd.cbi.pku.edu.cn/)
- constructed secretomes using both established, biologically verified data and improved secreted protein prediction approaches 
- total of 18152 secreted proteins ranked according to the prediction confidence.
0=known secreted protein
1=predicted secreted protein by two independent algorithmic approaches (predicted to be ~90% accurate)
- there two lower ranks, but I have excluded them here

[Publication](http://nar.oxfordjournals.org/content/33/suppl_1/D169.full)


```{r datload_secdb_ids}
rank0 <- as.data.frame(scan(paste(dataDir,"rank0.genenames", sep=""), what="character"))
rank1 <- as.data.frame(scan(paste(dataDir,"rank1.genenames", sep=""), what="character"))
names(rank0)="symbol"
names(rank1)="symbol"
rank0$type=0
rank1$type=1
sec.prot.ids <- rbind(rank0, rank1)
rm(rank1, rank0)
```

## ANALYSES

- intersected data to find genes that were both predicted to have  Hif1 binding sites in their promoters and be secreted proteins
- many genes have multiple Hif1 binding sites, and large numbers of weak Hif1 sites can be functional, so I attempted to factor in the number of sites as well as the likelihood of a site being present in the gene

```{r cisred_extract_symbols}
#### get gene symbols for cisred data (already included in cisred data, but need to be parsed)
cisred$ensemblID <- do.call(rbind,strsplit(as.vector(cisred$Genes.in.region), " "))[,1]
cisred$symbol <- toupper(sub("\\)","", sub("\\(", "", do.call(rbind,strsplit(as.vector(cisred$Genes.in.region), " "))[,2])))
### Intersect the datasets
HIF1.sec.pr.cisred <- cisred[cisred$symbol %in% sec.prot.ids$symbol,]
HIF1.sec.pr.cisred <- merge(HIF1.sec.pr.cisred, sec.prot.ids, by.x="symbol", by.y="symbol")
HIF1.sec.pr.possum <- possum.data[possum.data$symbol %in% sec.prot.ids[,"symbol"],]
HIF1.sec.pr.possum <- merge(HIF1.sec.pr.possum, sec.prot.ids, by.x="symbol", by.y="symbol")
### Summarize the datasets
cisred.means <- aggregate(HIF1.sec.pr.cisred$Discovery.pvalue, by=list(HIF1.sec.pr.cisred$symbol), mean)
cisred.ranks <- aggregate(HIF1.sec.pr.cisred$type, by=list(HIF1.sec.pr.cisred$symbol), mean)
cisred.nums <- as.data.frame(table(HIF1.sec.pr.cisred$symbol))
HIF1.sec.pr.cisred.summary <- merge(cisred.nums, merge(cisred.means, cisred.ranks, by.x="Group.1", by.y="Group.1"), by.x="Var1", by.y="Group.1")
names(HIF1.sec.pr.cisred.summary) <- c("symbol", "num.sites", "pvalue", "secretory.protein.rank")
possum.means <- aggregate(HIF1.sec.pr.possum$rel_score, by=list(HIF1.sec.pr.possum$symbol), mean)
possum.ranks <- aggregate(HIF1.sec.pr.possum$type, by=list(HIF1.sec.pr.possum$symbol), mean)
possum.nums <- as.data.frame(table(HIF1.sec.pr.possum$symbol))
HIF1.sec.pr.possum.summary <- merge(possum.nums, merge(possum.means, possum.ranks, by.x="Group.1", by.y="Group.1"), by.x="Var1", by.y="Group.1")
names(HIF1.sec.pr.possum.summary) <- c("symbol", "num.sites", "score", "secretory.protein.rank")
```

## RESULTS

### cisred

```{r generate_plot_cisred}
colors <- factor(HIF1.sec.pr.cisred.summary$secretory.protein.rank)
p <- ggplot(data=HIF1.sec.pr.cisred.summary, aes(x=num.sites, y=-log10(pvalue), label=symbol, color=colors)) + geom_point(size=8, alpha=0.7) + scale_colour_discrete(name = "Secretory Protein Confidence Rank\n0 => highest confidence ") + xlab("Number of predicted Hif1 sites") + ylab("Probability of Hif1 site presence (-log10(pvalue))") + geom_text(color=grey, size=3, alpha=0.5,position=position_jitter(w=0.5, h=0))
```

#### Figure 1 - Assessing likelihood of protein having HIF1 site in promoter (cisred) and being a secreted protein

```{r print_plot_cisred, fig.cap=""}
print(p)
```

```{r likely.HIF1.cisred.secreted.proteins}
cisred.likely.secreted.HIF1.targets <- HIF1.sec.pr.cisred.summary[which(HIF1.sec.pr.cisred.summary$pvalue<0.05 & HIF1.sec.pr.cisred.summary$secretory.protein.rank<2),]
cisred.likely.secreted.HIF1.targets$symbol=toupper(cisred.likely.secreted.HIF1.targets$symbol)
cisred.likely.secreted.HIF1.targets <- merge(cisred.likely.secreted.HIF1.targets, getBM(attributes=c("uniprot_genename", "description"), filters='uniprot_genename', values=unlist(cisred.likely.secreted.HIF1.targets$symbol), mart=ensembl), by.x="symbol", by.y="uniprot_genename")
cisred.likely.secreted.HIF1.targets$description <-sub("\\[.+$", "", cisred.likely.secreted.HIF1.targets$description)
```

#### Table1 - Secreted proteins most likely to have Hif1 binding sites in their promoters (cisred)

```{r table, results='asis'}
out1 <- xtable(cisred.likely.secreted.HIF1.targets, label=NULL, digits=c(1,1,1,6,0,1))
print(out1, type="html",include.rownames=FALSE)
```

---

### oPOSSUM 

```{r generate_plot_possum}
colors2 <- factor(HIF1.sec.pr.possum.summary$secretory.protein.rank)
p2 <- ggplot(data=HIF1.sec.pr.possum.summary, aes(x=num.sites, y=score, label=symbol, color=colors2)) + geom_point(size=4, alpha=0.7) + scale_colour_discrete(name = "Secretory Protein Confidence Rank\n0 => highest confidence ") + xlab("Number of predicted Hif1 sites") + ylab("Probability of Hif1 site presence (mean oPOSSUM matrix match score)") + geom_text(color=grey, size=3, alpha=0.5,position=position_jitter(w=0.5, h=0)) + geom_hline(aes(yintercept=as.vector(quantile(HIF1.sec.pr.possum.summary$score,0.95))), color="cyan") + geom_vline(aes(xintercept=as.vector(quantile(HIF1.sec.pr.possum.summary$num.sites,0.95))), color="cyan")
```
#### Figure 2 - Assessing likelihood of protein having HIF1 site in promoter (oPOSSUM) and being a secreted protein

```{r print_plot_possum, fig.cap=""}
print(p2)
```

```{r likely.HIF1.possum.secreted.proteins}
possum.likely.secreted.HIF1.targets <- HIF1.sec.pr.possum.summary[which((HIF1.sec.pr.possum.summary$score>quantile(HIF1.sec.pr.possum.summary$score, 0.95) | HIF1.sec.pr.possum.summary$num.sites>quantile(HIF1.sec.pr.possum.summary$num.sites, 0.95)) & HIF1.sec.pr.possum.summary$secretory.protein.rank<2),]
possum.likely.secreted.HIF1.targets$symbol=toupper(possum.likely.secreted.HIF1.targets$symbol)
possum.likely.secreted.HIF1.targets <- merge(possum.likely.secreted.HIF1.targets, getBM(attributes=c("uniprot_genename", "description"), filters='uniprot_genename', values=unlist(possum.likely.secreted.HIF1.targets$symbol), mart=ensembl), by.x="symbol", by.y="uniprot_genename")
possum.likely.secreted.HIF1.targets$description <-sub("\\[.+$", "", possum.likely.secreted.HIF1.targets$description)
```


- oPOSSUM finds a large number of potential Hif1 targets; to get the most likely targets, I filtered the results for either a high score or a high number of sites, grabbing only targets within the top 5% for either of these values (beyond blue lines on the plot)

#### Table2 - Secreted proteins mostly likely to have Hif1 binding sites in their promoters (oPOSSUM)

```{r table2, results='asis'}
out2 <- xtable(possum.likely.secreted.HIF1.targets, label=NULL, digits=c(1,1,1,6,0,1))
print(out2, type="html")
```

---

### Are there any common Hif1 target secreted proteins in both the Cisred and oPOSSUM databases?  

#### Figure 3 - Plot of potentially secreted proteins having HIF1 site in promoter as assessed by oPOSSUM and cisred (pvalue cutoff of 0.05)

```{r merged_plot}
mergedbs <- merge(HIF1.sec.pr.cisred.summary, HIF1.sec.pr.possum.summary, by=c("symbol", "secretory.protein.rank"))
p3 <- ggplot(data=mergedbs[which(mergedbs$pvalue<0.05),], aes(x=score, y=-log10(pvalue), label=symbol, size=num.sites.x, color=num.sites.y)) + geom_point(alpha=0.7) + xlab("oPOSSUM score") + ylab("-log10(cisred pvalue)") + geom_text(color=grey, size=6, alpha=0.5, position=position_jitter(w=0.01, h=0.01))+ scale_colour_continuous(name = "Number of Hif1 sites predicted by cisred")+ scale_size_continuous(name="Number of Hif1 sites predicted by oPOSSUM") + scale_area(range = c(2, 16))
```
```{r print_plot_merged, fig.cap=""}
print(p3)
```

#### Table 3 - Secreted proteins mostly likely to have Hif1 binding sites in their promoters as assessed by oPOSSUM and cisred (pvalue cutoff of 0.05)

```{r printmerge_dbs, results='asis'}
mergedbs$symbol <- toupper(mergedbs$symbol)
mergedbs <- merge(mergedbs, getBM(attributes=c("uniprot_genename", "description"), filters='uniprot_genename', values=unlist(mergedbs$symbol), mart=ensembl), by.x="symbol", by.y="uniprot_genename")
mergedbs$description <-sub("\\[.+$", "", mergedbs$description)
out3 <- xtable(mergedbs[which(mergedbs$secretory.protein.rank<2 & mergedbs$pvalue<0.05),])
print(out3, type='html',include.rownames=FALSE)
```

---

### Looking at experimentally suggestive targets
- targets of Hif1 that are secreted as suggested by gene expression experiments are VegfA, Cxcl12 (aka SDF1), Epo, Leptin, EprhinA and Pdgf
- how do they look in the Cisred, oPOSSUM and Secreted Protein Databases?

```{r finding_exp_targets_in_data}
exp.targets <- as.data.frame(sort(c("VEGFA", "VEGFB", "VEGFC", "SDF1", "EPO", "LEP", paste("EFNA", seq(1,10), sep=""),"PDGFA")))
names(exp.targets) <- "symbol"
row.names(exp.targets) <- exp.targets$symbol
exp.targets$in.sec.db <- as.vector(exp.targets$symbol %in% sec.prot.ids$symbol)
exp.targets$in.cisred <- exp.targets$symbol %in% cisred$symbol
exp.targets$mean.cisred.pvalue <- ifelse(exp.targets$symbol %in% HIF1.sec.pr.cisred.summary$symbol, HIF1.sec.pr.cisred.summary$pvalue, NA)
exp.targets$num.cisred.sites <- ifelse(exp.targets$symbol %in% HIF1.sec.pr.cisred.summary$symbol, HIF1.sec.pr.cisred.summary$num.sites, NA)
exp.targets$in.opossum <- exp.targets$symbol %in% possum.data$symbol
exp.targets$mean.opossum.score <- ifelse(exp.targets$symbol %in% HIF1.sec.pr.possum.summary$symbol, HIF1.sec.pr.possum.summary$score, NA)
exp.targets$mean.opossum.score.percentile <- round(ecdf(HIF1.sec.pr.possum.summary$score)(exp.targets$mean.opossum.score)*100, digits=2)
exp.targets$num.opossum.sites <- ifelse(exp.targets$symbol %in% HIF1.sec.pr.possum.summary$symbol, HIF1.sec.pr.possum.summary$num.sites, NA)
exp.targets$num.opossum.sites.percentile <- round(ecdf(HIF1.sec.pr.possum.summary$num.sites)(exp.targets$num.opossum.sites)*100, digits=2)
```

#### Table 4 - Experimentally suggested secreted targets of Hif1

```{r printexp, results='asis'}
out4 <- xtable(exp.targets)
print(out4, type='html',include.rownames=FALSE)
```
