---
title: Batch analysis of gender in RRBS data
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: yeti
    code_folding: hide
---


```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Kobzik - RRBS Batch correction"
clientname=" "
clientemail=""
labPI=""
lablocation=""
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

```


```{r knitrsetup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='', tidy.opts=list(keep.blank.line=FALSE, width.cutoff=200), fig.width = 16, fig.height = 14)

```


---

# Kobzik methylation

Examining differences between individual and pooled samples in RRBS data set. 

For this analysis, I kept the samples together for filtering, normalization and unification.

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Setup

## Libraries

R libraries needed to perform analysis

```{r setup}
library(gridExtra)
library(knitr)
library(stringr)
library(ggplot2)
library(CHBUtils)
library(RColorBrewer)
library(pheatmap)
library(GenomicRanges)
library(devtools)
#install_git("https://github.com/druvus/methylKit")
library(methylKit)
library(dplyr)
library(org.Mm.eg.db)
library(readr)

```

## Variables
Working directories, files and other variables necessary to the analysis.

```{r load}
## Setup Data and Results directory variables
if(file.exists("/home/jnh7/projects/kobzik-rrbs")){
	baseDir="/home/jnh7/projects/kobzik-rrbs"
} else{ 
baseDir="/Users/johnhutchinson/Work/projects/kobzik-rrbs/"
}
dataDir <- file.path(baseDir, "data/bed")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

lo.count.cutoff <- 10
q.cutoff <- 0.01
alpha <- 1e-5 #(for logit transforms)

numcores <- parallel::detectCores()
```


# Preparations

## Import Data and Metadata

Used all samples here, including the previous outlier detected by PCA analysis. 

### Metadata

Code needed to load metadata

```{r metadataload, bootstrap.show.output=TRUE}
metadata <- read.table(file.path(metaDir, "meta.txt"),sep="\t",header=T)

metadata.out <- metadata[,c("samplename", "exposure", "type" ,"cohort", "generation")]
kable(metadata.out)
``` 

### Data

Code needed to load methylation data

```{r dataload}
filelist10 <- paste("cpg", metadata$sampleid, "mincov10.txt", sep=".")

testdata <- ifelse(metadata$generation=="cntr", 1, 0)

setwd(dataDir)
meth.quants <- read(as.list(filelist10), sample.id=as.list(as.character(metadata$samplename)),  assembly="mm10", treatment=testdata, context="CpG" )
```
# Analysis

## Data Consolidation

### Filter by coverage 

- removes bases with very high coverage (above 99.9th percentile) to adjust for potential PCR biases
- also remove bases with very low coverage (<10)

```{r methcoveragefilter}
meth.quants.filtered <- filterByCoverage(meth.quants, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)    # methylBase object
```

### Normalize 

Normalizes coverage values between samples using a scaling factor derived from differences between median of coverage distributions

```{r methnormalization}
meth.quants.filtered.norm <- normalizeCoverage(meth.quants.filtered )
```

### Merge samples

- only compare samples at sites where they all have coverage
- reduce data to only these sites 

```{r methmerge, eval=TRUE, cache=TRUE, cache.lazy=FALSE}
meth.quants.filtered.merged <- unite(meth.quants.filtered.norm, destrand = TRUE)
```

## ChrX and ChrY exploration
- looking for potential gender swaps

```{r swapsprep}
meth.quants.filtered.merged.X.1 <- meth.quants.filtered.merged[which(getData(meth.quants.filtered.merged)$chr=="chrX" | getData(meth.quants.filtered.merged)$chr=="chr1"),]

meth.quants.filtered.merged.X <- meth.quants.filtered.merged[which(getData(meth.quants.filtered.merged)$chr=="chrX") ,]
```

### Y chromosome
- coverage and number of sites with at least 10 fold coverage on the Y chromosome

```{r Ycoverage}
median.coverage.Y <- as.data.frame(unlist(lapply(meth.quants.filtered.norm, function(x) median(subset(getData(x), chr=="chrY")$coverage)))) %>% tbl_df()
names(median.coverage.Y) <- "median.coverage"
median.coverage.Y$number.sites <- unlist(lapply(meth.quants.filtered.norm, function(x) length(subset(getData(x), chr=="chrY")$coverage)))
median.coverage.Y$sampleids <-  unlist(lapply(meth.quants.filtered.norm, function(x) x@sample.id))

ggplot(median.coverage.Y, aes(y=median.coverage, x=number.sites,  color=sampleids, label=sampleids))+geom_text(size=8)+theme_bw()+theme(legend.position="none")

```

### PCA of X Chromosome

```{r PCAX}
PCASamples(meth.quants.filtered.merged.X, main="Methylkit Default PCA - chrX sites", scale=FALSE)
```

### Methylation percentage distribution on X chromosome

- expect the X chromosome of female samples to be shifted towards the middle

```{r Xbetavaldist}
sampleids <- meth.quants.filtered.merged.X.1@sample.ids
meth.quant.mat <- getData(meth.quants.filtered.merged.X.1)
meth.quant.mat.m <- reshape2::melt(meth.quant.mat, id.vars=c("chr", "start", "end", "strand")) %>% tbl_df()
rm(meth.quant.mat)
meth.quant.mat.m <- mutate(meth.quant.mat.m, samplenum=unlist(str_extract_all(as.character(meth.quant.mat.m$variable), "(\\d)+$")))
meth.quant.mat.m <- mutate(meth.quant.mat.m, variable=str_replace_all(variable, "(\\d)+$", "" ))
meth.quant.mat.m <- mutate(meth.quant.mat.m, samplid=sampleids[as.numeric(samplenum)])
meth.quant.mat.m <- mutate(meth.quant.mat.m, batch=unlist(lapply(str_split(samplid, "-"), function(x) x[[2]])))
meth.quant.mat.m <- mutate(meth.quant.mat.m, generation=unlist(lapply(str_split(samplid, "-"), function(x) x[[3]])))
numCs.m <- filter(meth.quant.mat.m, variable=="numCs")
numTs.m <- filter(meth.quant.mat.m, variable=="numTs")
# merge the berta value data 
meth.quant.mat.betas.m <- inner_join(numCs.m, numTs.m, by=c("chr", "start", "end", "strand", "samplid", "samplenum", "batch", "generation"))
# calculate beta values
meth.quant.mat.betas.m <-  mutate(meth.quant.mat.betas.m, betaval=value.x/(value.x+value.y)) 
meth.quant.mat.betas.m <- mutate(meth.quant.mat.betas.m, mval=log2(value.x/value.y))
# plot distribtuions of beta values for all samples, spearateluy for each sample "type/generation"" 
ggplot(meth.quant.mat.betas.m, aes(x=betaval, color=samplid, linetype=batch))+geom_density()+facet_wrap(~chr)+ggtitle("All Samples")+theme_bw() + theme(plot.title = element_text(size=22))
ggplot(subset(meth.quant.mat.betas.m,generation=="cntr"), aes(x=betaval, color=samplid, linetype=batch))+geom_density()+facet_wrap(~chr)+scale_color_brewer(type="qual", palette="Paired")+ggtitle("Controls")+theme_bw() + theme(plot.title = element_text(size=22))
ggplot(subset(meth.quant.mat.betas.m,generation=="F1"), aes(x=betaval, color=samplid, linetype=batch))+geom_density()+facet_wrap(~chr)+scale_color_brewer(type="qual", palette="Paired")+ggtitle("F1s")+theme_bw() + theme(plot.title = element_text(size=22))
ggplot(subset(meth.quant.mat.betas.m,generation=="F2"), aes(x=betaval, color=samplid, linetype=batch))+geom_density()+facet_wrap(~chr)+scale_color_brewer(type="qual", palette="Paired") + ggtitle("F2s")+theme_bw() + theme(plot.title = element_text(size=22))
ggplot(subset(meth.quant.mat.betas.m,generation=="F3"), aes(x=betaval, color=samplid, linetype=batch))+geom_density()+facet_wrap(~chr)+scale_color_brewer(type="qual", palette="Paired")+ggtitle("F3s")+theme_bw() + theme(plot.title = element_text(size=22))


temp <- meth.quant.mat.betas.m %>% group_by(.,samplid) %>% summarise(., median(betaval)) 
plot(density(temp$`median(betaval)`))

```

### Sex-specific autosomal sites

From the [Meta-analysis of human methylation data for evidence of sex-specific autosomal patterns, McCarthy et al. 2014, BMC Genomics. 15:981](http://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-15-981) manuscript.

```{r autosomalsites}
autosites <- read_csv(file.path(metaDir, "12864_2014_6710_MOESM4_ESM.csv"))



```
# Conclusions

Although 14-b1-cntr looks fine here, we should still exlcude it on the basis of previous PCA analysis.

Multiple other samples look like they are not purely female,  based on all measures here (with some cases being clearer than others):

#### High number of sites with at least 10 fold coverage on the Y chromosome.  (listed in order of highest to lowest number of sites)
**125-b2-cntr**  
**15-b1-cntr**  
**121-b2-f3**  
92-b2-f2  
16-b1-cntr  
11-b1-f3  
10-b1-f3  
91-b2-f2  

#### Outliers on PCA of X-chromosome (listed in order of biggest to least outlier)
**15-b1-cntr**  
**121-b2-f3**  
**125-b2-cntr**  
91-b2-f2  
92-b2-f2  
10-b1-f3  
11-b1-f3  
16-b1-cntr  

#### X-chromosome methylation distribution outliers 
**121-b2-f3**  
**125-b2-f3**  
**15-b1-cntr**  
91-b2-f2  
92-b2-f2  
10-b1-f3  
11-b1-f3  
16-b1-cntr  


**121-b2-f3, 125-b2-f3 and 15-b1-cntr** consistently show the most evidence for presence of male DNA (high number of Y chromosomes sites, X-chromosome outliers with a methylation percentage distribution more similar to an autosome).

91-b2-f2, 92-b2-f2, 10-b1-f3, 11-b1-f3 &7 16-b1-cntr show evidence of presence of male DNA, but the picture is not as clear.


# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```

---

