---
title: Analysis of RRBS data
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    collapsed: true
    theme: simplex
---


```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Fedulov - RRBS Batch analysis - alternative statistics"
clientname=" "
clientemail=""
labPI=""
lablocation=""
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```

```{r knitrsetup, echo=FALSE}
library(knitr)
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='', tidy.opts=list(keep.blank.line=FALSE, width.cutoff=200), fig.width = 16, fig.height = 14)
```

---

# Fedulov methylation

Trying out different statistical approach for individual and pooled samples in RRBS data set to see if it helps with differences in hypo/hyper methylated numbers.

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Setup

## Libraries

R libraries needed to perform analysis

```{r setup}
library(org.Mm.eg.db)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(stringr)
library(ggplot2)
library(CHBUtils)
library(RColorBrewer)
library(pheatmap)
library(devtools)
library(readr)
library(dplyr)
library(bsseq)
library(DSS)
library(ggbio)
library(biovizBase)

```

## Variables
Working directories, files and other variables necessary to the analysis.

```{r load}
## Setup Data and Results directory variables
if(file.exists("/home/jnh7/projects/kobzik-rrbs")){
	baseDir="/home/jnh7/projects/kobzik-rrbs"
} else{ 
baseDir="/Users/johnhutchinson/Work/projects/kobzik-rrbs/"
}
dataDir <- file.path(baseDir, "data/bed")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

q.cutoff <- 0.05
alpha.cutoff <- 0.05 #(for logit transforms)
beta.delta.cutoff <- 0.25

numcores <- parallel::detectCores()
```

# Preparations

## Import Data and Metadata

Previous analyses identified sample 14-b1-cntr as an outlier so it was not used for this analysis. Also dropped 10-b1-F3, 11-b1-F3, 15-b1-cntr, 16-b1-cntr, 91-b2-F2, 92-b2-F2, 121-b2-F3 and 125-b2-cntr on suspicion of male DNA contamination.

### Metadata

Code needed to load metadata

```{r metadataload, bootstrap.show.output=TRUE}
metadata <- read.table(file.path(metaDir, "meta.txt"),sep="\t",header=T)
# drop questionable samples
drops <- c("10-b1-F3", "11-b1-F3", "14-b1-cntr", "15-b1-cntr", "16-b1-cntr", "91-b2-F2", "92-b2-F2", "121-b2-F3","125-b2-cntr") 
metadata <- metadata[!metadata$samplename %in% drops,]

metadata.out <- metadata[,c("samplename", "exposure", "type" ,"cohort", "generation")]
kable(metadata.out)
``` 

### Data

Code needed to load methylation data
- will also format for bsseq and aggregate strands

```{r dataload, cache=TRUE, cache.lazy=FALSE}
setwd(dataDir)
for (n in 1:length(metadata$sampleid)) {
  sampleid <- metadata$sampleid[n]
  file <- paste("methylcall.CpG", sampleid, "mincov0.txt", sep=".")
  
  dat <- read_delim(file, delim="\t") 
  dat <- mutate(dat, methylated=round(coverage*freqC/100)) %>%
    mutate(., unmethylated=coverage-methylated) %>% 
    dplyr::select(., chr, base, strand, freqC,  methylated, unmethylated, coverage) %>%
    mutate(., methylated=as.integer(methylated), unmethylated=as.integer(unmethylated)) %>%
    arrange(., chr, base, strand)
  
  # collapsing strands
  ## find how far apart the CpG base calls are
  dat <- mutate(dat, diff=c(NA,diff(base)))
  dat.df <- as.data.frame(dat)
  ## Find indices of CpG dyads for which we have both positive and negative strand information
  ### rows should be one base apart (as based on data.dyad$diff), first on positive strand and next on negative strand
  dyad.neg.strand.indices=which(dat.df$diff==1 & dat$strand=="R")
  dyad.pos.strand.indices=dyad.neg.strand.indices-1
  ## get combined coverage from positive and negative strands at dyads
  dyad.combined.numreads <- dat.df[dyad.pos.strand.indices, "coverage"]+dat.df[dyad.neg.strand.indices, "coverage"]
  dyad.combined.numCs <- dat.df[dyad.pos.strand.indices, "methylated"]+dat.df[dyad.neg.strand.indices, "methylated"]
  dyad.combined.numTs <- dat.df[dyad.pos.strand.indices, "unmethylated"]+dat.df[dyad.neg.strand.indices, "unmethylated"]
  ## recalculate percent methylation
  dyad.combined.percmeth <- dyad.combined.numCs/(dyad.combined.numCs+dyad.combined.numTs)*100
  ## load combined coverage into the positive strand position of dyad into table
  dat.df[dyad.pos.strand.indices, "coverage"] <- dyad.combined.numreads
  dat.df[dyad.pos.strand.indices, "methylated"] <- dyad.combined.numCs
  dat.df[dyad.pos.strand.indices, "unmethylated"] <- dyad.combined.numTs
  dat.df[dyad.pos.strand.indices, "freqC"] <- dyad.combined.percmeth
  ## remove negative strand rows of the combined dyads above from table
  dat.df <- dat.df[-dyad.neg.strand.indices,]
  ## find dyads that only have info for the negative strand
  neg.strand.indices <- which(dat.df$strand=="R")
  ## adjust start and end of unpaired negative strand dyads to the positive strand position
  dat.df[neg.strand.indices,"base"] <- dat.df[neg.strand.indices,"base"]-1
  dat.df[neg.strand.indices, "strand"] <- "F"
  ## No need to adjust any dyads for which you only have info from the positive strand
  dat.df$strand <- ifelse(dat.df$strand=="F", "+", "-")
  if (n==1){
    BS <- BSseq( M = as.matrix(dat.df$methylated,ncol=1), Cov = as.matrix(dat.df$coverage,ncol=1), pos = dat.df$base, chr = dat.df$chr,  sampleNames = sampleid) 
  } else {
    BS.temp <- BSseq( M = as.matrix(dat.df$methylated,ncol=1), Cov = as.matrix(dat.df$coverage,ncol=1), pos = dat.df$base, chr = dat.df$chr,  sampleNames =sampleid)
    BS <- bsseq::combine(BS, BS.temp)
  }
}

identical(sampleNames(BS),as.character(unlist(metadata$sampleid)))
pData(BS) <- metadata
sampleNames(BS) <- pData(BS)$sampleid
save(file=file.path(resultsDir, "BSeq.object"), list="BS")
```

# Analysis

**Only for batch 2 (individual samples)**

```{r separatebybatch, cache=TRUE, cache.lazy=FALSE}
# drop chrM
BS <- chrSelectBSseq(BS, seqnames = paste("chr", c(1:19,"X", "Y"), sep=""), order = TRUE)

BS.b2.f1 <- BS[,which(pData(BS)$cohort==2 & (pData(BS)$generation=="F1" | pData(BS)$generation=="cntr"))]
BS.b2.f2 <- BS[,which(pData(BS)$cohort==2 & (pData(BS)$generation=="F2" | pData(BS)$generation=="cntr"))]
BS.b2.f3 <- BS[,which(pData(BS)$cohort==2 & (pData(BS)$generation=="F3" | pData(BS)$generation=="cntr"))]
```

## Statistical Tests
- using [DSS](http://www.ncbi.nlm.nih.gov/pubmed/24561809) based Wald test
- non-smoothed, individual sites only

```{r DSStwogroup, cache=TRUE, cache.lazy=FALSE}
#use pData to subset and get samplenames 
dmlTest.b2.f1 <- DMLtest(BS.b2.f1, group1=as.character(pData(BS.b2.f1)$sampleid[which(pData(BS.b2.f1)$generation=="F1")]),group2=as.character(pData(BS.b2.f1)$sampleid[which(pData(BS.b2.f1)$generation=="cntr")]))

dmlTest.b2.f2 <- DMLtest(BS.b2.f2, group1=as.character(pData(BS.b2.f2)$sampleid[which(pData(BS.b2.f2)$generation=="F2")]),group2=as.character(pData(BS.b2.f2)$sampleid[which(pData(BS.b2.f2)$generation=="cntr")]))

dmlTest.b2.f3 <- DMLtest(BS.b2.f3, group1=as.character(pData(BS.b2.f3)$sampleid[which(pData(BS.b2.f3)$generation=="F3")]),group2=as.character(pData(BS.b2.f3)$sampleid[which(pData(BS.b2.f3)$generation=="cntr")]))

save(file=file.path(resultsDir, "RDATA.dmlTests"), list=c("dmlTest.b2.f1",  "dmlTest.b2.f2", "dmlTest.b2.f3"))
```

**Differentially methylated sites were defined as those with a change in methyulation of at least `r beta.delta.cutoff`, pvalue of less than `r alpha.cutoff` and an FDR of less than  `r q.cutoff`.**

```{r sigdiffs}
diffs.b2.f1 <- callDML(dmlTest.b2.f1, p.threshold = alpha.cutoff, delta=beta.delta.cutoff)
diffs.b2.f1 <- subset(diffs.b2.f1, fdr<q.cutoff)

diffs.b2.f2 <- callDML(dmlTest.b2.f2, p.threshold = alpha.cutoff, delta=beta.delta.cutoff)
diffs.b2.f2 <- subset(diffs.b2.f2, fdr<q.cutoff)

diffs.b2.f3 <- callDML(dmlTest.b2.f3, p.threshold = alpha.cutoff, delta=beta.delta.cutoff)
diffs.b2.f3 <- subset(diffs.b2.f3, fdr<q.cutoff)
```

## Results

### Hyper and Hypo methylation levels

- the different batches now have more similar distributions of hypo and hyper-methylated sites, with the individual samples still showing more hypermethylated sites

**HYPERmethylation** - methylation is *higher* in the F1/F2/F3 over controls

**HYPOmethyation** -  methylation is *lower* in the F1/F2/F3 samples compared to controls

```{r hypohyper}
diffs.b2.f1$batch <- "individual"
diffs.f1 <- diffs.b2.f1 %>% tbl_df()
ggplot(diffs.f1, aes(x=diff,color=batch))+geom_density(lwd=1.1)

chr.hypo.hyper.summary.f1 <- diffs.f1 %>% 
  dplyr::group_by(., chr, batch) %>%
  summarize(., hyper=length(which(diff>0)), hypo=length(which(diff<0))) %>%  
  mutate(., percent.hyper=hyper/(hyper+hypo)*100) %>% mutate(., percent.hypo=100-percent.hyper) %>%
  dplyr::select(., chr, batch, contains("percent")) %>% 
  reshape2::melt(., id.vars=c("chr", "batch"))  
ggplot(chr.hypo.hyper.summary.f1, aes(x=chr, y=value, fill=variable))+geom_bar(stat='identity')+ facet_wrap(~batch) + theme(axis.text.x = element_text(angle = 60, hjust = 1))+ggtitle("F1 vs. Control")

diffs.b2.f2$batch <- "individual"
diffs.f2 <-  diffs.b2.f2 %>% tbl_df()
ggplot(diffs.f2, aes(x=diff,color=batch))+geom_density()

chr.hypo.hyper.summary.f2 <- diffs.f2 %>% 
  dplyr::group_by(., chr, batch) %>%
  summarize(., hyper=length(which(diff>0)), hypo=length(which(diff<0))) %>%  
  mutate(., percent.hyper=hyper/(hyper+hypo)*100) %>% mutate(., percent.hypo=100-percent.hyper) %>%
  dplyr::select(., chr, batch, contains("percent")) %>% 
  reshape2::melt(., id.vars=c("chr", "batch"))  
ggplot(chr.hypo.hyper.summary.f2, aes(x=chr, y=value, fill=variable))+geom_bar(stat='identity')+ facet_wrap(~batch) + theme(axis.text.x = element_text(angle = 60, hjust = 1))

diffs.b2.f3$batch <- "individual"
diffs.f3 <- diffs.b2.f3 %>% tbl_df()
ggplot(diffs.f3, aes(x=diff,color=batch))+geom_density()

chr.hypo.hyper.summary.f3 <- diffs.f3 %>% 
  dplyr::group_by(., chr, batch) %>%
  summarize(., hyper=length(which(diff>0)), hypo=length(which(diff<0))) %>%  
  mutate(., percent.hyper=hyper/(hyper+hypo)*100) %>% mutate(., percent.hypo=100-percent.hyper) %>%
  dplyr::select(., chr, batch, contains("percent")) %>% 
  reshape2::melt(., id.vars=c("chr", "batch"))  
ggplot(chr.hypo.hyper.summary.f3, aes(x=chr, y=value, fill=variable))+geom_bar(stat='identity')+ facet_wrap(~batch) + theme(axis.text.x = element_text(angle = 60, hjust = 1))
```

### Differences in Genomic Context 

```{r karyograms}
data(ideoCyto, package = "biovizBase")

##F1
#dump differences into genomicranges object
diffs.gr.f1 <- with(diffs.f1, GRanges(chr, IRanges(pos, pos),  mu1=mu1, mu2=mu2, diff=diff, diff.se=diff.se, stat=stat,phi1=phi1, phi2=phi2, pval=pval, fdr=fdr, batch=batch))
# add sequence lengths for chromsomes
seqlengths(diffs.gr.f1) <- seqlengths(ideoCyto$mm10)[names(seqlengths(diffs.gr.f1))]
diffs.gr.f1 <- keepSeqlevels(diffs.gr.f1, paste0("chr", c(1:19, "X", "Y")))
#plot with each batch a different color and position so they don't overlap
print("All Differentially Methylated Sites, F1")
autoplot(seqinfo(diffs.gr.f1), layout = "karyogram") +
  layout_karyogram(data = diffs.gr.f1, aes(color=diff), geom = "rect") 
# now only pl9t the hypermethylated differences from each batch
diffs.hyper.gr.f1 <- subset(diffs.gr.f1,  diff>0)
print("HyperMethylated Sites, F1")
autoplot(seqinfo(diffs.hyper.gr.f1), layout = "karyogram") +
  layout_karyogram(data = diffs.hyper.gr.f1, aes(color=diff), geom = "rect")
# now only pl9t the hypomethylated differences from each batch
diffs.hypo.gr.f1 <- subset(diffs.gr.f1,  diff<0)
print("HypoMethylated Sites, F1")
autoplot(seqinfo(diffs.hyper.gr.f1), layout = "karyogram") +
  layout_karyogram(data = diffs.hypo.gr.f1, aes( color=diff), geom = "rect")

##F2
#dump differences into genomicranges object
diffs.gr.f2 <- with(diffs.f2, GRanges(chr, IRanges(pos, pos),  mu1=mu1, mu2=mu2, diff=diff, diff.se=diff.se, stat=stat,phi1=phi1, phi2=phi2, pval=pval, fdr=fdr, batch=batch))
# add sequence lengths for chromsomes
seqlengths(diffs.gr.f2) <- seqlengths(ideoCyto$mm10)[names(seqlengths(diffs.gr.f2))]
diffs.gr.f2 <- keepSeqlevels(diffs.gr.f2, paste0("chr", c(1:19, "X", "Y")))
#plot with each batch a different color and position so they don't overlap
print("All Differentially Methylated Sites, F2")
autoplot(seqinfo(diffs.gr.f2), layout = "karyogram") +
  layout_karyogram(data = diffs.gr.f2, aes(color=diff), geom = "rect") 
# now only pl9t the hypermethylated differences from each batch
diffs.hyper.gr.f2 <- subset(diffs.gr.f2,  diff>0)
print("HyperMethylated Sites, F2")
autoplot(seqinfo(diffs.hyper.gr.f2), layout = "karyogram") +
  layout_karyogram(data = diffs.hyper.gr.f2, aes(color=diff), geom = "rect")
# now only pl9t the hypomethylated differences from each batch
diffs.hypo.gr.f2 <- subset(diffs.gr.f2,  diff<0)
print("HypoMethylated Sites, F2")
autoplot(seqinfo(diffs.hyper.gr.f2), layout = "karyogram") +
  layout_karyogram(data = diffs.hypo.gr.f2, aes( color=diff), geom = "rect")

##F3
#dump differences into genomicranges object
diffs.gr.f3 <- with(diffs.f3, GRanges(chr, IRanges(pos, pos),  mu1=mu1, mu2=mu2, diff=diff, diff.se=diff.se, stat=stat,phi1=phi1, phi2=phi2, pval=pval, fdr=fdr, batch=batch))
# add sequence lengths for chromsomes
seqlengths(diffs.gr.f3) <- seqlengths(ideoCyto$mm10)[names(seqlengths(diffs.gr.f3))]
diffs.gr.f3 <- keepSeqlevels(diffs.gr.f3, paste0("chr", c(1:19, "X", "Y")))
#plot with each batch a different color and position so they don't overlap
print("All Differentially Methylated Sites, F3")
autoplot(seqinfo(diffs.gr.f3), layout = "karyogram") +
  layout_karyogram(data = diffs.gr.f3, aes(color=diff), geom = "rect") 
# now only pl9t the hypermethylated differences from each batch
diffs.hyper.gr.f3 <- subset(diffs.gr.f3,  diff>0)
print("HyperMethylated Sites, F3")
autoplot(seqinfo(diffs.hyper.gr.f3), layout = "karyogram") +
  layout_karyogram(data = diffs.hyper.gr.f3, aes(color=diff), geom = "rect")
# now only pl9t the hypomethylated differences from each batch
diffs.hypo.gr.f3 <- subset(diffs.gr.f3,  diff<0)
print("HypoMethylated Sites, F3")
autoplot(seqinfo(diffs.hyper.gr.f3), layout = "karyogram") +
  layout_karyogram(data = diffs.hypo.gr.f3, aes( color=diff), geom = "rect")
```


```{r annotatedist}
require(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
library(ChIPseeker)
# Hack chipseeker function to add annotaiton
getGenomicAnnoStat <- function(peakAnno) {
    if ( class(peakAnno) == "GRanges" )
        peakAnno <- as.data.frame(peakAnno)
    anno <- peakAnno$annotation
    ## anno <- sub(" \\(.+", "", anno)

    anno[grep("exon 1 of", anno)] <- "1st Exon"
    anno[grep("intron 1 of", anno)] <- "1st Intron"
    anno[grep("Exon \\(", anno)] <- "Other Exon"
    anno[grep("Intron \\(", anno)] <- "Other Intron"
    anno[grep("Downstream", anno)] <- "Downstream (<=3kb)"
    anno[grep("Promoter", anno)] <- "Promoter"
    ## count frequency
    anno.table <- table(anno)
    
    ## calculate ratio
    anno.ratio <- anno.table/ sum(anno.table) * 100
    anno.df <- as.data.frame(anno.ratio)
    colnames(anno.df) <- c("Feature", "Frequency")
    anno.df$Numbers <- anno.table
    lvs <- c(
             "Promoter",
             "CpGIsland", #here  new class
             "5' UTR",
             "3' UTR",
             "1st Exon",
             "Other Exon",
             "1st Intron",
             "Other Intron",
             "Downstream (<=3kb)",
             "Distal Intergenic",
             "Others")
    anno.df$Feature <- factor(anno.df$Feature, levels=lvs[lvs %in% anno.df$Feature])
    anno.df <- anno.df[order(anno.df$Feature),]
    return(anno.df)
}

# get CpG
cpg = readr::read_tsv("http://hgdownload.soe.ucsc.edu/goldenPath/mm10/database/cpgIslandExt.txt.gz", col_names = FALSE,progress = FALSE)
cpg_r = makeGRangesFromDataFrame(cpg, 
                                 keep.extra.columns = FALSE, 
                                 ignore.strand = TRUE, end.field = "X4",
                                 start.field = "X3", seqnames.field = "X2")

plot_annotations <- function(df_gr, samples=""){
  # here we create the annotation
  an = annotatePeak(df_gr, TxDb = txdb, tssRegion = c(-1000,1000), 
                    annoDb = "org.Mm.eg.db", verbose = FALSE)
  
  idx = findOverlaps(an@anno, cpg_r)
  an@detailGenomicAnnotation[,"cpg"] = FALSE
  an@detailGenomicAnnotation[queryHits(idx),"cpg"] = TRUE
  an@anno$annotation[queryHits(idx)] = "CpGIsland"
  .df = getGenomicAnnoStat(an@anno)
  slot(an, "annoStat") <- .df
  plotAnnoBar(an, title=paste("Feature Distribution", samples, sep=" - "))
}

plot_annotations(df_gr=diffs.gr.f1, samples="Batch1, F1 vs. Control")

plot_annotations(df_gr=diffs.gr.f2, samples="Batch1, F2 vs. Control")

plot_annotations(df_gr=diffs.gr.f3, samples="Batch1, F3 vs. Control")
```


### Output Results

#### Differentially Methylated CpGs

```{r annotategene, results="hide"}
library(bumphunter)
library(xlsx)
genes.gr <- annotateTranscripts(TxDb.Mmusculus.UCSC.mm10.knownGene, annotationPackage = "org.Mm.eg.db", by = c("gene"), codingOnly=FALSE, verbose = TRUE, requireAnnotation = TRUE)

genes.diffs.f1.gr <- matchGenes(diffs.gr.f1, genes.gr, type = c("any", "fiveprime"), promoterDist = 2500, skipExons = FALSE, verbose = TRUE)
genes.diffs.f1.gr <- cbind(diffs.gr.f1, genes.diffs.f1.gr)
write.csv(as.data.frame(genes.diffs.f1.gr), file=file.path(resultsDir, "differentially.methylated.CpGs.f1.vs.control.csv"))
write.xlsx(as.data.frame(genes.diffs.f1.gr), file=file.path(resultsDir, "differentially.methylated.CpGs.f1.vs.control.xlsx"), sheetName="Sheet1")

genes.diffs.f2.gr <- matchGenes(diffs.gr.f2, genes.gr, type = c("any", "fiveprime"), promoterDist = 2500, skipExons = FALSE, verbose = TRUE)
genes.diffs.f2.gr <- cbind(diffs.gr.f2, genes.diffs.f2.gr)
write.csv(as.data.frame(genes.diffs.f2.gr), file=file.path(resultsDir, "differentially.methylated.CpGs.f2.vs.control.csv"))
write.xlsx(as.data.frame(genes.diffs.f2.gr), file=file.path(resultsDir, "differentially.methylated.CpGs.f2.vs.control.xlsx"), sheetName="Sheet1")

genes.diffs.f3.gr <- matchGenes(diffs.gr.f3, genes.gr, type = c("any", "fiveprime"), promoterDist = 2500, skipExons = FALSE, verbose = TRUE)
genes.diffs.f3.gr <- cbind(diffs.gr.f3, genes.diffs.f3.gr)
write.csv(as.data.frame(genes.diffs.f3.gr), file=file.path(resultsDir, "differentially.methylated.CpGs.f3.vs.control.csv"))
write.xlsx(as.data.frame(genes.diffs.f3.gr), file=file.path(resultsDir, "differentially.methylated.CpGs.f3.vs.control.xlsx"), sheetName="Sheet1")
```

**Table legend**

 - *chr* chromosome number
 - *start* chromosome position
 - *end* chromosome position
 - *width* width of CpG feature
 - *strand* chromosomal strand (not relevant here as all CpG dyads were destranded for the analysis)
 - *mu1* mean methylation of group1 (here, F1, F2 or F3)
 - *mu2* mean methylaiton of group2 (here, controls)
 - *diff* difference in mean methlation levels between the two groups
 - *diff.se* standard error of the methylation difference
 - *stat*, *phi1*, *phi2* wald statistics
 - *pval* pvalue
 - *fdr* multiple testing adjusted pvalue
 - *batch* batch of experiment (should all be individual here)
 - *name* gene symbol of closest gene
 - *annotation* refseq ID of closeset gene
 - *description* position of CpG relative to gene
 - *region* general position of CpG relative to gene
 - *distance* distance before 5' end of gene
 - *subregion*  information about position of gene if inside exon/intron
 - *insidDistance* distance past 5' end of gene
 - *exonnumber* which exon CpG is in (if relevant)
 - *nexons* number of exons in gene
 - *UTR*  information about position of gene if inside UTR
 - *strand* strand gene transcript is on
 - *geneL* gene length
 - *condingL* coding length of gene
 - *Entrez* entrez id of gene
 - *subjectHits* index of gene in database that was queried (not relevant)


 
[F1 vs. control differentially methylated sites CSV file](../results/differentially.methylated.CpGs.f1.vs.control.csv)

[F1 vs. control differentially methylated sites Excel file](../results/differentially.methylated.CpGs.f1.vs.control.xlsx)

[F2 vs. control differentially methylated sites CSV file](../results/differentially.methylated.CpGs.f2.vs.control.csv)

[F2 vs. control differentially methylated sites Excel file](../results/differentially.methylated.CpGs.f2.vs.control.xlsx)

[F3 vs. control differentially methylated sites CSV file](../results/differentially.methylated.CpGs.f3.vs.control.csv)

[F3 vs. control differentially methylated sites Excel file](../results/differentially.methylated.CpGs.f3.vs.control.xlsx)



# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```

---

