---
output:
  knitrBootstrap::bootstrap_document:
    title: "Kobzik - RRBS Batch correction"
    theme: readable
    highlight: zenburn
    theme.chooser: FALSE
    highlight.chooser: FALSE
---


```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Kobzik - RRBS Batch correction"
clientname=" "
clientemail=""
labPI=""
lablocation=""
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

```


```{r knitrsetup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", bootstrap.show.code=FALSE, bootstrap.show.output=FALSE, bootstrap.thumbnail.size="col-md-10", cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='', tidy.opts=list(keep.blank.line=FALSE, width.cutoff=200), fig.width = 16, fig.height = 14)

```


---

# Kobzik methylation

Examining differences between individual and pooled samples in RRBS data set. 

For this analysis, I kept the samples together for filtering, normalization and unification.

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Setup

## Libraries

R libraries needed to perform analysis

```{r setup}
library(gridExtra)
library(knitr)
library(ggplot2)
library(CHBUtils)
library(RColorBrewer)
library(pheatmap)
library(GenomicRanges)
library(devtools)
#install_git("https://github.com/druvus/methylKit")
library(methylKit)
library(dplyr)
library(org.Mm.eg.db)

```

## Variables
Working directories, files and other variables necessary to the analysis.

```{r load}
## Setup Data and Results directory variables
if(file.exists("/home/jnh7/projects/kobzik-rrbs")){
		baseDir="/home/jnh7/projects/kobzik-rrbs"
} else{ 
	baseDir="/Users/johnhutchinson/Work/projects/kobzik-rrbs/"
}
dataDir <- file.path(baseDir, "data/bed")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

lo.count.cutoff <- 10
q.cutoff <- 0.01
```

## Functions 

``` {r functions}
rowSds <- function(x, center=NULL, ...) {
  n <- !is.na(x);
  n <- rowSums(n);
  n[n <= 1] <- NA;
    if (is.null(center)) {
    center <- rowMeans(x, ...);
  }
  x <- x - center;
  x <- x*x;
  x <- rowSums(x, ...);
  x <- x/(n-1);
  sqrt(x);
}

prepPCA <- function(x,  
                    sd.filter=TRUE,
                    sd.threshold=0.5, 
                    filterByQuantile=TRUE, 
                    scale=TRUE,
                    center=TRUE, 
                    logit=FALSE,
                    alpha=0.01){
  mat = getData(x)
  # remove rows containing NA values, they might be introduced at unite step
  mat      = mat[ rowSums(is.na(mat))==0, ]
  
  if(logit){
    meth.mat <- log2((mat[,x@numCs.index]+1e-5)/(mat[,x@numTs.index]+1e-5))
  } else {
    meth.mat = mat[, x@numCs.index]/(mat[,x@numCs.index] + mat[,x@numTs.index] )
  }
  names(meth.mat)=x@sample.ids
  
  if(sd.filter){
    if(filterByQuantile){
      sds=rowSds(as.matrix(meth.mat))
      cutoff=quantile(sds,sd.threshold)
      meth.mat=meth.mat[sds>cutoff,]
    }else{
      meth.mat=meth.mat[rowSds(as.matrix(meth.mat))>sd.threshold,]
    }
  }
  sample.ids=x@sample.ids
  meth.pr = prcomp((meth.mat),scale=TRUE,center=TRUE)
  return(meth.pr)
}

methylPCA <- function(meth.pr,  
                      meth.prext.metadata="NULL", 
                      metadata.merge.col="NULL",
                      comp1=1,
                      comp2=2, 
                      adj.lim=c(0.001,0.1), 
                      obj.return=FALSE, 
                      title="PCA Plot"){
  loads = meth.pr$rotation
  expvar <- (meth.pr$sdev)^2 / sum(meth.pr$sdev^2)
  
  df <- data.frame(pc1=loads[,comp1], pc2=loads[,comp2], id=rownames(loads))
  df <- merge(df, metadata, by.x="id", by.y=metadata.merge.col)
  
  p <- ggplot(df, aes(pc1, pc2, col=generation, shape=as.factor(cohort))) +
    geom_point(size=5) +
    labs(list(x =paste("loadings for PC",comp1," (", round(expvar[comp1]*100,1), "%)", sep=""),
              y = paste("loadings for PC",comp2," (", round(expvar[comp2]*100,1), "%)", sep="")))+
    scale_x_continuous() +
    ggtitle(title) +
    theme(plot.title = element_text(size=20, face="bold", vjust=2)) +
    theme_bw()
  return(p)
}  
  
  
cor.meth <- function(object, method="spearman"){
  meth.mat = getData(object)[, object@numCs.index]
  (getData(object)[,object@numCs.index] + getData(object)[,object@numTs.index] )    
  names(meth.mat) = object@sample.ids
  cor.mat <- ( cor(meth.mat,method=method) )
  cor.mat
}

plot.targetanno.cpgi <- function(mydiff, cpg, id){
  diffCpGann=annotate.WithFeature.Flank(mydiff,cpg$CpGi,cpg$shores,feature.name="CpGi",flank.name="shores")
  #getFeatsWithTargetsStats(diffCpGann,percentage=TRUE)
  #plotTargetAnnotation(diffCpGann,col=c("green","gray","white"), main=paste(id, "- differential methylation annotation"))
  plotTargetAnnotation(diffCpGann,col="Set1", main=paste(id, "- differential methylation annotation"))
 }

plot.targetanno.pei <- function(mydiff, gene, id){
  diffAnn=annotate.WithGenicParts(mydiff,gene)
  #getTargetAnnotationStats(diffAnn,percentage=TRUE,precedence=TRUE)
  #props=getTargetAnnotationStats(diffAnn,percentage=TRUE,precedence=T)
  #names(props)=paste( paste(round(props),"%"),sep=" ")
  plotTargetAnnotation(diffAnn,precedence=TRUE, col="Set1", main=paste(id, "- differential methylation annotation"))
}

annodiff2 <- function(diff, idx, gene, cpg, all=F){
  if (all){
    idx <- 1:nrow(diff)
  }
  dselect <- methylKit::select(diff, idx)
  anno <- annotate.WithGenicParts(dselect,gene)
  gene.info <- getMembers(anno)
  cpgi.info <- getMembers(annotate.WithFeature.Flank(dselect,cpg$CpGi,cpg$shores,feature.name="CpGi",flank.name="shores"))
  
  assoc <- getAssociationWithTSS(anno)
  
  if (all){
    out <- data.frame(getData(dselect), assoc[,-1])
  } else {
    out <- data.frame(getData(dselect)[,1:3], assoc[,-1])
  }
  symbol <- convertIDs(out$feature.name, "ENSEMBL", "SYMBOL", org.Mm.eg.db, "useFirst")
  entrez <- convertIDs(out$feature.name, "ENSEMBL", "ENTREZID", org.Mm.eg.db, "useFirst")
  description <- convertIDs(out$feature.name, "ENSEMBL", "GENENAME", org.Mm.eg.db, "useFirst")
  refseq <- convertIDs(out$feature.name, "ENSEMBL", "REFSEQ", org.Mm.eg.db, "useFirst")
  
  out <- data.frame(out, gene.info, cpgi.info, symbol=symbol, entrez=entrez, description=description, refseq=refseq)
  return(out)
}

annodiff <- function(diff, idx, gene, cpg, all=F){
  if (all){
    idx <- 1:nrow(diff)
  }
  dselect <- methylKit::select(diff, idx)
  anno <- annotate.WithGenicParts(dselect,gene)
  gene.info <- getMembers(anno)
  cpgi.info <- getMembers(annotate.WithFeature.Flank(dselect,cpg$CpGi,cpg$shores,feature.name="CpGi",flank.name="shores"))
  
  assoc <- getAssociationWithTSS(anno)
  
  if (all){
    out <- data.frame(getData(dselect), assoc[,-1])
  } else {
    out <- data.frame(getData(dselect)[,1:3], assoc[,-1])
  }
  symbol <- convertIDs(out$feature.name, "REFSEQ", "SYMBOL", org.Mm.eg.db, "useFirst")
  entrez <- convertIDs(out$feature.name, "REFSEQ", "ENTREZID", org.Mm.eg.db, "useFirst")
  description <- convertIDs(out$feature.name, "REFSEQ", "GENENAME", org.Mm.eg.db, "useFirst")
  ensembl <- convertIDs(out$feature.name, "REFSEQ", "ENSEMBL", org.Mm.eg.db, "useFirst")
  
  out <- data.frame(out, gene.info, cpgi.info, symbol=symbol, entrez=entrez, description=description, ensembl=ensembl)
  return(out)
}


convertIDs <- function( ids, from, to, db, ifMultiple=c("putNA", "useFirst")) {
  stopifnot( inherits( db, "AnnotationDb" ) )
  ifMultiple <- match.arg( ifMultiple )
  suppressWarnings( selRes <- AnnotationDbi::select(
    db, keys=ids, keytype=from, columns=c(from,to) ) )
  if ( ifMultiple == "putNA" ) {
    duplicatedIds <- selRes[ duplicated( selRes[,1] ), 1 ]
    selRes <- selRes[ ! selRes[,1] %in% duplicatedIds, ]
  }
  return( selRes[ match( ids, selRes[,1] ), 2 ] )
}

manhattan.plot <- function(x, main="", col= c("red", "blue"), cutoff = 3, ymax=NULL) {
  dm <- getData(x)
  dm$chr <- gsub("chr", "", dm$chr)
  dm <- dm[dm$chr != "X" & dm$chr != "Y",]
  dm$chr <- as.numeric(dm$chr)
  dm <- dm[,c(1,2,6,7)]
  colnames(dm)<- c("CHR", "S", "Q", "MD")
  gr.dm <- transformDfToGr(dm, seqnames = "CHR", start = "S", width=1)
  gr.dm <- keepSeqlevels(gr.dm, as.character(1:19))
  data(ideoCyto, package = "biovizBase")
  seqlengths(gr.dm) <- as.numeric(seqlengths(ideoCyto$mm10)[1:19])
  gr.dm <- gr.dm[!is.na(gr.dm$Q)]
  ## transform pvalue
  values(gr.dm)$pvalue <- -log10(values(gr.dm)$Q)
  if (is.null(ymax) ) {
    ymax <- max(values(gr.dm)$pvalue)
  }  
  ylim <- range(0, ymax)
  values(gr.dm)$pvalue[values(gr.dm)$pvalue > ymax] <- Inf
  plotGrandLinear(gr.dm, aes(y = pvalue), color = col, cutoff = cutoff, cutoff.color = "blue", cutoff.size = 0.2, main=main, ylim=ylim)
}
```

# Preparations

## Import Data and Metadata

Previous analyses identified sample 14-b1-cntr as an outlier so it was not used for this analysis.
Also dropped 15-b1-cntr due to odd beta value distribution

### Metadata

Code needed to load metadata

```{r metadataload, bootstrap.show.output=TRUE}
metadata <- read.table(file.path(metaDir, "meta.txt"),sep="\t",header=T)
# drop male sample
metadata <- metadata[-grep("14-b1-cntr", metadata$samplename),]
metadata <- metadata[-grep("15-b1-cntr", metadata$samplename),]

metadata.out <- metadata[,c("samplename", "exposure", "type" ,"cohort", "generation")]
kable(metadata.out)
``` 

### Data

Code needed to load methylation data

```{r dataload, eval=FALSE}
filelist10 <- paste("cpg", metadata$sampleid, "mincov10.txt", sep=".")

testdata <- ifelse(metadata$generation=="cntr", 1, 0)

setwd(dataDir)
meth.quants <- read(as.list(filelist10), sample.id=as.list(as.character(metadata$samplename)),  assembly="mm10", treatment=testdata, context="CpG" )
```

# Analysis

## Data Consolidation

### Filter by coverage 

- removes bases with very high coverage (above 99.9th percentile) to adjust for potential PCR biases
- also remove bases with very low coverage (<10)

```{r methcoveragefilter, eval=FALSE}
meth.quants.filtered <- filterByCoverage(meth.quants, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)    # methylBase object
```

### Normalize 

Normalizes coverage values between samples using a scaling factor derived from differences between median of coverage distributions


```{r methnormalization, eval=FALSE}
meth.quants.filtered.norm <- normalizeCoverage(meth.quants.filtered )
```

### Merge samples

- only compare samples at sites where they all have coverage
- reduce data to only these sites 

```{r methmerge, eval=FALSE}
meth.quants.filtered.merged <- unite(meth.quants.filtered.norm, destrand = TRUE)
rm(meth.quants, meth.quants.filtered, meth.quants.filtered.norm)
save(file=file.path(resultsDir, "meth.quants.filtered.merged"), list="meth.quants.filtered.merged")
```

```{r methmergeload, echo=FALSE}
load(file=file.path(resultsDir, "meth.quants.filtered.merged"))
```
## Data Exploration

### Sample correlations

#### Correlation heatmap

- distance = pearson correlation
- agglomeration methods = ward

```{r correlations, fig.cap="Correlation matrix", fig.width = 8, fig.height = 8}
cor.mat  <- cor.meth(meth.quants.filtered.merged, method="pearson")
#gpairs(cor.mat, lower.pars=list(scatter="lm"), upper.pars=list(scatter="stats"), stat.pars=list(fontsize=16, use.color=FALSE), scatter.pars=list(pch=20, col="#00000022"), diag.pars=list(fontsize=10, show.hist=FALSE))
annotation_col <- data.frame(generation=metadata$generation, batch=factor(metadata$cohort))
rownames(annotation_col) <- metadata$samplename
pheatmap(cor.mat, clustering_method="ward.D2", annotation_col=annotation_col)

rm(cor.mat) ## cleanup
```

#### Dendrogram clustering

- distance = correlation
- methods = ward

```{r clustering, fig.cap="Methylation-based sample dendrogram", fig.width = 8, fig.height = 8}
meth.dendrogram <- clusterSamples(meth.quants.filtered.merged, dist="correlation", method="ward", plot=FALSE)

plot(meth.dendrogram)

rm(meth.dendrogram) ## cleanup
```

#### Principal Component Analysis

__All samples__

Overview of all included samples

```{r PCA.dimension.plot.all, fig.cap="Plot of First Two Principals Components", fig.width = 8, fig.height = 8}
meth.pr <- prepPCA(meth.quants.filtered.merged)
methylPCA(meth.pr, meth.prext.metadata = metadata, metadata.merge.col = "samplename", title="PCA", comp1=1, comp2=2)
```

There does appear to be some level of batch effect in the data, as generally speaking the individual samples cluster with each other and the pooled samples cluster with each other.

This is highly evident when you look at hypo and hyper methylated regions for the individual batches in the previous results. Though this might also be due each batch (and genearation) being treated separately i.e. they could be looking at different sites. How does the data look if you only split up the samples after filtering, normalization and unification?

```{r batches}
#batch 1
meth.quants.b1F1.filtered.merged <- reorganize(meth.quants.filtered.merged ,sample.ids=c("16-b1-cntr", "17-b1-cntr", "1-b1-F1", "2-b1-F1", "3-b1-F1", "4-b1-F1"), treatment=c(0,0,1,1,1,1) )
meth.quants.b1F2.filtered.merged <- reorganize(meth.quants.filtered.merged ,sample.ids=c( "16-b1-cntr", "17-b1-cntr",  "5-b1-F2", "6-b1-F2", "7-b1-F2", "8-b1-F2"), treatment=c(0,0,1,1,1,1) )
meth.quants.b1F3.filtered.merged <- reorganize(meth.quants.filtered.merged ,sample.ids=c( "16-b1-cntr", "17-b1-cntr",  "10-b1-F3", "11-b1-F3", "12-b1-F3", "13-b1-F3"), treatment=c(0,0,1,1,1,1) )

#batch 2
meth.quants.b2F1.filtered.merged <- reorganize(meth.quants.filtered.merged  ,sample.ids=c("123-b2-cntr","124-b2-cntr", "125-b2-cntr", "126-b2-cntr", "60-b2-F1", "61-b2-F1", "62-b2-F1", "63-b2-F1"), treatment=c(0,0,0,0,1,1,1,1) )
meth.quants.b2F2.filtered.merged <- reorganize(meth.quants.filtered.merged  ,sample.ids=c("123-b2-cntr","124-b2-cntr", "125-b2-cntr", "126-b2-cntr", "90-b2-F2", "91-b2-F2", "92-b2-F2", "93-b2-F2"), treatment=c(0,0,0,0,1,1,1,1) )
meth.quants.b2F3.filtered.merged <- reorganize(meth.quants.filtered.merged  ,sample.ids=c("123-b2-cntr","124-b2-cntr", "125-b2-cntr", "126-b2-cntr", "119-b2-F3", "120-b2-F3", "121-b2-F3", "122-b2-F3"), treatment=c(0,0,0,0,1,1,1,1) )
```


## Differential methylation                           

### Logistic regression test

Significance wwas calculated by applying Logistic regression tests.
  
```{r logreg, eval=FALSE}
#batch1
b1F1.diffMeth <- calculateDiffMeth(meth.quants.b1F1.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=4)
b1F2.diffMeth <- calculateDiffMeth(meth.quants.b1F2.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=4)
b1F3.diffMeth <- calculateDiffMeth(meth.quants.b1F3.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=4)

#batch2
b2F1.diffMeth <- calculateDiffMeth(meth.quants.b2F1.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=4)
b2F2.diffMeth <- calculateDiffMeth(meth.quants.b2F2.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=4)
b2F3.diffMeth <- calculateDiffMeth(meth.quants.b2F3.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=4)

save(file=file.path(resultsDir, "batch.diffs"), list=ls()[grep("b.+diffMeth", ls())])
```

```{r loaddiffs, echo=FALSE}
load(file=file.path(resultsDir, "batch.diffs"))
```

### Differential methylation 


```{r methyldiffs}
# get differentially methylated regions with 25% difference and qvalue<0.01
diff.cutoff <- 25

#batch1
b1F1.Diff25p <- get.methylDiff(b1F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b1F1.Diff25pHypo <- get.methylDiff(b1F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b1F1.Diff25pHyper <- get.methylDiff(b1F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b1F2.Diff25p <- get.methylDiff(b1F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b1F2.Diff25pHypo <- get.methylDiff(b1F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b1F2.Diff25pHyper <- get.methylDiff(b1F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b1F3.Diff25p <- get.methylDiff(b1F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b1F3.Diff25pHypo <- get.methylDiff(b1F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b1F3.Diff25pHyper <- get.methylDiff(b1F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

#batch2
b2F1.Diff25p <- get.methylDiff(b2F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b2F1.Diff25pHypo <- get.methylDiff(b2F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b2F1.Diff25pHyper <- get.methylDiff(b2F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b2F2.Diff25p <- get.methylDiff(b2F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b2F2.Diff25pHypo <- get.methylDiff(b2F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b2F2.Diff25pHyper <- get.methylDiff(b2F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b2F3.Diff25p <- get.methylDiff(b2F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b2F3.Diff25pHypo <- get.methylDiff(b2F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b2F3.Diff25pHyper <- get.methylDiff(b2F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")
```


```{r overlaps}
b1F1.gr = as(b1F1.Diff25p, "GRanges")
b1F2.gr = as(b1F2.Diff25p, "GRanges")
b1F3.gr = as(b1F3.Diff25p, "GRanges")
b2F1.gr = as(b2F1.Diff25p, "GRanges")
b2F2.gr = as(b2F2.Diff25p, "GRanges")
b2F3.gr = as(b2F3.Diff25p, "GRanges")

b1F1F2.overlap <- findOverlaps(b1F1.gr, b1F2.gr)
b1F1F3.overlap <- findOverlaps(b1F1.gr, b1F3.gr)
b1F1F2F3.overlap <- intersect(as.matrix(b1F1F2.overlap)[, 1], as.matrix(b1F1F3.overlap)[, 1])
b1F1F2.num <- nrow(as.matrix(b1F1F2.overlap))
b1F1F3.num <- nrow(as.matrix(b1F1F3.overlap))
b1F1F2F3.num <- length(b1F1F2F3.overlap)

b2F1F2.overlap <- findOverlaps(b2F1.gr, b2F2.gr)
b2F1F3.overlap <- findOverlaps(b2F1.gr, b2F3.gr)
b2F1F2F3.overlap <- intersect(as.matrix(b2F1F2.overlap)[, 1], as.matrix(b2F1F3.overlap)[, 1])
b2F1F2.num <- nrow(as.matrix(b2F1F2.overlap))
b2F1F3.num <- nrow(as.matrix(b2F1F3.overlap))
b2F1F2F3.num <- length(b2F1F2F3.overlap)

b1F1F2F3.coords <- getData(methylKit::select(b1F1.Diff25p, b1F1F2F3.overlap))[, 1:3]
b2F1F2F3.coords <- getData(methylKit::select(b2F1.Diff25p, b2F1F2F3.overlap))[, 1:3]

b1F1b2F1.overlap <- findOverlaps(b1F1.gr, b2F1.gr)
b1F1b2F1.num <- nrow(as.matrix(b1F1b2F1.overlap))
b1F1F2.idx <- as.matrix(b1F1F2.overlap)[, 1]
b1F1F3.idx <- as.matrix(b1F1F3.overlap)[, 1]
b1F1F2F3.idx <- b1F1F2F3.overlap
b2F1F2.idx <- as.matrix(b2F1F2.overlap)[, 1]
b2F1F3.idx <- as.matrix(b2F1F3.overlap)[, 1]
b2F1F2F3.idx <- b2F1F2F3.overlap
```

Comparing the overlap methylated positions between the different generations gives the following results..

__Batch 1__

* Overlap betwen F1 and F2: `r b1F1F2.num` nucleotides
* Overlap betwen F1 and F3: `r b1F1F3.num` nucleotides
* Overlap betwen F1, F2  and F3: `r b1F1F2F3.num` nucleotides

__Batch 2__

* Overlap betwen F1 and F2: `r b2F1F2.num` nucleotides
* Overlap betwen F1 and F3: `r b2F1F3.num` nucleotides
* Overlap betwen F1, F2  and F3: `r b2F1F2F3.num` nucleotides

__combined__

* Overlap betwen batch 1 F1 and batch 2 F1: `r b1F1b2F1.num` nucleotides

# Export

Loading genes and CpG islands. Promoters are defined as 5000bp upstreams. Shores are 200bp flanks of CpG islands. 

```{r loadanno}
# annotation by genes
gene.obj=read.transcript.features(file.path(metaDir, "mm10_refseq.bed"), up.flank = 5000, down.flank = 0)

# annotation by CpG islands
cpg.obj=read.feature.flank(location=file.path(metaDir, "mm10_CpG_Islands.bed") ,feature.flank.name=c("CpGi","shores"), flank=2000)
```


# Visualization  

## Differentially methylated Cs with promoter/exon/intron

__Batch 1__

```{r pei.plot.1, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.pei(b1F1.Diff25p,gene.obj, "b1F1")
p2 <- plot.targetanno.pei(b1F2.Diff25p,gene.obj, "b1F2")
p3 <- plot.targetanno.pei(b1F3.Diff25p,gene.obj, "b1F3")

grid.arrange(p1, p2, p3, ncol=2)
```

__Batch 2__

```{r pei.plot.2, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.pei(b2F1.Diff25p,gene.obj, "b2F1")
p2 <- plot.targetanno.pei(b2F2.Diff25p,gene.obj, "b2F2")
p3 <- plot.targetanno.pei(b2F3.Diff25p,gene.obj, "b2F3")

grid.arrange(p1, p2, p3, ncol=2)
```


## Differentially methylated Cs with CpG islands/shores

__Batch 1__

```{r cpgi.plot.1, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.cpgi(b1F1.Diff25p,cpg.obj, "b1F1")
p2 <- plot.targetanno.cpgi(b1F2.Diff25p,cpg.obj, "b1F2")
p3 <- plot.targetanno.cpgi(b1F3.Diff25p,cpg.obj, "b1F3")

grid.arrange(p1, p2, p3, ncol=2)
```


__Batch 2__

```{r cpgi.plot.2, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.cpgi(b2F1.Diff25p,cpg.obj, "b2F1")
p2 <- plot.targetanno.cpgi(b2F2.Diff25p,cpg.obj, "b2F2")
p3 <- plot.targetanno.cpgi(b2F3.Diff25p,cpg.obj, "b2F3")

grid.arrange(p1, p2, p3, ncol=2)
```

### Number of hyper/hypo methylated regions/bases per chromosome

__Batch 1__

```{r diffmethperchr.1}
#diffMethPerChr(myDiff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL)

p1 <- diffMethPerChr(b1F1.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b1F1")
p2 <- diffMethPerChr(b1F2.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b1F2")
p3 <- diffMethPerChr(b1F3.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b1F3")

grid.arrange(p1, p2, p3, ncol=2)
```

__Batch 2__

```{r diffmethperchr.2}

p1 <- diffMethPerChr(b2F1.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b2F1")
p2 <- diffMethPerChr(b2F2.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b2F2")
p3 <- diffMethPerChr(b2F3.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b2F3")

grid.arrange(p1, p2, p3, ncol=2)

```


**Seeing the same issues even if you filter, normalize and unify the samples all together.**





# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```

---

