```{r setup, echo=FALSE}
library(knitr)
opts_chunk$set(tidy=TRUE, cache=TRUE, highlight=TRUE, figalign="center", echo=TRUE, warning=FALSE, error=TRUE, message=FALSE)
```

Array analysis for [Anju Singh](mailto:asingh150@gmail.com), Glimcher group at HSPH.  Contact [Oliver Hofmann](mailto:ohofmann@hsph.harvard.edu) for additional details.

## Methods summary
All MoGene 1.0 ST arrays were processed using the 'oligo' BioConductor package[^1], quality-controlled with arrayQualityMetrics[^2] and corrected for batch effects with ComBat[^3] after RMA normalization[^4]. Differentially expressed genes were identified using limma (Anju to list FDR, fold-change cutoffs) and manually curated for any association with chondrocytes or osteoblasts.

## Workflow

The sections below provide minimal code to reproduce the included results and plots. Libraries used to process and visualize data:

```{r libraries}
library(Biobase)
library(xtable)
library(RColorBrewer)
library(ggplot2)
library(pheatmap)
library(genefilter)
library(plyr) 
library(reshape)
library(limma)
library(sva)
library(oligo)
```

### Primary data processing

Loading of CEL files based on prefix patterns:

```{r parseArrays}
#
# CEL-file handling
#
basepath <- '.'
resultspath <- file.path(basepath)
# make sure this points at the CEL files
dataDir <- file.path('..', 'data')

celFiles <- c(list.celfiles(dataDir, pattern='LG2011', full.names=T),
              list.celfiles(dataDir, pattern='LG201202', full.names=T))
celFiles
affy <- read.celfiles(celFiles, verbose=F)
```

Covariate information for both experiments, as provided by Anju. Samples were generated in at least two independent batches:

```{r covariateInformation, results="asis"}
pDataFile <- file.path(datadir, 'sampleInfo_Combined.txt')
pDataObj <- read.table(pDataFile, row.names=1, header=T, sep='\t')

# Ensure CEL file names and covariate row names match
all(rownames(pDataObj) == colnames(exprs(affy)))
pData(affy) <- pDataObj

# Sampe information table
pDataTable <- xtable(pData(affy))
print(pDataTable, type='html')
```

Summarize probe information on the gene (`core`) level:

```{r normalization}
# Transcript (gene) level normalization using RMA
allArrays <- rma(affy, target='core')
allArrays
```

### Batch correction

Correct for batch effects with ComBat from the `sva` package:

```{r batchCorrection}
#
# Correct for batch effects
#
# Create model with Condition as factor variable
mod <- model.matrix(~as.factor(Condition), data=pDataObj)
batch <- pData(allArrays)$Batch

# Modify expression matrix
edata <- exprs(allArrays)
combat_edata <- ComBat(dat=edata,
                       batch=batch,
                       mod=mod,
                       numCovs=NULL,
                       par.prior=TRUE, 
                       prior.plots=TRUE)
allArraysBatch <- allArrays
exprs(allArraysBatch) <- combat_edata

# Get rid of the old data set, no longer needed
rm(allArrays)
```

### Gene annotation

Generate basic annotation (gene symbols in particular):

```{r annotation}
# Retrieving NetAffx Biological Annotation
featureData(allArraysBatch) <- getNetAffx(allArraysBatch, 'transcript')
varLabels(featureData(allArraysBatch))

# Extract the 'gene assignment' annotation
annot <- pData(featureData(allArraysBatch)[, c('geneassignment')])
head(annot[!is.na(annot), ], 1)

# Generate a list of gene symbols from the gene assignment
desc <- annot[, 1]
symbols <- unlist(lapply(desc, function(x) strsplit(x, ' // ')[[1]][2]))
length(featureData(allArraysBatch)$probesetid) == length(symbols)
head(symbols[!is.na(symbols)])
```

### Finding differentialy expressed genes

Test for differential expression with limma:

```{r limma, results='asis'}
#
# limma comparison
#
design.pairs <- function(levels) {
  n <- length(levels)
  design <- matrix(0, n, choose(n, 2))
  rownames(design) <- levels
  colnames(design) <- 1:choose(n, 2)
  
  k <- 0
  
  for (i in 1:(n - 1))
    for (j in (i + 1):n) {
      k <- k+1
      design[i, k] <- 1
      design[j, k] <- -1
      colnames(design)[k] <- paste(levels[i],
                                   "-",
                                   levels[j],
                                   sep="")
    }
}

# Create appropriate design matrix
design <- model.matrix(~0 + factor(pData(allArraysBatch)$Condition))
colnames(design) <- c('A', 'B', 'C', 'D')

designTable <- xtable(design)
print(designTable, type='html')
```

Contrast matrix and limma:

```{r comp}
# Second comparison, this time relative to *B*
contrast.matrix <- makeContrasts(B-A, B-C, B-D, levels=c('A', 'B', 'C', 'D'))

# Fits a linear model for each gene based on the given series of arrays
fit <- lmFit(exprs(allArraysBatch), design) 

# Computes estimated coefficients and standard errors for contrasts
fit2 <- contrasts.fit(fit, contrast.matrix) 

# Computes moderated t-statistics and log-odds of differential expression 
# by empirical Bayes shrinkage of the standard errors towards a common value.
fit2 <- eBayes(fit2) 
```

Generate results for manual curation:

```{r output, results='asis'}
# Obtain _all_ results, no cutoffs
ba <- topTable(fit2, coef=1, adjust="fdr", number=35556, genelist=symbols) 
colnames(ba) <- c('ID_ba', 'logFC_ba', 'AveExpr_ba', 't_ba', 'pValue_ba',
                  'adj_pValue_ba', 'B_ba')
bc <- topTable(fit2, coef=2, adjust="fdr", number=35556, genelist=symbols) 
colnames(bc) <- c('ID_bc', 'logFC_bc', 'AveExpr_bc', 't_bc', 'pValue_bc',
                  'adj_pValue_bc', 'B_bc')
bd <- topTable(fit2, coef=3, adjust="fdr", number=35556, genelist=symbols) 
colnames(bd) <- c('ID_bd', 'logFC_bd', 'AveExpr_bd', 't_bd', 'pValue_bd',
                  'adj_pValue_bd', 'B_bd')

cd <- merge(bc, bd, by='row.names', sort=F)
rownames(cd) <- cd$Row.names                  
acd <- merge(ba, cd, by='row.names', sort=F)
acd <- acd[, c(1, 2, 3, 4, 7,
               11, 12, 15,
               18, 19, 22)]
colnames(acd) <- c('ProbeID', 'Symbol', 
                   "logFC_BA", "AveExpr_BA", "adj_pValue_BA",
                   "logFC_BC", "AveExpr_BC", "adj_pValue_BC",
                   "logFC_BD", "AveExpr_BD", "adj_pValue_BD")

acdTable <- xtable(head(acd))
print(acdTable, type='html')

#write.table(acd, file=file.path(resultspath,
#                                'relativeToB_combined.txt'),
#            sep='\t',
#            quote=F, row.names=F)
```

## Gene list visualization

Visualization of manually selected gene lists with a basic heatmap, clustered hierarchically both for genes and samples:

```{r geneCuration}
#
# Grabbing an _annotated_ gene list and order from Anju's data
#
heatmapGenes <- read.table('GeneList.txt', sep='\t', header=T)

# Extract the probes
heatmapProbes <- heatmapGenes$Probe

#
# Subset expression set by probe ID
#
heatmapArray <- exprs(allArraysBatch[heatmapProbes, ])
colnames(heatmapArray) <- pData(allArraysBatch)$Sample
rownames(heatmapArray) <- paste(heatmapGenes$Probe,
                                heatmapGenes$Symbol,
                                sep='.')


# Visualization
colors <- brewer.pal(9, 'RdYlGn')
pal <- rev(colorRampPalette(colors)(50))

# Define colors for the annotation
annotation <- data.frame(Var1=pData(allArraysBatch)$Condition)
rownames(annotation) <- pData(allArraysBatch)$Sample
Var1 <- brewer.pal(4, 'Spectral')
names(Var1) <- levels(pData(allArraysBatch)$Condition)
ann_colors <- list(Var1=Var1)

# Plot the actual heatmap
pheatmap(heatmapArray,
         color=pal,
#         cellwidth=25,
#         cellheight=12,
#         fontsize_col=10,
#         filename='Sample1_Sorted.pdf',
         scale='none',
         cluster_rows=T,
         cluster_cols=T,
         legend=T,
         annotation=annotation,
         annotation_legend=T,
         annotation_colors=ann_colors)
```

## Functional comparison

The idea here is that genes associated with chondrocytes or osteoblasts exhibit different expression patterns in the four sample distributions. Using a slightly different gene lists for this part as Anju did not assign the classification to all genes used for the heatmap -- another way would have been to exclude unannotated genes at this point:


```{r annotatedGenes}
#
# Grabbing an _annotated_ gene list and order from Anju's data
#
genelist <- read.table('GeneAnnotation.txt',
                     sep='\t', header=T)

# Extract the probes
probelist <- genelist$Probe

#
# Subset expression set by probe ID
#
arraySubset <- exprs(allArraysBatch[probelist, ])
colnames(arraySubset) <- pData(allArraysBatch)$Sample
rownames(arraySubset) <- paste(genelist$Probe,
                               genelist$Symbol,
                               sep='.')
```

Modify the data for plotting:

```{r ggPrep}
# Prepare data for ggplot2
data <- arraySubset
colnames(data) <- pData(allArraysBatch)$Condition
data <- cbind(data, genelist)
data <- data[, !colnames(data) %in% c('Symbol')]

# Convert into long form
p <- melt(data, c('Probe', 'Annotation'))

# Rename columns
colnames(p) <- c('Probe', 'Annotation', 'Condition', 'Expression')

# Re-create four conditions
p$Condition <- as.factor(substring(p$Condition, 1, 1))

# Minimal theme
theme_set(theme_grey())
```

Start with the overall expression distribution between the two cell type associations:

```{r boxplot}
# Basic plot data and labels
g0 <- ggplot(p, aes(x=Annotation, y=Expression)) +  
  scale_y_continuous(name='Probe intensity (log)') + 
  opts(title='Expression changes between chondrocytes / osteoblasts',
       axis.text.x=theme_text(angle=0, hjust=0.5, size=14),
       axis.text.y=theme_text(size=14),
       axis.title.x=theme_blank(),
       axis.title.y=theme_text(angle=90, size=14, vjust=0.4),
       legend.title=theme_text(size=14),
       legend.text=theme_text(size=12)
  )

# Basic boxplot
g_boxplot <- g0 + 
  geom_boxplot() 
g_boxplot  
```

Repeat, but break up the boxplots by cells (condition):

```{r boxplotFacets}
# Slight tweaks to the axis labels
g0 <- ggplot(p, aes(x=Annotation, y=Expression)) +  
  scale_y_continuous(name='Probe intensity (log)') + 
  opts(title='Expression changes between chondrocytes / osteoblasts',
       axis.text.x=theme_text(angle=90, hjust=1, size=14),
       axis.text.y=theme_text(size=14),
       axis.title.x=theme_blank(),
       axis.title.y=theme_text(angle=90, size=14, vjust=0.4),
       plot.title=theme_text(size=18, vjust=0.9),
       legend.position='none'
  )

# Boxplots with the individual data points overlayed
g_boxplot <- g0 + 
  geom_boxplot() +
  scale_fill_brewer(palette = "Set1") + 
  geom_point(colour='grey',
             size=3,
             alpha=0.25) +
  facet_grid(~Condition) +
  opts(strip.text.x=theme_text(size=18, face='bold'))
g_boxplot
```

Finally try this with the unlogged expression data:

```{r boxplotUnlogged}
pUnlog <- p
pUnlog$Expression <- 2**pUnlog$Expression

g0 <- ggplot(pUnlog, aes(x=Annotation, y=Expression)) +  
  scale_y_continuous(name='Probe intensity') + 
  opts(title='Expression changes between chondrocytes / osteoblasts',
       axis.text.x=theme_text(angle=90, hjust=1, size=14),
       axis.text.y=theme_text(size=14),
       axis.title.x=theme_blank(),
       axis.title.y=theme_text(angle=90, size=14, vjust=0.2),
       plot.title=theme_text(size=18, vjust=0.9),
       legend.position='none'
  )

# Boxplots with the individual data points overlayed
g_boxplot <- g0 + 
  geom_boxplot() +
  scale_fill_brewer(palette = "Set1") + 
  geom_point(colour='grey',
             size=3,
             alpha=0.25) +
  facet_grid(~Condition) +
  opts(strip.text.x=theme_text(size=18, face='bold'))
g_boxplot
```



[^1]: BioC
[^2]: arrayQualityMetrics
[^3]: SVA
[^4]: RMA

