 
```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=TRUE, echo=TRUE, highlight=TRUE, figalign="center", fig.height=8.5, fig.width=8.5, message=FALSE, error=FALSE, warnings=FALSE, comment=NA)
```

# Methylation and Expression analyses of Microarray and RRBS data from Kenneth Chien's lab
- for [Huansheng Xu](mailto:HXU7@PARTNERS.ORG), Chien group at MGH.  Contact [John Hutchinson](mailto:jhutchin@hsph.harvard.edu) for additional details.
- see [Xu et al.](http://www.ncbi.nlm.nih.gov/pubmed/22064699) for sample and experimental details
- post read trimming QC and RRBS based alignment with Bismark[^1]

**The R code below should allow reproduction of results**

---

# GENERAL SETUP

## LIBRARIES

```{r libraries1}
library(affy)
library(simpleaffy)
library(arrayQualityMetrics)
library(limma)
library(mouse430a2.db)
library(pheatmap)
library(RColorBrewer)
library(xtable)
library(biomaRt)
ensembl = useMart("ensembl",dataset="mmusculus_gene_ensembl")
library(plyr)
library(ggplot2)
library(reshape)
library(methylKit)
library(parallel)
library(IRanges)
library(GenomicRanges)
library(gpairs)
```


## FUNCTIONS

```{r functions1}
rbind.fill <- function(l) {
  do.call(rbind, lapply(lapply(l, unlist), "[", unique(unlist(c(sapply(l,names))))))
  }

bedTools.2in<-function(functionstring="intersectBed",bed1,bed2,opt.string="")
{
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) ## do not not to use scientific notation when writing out
   ## write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
   ## create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  print("pasg")
  cat(command,"\n")
  try(system(command))
   res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}

getCorr = function(obj){
   meth.mat  <-  getData(obj)[,obj@numCs.index]/(obj[,obj@numCs.index]+obj[,obj@numTs.index])  
   names(meth.mat) <- obj@sample.ids
   return(meth.mat)
}
```

## DATA DIRECTORIES

```{r datadirectories}
baseDir="/n/hsphS10/hsphfs1/chb/projects/JH/kc_rrbs"
annotDir="/n/hsphS10/hsphfs1/chb/projects/JH/kc_rrbs/data/annotations"
exprdataDir=paste(baseDir, "data/RNA_Microarray", sep="/")
methdataDir <- paste(baseDir, "data/fastq/trimmed", sep="/")
resultsDir=paste(baseDir, "results", sep="/")
```

---

# Array-based expression analysis component of methylation-expression analysis 

## EXPRESSION MICROARRAY METHODS SUMMARY
All Mouse 430A2 arrays were processed using the 'affy' BioConductor package[^2], quality-controlled with arrayQualityMetrics[^3] and RMA normalized[^4]. Differentially expressed genes were identified using limma[^5].

## EXPRESSION MICROARRAY DATA IMPORT
- load in phenotype data from covars.desc in "data" subdirectory
  - this file contains the names and descriptions of CEL files contained in same directory 
    - these CEL file names will then be used to load in the CEL files
- stores all this data in a single microarray "Affybatch" object

```{r exprdataload}
mic.data.raw <- read.affy('covars.desc', path=exprdataDir, verbose=T)
```

## EXPRESSION MICROARRAY SAMPLE INFO
- metadata for the samples

```{r metadata, results='asis'}
pData(mic.data.raw)$group <- paste(pData(mic.data.raw)$origin , pData(mic.data.raw)$tissue, sep="_")
pd <- pData(mic.data.raw)
metadataTable <- xtable(pd)
print(metadataTable, type='html')
rm(metadataTable) ## cleanup
```

## EXPRESSION MICROARRAY NORMALIZATION 
- RMA (Robust Multichip Average) based

```{r normalize_and_extract_expression_values}
mic.data.norm <- call.exprs(mic.data.raw, "rma")
eset.norm <- exprs(mic.data.norm)
colnames(eset.norm) <- pData(mic.data.norm)$Sample
```

## EXPRESSION MICROARRAY QA/QC 
### Raw data QA/QC report

```{r rawQC}
arrayQualityMetrics(expressionset=mic.data.raw, outdir=paste(resultsDir, "report_raw", sep="/"), force=TRUE, do.logtransform=TRUE)
rm(mic.data.raw) ## cleanup
```

[raw data QC](./results/report_raw/index.html)

- some outliers in the raw data, particularly the somatic samples

### Normalized data QA/QC report

```{r normQC}
arrayQualityMetrics(expressionset=mic.data.norm, outdir=paste(resultsDir, "report_rma", sep="/"), force=TRUE, do.logtransform=FALSE)
```

[normalized data QC](./results/report_rma/index.html)

- the somatic samples are still outliers for the distance between arrays metric
  - this is not alarming as it is expected that they should be fairly different from the embryonic samples 
- one of the somatic samples is still an outlier for the boxplot summary of signal intensity distribution, which is more worrisome

### Expression Microarray Primary Clustering - Principal Component Analysis
- the goal of this analysis is to blindly evaluate the variability within the raw data and determine whether this variability could predict the different sample categories by itself
- in basic terms, we are checking whether the samples cluster by sample type; its a good sign if they do
 

```{r exprPCA, fig.show='hold', fig.cap="" }
myPca <- prcomp(t(eset.norm))
# SD of components
plot(myPca$sdev, xlab="PCA", ylab="sddev", main="Variance explained by each principal component")
# Plot samples in all combinations in the 5 first primary components
tmpPCAData <- as.data.frame(myPca$x[,1:5])
colors <- brewer.pal(length(unique(paste(pData(mic.data.norm)$tissue, pData(mic.data.norm)$origin, sep="_"))), "Set1")[factor(paste(pData(mic.data.norm)$tissue, pData(mic.data.norm)$origin, sep="_"))]
plot(tmpPCAData, col=colors, pch=row.names(myPca$x), main="PCA plot of first 5 components")
rm(myPca, tmpPCAData, colors) ## cleanup
```

The first plot shows how much variation is explained by each principal component.

The second plot, shows a scatterplot of the arrays along all dual combinations of the first five principal components. Each sample group is represented by a separate color, the symbol letter used corresponds to the first letter of the "treatment" column in the Sample metadata above. You can use this plot to explore if the arrays cluster, and whether this is according to an intended experimental factor or according to unintended causes such as batch effects. Principal component analysis is a dimension reduction and visualisation technique that is here used to project the multivariate data vector of each array into a two-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the arrays.

- sample groups are color coded
- C=iPS_VM cells, T(blue)=iPS_TTF cells, T(green)=somatic_TTF, V(red)=ES cells, V(orange)=somatic_VM
- overall, the samples cluster fairly well by group, the main PC1 and PC2 components still show clustering of the embryonic samples together, but these samples separate better with the the PC3 and PC4 componenets


## EXPRESSION MICROARRAY ANNOTATIONS
- get the gene symbols for the array probes, which will be used to summarize expression at the genic level
- get the ensembl gene ids for the array probes, which will be used later to link gene identity to CpG Islands in their promoters

```{r annotations}
symbols=mget(row.names(eset.norm), mouse430a2SYMBOL, ifnotfound=NA)
ensembls=mget(row.names(eset.norm),mouse430a2ENSEMBL, ifnotfound=NA)
gene.annots=as.data.frame(cbind(symbols, ensembls))
nrow(gene.annots)==nrow(eset.norm) ## quality check, if TRUE, annotations are good
all(row.names(eset.norm)==row.names(gene.annots)) ## another quality check, if TRUE, annotations are good
```

## EXPRESSION MICROARRAY ANALYSIS
#### Summarize data at the gene level
- summarizes probes for an individual gene by taking their median intensity (took median value instead of mean to reduce effects of outliers)

```{r mediansummarize}
eset.genes.norm <- aggregate(eset.norm, by=list(symbols=as.vector(unlist(gene.annots$symbols))), median)
row.names(eset.genes.norm) <- eset.genes.norm[,"symbols"]
eset.genes.norm <- eset.genes.norm[,-(grep("symbols", names(eset.genes.norm)))]
```

### Limma

A linear model for microarray data analysis (Limma) was performed on the samples to identify differentially expressed genes between all pair-wise comparisons of the sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we constructed two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.


#### Create appropriate design matrix 
- make a matrix with arrays as rows, sample groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group, 
- in this case it looks like the following table, where 1-11 are the 11 microarrays named in the covars.desc file in the same order as in the file

```{r design, results="asis"}
design <- model.matrix(~ -1+factor(pData(mic.data.norm)$group))
## always make sure the headings match
colnames(design) <- c("ES_ES", "iPS_TTF", "iPS_VM", "somatic_TTF", "somatic_VM")

designTable <- xtable(design)
print(designTable, type='html')
rm(designTable) ## cleanup
```


#### Create contrast matrix
- chose to perform all possible pairwise comparisons, just in case
- in this table, columns are contrasts/comparisons and rows are sample groups
- a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
design.pairs <- function(levels) {
 n <- length(levels)
 design <- matrix(0, n, choose(n, 2))
 rownames(design) <- levels
 colnames(design) <- 1:choose(n, 2)
 k <- 0
 for (i in 1:(n - 1))
   for (j in (i + 1):n) {
     k <- k+1
     design[i, k] <- 1
     design[j, k] <- -1
     colnames(design)[k] <- paste(levels[i], "-", levels[j], sep="")
     }
 design
}
contrast.matrix <- design.pairs(levels(as.factor(pData(mic.data.norm)$group)))

contrastmatrixTable <- xtable(contrast.matrix)
print(contrastmatrixTable, type='html')
rm(contrastmatrixTable, design.pairs) ## cleanup
```

These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix for each gene based on the given series of arrays
- using the contrast matrix, compute estimated coefficients and standard errors for contrasts
- compute moderated t-statistics and log-odds of differential expression by empirical Bayes shrinkage of the standard errors towards a common value

#### Fit a linear model
- for each gene based on the given series of arrays

```{r linearmodel}
fit <- lmFit(eset.genes.norm, design) 
```

#### Compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2 <- contrasts.fit(fit, contrast.matrix)

```

#### Compute moderated t-statistics and log-odds of differential expression 

```{r bayes}
fit2 <- eBayes(fit2) 
```

## EXPRESSION MICROARRAY RESULTS

### More QA/QC 
#### some evidence from thea QC plots above that the somatic arrays have more RNA on then than the stem cell arrays
- even with normalization, this can result in seeing falsely high log fold changes for genes that are below background in the stem cell arrays and above background in the somatic arrays 
- check to see if you consistenly see higher expression of genes in the somatic arrays
- require at least a sixteen-fold change (log2FC>4) in expression
- this function will assign a value of -1,0 or 1 depending on how well the data at a probeset fits with the contrast, a 0 indicates less than a sixteen-fold difference between the two groups, a 1 indicates a difference *with the first group in the contrast having higher expression* and a -1 indicates a difference in the opposite direction 
- seeing a larger number of 1s than -1s for a contrast indicates that the first sample group consistently has higher gene expression than the latter, which could be biological, but could also be the result of more RNA on the first array/group

```{r balance, results='asis'}
diffs.temp=as.data.frame(rbind.fill(apply(decideTests(fit2,method="separate",  lfc=4),2,table)))
names(diffs.temp)=c("0","1","-1")
diffs.temp$log2ratio=log2(diffs.temp[,"1"]/diffs.temp[,"-1"])
balanceTable <- xtable(diffs.temp)
print(balanceTable, type='html')
rm(balanceTable, diffs.temp) ## cleanup
```

- these results show consistently higher transcript levels in the somatic cell types  (eg. ES_ES-somatic_TTF shows an almost two fold number of genes with at least 16-fold higher expression in the somatic_TTF sample)
- together with the lack of replicates for the somatic cells these results suggest using caution when interpreting expression differences between the embryonic cell and somatic cell groups

### Volcanoplots
- to visualize relationship between fold change and signifcance of changes in expression between groups
- genes outside the two vertical orange lines show more than 2-fold changes in expression between groups
- genes which pass an FDR  (false discovery rate) multiple testing correction of 0.05 (5% FDR) are colored blue
- the larger the pointsize, the more likely there is significant expression change (B = log-odds of differential expression) between the sample groups
- only did plots for comparisons where there at least two replicates for each sample group (i.e. left out the somatic samples)

```{r volcanoplot, fig.cap="Volcano plot", fig.width=8.5, fig.height=8.5, fig.show='hold', results='hide'}
llply(setdiff(seq(1,ncol(fit2$contrasts), 1),grep("somatic", dimnames(fit2$contrasts)[[2]])), function(n){
  n <- as.numeric(unlist(n))
  stats.core <- topTable(fit2, coef=n, sort.by="B",number=length(symbols), genelist=fit2$genes)
  stats.core$Passes.0.05_FDR.threshold  <-  as.factor(stats.core$adj.P.Val<0.05)
  g <- ggplot(data=stats.core, aes(x=logFC, y=-log10(P.Value), color=Passes.0.05_FDR.threshold, size=B)) +
    geom_point(alpha=0.5) + 
    geom_vline(xintercept=c(-1,1), color="orange", alpha=0.8) +
    xlab("log2 fold change") + ylab("-log10 p-value") +
    labs(title=dimnames(fit2$contrasts)[[2]][n])
  return(g)
  })
```

### Statistics for all comparisons
- generate list of expressed genes,  as sorted by "B", the log-odds that a gene is differentially expressed

```{r allexprstats}
all.stats=lapply(seq(1,ncol(contrast.matrix),1), function(x) {
  contrast=colnames(contrast.matrix)[x]
  stats.all.core=topTable(fit2, coef=x, adjust="fdr", p.value=1,sort="B",number=nrow(fit2$genes), genelist=fit2$genes)
  eset.all.core  <-  eset.genes.norm[stats.all.core[, "ID"], ]
  stats=cbind(stats.all.core, eset.all.core)
  return(list(stats=stats, contrast=contrast))
})
names(all.stats)=dimnames(fit2$contrasts)[[2]]
```

### Output pvalues and expression levels to files

#### The summary table contains the following information: 
- ID is the gene symbol
- logFC is the log2-fold change
- the AveExpr is the average expression value accross all arrays
- the moderated t-statistic (t) is the logFC to its standard error, the P.Value is the associated p-value
- the adj.P.Value is the p-value adjusted for multiple testing (by FDR) 
- the B-value (B) is the log-odds that a gene is differentially expressed (the-higher-the-better)
- the last 11 columns contain the log-transformed normalized expression levels for these genes in each sample

```{r outputs, results='hide'}
lapply(all.stats, function(x) {
  contrast=x$contrast
  write.table(x$stats, file=paste("../results/all.genes.stats", contrast, "xls",sep="."), sep="\t", row.names=F, col.names=T)
  })
```


---

---

# Methylation Analysis
- using Methylkit[^6] on trimmed fastq files aligned using Bismark[^7]

## METHYLATION SAMPLE INFO

```{r methsampleinfo}
sampleinfo <- read.table(paste(methdataDir, "sampleIDs.tab", sep="/"), header=T)
datafiles <- list.files(methdataDir, pattern="CpG.txt")
sampleinfo$files <- datafiles[as.vector(apply(sapply(as.vector(sampleinfo$fileprefix), regexpr, datafiles), 2, function(n) which(n==1)))]
```
  
## METHYLATION DATA IMPORT

```{r methdataimport}
sampleIDs <- as.list(as.vector(sampleinfo$sampleID))
locations <- as.list(paste(methdataDir, sampleinfo$files, sep="/"))
treatments <- as.vector(sampleinfo$treatment)
meth.quants <-read(location=locations, sample.id=sampleIDs, assembly="mm9", context="CpG", pipeline="bismark", resolution="base", treatment=treatments) ##long step
rm(sampleIDs, locations, treatments, methdataDir)
```

## METHYLATION ANALYSES
### Descriptive statistics
#### Percent methylation distribution in all samples

```{r methstats1}
lapply(meth.quants, function(n) {
  print(n@sample.id)
  getMethylationStats(n, plot=FALSE)
  })
lapply(meth.quants, function(n) getMethylationStats(n, plot=TRUE, labels=FALSE))
```

#### CpG coverage in all samples

```{r methstats2}
lapply(meth.quants, function(n) getCoverageStats(n, plot=TRUE, labels=TRUE))
```

### Filter by coverage 
- removes bases with very high coverage (above 99.9th percentile) to adjust for potential PCR biases
- also remove bases with very low coverage (<10)

```{r methcoveragefilter}
meth.quants.filtered <- filterByCoverage(meth.quants, lo.count = 10, lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)
rm(meth.quants) ## cleanup
```

### Merge samples
- only compare samples at sites where they all have coverage
- reduce data to only these sites 

```{r methmerge}
meth.quants.filtered.merged <- unite(meth.quants.filtered, destrand = TRUE)
meth.quants.filtered.merged@sample.ids=sub("_p8", "", sub("_derived", "", sub("Mouse_", "", meth.quants.filtered.merged@sample.ids)))
```

### Sample correlations
#### Correlation matrix
- upper quadrant=correlation coefficient
- diagonal=sampleid
- lower quadrant=data pointswith linear regression in red)

```{r correlations, fig.cap="Correlation matrix"}
cor.mat=getCorr(meth.quants.filtered.merged)
gpairs(cor.mat[sample(1:nrow(cor.mat),10000 ),], lower.pars=list(scatter="lm"), upper.pars=list(scatter="stats"), stat.pars=list(fontsize=16, use.color=FALSE), scatter.pars=list(pch=20, col="#00000022"), diag.pars=list(fontsize=10, show.hist=FALSE))
rm(cor.mat) ## cleanup
```

#### Dendrogram clustering
- distance = correlation
- methods = ward

```{r clustering, fig.cap="Methylation-based sample dendrogram"}
meth.dendrogram=clusterSamples(meth.quants.filtered.merged, dist="correlation", method="ward", plot=FALSE)
plot(meth.dendrogram)
rm(meth.dendrogram) ## cleanup
```

#### Principal Component Analysis
- same concept as for expression analyses

```{r PCA.screeplot, fig.cap="Variance Explained by Principal Components"}
PCASamples(meth.quants.filtered.merged, screeplot = TRUE)
```


```{r PCA.dimension.plot, fig.cap="Plot of First Two Principals Components"}
PCASamples(meth.quants.filtered.merged)
```
(sorry about the colors, they are hardcoded into the function right now)

### Regional Analyses
- can save on memory by grouping CpGs

#### Tiled Windows
- window size=1000, step size=1000  

```{r tile}
tiles <- tileMethylCounts(meth.quants.filtered, win.size = 1000, step.size = 1000)
tiles.merged <- unite(tiles)
tiles.merged@sample.ids=sub("_p8", "", sub("_derived", "", sub("Mouse_", "", tiles.merged@sample.ids)))
rm(tiles) ## cleanup
```

#### CpGIslands
- downloaded as BED files from UCSC genome browser
- convert to GRanges objects
- get methylation counts for those regions

```{r functionalregions}
CpGIslands.bed=read.delim(paste(annotDir, "mm9_CpG_Islands.bed", sep="/"), header=F)
names(CpGIslands.bed)=c("chr", "start", "end", "id")
## subset to chromosomes present in methylation reads
CpGIslands.bed=CpGIslands.bed[CpGIslands.bed$chr %in% meth.quants.filtered[[1]]@.Data[[2]],]
CpGIslands.bed=droplevels(CpGIslands.bed)
## convert to GRanges objects
CpGIslands.GR=GRanges(seqnames=CpGIslands.bed$chr, ranges=IRanges(start=CpGIslands.bed$start, end=CpGIslands.bed$end), ids=CpGIslands.bed$id)
## get methylation counts for regions
meth.quants.CpGIslands=regionCounts(meth.quants.filtered,  CpGIslands.GR)
rm(meth.quants.filtered, CpGIslands.GR) ## cleanup
## merge 
meth.quants.CpGIslands.merged <- unite(meth.quants.CpGIslands)
## change ids
meth.quants.CpGIslands.merged@sample.ids=sub("_p8", "", sub("_derived", "", sub("Mouse_", "", meth.quants.CpGIslands.merged@sample.ids)))

```

### Plot percent methylation distributions of CpGIslands

```{r plot_percent_meth_dists, fig.cap="Percent Methylation Distributionsof CpG Islands for All Samples"}
data.perc.meth=meth.quants.CpGIslands.merged[,c("id", "chr", "start", "end","strand")]
for (n in 1:length(grep("numCs", names(meth.quants.CpGIslands.merged)))) {
  coverage=meth.quants.CpGIslands.merged[,paste("coverage",n, sep="")]
  numCs=meth.quants.CpGIslands.merged[,paste("numCs",n, sep="")]
  perc.meth=numCs/coverage
  data.perc.meth=cbind(data.perc.meth, perc.meth) 
  names(data.perc.meth)[ncol(data.perc.meth)]=sub("$", n,  names(data.perc.meth)[ncol(data.perc.meth)])
} 
rm(coverage, numCs, perc.meth)
data.perc.meth.m=melt(data.perc.meth, id.vars=c("id", "chr", "start", "end", "strand"))

ggplot(data.perc.meth.m, aes(x=value, color=variable))+geom_density() +
  scale_x_log10() +
  xlab("% methylation")+
  scale_color_hue( name="Sample",  breaks=c("perc.meth1","perc.meth2","perc.meth3","perc.meth4","perc.meth5","perc.meth6", "perc.meth7", "perc.meth8", "perc.meth9"),labels=unlist(lapply(meth.quants.CpGIslands, function(n) n@sample.id)))

```

### Calculate differential methylation

#### Tiled Regions
- window size=1000, step size=1000  

```{r diffmeth.tiled}
## tiled regions
diffs.all.tiled=alply(combos, 2, function(n) {
  indices=which(tiles.merged@treatment %in% n)
  sample.ids.subset=tiles.merged@sample.ids[indices]
  treatments.subset=c(0,1)[factor(tiles.merged@treatment[indices])]
  tiles.merged.subset=reorganize(tiles.merged, sample.ids=sample.ids.subset,treatment=treatments.subset)
  diffs=calculateDiffMeth(tiles.merged.subset, num.cores=detectCores())
  return(diffs)
  })
names(diffs.all.tiled)=combos.by.name
```

#### CpG Islands

```{r diffmeth.CpGs}
diffs.all.CpGIslands=alply(combos, 2, function(n) {
  indices=which(meth.quants.CpGIslands.merged@treatment %in% n)
  sample.ids.subset=meth.quants.CpGIslands.merged@sample.ids[indices]
  treatments.subset=c(0,1)[factor(meth.quants.CpGIslands.merged@treatment[indices])]
  meth.quants.CpGIslands.subset=reorganize(meth.quants.CpGIslands.merged, sample.ids=sample.ids.subset, treatment=treatments.subset)
  diffs=calculateDiffMeth(meth.quants.CpGIslands.subset, num.cores=detectCores())
  return(diffs)
  })
names(diffs.all.CpGIslands)=combos.by.name
```

### Find differentially methylated regions, hypo and hyper, tiled CpGs

```{r diffmethresults.tiled, fig.cap="Hypo/Hyper Methylated Tiled  Regions by Chromosome"}
lapply(diffs.all.tiled, function(n) {
  diffs.all.tiled[[1]]@sample.id
diffMethPerChr(n, meth.cutoff=25, qvalue=0.05, plot=TRUE)
})
myDiffs25p.tiled <- lapply(diffs.all.tiled, function(x) get.methylDiff(x, difference = 25,    qvalue = 0.01))
myDiffs25p.hypo.tiled <- lapply(diffs.all.tiled, function(x) get.methylDiff(x, difference = 25,    qvalue = 0.01, type="hypo"))
myDiffs25p.hyper.tiled <- lapply(diffs.all.tiled, function(x) get.methylDiff(x, difference = 25,    qvalue = 0.01, type="hyper"))
```

### Find differentially methylated regions, hypo and hyper,  CpG Islands

```{r diffmethresults.CpGIslands, fig.cap="Hypo/Hyper Methylated CpG Islands by Chromosome"}
lapply(diffs.all.CpGIslands, function(n) {
  print(diffs.all.CpGIslands[[1]]@sample.id)
diffMethPerChr(n, meth.cutoff=25, qvalue=0.05, plot=TRUE)
})
myDiffs25p.CpGI <- lapply(diffs.all.CpGIslands, function(x) get.methylDiff(x, difference = 25,    qvalue = 0.01))
myDiffs25p.hypo.CpGI <- lapply(diffs.all.CpGIslands, function(x) get.methylDiff(x, difference = 25,    qvalue = 0.01, type="hypo"))
myDiffs25p.hyper.CpGI <- lapply(diffs.all.CpGIslands, function(x) get.methylDiff(x, difference = 25,    qvalue = 0.01, type="hyper"))

```

### Map CpG differences to genomic features
- tiled regions

```{r get_genomic_features}
gene.obj=read.transcript.features(paste(exprdataDir, "/mm9.knowngenes.bed", sep=""))
CpGI.obj=read.feature.flank(paste(exprdataDir, "/mm9CpGIslands.bed", sep=""), feature.flank.name = c("CpGi","shores"))
lapply(myDiffs25p.tiled, function(n) annotate.WithGenicParts(n, gene.obj))
lapply(myDiffs25p.tiled, function(n) annotate.WithFeature.Flank(n, CpGI.obj$CpGi, CpGI.obj$shores, feature.name = "CpGi",flank.name = "shores"))
```

---

## Integrating Expression and Methylation
- used CpGs within CpG islands

### Assign CpG Islands to genes
- CpG Island assigned to gene if it is contained within the 5kb upstream (promoter) of a known enesmbl annotated gene

```{r assign_CpGIs_to_genes}
known_genes.bed=read.delim(paste(annotDir, "mm9_known_genes.bed", sep="/"), header=F)[,c(1:6)]
promoters.bed=read.delim(paste(annotDir,"5kb_upstream_of_mm9_known_genes.bed", sep="/"), header=F)[,c(1:6)]
names(promoters.bed)=c("chr", "start", "end", "id", "score", "strand")
names(known_genes.bed)=c("chr", "start", "end", "id", "score", "strand")
## convert ensembl transcriptids to ensembl geneids and remake bed file
known_genes.bed=merge(getBM(attributes=c("ensembl_gene_id", "ensembl_transcript_id"), filters="ensembl_transcript_id", value=known_genes.bed$id, mart=ensembl), known_genes.bed, by.x="ensembl_transcript_id", by.y="id")
known_genes.bed=known_genes.bed[,c(3,4,5,2,6,7)]
promoters.bed=merge(getBM(attributes=c("ensembl_gene_id", "ensembl_transcript_id"), filters="ensembl_transcript_id", value=promoters.bed$id, mart=ensembl), promoters.bed, by.x="ensembl_transcript_id", by.y="id")
promoters.bed=promoters.bed[,c(3,4,5,2,6,7)]
## relate CpG islands to promoters
CpGIslands.promoters<-bedTools.2in( bed1=CpGIslands.bed, bed2=promoters.bed, opt.string="-wo -f 1")
names(CpGIslands.promoters)<-c("chr.CpGI", "start.CpGI", "end.CpGI", "id.CpGI", "chr.prom", "start.prom", "end.prom", "ensembl_gene_id.prom", "score.prom", "strand.prom", "bp.overlap")
CpGIslands.promoters$CpGI.loc.ID=paste(CpGIslands.promoters$chr.CpGI, CpGIslands.promoters$start.CpGI, CpGIslands.promoters$end.CpGI, sep=".")
```

### Plot the relationship between CpG island methylation differences and changes in gene expression for all pairwise comparisons

```{r plot_relationships, fig.cap=""}
## merge CpG Island methylation and gene expression data for all comparisons through "CpGIslands.promoters" table
lapply(as.list(names(all.stats)), function(n) {
  stats.exprs=all.stats[[n]]$stats
  stats.exprs=merge(stats.exprs, gene.annots, by.x="ID", by.y="symbols", all=FALSE)
  stats.CpGmeth=getData(diffs.all.CpGIslands[[n]])
  names(stats.CpGmeth)=paste(names(stats.CpGmeth), "CpGI", sep=".")
  stats.CpGmeth=merge(stats.CpGmeth, CpGIslands.promoters, by.x=c("id.CpGI","chr.CpGI", "start.CpGI", "end.CpGI"),  by.y=c("CpGI.loc.ID", "chr.CpGI", "start.CpGI", "end.CpGI"), all=FALSE)
  stats=merge(stats.CpGmeth, stats.exprs, by.x="ensembl_gene_id.prom", by.y="ensembls", all=FALSE)
ggplot(data=subset(stats, qvalue.CpGI<0.05 & adj.P.Val<0.05), aes(x=logFC, y=meth.diff.CpGI, color=-log10(P.Value), size=-log10(pvalue.CpGI), label=ID))+geom_point(alpha=0.4)+scale_color_gradient(low="blue", high="red")+scale_size_continuous(range=c(2,12))+geom_text(size=5, vjust=-1.1)+geom_hline(yintercept=0, color="darkgrey")+geom_vline(xintercept=0, color="darkgrey")+labs(title=n)
  })
```

### Examine relationship between methylation differences in somatic and iPS samples
-are there CpGIslands taht show differential methylatioi between TTF and VM cells at both the somatic and iPS levels? 
-if so, what genes are affected? 
-are these differences also seen when you compare ES cells to TTFs or VMs?
-plots are restricted to CpGIslands with and FDR<0.01 for both differential methylation and gene expression
 
```{r relate_somatic_differences_to_embryonic, fig.cap="", fig.show='hold', fig.width=8.5, fig.height=8.5}
#bring together expression data for all comparisons into one large table
all.merged.stats.exprs=all.stats[[1]]$stats[,1:7]
for (n in 2:length(names(all.stats))){
  all.merged.stats.exprs=merge(all.merged.stats.exprs, all.stats[[n]]$stats[,1:7], by="ID", suffixes=c(paste(".", names(all.stats)[n-1], sep=""), paste(".", names(all.stats)[n], sep="")))
  }
all.merged.stats.exprs=merge(all.merged.stats.exprs, gene.annots, by.x="ID", by.y="symbols", all=FALSE)
#bring together methylation data for all comparisons into one large table
all.merged.stats.CpGI=diffs.all.CpGIslands[[1]]
for (n in 2:length(names(all.stats))){
  all.merged.stats.CpGI=merge(all.merged.stats.CpGI, diffs.all.CpGIslands[[n]], by=c("id", "chr","start","end","strand"), suffixes=c(paste(".", names(all.stats)[n-1], sep=""), paste(".", names(all.stats)[n], sep="")))
  }
names(all.merged.stats.CpGI)=sub("$", ".CpGI", names(all.merged.stats.CpGI))
all.merged.stats.CpGI=merge(all.merged.stats.CpGI, CpGIslands.promoters, by.x=c("id.CpGI","chr.CpGI", "start.CpGI", "end.CpGI"),  by.y=c("CpGI.loc.ID", "chr.CpGI", "start.CpGI", "end.CpGI"), all=FALSE)
#merge expression and methylation data for all comparisons into one large table
all.merged.stats=merge(all.merged.stats.CpGI, all.merged.stats.exprs, by.x="ensembl_gene_id.prom", by.y="ensembls", all=FALSE)
names(all.merged.stats)=sub("-", "_", names(all.merged.stats))



#plot relationship between somatic methylation differences and iPS methylation differences
ggplot(subset(all.merged.stats, qvalue.iPS_TTF_iPS_VM.CpGI<0.01 & qvalue.somatic_TTF_somatic_VM.CpGI<0.01), aes(x=meth.diff.iPS_TTF_iPS_VM.CpGI, y=meth.diff.somatic_TTF_somatic_VM.CpGI, size=-log10(qvalue.iPS_TTF_iPS_VM.CpGI), color=-log10(qvalue.somatic_TTF_somatic_VM.CpGI), label=ID))+
  geom_point()+ 
  scale_color_gradient(low="blue", high="red", guide="colourbar")+
  scale_size_continuous(range=c(2,12))+
  geom_text(size=5, vjust=-1.1)+
  ggtitle("CpGIsland Methylation differences \n between TTF and VM samples in both somatic and iPS cells")+
  ylab("somatic")+xlab("iPS")+
  geom_hline(yintercept=0)+geom_vline(xintercept=0)


#plot relationship between ES-iPS_TTF methylation differences and ES-somatic_TTF methylation differences
ggplot(subset(all.merged.stats, qvalue.ES_ES_iPS_TTF.CpGI<0.01 & qvalue.ES_ES_iPS_VM.CpGI<0.01), aes(x=meth.diff.ES_ES_iPS_TTF.CpGI, y=meth.diff.ES_ES_iPS_VM.CpGI, size=-log10(qvalue.ES_ES_iPS_TTF.CpGI), color=-log10(qvalue.ES_ES_iPS_VM.CpGI), label=ID))+
  geom_point()+ 
  scale_color_gradient(low="blue", high="red", guide="colourbar")+
  scale_size_continuous(range=c(2,12))+
  geom_text(size=5, vjust=-1.1)+
  ggtitle("CpGIsland Methylation differences \n between iPS/somatic TTF and ES cells")+
  ylab("ES vs. iPS_VM")+xlab("ES vs. iPS_TTF")+  
  geom_hline(yintercept=0)+geom_vline(xintercept=0)

#plot relationship between ES-iPS_VM methylation differences and ES-somatic_VM methylation differences
ggplot(subset(all.merged.stats, qvalue.ES_ES_iPS_TTF.CpGI<0.01 & qvalue.ES_ES_somatic_TTF.CpGI<0.01), aes(x=meth.diff.ES_ES_iPS_TTF.CpGI, y=meth.diff.ES_ES_somatic_TTF.CpGI, size=-log10(qvalue.ES_ES_iPS_TTF.CpGI), color=-log10(qvalue.ES_ES_somatic_TTF.CpGI), label=ID))+
  geom_point()+ 
  scale_color_gradient(low="blue", high="red", guide="colourbar")+
  scale_size_continuous(range=c(2,12))+
  geom_text(size=5, vjust=-1.1)+
  ggtitle("CpGIsland Methylation differences \n between iPS/somatic VM and ES cells")+
  ylab("ES vs. somatic_TTF")+xlab("ES vs. iPS_TTF")+
  geom_hline(yintercept=0)+geom_vline(xintercept=0)
save.image()
```



---

[^1]: [Bismark](http://www.bioinformatics.babraham.ac.uk/projects/bismark/)
[^2]: [BioC](http://www.bioconductor.org)
[^3]: [arrayQualityMetrics](http://www.bioconductor.org/packages/2.10/bioc/html/arrayQualityMetrics.html)
[^4]: [RMA](http://bioinformatics.oxfordjournals.org/content/19/2/185)
[^5]: [limma](http://www.bioconductor.org/packages/release/bioc/html/limma.html)
[^6]: [Methylkit](http://code.google.com/p/methylkit/)
[^7]: [Bismark](http://www.bioinformatics.babraham.ac.uk/projects/bismark/)

## Session Info

```{r sessioninfo}
print(sessionInfo())
}
```


