```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, figalign="center", echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="png")
options(width=200)

## setup report details
clientname="Owen Tamplin"
clientemail="otamplin@enders.tch.harvard.edu"
labPI="Zon"
lablocation="Enders/HSCI"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

Second analysis, using increased Lycorine concentration and matched control.

---

## Methods Summary and Workflow

All ZebGene 1.0 ST arrays were processed using the [BioConductor][1] package [oligo][2], quality-controlled with [arrayQualityMetrics][3] and [RMA normalized][4]. Batch correction was performed with the [sva][5] package via the [ComBat][6] method. Differentially expressed genes were identified using [limma][7].

The sections below provide descriptions of the included results and include code to reproduce them.

---

# Setup

## Libraries

Bioconductor and R libraries used to process and visualize the data

```{r libraries}
library(oligo)
library(arrayQualityMetrics)
library(limma)
library(pheatmap)
library(RColorBrewer)
library(googleVis)
library(Biobase)
library(pd.zebgene.1.0.st)
library(plyr)
library(ggplot2)
library(biomaRt)
library(genefilter)
library(ggdendro)
```

## Functions

Miscellaneous convenience functions

```{r functions}
source("http://dl.dropboxusercontent.com/u/4253254/Resources/functions.r")
```

## Variables

```{r VARIABLES}
## Setup Data and Results directory variables
if(file.exists("/n/home08/jhutchin/")){
  baseDir="/n/hsphS10/hsphfs1/chb/projects/lz_HSC_niche_effects/"
  } else if (file.exists("/Volumes/ody/")){
  baseDir="/Volumes/ody/projects/lz_HSC_niche_effects/"
}
metaDir=file.path(baseDir, "meta")
dataDir=file.path(baseDir, "data")
resultsDir=file.path(baseDir, "results", "hilyco")
covarsfilename="Owen_2013_array_list.tab" # do not use full path
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
ensemblidentifier="ENSDAR"
pvalcutoff=0.1
logFCcutoff=1
exprfilter <- 2.5
sdfilter <- 0.1
```

---

# Data Access

Load the metadata and use the  phenotypes and array names from the metadata to load in the raw data.

```{r dataload, results='hide'}
covars <- read.delim(file.path(metaDir, covarsfilename), row.names=1)
covars <- covars[order(covars$sample),]
celFiles <- file.path(dataDir, row.names(covars))
affyRaw <- read.celfiles(celFiles)
## simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in
#subsequent columns
identical(basename(celFiles), row.names(covars)) ## sanity check
pData(affyRaw) <- covars
validObject(affyRaw) ## sanity check
````

The table below details the metadata for this analysis.

```{r covars, results='asis'}
## Sample information table
pd.gvis <- gvisTable(as.data.frame(apply(rownames2col(pData(affyRaw), "CEL"),2, as.character)), options=list(width=960))
print(pd.gvis, "chart")
```

---

# Basic QA/QC

First, I used the arrayQualityMetrics library to examine the raw CEL file data. Here I grouped the samples by cell type, treatment and batch.

```{r rawQC, eval=FALSE}
arrayQualityMetrics(expressionset=affyRaw, outdir=file.path(resultsDir, 'report_raw'), force=TRUE, do.logtransform=TRUE, intgroup=c("celltype", "treatment", "Batch"))
```

[Raw data QC report](../results/hilyco/report_raw/index.html)
*this time around the array# matches the sample#*

The new arrays look fine. There is still good clustering by cell type. There doesn't appear to be too much overall clustering by batch (batch 5 are the new arrays).

I kept the old lyrcorine samples in the dataset  for now, to compare with the new samples.

```{r drop_arrays, results='hide'}
## simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in
#subsequent columns
covars <- read.delim(file.path(metaDir, covarsfilename), row.names=1)
covars <- covars[-(grep("3PR|4PF", covars$array_name)),]
covars <- covars[order(covars$sample),]
celFiles <- file.path(dataDir, row.names(covars))
affyRaw <- read.celfiles(celFiles)
## simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in
#subsequent columns
identical(basename(celFiles), row.names(covars)) ## sanity check
pData(affyRaw) <- covars
validObject(affyRaw) ## sanity check
```

---

# Background Correct and Normalize
The data was normalized for differential gene expression analysis using Robust Multichip Average ([RMA][3]) in the oligo BioConductor package. Here, RMA normalizes the intensity values at the probe level, and  collapses probes into "core" transcripts based on annotations provided by Affymetrix

```{r normalize, results='hide', cache=TRUE}
affyNorm <- rma(affyRaw, target="core", background=T, normalize=T)
pData(affyNorm)$samplelabel <- paste(pData(affyNorm)$celltype, pData(affyNorm)$treatment, pData(affyNorm)$replicate, sep="_")
```

---

# Post-Normalization Data Exploration

```{r normQC, eval=FALSE}
arrayQualityMetrics(expressionset=affyNorm, outdir=file.path(resultsDir, 'report_rma.core'), force=TRUE, do.logtransform=FALSE,intgroup=c("celltype", "treatment", "Batch"))
```

[Normalized data QC report](../results/hilyco/report_rma.core/index.html)

All the samples look fine after normalization, the two arrays tagged as outlier are only just over the threshold, so I decided to keep them.

## Basic Clustering

The new lycorine samples appear to cluster much better, though this may be a batch driven  (observe how the Negative control DMSO sample from the same batch (batch 5 here) clusters with the new Lycorine samples). *PGE2 samples were retained for the batch correction but not shown in this plot.*

```{r basic_clustering2, out.width='50%', fig.show='hold',dev="CairoSVG"}
plot_dendro(affyNorm[,affyNorm@phenoData$celltype=="NEG" & affyNorm@phenoData$treatment!="PGE2"], labels.colname="treatment", colors.colname="Batch", title="Treatments and Batches, NEG Cells - All Lycorine Concentrations")
```

In any case, going forward, I excluded the lower concentration lycorine samples.

```{r excludelolycorine}
affyNorm <- affyNorm[,which(affyNorm@phenoData$lycorine_conc!="LO")]
```

---

# Batch Correction

Correct for batch effects with ComBat from the [sva][4] package:

```{r batchcorrect, cache=TRUE, fig.keep='none'}
library(sva)
# Create model with cell type as factor variable
mod <- model.matrix(~treatment*celltype, data=pData(affyNorm))
batch <- pData(affyNorm)$Batch
# Modify expression matrix
eset.core <- exprs(affyNorm)
combat_edata <- ComBat(dat=eset.core,
                       batch=batch,
                       mod=mod,
                       numCovs=NULL,
                       par.prior=TRUE,
                       prior.plots=TRUE)
affyNorm.batch <- affyNorm
exprs(affyNorm.batch) <- combat_edata
```

Remove PGE2 samples.
```{r removePGE2}
affyNorm.batch <- affyNorm.batch[,which(pData(affyNorm.batch)$treatment!="PGE2")]
pData(affyNorm.batch)$treatment <- droplevels(pData(affyNorm.batch)$treatment)
```

---

# Post Batch Correction Data Exploration

## Basic Clustering

```{r batch.basic_clustering1, out.width='50%', fig.show='hold',dev="CairoSVG"}
plot_dendro(affyNorm.batch[,which(pData(affyNorm.batch)$lycorine_conc!="LO")], labels.colname="celltype", colors.colname="treatment", title="Cell Types and Treatments, All Batch Corrected Samples - High Lycorine Concentration")
plot_dendro(affyNorm.batch[,which(pData(affyNorm.batch)$lycorine_conc!="LO")], labels.colname="treatment", colors.colname="celltype", title="Cell Types and Treatments, All Batch Corrected Samples - High Lycorine Concentration")
```

The endothelial cells cluster together well. There is still less separation between the HSC and negative cells. There is definite improvement of clustering based on treatment.

## Principal Component Analyses

Principal component analysis is a dimension reduction and visualisation technique that is here used to project the multivariate data vector of each array into a two-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the arrays. The goal of this analysis is to blindly evaluate the variability within the normalized data and determine whether this variability could predict the different sample categories by itself. In basic terms, we are checking whether the samples cluster by sample type. Due to the number of overlapping sample types to consider, I ran the PCA on the different subsets (cell type or treatment) of the data.

```{r pca.batch1, out.width='50%', fig.show='hold',dev="CairoSVG"}
PCAplot(affyNorm.batch, categories="celltype", colorpalette=cbPalette[1:3], alpha=0.8, title="All Samples - Cell Type")
PCAplot(affyNorm.batch, categories="treatment", colorpalette=cbPalette[5:7], alpha=0.8, title="All Samples - Treatment")
```

These figures shows a scatterplot of the arrays along all dual combinations of the first four principal components. Each sample group is represented by a separate color.  For the first plot, the sample types represent cell types. In the second  plot, the sample types represent the three treatments.

Interestingly, after you exlude the PGE2 samples, even though the ENDO vs. other cell types is still a dominant source of variation, they seem to separate into distinct clusters by treatment better than previously. This suggests that the cell type has a more significant effect on phenotype than the drug treatment but also that the PGE2 treatment had little effect .

To see the effect of the treatments, we can examine each cell type individually. In these plots, each plot is  for a single treatment, and the sample groups (colors) represent cell types.

```{r pca.batch2, out.width='33%', fig.show='hold',dev="CairoSVG"}
PCAplot(affyNorm.batch[,which(affyNorm.batch@phenoData$celltype=="HSC")], categories="treatment", colorpalette=cbPalette[5:7], alpha=0.8, title="HSC Cells - Treatments")
PCAplot(affyNorm.batch[,which(affyNorm.batch@phenoData$celltype=="ENDO")], categories="treatment", colorpalette=cbPalette[5:7], alpha=0.8, title="ENDO Cells - Treatments")
PCAplot(affyNorm.batch[,which(affyNorm.batch@phenoData$celltype=="NEG")], categories="treatment", colorpalette=cbPalette[5:7], alpha=0.8, title="NEG Cells - Treatments")
```

For all three cell types, excluding the PGE2 samples allows us to more clearly see that the the higher concentration lycorine treatment has a strong effect.

---

# Pre-filtering

Now that we're done normalizing and batch correcting, we can get rid of the control probes (nearly 2/3 of the probes on the array).

We can also use the standard deviations and mean expression levels to filter out probes that are poorly expressed and/or don't vary much between samples. We're unlikely to to see significant differences for these genes and by using fewer probes in the [multiple test correction](http://en.wikipedia.org/wiki/Multiple_comparisons) of our statistical tests we increase our chances of seeing significant results.

```{r genefilter}
# retrieve NetAffx Biological Annotation
featureData(affyNorm.batch) <- getNetAffx(affyNorm.batch, "transcript")
## remove control and unannotated probes, keep "main" category probes
affyNorm.batch <- affyNorm.batch[which(pData(featureData(affyNorm.batch))$category=="main"),]
## remove probes with low expression levels
means <- apply(exprs(affyNorm.batch), 1, mean)
affyNorm.batch <- affyNorm.batch[which(means>exprfilter),]
## remove probes with little variance among samples
sds <- apply(exprs(affyNorm.batch),1, sd)
affyNorm.batch <- affyNorm.batch[which(sds>sdfilter),]
```

Here I filtered out probes that show a mean intensity level (after log transformation) of less than `r exprfilter` and a standard deviation of less than `r sdfilter`. After filtering in this manner reduced the number of probes from `r length(featureNames(affyNorm))` to `r length(featureNames(affyNorm.batch))`.

---

# Translating Annotations

Translate the array feature data (i.e. information about the probes) to gene symbols that correspond to each probeset group on the array.

The annotations I use here are obtained from two sources.
First, using the *getNetAffx* command from the [oligo][2] package, retrieve the annotations from the Bioconductor package pd.zebgene.1.0.st, which are auto-generated by BioConductor from the Affymetrix supplied *cdf* microarray library file. I used this information above, to determine which probes were in the 'main' category.

Full probe-gene/mrna annotations were then imported and parsed from the [annotation](../meta/ZebGene-1_0-st-v1.na33.3.zv9.transcript.csv) file downloaded from the NetAffx site.
The annotations are a bit confusing, so I tried to simplify as best I could. Most probes bind to multiple annotated accessions from different databases with varying degrees of evidence; often these accessions describe slight variations of the same gene(s) (listed under *symbols*, which appear to include the ZFIN IDs). I listed the gene accession numbers genes in order of evidence, comma separated, in the *accessions* column. The accession (regardless of database origin) with the most evidence is separately listed in the *bestgene* column. Many probes don't bind to any annotated genes, so I also included the bound mRNAs in a similar manner to the genes.

```{r features, results='hide'}
# Import the annotations
dat <- read.csv(file.path(metaDir, "ZebGene-1_0-st-v1.na33.3.zv9.transcript.csv"), comment.char = "#", stringsAsFactors=FALSE, na.string = "---")
dat <- col2rownames(dat, "probeset_id")
#extract and reorder to match the array features
dat <- dat[row.names(fData(affyNorm.batch)),]
dat <- dat[,c("probeset_id", "seqname", "strand", "start", "stop", "gene_assignment", "mrna_assignment")]
dat <- as.matrix(dat)
# parse mrna_assignments
headercol <- "mrna_assignment"
mrnas <- t(sapply(strsplit(dat[, headercol], " /// "), function(x) {
  dat.probe.df <- do.call(rbind, strsplit(x, " // "))
  bestrna <- dat.probe.df[1,1]
  rnas <- paste(dat.probe.df[,1], collapse=",")
  c(bestrna, rnas)
  }))
mrnas <- as.data.frame(mrnas)
names(mrnas) <- c("best.mrna", "mrnas")
# parse gene assignments
headercol <- "gene_assignment"
genes <- t(sapply(strsplit(dat[, headercol], " /// "), function(x) {
  if(is.na(x[1])){
    out <- rep("NA", 6)
    } else {
      dat.probe.mat <- as.matrix(do.call(rbind, strsplit(x, " // ")))
      bestgene <- as.character(dat.probe.mat[1,1])
      dat.probe.vec <- apply(dat.probe.mat, 2, function(y) {
        paste(unique(y), collapse=",")
        })
      out <- as.character(c(bestgene,dat.probe.vec))
      }
  return(out)
  }))

genes <- as.data.frame(genes[,c(1,2,3,4,6)])
names(genes) <- c("bestgene", "accessions", "symbols", "descriptions", "entrezIDs")
genes <- rownames2col(genes, "probeids")
#combo mrna and gene assigments
gene.annots <- cbind(genes, mrnas)
# check to make sure data is correct
identical(length(featureData(affyNorm.batch)$probesetid), nrow(gene.annots)) # sanity check, sane=TRUE
identical(as.character(gene.annots$probeids), as.character(featureData(affyNorm.batch)$probesetid)) # sanity check, sane=TRUE
```

---

# Expression values for **ALL** genes in **ALL** samples

```{r allexprs}
#output all expression values for all samples
eset.all <- exprs(affyNorm.batch)
colnames(eset.all) <- pData(affyNorm.batch)$samplelabel
eset.all=merge(gene.annots,rownames2col(eset.all, "probeids"))
write.table(eset.all, file=file.path(resultsDir, paste("all.exprs", "xls",sep=".")), row.names=F, col.names=T, sep='\t')
```

[Link to Excel file with expression values for all genes and samples](../results/hilyco/all.exprs.xls)

---

# Differential Expression Analysis Methods

## Limma

A linear model for microarray data analysis (Limma) was performed on the samples to identify differentially expressed genes between pair-wise comparisons of sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we constructed two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.

### Design matrix
- make a matrix with arrays as rows, treatment groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group

```{r design, results="asis"}
design <- model.matrix(~0+celltype:treatment, data=model.frame(pData(affyNorm.batch)))
# make sure the headings match
colnames(design) <- sub(":", "_", gsub("celltype|treatment|", "", colnames(design)))
design.gvis <- gvisTable(as.data.frame(apply(rownames2col(design,"id"), 2, as.character)), options=list(width=960))
print(design.gvis, "chart")
```

### Contrast matrix
- to perform specified pairwise comparisons
 - can also setup more complex comparisons, such as to look for genes that change in HSC cells upon PGE2 treatment but not in negative cells
- in this table, columns are contrasts/comparisons and rows are sample groups
-  generally speaking, a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

### Comparisons

  **Differences between cell types**
  - differences between HSC and ENDO cell types from negative and from each other, after DMSO treatment (ENDO.celltype,  HSC.celltype and ENDO.HSC.celltype)

  **Effect of drugs on individual cell types**
  - changes in HSC ("HSC.Lyc") cells after drug treatments
  - changes in ENDO  ("ENDO.Lyc") cells after lyrcorine treatment

  **Differences in how cell types respond to drug treatment**
    - "interaction terms"
    - differences in how HSC cells respond to lycorine as compared to NEG (Diff.Lyc.HSC.Neg) cells
    - differences in how ENDO cells respond to lyrcorin as compared to NEG (Diff.Lyc.ENDO.Neg) cells

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts(ENDO.celltype=ENDO_DMSO - NEG_DMSO,
                                 HSC.celltype=HSC_DMSO - NEG_DMSO,
                                 ENDO.HSC.celltype=ENDO_DMSO - HSC_DMSO,
                                 ENDO.Lyc=ENDO_Lycorine - ENDO_DMSO,
                                 HSC.Lyc=HSC_Lycorine - HSC_DMSO,
                                 Diff.Lyc.HSC.Neg = (HSC_Lycorine - HSC_DMSO) - (NEG_Lycorine - NEG_DMSO),
                                 Diff.Lyc.ENDO.Neg = (ENDO_Lycorine - ENDO_DMSO) - (NEG_Lycorine - NEG_DMSO),
                                 levels=c("ENDO_DMSO", "HSC_DMSO", "NEG_DMSO", "ENDO_Lycorine", "HSC_Lycorine", "NEG_Lycorine"))
contrast.gvis <- gvisTable(as.data.frame(apply(rownames2col(contrast.matrix,"class"), 2, as.character)), options=list(width=960,height=320 ))
print(contrast.gvis, "chart")
```

### Linear model

These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix
  - for each gene based on the given series of arrays

```{r linearmodel}
eset.batch <- exprs(affyNorm.batch)
fit.core <- lmFit(eset.batch, design)
```

- using the contrast matrix, compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2.core <- contrasts.fit(fit.core, contrast.matrix)
```

- compute moderated t-statistics and log-odds of differential expression
  - by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.core <- eBayes(fit2.core)
```

---

# Differential Expression Analysis Results

## Volcanoplots

These plots help visualize relationship between fold change and signifcance of changes in expression between groups.

- genes outside the two vertical orange lines show more than `r 2^logFCcutoff`-fold changes in expression between groups
- genes which pass an FDR  (false discovery rate) multiple testing correction of `r pvalcutoff` (`r pvalcutoff*100`% FDR) and show at least a `r 2^logFCcutoff`-fold change in expression are colored blue
- the larger the pointsize, the more likely there is significant expression change (B = log-odds of differential expression) between the sample groups

```{r volcanoplot, fig.cap="", fig.width=12, fig.height=12, out.width='50%', fig.show='hold', results='hide', cache=TRUE, dev='png'}
all.stats <- llply(seq(1,ncol(contrast.matrix),1), function(n) {
  contrast <- dimnames(fit2.core$contrasts)$Contrasts[n]
  stats.core <- topTable(fit2.core, coef=n, sort.by="B",number=nrow(gene.annots), genelist=cbind(gene.annots, fit2.core$genes))
  stats.core$Passes.FDR.logFC.threshold  <-  as.factor(stats.core$adj.P.Val<pvalcutoff & abs(stats.core$logFC)>logFCcutoff)
  g <- ggplot(data=stats.core, aes(x=logFC, y=-log10(P.Value), color=Passes.FDR.logFC.threshold, size=B)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept=c(-logFCcutoff,logFCcutoff), color="orange", alpha=0.7, lwd=1.25) +
    xlab("log2 fold change") + ylab("-log10 p-value") +
    labs(title=contrast)
  return(list(g=g, stats.core=stats.core, contrast=contrast))
  })
lapply(all.stats, function(x) {
  x$g
  })
```

## Statistics and expression levels for ALL genes for each comparison

These Excel files contain the following information:
- logFC is the log2-fold change
- the AveExpr is the average expression value across this the contrast's set of arrays
- the moderated t-statistic (t) is the logFC to its standard error, the P.Value is the associated p-value
- the adj.P.Value is the p-value adjusted for multiple testing (by FDR)
- the B-value (B) is the log-odds that a gene is differentially expressed (the-higher-the-better)
- the last columns contain the log-transformed normalized expression levels for these genes in the relevant samples

```{r all_exprs_stats, results='hide'}
all.exprs.stats <- lapply(all.stats, function(x) {
  contrast <- x$contrast
  stats.core <- x$stats.core
  eset.all <- exprs(affyNorm.batch)[stats.core$probeids,]
  colnames(eset.all) <- pData(affyNorm.batch)$samplelabel
  # parse the contrast matrix and design matrix to find the sample indices for the samples contained within this contrast and subset the eset to those samples
  eset.stats.all <- cbind(stats.core, eset.all)
  return(list(eset.stats.all=eset.stats.all, contrast=contrast ))
  })

lapply(all.exprs.stats, function(x){
  contrast=x$contrast
  write.table(x$eset.stats.all, file=file.path(resultsDir, paste("all.stats", contrast, "xls",sep=".")), row.names=F, col.names=T, sep='\t')
  })
```

`r index=1`

[Link to Excel files with results for genes in `r all.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("all.stats", all.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Link to Excel file with results for genes in `r all.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("all.stats", all.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Link to Excel file with results for genes in `r all.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("all.stats", all.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Link to Excel file with results for genes in `r all.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("all.stats", all.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Link to Excel file with results for genes in `r all.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("all.stats", all.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Link to Excel file with results for genes in `r all.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("all.stats", all.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Link to Excel file with results for genes in `r all.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("all.stats", all.stats[[index]]$contrast, "xls",sep=".")`)

---

## Top results

The top 200 differentially expressed genes as sorted by "B", the log-odds that a gene is differentially expressed

```{r toptable, cache=TRUE}
top.stats <- lapply(seq(1,ncol(contrast.matrix),1), function(n) {
  contrast <- dimnames(fit2.core$contrasts)$Contrasts[n]
  if (length(which(all.stats[[n]]$stats.core$adj.P.Val<pvalcutoff & abs(all.stats[[n]]$stats.core$logFC)>logFCcutoff))>=200) {
    print(paste(contrast, "has at least 200 probes with an ajusted p.value of at least", pvalcutoff,  "and a log fold change of at least", logFCcutoff, sep=" "))
    stats.top.core <- topTable(fit2.core, coef=n, adjust="fdr", p.value=pvalcutoff, lfc=logFCcutoff, sort.by="B", number=nrow(gene.annots), genelist=cbind(gene.annots, fit2.core$genes))
    stats.top.core <- stats.top.core[!is.na(stats.top.core$symbol),]
    } else {
      stats.top.core <- topTable(fit2.core, coef=n, adjust="fdr", sort.by="B", number=nrow(gene.annots), genelist=gene.annots)
      stats.top.core <- stats.top.core[!is.na(stats.top.core$symbol),]
      }
  if (nrow(stats.top.core)>=200){
    stats.top.core <- stats.top.core[1:200,]
    }
  # incorporate expression levels for these genes in all samples
  if(nrow(stats.top.core)>0) {
    eset.top.core  <-  exprs(affyNorm.batch[stats.top.core[, "probeids"], ])
    colnames(eset.top.core)  <-  pData(affyNorm.batch)[,"samplelabel"]
    eset.top.core=eset.top.core[,apply(design[,colnames(design) %in% names(which(abs(contrast.matrix[,contrast])==1))]==1, 1, any)]
    eset.stats.top.core <- cbind(stats.top.core, eset.top.core)
    } else { eset.stats.top.core<-"NA"}
  return(list(eset.stats.top.core=eset.stats.top.core, contrast=contrast))
  })
```

### Output statistics and expression levels for these genes to Excel files
*these files contain the same information as the previous Excel files, and are merely subsetted to the top differentially expressed genes for each comparison*

```{r output_top, results='hide'}
lapply(top.stats, function(x){
  contrast=x$contrast
  write.table(x$eset.stats.top.core, file=file.path(resultsDir, paste("top200.stats", contrast, "xls",sep=".")), row.names=F, col.names=T, sep='\t')
  })
```

`r index=1`

[Links to Excel files with results for genes in top 200 log-odds of differential expression in `r top.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("top200.stats", top.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Links to Excel file with results for genes in top 200 log-odds of differential expression in `r top.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("top200.stats", top.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Links to Excel file with results for genes in top 200 log-odds of differential expression in `r top.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("top200.stats", top.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Links to Excel file with results for genes in top 200 log-odds of differential expression in `r top.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("top200.stats", top.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Links to Excel file with results for genes in top 200 log-odds of differential expression in `r top.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("top200.stats", top.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Links to Excel file with results for genes in top 200 log-odds of differential expression in `r top.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("top200.stats", top.stats[[index]]$contrast, "xls",sep=".")`)
`r index=index+1`

[Links to Excel file with results for genes in top 200 log-odds of differential expression in `r top.stats[[index]]$contrast` comparison](../results/hilyco/`r paste("top200.stats", top.stats[[index]]$contrast, "xls",sep=".")`)

### Heatmaps
- of top 100 results for each comparison
- as sorted by "B", the log-odds that a gene is differentially expressed
- clustered for similarity by probe/gene and sample

```{r heatmaps, fig.width=16, fig.height=16, fig.show='hold', out.width='50%', dev="CairoSVG"}
for(n in 1:ncol(contrast.matrix)){
  # get the names of the stages that are being compared in this comparison
  contrast <- top.stats[[n]]$contrast
  stages <- names(which(abs(contrast.matrix[,contrast])>0))
  # subset the samples and the data to those in the stages compared in this comparison
  sample.indices <- which(design[,stages[1]]>0 | design[,stages[2]]>0)
  sample.pd <- pData(affyNorm.batch)[sample.indices,]
  top.exprs.contrast <- top.stats[[n]]$eset.stats.top.core
  # setup row names for the heatmap, paste probeset ID to gene symbol
    row.names(top.exprs.contrast) <- paste(top.exprs.contrast$probeids, top.exprs.contrast$symbols, sep="_")
  # subset to relevant samples
  top.exprs.contrast <- top.exprs.contrast[,unlist(sample.pd$samplelabel)]
  # heatmap annotations
  heatmap.annots <- sample.pd[,c("samplelabel", "treatment", "celltype", "Batch")]
  heatmap.annots <- as.data.frame(apply(heatmap.annots, 2, unlist))
  row.names(heatmap.annots) <- heatmap.annots$samplelabel
  heatmap.annots$samplelabel <- NULL
  # heatmap annotation colors
  treatment_colors <- cbPalette[5:7]
  names(treatment_colors) <- unique(unlist(pData(affyNorm.batch)$treatment))
  celltype_colors <- cbPalette[1:3]
  names(celltype_colors) <- unique(unlist(pData(affyNorm.batch)$celltype))
  Batch_colors <- cbPalette[c(4,6,9)]
  names(Batch_colors) <- unique(unlist(pData(affyNorm.batch)$Batch))
  ann_colors = list(treatment = treatment_colors, celltype = celltype_colors, Batch=Batch_colors)
  ## Heatmaps
  pheatmap(as.matrix(top.exprs.contrast[1:100,]), annotation=heatmap.annots, color=rev(brewer.pal(11,"RdBu")), main=contrast, cluster_rows=T, cluster_cols=T, show_colnames=F, fontsize=14, fontsize_row=10,annotation_colors=ann_colors)
  }
```

[1]: http://www.bioconductor.org (BioC)
[2]: http://www.bioconductor.org/packages/2.12/bioc/html/oligo.html (oligo)
[3]: http://www.bioconductor.org/packages/2.10/bioc/html/arrayQualityMetrics.html (arrayQualityMetrics)
[4]: http://bioinformatics.oxfordjournals.org/content/19/2/185 (RMA)
[5]: http://www.bioconductor.org/packages/2.12/bioc/html/sva.html (sva)
[6]: http://www.bu.edu/jlab/wp-assets/ComBat/Abstract.html (ComBat)
[7]: http://www.bioconductor.org/packages/release/bioc/html/limma.html (limma)


