---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

```{r setup0, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
```

```{r setup, echo=FALSE}
# Setup report details
clientname="David Breault"
clientemail="David.Breault@childrens.harvard.edu"
contactname="Emanuele Pignatti"
contactemail="emanuele.pignatti@childrens.harvard.edu"
analystname="Radhika Khetani"
analystemail="rkhetani@hsph.harvard.edu"
```

#Breault lab RNA-seq (beta-catenin expression in Adrenal Glands)

RNA-Seq data analysis for `r contactname` (`r contactemail`), from  `r clientname`'s (`r clientemail`) group. 
<br>Contact `r analystname` (`r analystemail`) for additional details. 
</br>

Request from client was:

> Get lists of genes that are differentially expressed in the adrenal gland (outer layer) in response to changes in beta-catenin expression.

```{r setup1, echo=FALSE}
library(knitr)
library(rmarkdown)
library(knitrBootstrap)
```

## Setup
### Bioconductor and CRAN libraries used

```{r setup2, echo=TRUE}
library(ggplot2)
library(gplots)
library(RColorBrewer)
library(CHBUtils)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(DESeq2)
library(Biobase)
library(gProfileR)
library(pheatmap)
source("~/Documents/revigo.R")
library(treemap)
```

### Set variables
```{r setup3}
baseDir <- "/Users/rkhetani/Dropbox/HBC consults/Breault_betacatenin_RNA-Seq//breault-betac-bcbio/DESeq2/"
dataDir <- paste(baseDir, "/data", sep="")
resultsDir <- paste(baseDir, "/results", sep="")
metaDir <- paste(baseDir, "/meta", sep="")

heatcolors.2 <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
```

### Load data and metadata
```{r load}
data <- read.delim(file.path(dataDir, 'combined.counts'), header=T, sep="\t", row.names=1, as.is=T) 
meta <- read.delim(file.path(metaDir, 'metadata.csv'), header=T, sep=",", row.names=1)
```

## Differential Expression Analysis (DE)
Differential expression analysis was performed in 2 ways using the DESeq2 package: 

1. Pairwise comparisons between the various groups.
2. A single comparison across all groups (Analysis of Deviation, ANODEV) usinf the LRT test. 

### Pairwise comparisons (DE)
```{r deseq2, fig.align='center'}
eset <- new("ExpressionSet", exprs=as.matrix(data))
meta$treatment <- relevel(meta$treatment, ref="Ctr")
pData(eset) <- meta

dds <- DESeqDataSetFromMatrix(countData = exprs(eset), colData = pData(eset), design = ~ treatment)
dds <- DESeq(dds)
```

For the following summaries, an adjusted p-value (padj) threshold of 0.01 was used as a cut-off for significance, since a resonable number of genes were found to be DE with this criteria. *Please note these numbers are only based on a p-value cut-off, and do not employ a fold change cut-off.*

```{r summarize1}
res.ko.ctr <- results(dds, contrast=c("treatment","KO","Ctr"))
res.ko.ctr$threshold <- as.logical(res.ko.ctr$padj < 0.01)
sig.ko.ctr <- res.ko.ctr[which(res.ko.ctr$threshold),]
list.ko.ctr <- rownames(sig.ko.ctr)
```

**Summary of differentially expressed genes between KO and control**
```{r summarize1.5}
summary(res.ko.ctr, alpha=0.01)
```
```{r summarize2}
res.ex3.ctr <- results(dds, contrast=c("treatment","ex3","Ctr"))
res.ex3.ctr$threshold <- as.logical(res.ex3.ctr$padj < 0.01)
sig.ex3.ctr <- res.ex3.ctr[which(res.ex3.ctr$threshold),]
list.ex3.ctr <- rownames(sig.ex3.ctr)
```

**Summary of differentially expressed genes between ex3 and control**
```{r summarize2.5}
summary(res.ex3.ctr, alpha=0.01)
```

```{r summarize3}
res.ko.ex3 <- results(dds, contrast=c("treatment","KO","ex3"))
res.ko.ex3$threshold <- as.logical(res.ko.ex3$padj < 0.01)
sig.ko.ex3 <- res.ko.ex3[which(res.ko.ex3$threshold),]
list.ko.ex3 <- rownames(sig.ko.ex3)
```

**Summary of differentially expressed genes between KO and ex3**
```{r summarize3.5}
summary(res.ko.ex3, alpha=0.01)
```

#### Volcano plots, pairwise
The plots represent each gene with a dot. The fold change (log2) is plotted on the x-axis and the adjusted p-value (log10, p-value adjusted/corrected for multiple testing) is on the y-axis. The purple dot represent the gene that is differentially expressed (p-value < 0.01). *Please note these coloring is only based on a p-value cut-off, and does not employ a fold change cut-off.*  

```{r volcano, fig.align='center'}
p.cutoff <- 0.01

res.ko.ctr$threshold <- as.logical(res.ko.ctr$padj < p.cutoff)
res.ex3.ctr$threshold <- as.logical(res.ex3.ctr$padj < p.cutoff)
res.ko.ex3$threshold <- as.logical(res.ko.ex3$padj < p.cutoff)

df.ko.ctr <- data.frame(res.ko.ctr)
plot.ko.ctr <- ggplot(data=df.ko.ctr, aes(x=log2FoldChange, y=-log10(padj), colour = threshold)) +
  scale_color_manual(values = c("grey", "purple")) +
  geom_point(alpha=0.75, pch=16) +
  xlim(c(-10,10)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.25)),
        axis.title = element_text(size = rel(1)),
        axis.text = element_text(size = rel(.75))) +
  ggtitle('KO vs Control') +
  xlab("log2 fold change") + ylab("-log10 p-value")

df.ex3.ctr <- data.frame(res.ex3.ctr)
plot.ex3.ctr <- ggplot(data=df.ex3.ctr, aes(x=log2FoldChange, y=-log10(padj), colour = threshold)) +
  scale_color_manual(values = c("grey", "purple")) +
  geom_point(alpha=0.75, pch=16) +
  xlim(c(-10,10)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.25)),
        axis.title = element_text(size = rel(1)),
        axis.text = element_text(size = rel(.75))) +
  ggtitle('ex3 vs Control') +
  xlab("log2 fold change") + ylab("-log10 p-value")


df.ko.ex3 <- data.frame(res.ko.ex3)
plot.ko.ex3 <- ggplot(data=df.ko.ex3, aes(x=log2FoldChange, y=-log10(padj), colour = threshold)) +
  scale_color_manual(values = c("grey", "purple")) +
  geom_point(alpha=0.75, pch=16) +
  xlim(c(-10,10)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.25)),
        axis.title = element_text(size = rel(1)),
        axis.text = element_text(size = rel(0.75))) +
  ggtitle('KO vs ex3') +
  xlab("log2 fold change") + ylab("-log10 p-value")

grid.arrange(grobs = list(plot.ko.ctr, plot.ex3.ctr, plot.ko.ex3), ncol=2, top="Volcano Plots")

```

#### MA plot, pairwise
Similar to the volcano plot, the MA plot is a great way to visualize the comparative expression metrics for a two group comparison. The x–axis is the average/mean expression over all the samples and the y axis is the log2 fold change between WT and KO. The red dots represent the genes that are differentially expressed.

```{r MAplot, fig.align='center'}
par(mfrow = c(2, 2))

plotMA(res.ko.ctr, alpha = 0.01, main="KO vs Control", ylim=c(-2,2))
plotMA(res.ex3.ctr, alpha = 0.01, main="ex3 vs Control", ylim=c(-2,2))
plotMA(res.ko.ex3, alpha = 0.01, main="KO vs ex3", ylim=c(-2,2))
```

#### Functional analysis, pairwise

We ran the list of significantly different genes in all 3 pairwise comparisons through [g:Profiler](http://biit.cs.ut.ee/gprofiler/), a tool that performs statistical enrichment analysis to find over-representation of information like Gene Ontology terms, biological pathways, regulatory DNA elements, human disease gene annotations, and protein-protein interaction networks.

We took the results from g:Profiler and used the significant (fdr>=0.05) GO terms as input to [REVIGO](http://revigo.irb.hr/) to visualize the most prominent processes and reduce redundancy of terms. In the REVIGO GO Treemap figures, each rectangle is a single cluster representative. The representatives are joined into ‘superclusters’ of loosely related terms, visualized with different colors. Size of the rectangles may be adjusted to reflect either the p-value, or the frequency of the GO term in the underlying GOA database.

**KO versus Control**
```{r functional1, fig.align='center'}
gprofiler_results  <-   gprofiler(query = list.ko.ctr, 
            organism = "mmusculus",
            ordered_query = F, 
            exclude_iea = F, 
            max_p_value = 0.05, 
            max_set_size = 0,
            correction_method = "fdr",
            hier_filtering = "none", 
            domain_size = "annotated",
            custom_bg = "")

#write.table(gprofiler_results, "results/KO_control-gprofiler.txt", sep="\t")
allterms <- gprofiler_results$term.id
GOs <- allterms[grep('GO:', allterms)]
pvals <- gprofiler_results$p.value[grep('GO:', allterms)]
GO.pval <- cbind(gprofiler_results$term.id,gprofiler_results$p.value)
runRevigo(GOs, pvals)
```

**ex3 vs Control**
```{r functional2, fig.align='center'}
gprofiler_results  <-   gprofiler(query = list.ex3.ctr, 
            organism = "mmusculus",
            ordered_query = F, 
            exclude_iea = F, 
            max_p_value = 0.05, 
            max_set_size = 0,
            correction_method = "fdr",
            hier_filtering = "none", 
            domain_size = "annotated",
            custom_bg = "")

#write.table(gprofiler_results, "results/ex3_control-gprofiler.txt", sep="\t")
allterms <- gprofiler_results$term.id
GOs <- allterms[grep('GO:', allterms)]
pvals <- gprofiler_results$p.value[grep('GO:', allterms)]
GO.pval <- cbind(gprofiler_results$term.id,gprofiler_results$p.value)
runRevigo(GOs, pvals)
```

**KO vs ex3**
```{r functional3, fig.align='center'}
gprofiler_results  <-   gprofiler(query = list.ko.ex3, 
            organism = "mmusculus",
            ordered_query = F, 
            exclude_iea = F, 
            max_p_value = 0.05, 
            max_set_size = 0,
            correction_method = "fdr",
            hier_filtering = "none", 
            domain_size = "annotated",
            custom_bg = "")

#write.table(gprofiler_results, "results/KO_ex3-gprofiler.txt", sep="\t")
allterms <- gprofiler_results$term.id
GOs <- allterms[grep('GO:', allterms)]
pvals <- gprofiler_results$p.value[grep('GO:', allterms)]
GO.pval <- cbind(gprofiler_results$term.id,gprofiler_results$p.value)
runRevigo(GOs, pvals)
```

### ANODEV using the LRT test
Your experimental design can also be considered a 3-factor design, and DESeq2 offers another test (LRT) for this type of a design. In this case the list of differentially expressed genes include genes that are differentially expressed in any one group with respect to the one or both of the other groups.

Description from the [DESeq2 manual](https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf):

    DESeq2 offers two kinds of hypothesis tests: the Wald test, where we use the estimated standard error of a log2 fold change to test if it is equal to zero, and the likelihood ratio test (LRT). The LRT examines two models for the counts, a full model with a certain number of terms and a reduced model, in which some of the terms of the full model are removed. The test determines if the increased likelihood of the data using the extra terms in the full model is more than expected if those extra terms are truly zero.
    
    The LRT is therefore useful for testing multiple terms at once, for example testing 3 or more levels of a factor at once, or all interactions between two variables. The LRT for count data is conceptually similar to an analysis of variance (ANOVA) calculation in linear regression, except that in the case of the Negative Binomial GLM, we use an analysis of deviance (ANODEV), where the deviance captures the difference in likelihood between a full and a reduced model.

``` {r ANODEV-DE}
dds2 <- DESeqDataSetFromMatrix(countData = exprs(eset), colData = pData(eset), design = ~ treatment)
dds2 <- DESeq(dds2, test = "LRT", reduced = ~ 1)
res2 <- results(dds2)
res2$threshold <- as.logical(res2$padj < 0.01)
sig.lrt <- res2[which(res2$threshold),]
list.lrt <- rownames(sig.lrt)
```

**Summary of differentially expressed genes**

*Please note here that the fold change value only takes into account 2 groups.*
```{r ANODEV-DE_summary}
summary(res2, alpha=0.01)
```

#### Comparing count patterns across the groups for the top 30 most significant genes
``` {r plot counts setup}
geneplot <- function(dd, res, n = 1) {
    top.res <- order(res$padj)
    genedata <- plotCounts(dd, gene = top.res[n], intgroup = "treatment", returnData = TRUE, normalized = TRUE)
    ggplot(genedata, aes(x = treatment, y = count, color = treatment, group = 1)) + geom_point() + scale_y_log10() + ggtitle(as.character(mcols(dd)$symbol[top.res[n]]))
}
```

These plots allow us to look at the distribution of counts across the 3 groups, but also the variance between the replicates. Interestingly, for all the top 30 genes, the ex3 group looks most different, and higher than control and KO.

Top 6 genes, ordered by adjusted p-value
```{r plotcounts1, fig.align='center'}
gp1 <- geneplot(dds2, res2, 1)
gp2 <- geneplot(dds2, res2, 2)
gp3 <- geneplot(dds2, res2, 3)
gp4 <- geneplot(dds2, res2, 4)
gp5 <- geneplot(dds2, res2, 5)
gp6 <- geneplot(dds2, res2, 6)
grid.arrange(gp1, gp2, gp3, gp4, gp5, gp6, ncol = 2)
```

Next 6 genes (7 - 12), ordered by adjusted p-value
```{r plotcounts2, fig.align='center'}
gp7 <- geneplot(dds2, res2, 7)
gp8 <- geneplot(dds2, res2, 8)
gp9 <- geneplot(dds2, res2, 9)
gp10 <- geneplot(dds2, res2, 10)
gp11 <- geneplot(dds2, res2, 11)
gp12 <- geneplot(dds2, res2, 12)

grid.arrange(gp7, gp8, gp9, gp10, gp11, gp12, ncol = 2)
```

Next 6 genes (13 - 18), ordered by adjusted p-value
```{r plotcounts3, fig.align='center'}
gp13 <- geneplot(dds2, res2, 13)
gp14 <- geneplot(dds2, res2, 14)
gp15 <- geneplot(dds2, res2, 15)
gp16 <- geneplot(dds2, res2, 16)
gp17 <- geneplot(dds2, res2, 17)
gp18 <- geneplot(dds2, res2, 18)

grid.arrange(gp13, gp14, gp15, gp16, gp17, gp18, ncol = 2)
```

Next 6 genes (19 - 24), ordered by adjusted p-value
```{r plotcounts4, fig.align='center'}
#Next 10 genes (21 - 30), ordered by adjusted p-value
gp19 <- geneplot(dds2, res2, 19)
gp20 <- geneplot(dds2, res2, 20)
gp21 <- geneplot(dds2, res2, 21)
gp22 <- geneplot(dds2, res2, 22)
gp23 <- geneplot(dds2, res2, 23)
gp24 <- geneplot(dds2, res2, 24)

grid.arrange(gp19, gp20, gp21, gp22, gp23, gp24, ncol = 2)
```

Next 6 genes (25 - 30), ordered by adjusted p-value
```{r plotcounts5, fig.align='center'}
gp25 <- geneplot(dds2, res2, 25)
gp26 <- geneplot(dds2, res2, 26)
gp27 <- geneplot(dds2, res2, 27)
gp28 <- geneplot(dds2, res2, 28)
gp29 <- geneplot(dds2, res2, 29)
gp30 <- geneplot(dds2, res2, 30)

grid.arrange(gp25, gp26, gp27, gp28, gp29, gp30, ncol = 2)
```


```{r plotcounts6, fig.align='center', echo=FALSE, eval=FALSE}
#Next 10 genes (31 - 40), ordered by adjusted p-value
gp31 <- geneplot(dds2, res2, 31)
gp32 <- geneplot(dds2, res2, 32)
gp33 <- geneplot(dds2, res2, 33)
gp34 <- geneplot(dds2, res2, 34)
gp35 <- geneplot(dds2, res2, 35)
gp36 <- geneplot(dds2, res2, 36)
gp37 <- geneplot(dds2, res2, 37)
gp38 <- geneplot(dds2, res2, 38)
gp39 <- geneplot(dds2, res2, 39)
gp40 <- geneplot(dds2, res2, 40)

grid.arrange(gp31, gp32, gp33, gp34, gp35, gp36, gp37, gp38, gp39, gp40, ncol = 2)
```

#### Heatmap of the 50 most significant genes (LRT)
Interestingly, unlike with all the genes (see [QC report](https://dl.dropboxusercontent.com/u/74036176/Breault_RNA-Seq/qc-summary.html), the replicates are clustering together really nicely when the normalized counts from just the top 50 genes are considered (see the dendogram above). Please note that the genes are sorted by p-value in this heatmap, the first gene (on top) has the lowest p-value.

```{r heatmap setup}
#get normalized counts and add gene names to the matrix.
norm.count <- counts(dds2, normalized=T)
library(biomaRt)
mart<- useDataset("mmusculus_gene_ensembl", useMart("ensembl"))
gene.names <- getBM(
  filters= "ensembl_gene_id", 
  attributes= c("ensembl_gene_id", "external_gene_name"),
  values= row.names(norm.count),
  mart= mart)

# Normalized counts
ens.id <- row.names(norm.count)
GeneName <- gene.names[match(ens.id,gene.names$ensembl_gene_id),"external_gene_name"]
norm.count.gene <- data.frame(norm.count,GeneName)

sig.lrt.order <- sig.lrt[order(sig.lrt$padj),]
top.genes <- rownames(sig.lrt.order[1:50,])
norm.count.sig50 <- norm.count.gene[rownames(norm.count.gene) %in% top.genes,]
rownames(norm.count.sig50) <- norm.count.sig50$GeneName
norm.count.sig50 <- norm.count.sig50[,-19]
norm.count.sig50 <- as.matrix(norm.count.sig50)
colnames(norm.count.sig50) <- colnames(norm.count)
pheatmap(norm.count.sig50, annotation = meta, color = colorRampPalette(c("navy", "white", "firebrick3"))(2000), cluster_rows = F)
```


#### Functional Analysis, ANODEV
```{r functional lrt, fig.align='center'}
gprofiler_results  <-   gprofiler(query = list.lrt, 
            organism = "mmusculus",
            ordered_query = F, 
            exclude_iea = F, 
            max_p_value = 0.05, 
            max_set_size = 0,
            correction_method = "fdr",
            hier_filtering = "none", 
            domain_size = "annotated",
            custom_bg = "")

#write.table(gprofiler_results, "results/ANODEV-gprofiler.txt", sep="\t")
allterms <- gprofiler_results$term.id
GOs <- allterms[grep('GO:', allterms)]
pvals <- gprofiler_results$p.value[grep('GO:', allterms)]
GO.pval <- cbind(gprofiler_results$term.id,gprofiler_results$p.value)
runRevigo(GOs, pvals)
```

## Files for download
```{r biomart, echo = FALSE, eval=FALSE}
library(biomaRt)
mart<- useDataset("mmusculus_gene_ensembl", useMart("ensembl"))
norm.count <- counts(dds, normalized=T)

gene.names <- getBM(
  filters= "ensembl_gene_id", 
  attributes= c("ensembl_gene_id", "external_gene_name"),
  values= row.names(norm.count),
  mart= mart)

# Normalized counts
ens.id <- row.names(norm.count)
GeneName <- gene.names[match(ens.id,gene.names$ensembl_gene_id),"external_gene_name"]
new <- data.frame(norm.count,GeneName)
write.table(new, "data/norm_counts_ALL.txt", sep="\t")

norm.count.ex3.ctr <- new[rownames(new) %in% list.ex3.ctr,]
write.table(norm.count.ex3.ctr, "data/norm_counts_ex3-ctr_SIG.txt", sep="\t")

norm.count.ko.ctr <- new[rownames(new) %in% list.ko.ctr,]
write.table(norm.count.ko.ctr, "data/norm_counts_ko-ctr_SIG.txt", sep="\t")

norm.count.ex3.ko <- new[rownames(new) %in% list.ko.ex3,]
write.table(norm.count.ex3.ko, "data/norm_counts_KO-ex3_SIG.txt", sep="\t")

norm.count.sig <- norm.count.gene[rownames(norm.count.gene) %in% list.lrt,]
write.table(norm.count.sig, "data/norm_counts_ANODEV_SIG.txt", sep="\t")

# For DE results
#KO-ctr
ens.id <- row.names(res.ko.ctr)
GeneName <- gene.names[match(ens.id,gene.names$ensembl_gene_id),"external_gene_name"]
new.ko.ctr <- data.frame(res.ko.ctr,GeneName)
new.ko.ctr$threshold <- NULL
new.ko.ctr.sig <- new.ko.ctr[rownames(new.ko.ctr) %in% list.ko.ctr,]

write.table(new.ko.ctr, "results/KO_control-results.txt", sep="\t")
write.table(new.ko.ctr.sig, "results/KO_control-SIG.results.txt", sep="\t")

#ex3-ctr
ens.id <- row.names(res.ex3.ctr)
GeneName <- gene.names[match(ens.id,gene.names$ensembl_gene_id),"external_gene_name"]
new.ex3.ctr <- data.frame(res.ex3.ctr,GeneName)
new.ex3.ctr$threshold <- NULL
new.ex3.ctr.sig <- new.ex3.ctr[rownames(new.ex3.ctr) %in% list.ex3.ctr,]

write.table(new.ex3.ctr, "results/ex3_control-results.txt", sep="\t")
write.table(new.ex3.ctr.sig, "results/ex3_control-SIG.results.txt", sep="\t")

#ex3-ko
ens.id <- row.names(res.ko.ex3)
GeneName <- gene.names[match(ens.id,gene.names$ensembl_gene_id),"external_gene_name"]
new.ko.ex3 <- data.frame(res.ko.ex3,GeneName)
new.ko.ex3$threshold <- NULL
new.ko.ex3.sig <- new.ko.ex3[rownames(new.ko.ex3) %in% list.ko.ex3,]

write.table(new.ko.ex3, "results/KO_ex3-results.txt", sep="\t")
write.table(new.ko.ex3.sig, "results/KO_ex3-SIG.results.txt", sep="\t")

#LRT
ens.id <- row.names(res2)
GeneName <- gene.names[match(ens.id,gene.names$ensembl_gene_id),"external_gene_name"]
new2 <- data.frame(res2,GeneName)
new2$threshold <- NULL
new2.sig <- new2[rownames(new2) %in% list.lrt,]

write.table(new2, "results/ANODEV-results.txt", sep="\t")
write.table(new2.sig, "results/ANODEV-SIG.results.txt", sep="\t")

```

Below are the links to the files generated in this report, which will be useful for any downstream analysis.

### Data Files
* Normalized expression matrix [here](./data/norm_counts_ALL.txt)
* Subsetted expression matrix for significantly DE genes:
      * [KO versus control](./data/norm_counts_KO-ctr_SIG.txt)
      * [ex3 versus control](./data/norm_counts_ex3-ctr_SIG.txt)
      * [KO versus ex3](./data/norm_counts_KO-ex3_SIG.txt)
      * [ANODEV](./data/norm_counts_ANODEV_SIG.txt)

### DE results
*Legend for DE results:*
<br>
**baseMean** - mean of normalized counts for all samples
<br>
**log2FoldChange** - log2 fold change
<br>
**lfcSE** -  standard error
<br>
**stat** - Wald statistic
<br>
**pvalue** - Wald test p-value
<br>
**padj** - BH adjusted p-values
</br>

* Complete tables of results:
      * [KO versus control](./results/KO_control-results.txt)
      * [ex3 versus control](./results/ex3_control-results.txt)
      * [KO versus ex3](./results/KO_ex3-results.txt)
      * [ANODEV](./results/ANODEV-results.txt) [Please note here that the fold change value only takes into account 2 groups so in this case, it is best to ignore the "log2FoldChange" column.]

* Tables with results of only the significantly DE genes:
      * [KO versus control](./results/KO_control-SIG.results.txt)
      * [ex3 versus control](./results/ex3_control-SIG.results.txt)
      * [KO versus ex3](./results/KO_ex3-SIG.results.txt)
      * [ANODEV](./results/ANODEV-SIG.results.txt) [Please note here that the fold change value only takes into account 2 groups so in this case, it is best to ignore the "log2FoldChange" column.]

### Functional enrichment results
* Tables of g:Profiler results:
      * [KO versus control](./results/KO_control-gprofiler.txt)
      * [ex3 versus control](./results/ex3_control-gprofiler.txt)
      * [KO versus ex3](./results/KO_ex3-gprofiler.txt)
      * [ANODEV](./results/ANODEV-gprofiler.txt)

