---    
output:
  knitrBootstrap::bootstrap_document:
  theme: readable
  highlight: zenburn
  theme.chooser: TRUE
  highlight.chooser: TRUE
html_document:
  toc: true
  highlight: zenburn
---
        
```{r setup0, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,message=FALSE, prompt=TRUE, comment='', fig.cap='', fig.height = 9, fig.width = 12, bootstrap.show.code=FALSE)
```

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Henske miR21 RNA-seq functional analysis"
clientname="Hilaire Lam"
clientemail="hclam@partners.org"
labPI="Elizabeth Henske"
analystname="Mary Piper"
analystemail="piper@hsph.harvard.edu"
```
---
        
# Henske miR21 RNA-Seq Functional Analysis Report
RNA-Seq functional analysis report for the Henske project investigating the pathways dysregulated upon treatment with miR21 inhibitor in the presence of Rapamycin. 

Client: `r clientname`, `r labPI` group.  

Analysts: `r analystname` (`r analystemail`)

The most recent update of this html document occurred: `r date()`

---
        
# Overview
Investigated the pathways dysregulated in response to miR21 inhibitor in the presence of Rapamycin. Specifically interested in significantly dysregulated pathways related to cell death. Genes to focus on: Ppif, Tgfbr3, Tgfbr3l, Pdcd4, Tp53bp2

```{r qc-setup}
library(CHBUtils)
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(RColorBrewer)
library(knitr)
library(knitrBootstrap)
library(tidyr)
library(reshape)
library(rmarkdown)
library(dplyr)
library(ggdendro)
library(reshape)
library(gridExtra)
library(Biobase)
library(scales)
library(DESeq2)
library(gProfileR)
library(DEGreport)
library(vsn)
library(ggrepel)
library(sleuth)
library(png)
library(genefilter)
library(gage)
library(pathview)
library(gageData)
library(dplyr)
library(biomaRt)
library(org.Mm.eg.db)
library(clusterProfiler)
library(DOSE)
library(SPIA)
library(R.devices)
library(grid)

de_results <- read.table(file = "data/mir_21_vs_control_e2.tsv", header = T, quote = "", sep = "\t")
```

# Functional Analysis - miR21 vs. control

## gProfiler and REVIGO
Using the significant DE genes (padj < 0.05) and a foldchange threshold of 1.25 for the miR21 vs. control samples, a list of statistically enriched gene ontology (GO) terms, human phenotype ontology (HP) terms, and KEGG pathways (keg) was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). Often a foldchange threshold of 2.0 or greater is used as a threshold for functional analysis when a large number of genes are returned, but since the foldchanges of the known gene targets of miR21 were relatively small, the fold change threshold was relaxed.

A list including only the significant GO terms was then used as input to [REViGO](http://revigo.irb.hr/), which collapsed redundant and semantically-related terms and output the most significantly enriched functional categories. Note that we did not include in this REVIGO analysis enriched GO terms for molecular functions or cellular components.

Expected affected processes related to cell death are represented among the significant GO terms, including 'autophagic cell death' and various terms related to apoptosis (visualized within the REVIGO figure below). 

The list of all significant GO/HP/keg terms and associated genes, as well as a high resolution copy of the REViGO figure can be downloaded using the links below the table. 

J. Reimand, T. Arak, P. Adler, L. Kolberg, S. Reisberg, H. Peterson, J. Vilo. g:Profiler -- a web server for functional interpretation of gene lists (2016 update). Nucleic Acids Research 2016; doi: 10.1093/nar/gkw199

Supek F, Bošnjak M, Škunca N, Šmuc T. REVIGO summarizes and visualizes long lists of Gene Ontology terms. PLoS ONE 2011. doi:10.1371/journal.pone.0021800

```{r gprofiler_revigo}

# Return all results below threshold of padj = 0.05
resSig <- subset(de_results, padj < 0.05 & (abs(log2FoldChange)) > 0.32)
DEG_genenames <- resSig[order(abs(resSig$log2FoldChange), decreasing = T),]
no_NAs <- complete.cases(DEG_genenames)
gene_list <- DEG_genenames[no_NAs, ]
rownames(gene_list) <- gene_list$qid
gene_list <- gene_list[order(gene_list$padj),]
gene_list <- gene_list[, c(7,8)]

#gprofileR
library(gProfileR)
gprofiler_results <- gprofiler(query = gene_list, organism = "mmusculus", ordered_query = F,
                               exclude_iea = F, max_set_size = 0, correction_method = "fdr",
                               hier_filtering = "none", domain_size = "annotated", custom_bg = "")
allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
#write(GOs, file ="results/henske_miR21_GOterms.csv")
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
#write.csv(GO_genes, file ="results/henske_miR21_GO_genes.csv", quote = FALSE, row.names=T)

# library(treemap) 								# treemap package by Martijn Tennekes
# 
# # Set the working directory if necessary
# # setwd("C:/Users/username/workingdir");
# 
# # --------------------------------------------------------------------------
# # Here is your data from REVIGO. Scroll down for plot configuration options.
# 
# revigo.names <- c("term_ID","description","freqInDbPercent","uniqueness","dispensability","representative");
# revigo.data <- rbind(c("GO:0010225","response to UV-C",0.001,0.888,0.000,"response to UV-C"),
# c("GO:0034405","response to fluid shear stress",0.001,0.883,0.271,"response to UV-C"),
# c("GO:0072331","signal transduction by p53 class mediator",0.007,0.732,0.363,"response to UV-C"),
# c("GO:0030330","DNA damage response, signal transduction by p53 class mediator",0.004,0.704,0.352,"response to UV-C"),
# c("GO:0071243","cellular response to arsenic-containing substance",0.001,0.840,0.368,"response to UV-C"),
# c("GO:0010165","response to X-ray",0.002,0.883,0.608,"response to UV-C"),
# c("GO:0048010","vascular endothelial growth factor receptor signaling pathway",0.003,0.733,0.305,"response to UV-C"),
# c("GO:0050848","regulation of calcium-mediated signaling",0.002,0.712,0.465,"response to UV-C"),
# c("GO:1990090","cellular response to nerve growth factor stimulus",0.000,0.823,0.646,"response to UV-C"),
# c("GO:1990089","response to nerve growth factor",0.000,0.850,0.656,"response to UV-C"),
# c("GO:0071378","cellular response to growth hormone stimulus",0.001,0.818,0.567,"response to UV-C"),
# c("GO:0002931","response to ischemia",0.001,0.884,0.260,"response to UV-C"),
# c("GO:1904321","response to forskolin",0.003,0.839,0.572,"response to UV-C"),
# c("GO:0051896","regulation of protein kinase B signaling",0.007,0.698,0.580,"response to UV-C"),
# c("GO:0014065","phosphatidylinositol 3-kinase signaling",0.005,0.735,0.319,"response to UV-C"),
# c("GO:0009749","response to glucose",0.007,0.831,0.597,"response to UV-C"),
# c("GO:0030521","androgen receptor signaling pathway",0.003,0.744,0.218,"response to UV-C"),
# c("GO:1900408","negative regulation of cellular response to oxidative stress",0.001,0.668,0.660,"response to UV-C"),
# c("GO:1902882","regulation of response to oxidative stress",0.003,0.738,0.286,"response to UV-C"),
# c("GO:0030177","positive regulation of Wnt signaling pathway",0.006,0.616,0.584,"response to UV-C"),
# c("GO:0046683","response to organophosphorus",0.005,0.840,0.407,"response to UV-C"),
# c("GO:0007584","response to nutrient",0.004,0.857,0.201,"response to UV-C"),
# c("GO:0071456","cellular response to hypoxia",0.004,0.803,0.521,"response to UV-C"),
# c("GO:0048678","response to axon injury",0.002,0.878,0.279,"response to UV-C"),
# c("GO:0042770","signal transduction in response to DNA damage",0.005,0.723,0.384,"response to UV-C"),
# c("GO:0036119","response to platelet-derived growth factor",0.000,0.852,0.629,"response to UV-C"),
# c("GO:0036120","cellular response to platelet-derived growth factor stimulus",0.000,0.828,0.509,"response to UV-C"),
# c("GO:0033683","nucleotide-excision repair, DNA incision",0.000,0.803,0.251,"response to UV-C"),
# c("GO:0034620","cellular response to unfolded protein",0.005,0.791,0.598,"response to UV-C"),
# c("GO:0034616","response to laminar fluid shear stress",0.001,0.887,0.184,"response to UV-C"),
# c("GO:0038034","signal transduction in absence of ligand",0.005,0.739,0.311,"response to UV-C"),
# c("GO:2001021","negative regulation of response to DNA damage stimulus",0.003,0.685,0.529,"response to UV-C"),
# c("GO:0019883","antigen processing and presentation of endogenous antigen",0.001,0.962,0.000,"antigen processing and presentation of endogenous antigen"),
# c("GO:0002478","antigen processing and presentation of exogenous peptide antigen",0.002,0.955,0.643,"antigen processing and presentation of endogenous antigen"),
# c("GO:0002428","antigen processing and presentation of peptide antigen via MHC class Ib",0.000,0.961,0.535,"antigen processing and presentation of endogenous antigen"),
# c("GO:0019884","antigen processing and presentation of exogenous antigen",0.002,0.960,0.584,"antigen processing and presentation of endogenous antigen"),
# c("GO:0019885","antigen processing and presentation of endogenous peptide antigen via MHC class I",0.001,0.958,0.592,"antigen processing and presentation of endogenous antigen"),
# c("GO:0002260","lymphocyte homeostasis",0.004,0.799,0.458,"antigen processing and presentation of endogenous antigen"),
# c("GO:0031648","protein destabilization",0.001,0.836,0.000,"protein destabilization"),
# c("GO:2000278","regulation of DNA biosynthetic process",0.002,0.794,0.199,"protein destabilization"),
# c("GO:0090559","regulation of membrane permeability",0.003,0.831,0.331,"protein destabilization"),
# c("GO:0032846","positive regulation of homeostatic process",0.004,0.733,0.477,"protein destabilization"),
# c("GO:1903332","regulation of protein folding",0.000,0.804,0.336,"protein destabilization"),
# c("GO:0051881","regulation of mitochondrial membrane potential",0.002,0.831,0.320,"protein destabilization"),
# c("GO:1903333","negative regulation of protein folding",0.000,0.769,0.163,"protein destabilization"),
# c("GO:0051353","positive regulation of oxidoreductase activity",0.002,0.740,0.449,"protein destabilization"),
# c("GO:0030307","positive regulation of cell growth",0.006,0.655,0.554,"protein destabilization"),
# c("GO:0050685","positive regulation of mRNA processing",0.001,0.692,0.536,"protein destabilization"),
# c("GO:1901983","regulation of protein acetylation",0.003,0.760,0.392,"protein destabilization"),
# c("GO:1903362","regulation of cellular protein catabolic process",0.007,0.725,0.492,"protein destabilization"),
# c("GO:0051984","positive regulation of chromosome segregation",0.000,0.747,0.134,"protein destabilization"),
# c("GO:0032091","negative regulation of protein binding",0.003,0.818,0.149,"protein destabilization"),
# c("GO:0010811","positive regulation of cell-substrate adhesion",0.005,0.716,0.499,"protein destabilization"),
# c("GO:0034248","regulation of cellular amide metabolic process",0.001,0.817,0.142,"protein destabilization"),
# c("GO:0034250","positive regulation of cellular amide metabolic process",0.001,0.735,0.506,"protein destabilization"),
# c("GO:2001056","positive regulation of cysteine-type endopeptidase activity",0.007,0.663,0.606,"protein destabilization"),
# c("GO:0001678","cellular glucose homeostasis",0.004,0.800,0.340,"protein destabilization"),
# c("GO:0048144","fibroblast proliferation",0.005,0.935,0.000,"fibroblast proliferation"),
# c("GO:0001935","endothelial cell proliferation",0.006,0.931,0.694,"fibroblast proliferation"),
# c("GO:0048145","regulation of fibroblast proliferation",0.004,0.792,0.685,"fibroblast proliferation"),
# c("GO:1903409","reactive oxygen species biosynthetic process",0.002,0.957,0.000,"reactive oxygen species biosynthesis"),
# c("GO:0045428","regulation of nitric oxide biosynthetic process",0.002,0.800,0.614,"reactive oxygen species biosynthesis"),
# c("GO:2000377","regulation of reactive oxygen species metabolic process",0.007,0.804,0.610,"reactive oxygen species biosynthesis"),
# c("GO:0043248","proteasome assembly",0.004,0.864,0.013,"proteasome assembly"),
# c("GO:0034333","adherens junction assembly",0.003,0.820,0.406,"proteasome assembly"),
# c("GO:0097435","fibril organization",0.001,0.864,0.273,"proteasome assembly"),
# c("GO:0044089","positive regulation of cellular component biogenesis",0.002,0.699,0.606,"proteasome assembly"),
# c("GO:0032459","regulation of protein oligomerization",0.002,0.734,0.484,"proteasome assembly"),
# c("GO:0043044","ATP-dependent chromatin remodeling",0.004,0.848,0.474,"proteasome assembly"),
# c("GO:0000462","maturation of SSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)",0.001,0.830,0.630,"proteasome assembly"),
# c("GO:0010991","negative regulation of SMAD protein complex assembly",0.000,0.632,0.673,"proteasome assembly"),
# c("GO:0000028","ribosomal small subunit assembly",0.002,0.848,0.500,"proteasome assembly"),
# c("GO:0030490","maturation of SSU-rRNA",0.001,0.828,0.638,"proteasome assembly"),
# c("GO:0033108","mitochondrial respiratory chain complex assembly",0.004,0.824,0.545,"proteasome assembly"),
# c("GO:0097178","ruffle assembly",0.001,0.844,0.389,"proteasome assembly"),
# c("GO:0006998","nuclear envelope organization",0.002,0.831,0.458,"proteasome assembly"),
# c("GO:0006997","nucleus organization",0.008,0.853,0.306,"proteasome assembly"),
# c("GO:0032231","regulation of actin filament bundle assembly",0.003,0.695,0.629,"proteasome assembly"),
# c("GO:0051029","rRNA transport",0.000,0.926,0.025,"rRNA transport"),
# c("GO:0070198","protein localization to chromosome, telomeric region",0.000,0.925,0.552,"rRNA transport"),
# c("GO:0010632","regulation of epithelial cell migration",0.008,0.636,0.507,"rRNA transport"),
# c("GO:0042073","intraciliary transport",0.003,0.766,0.532,"rRNA transport"),
# c("GO:0006892","post-Golgi vesicle-mediated transport",0.005,0.896,0.550,"rRNA transport"),
# c("GO:1903335","regulation of vacuolar transport",0.000,0.781,0.449,"rRNA transport"),
# c("GO:0034502","protein localization to chromosome",0.002,0.921,0.336,"rRNA transport"),
# c("GO:1903352","L-ornithine transmembrane transport",0.006,0.909,0.345,"rRNA transport"),
# c("GO:0042991","transcription factor import into nucleus",0.005,0.879,0.622,"rRNA transport"),
# c("GO:0007034","vacuolar transport",0.006,0.897,0.399,"rRNA transport"),
# c("GO:1990182","exosomal secretion",0.000,0.895,0.336,"rRNA transport"),
# c("GO:0006407","rRNA export from nucleus",0.000,0.824,0.534,"rRNA transport"),
# c("GO:1903541","regulation of exosomal secretion",0.001,0.755,0.642,"rRNA transport"),
# c("GO:0046825","regulation of protein export from nucleus",0.001,0.746,0.670,"rRNA transport"),
# c("GO:0032906","transforming growth factor beta2 production",0.001,0.830,0.026,"transforming growth factor beta2 production"),
# c("GO:0032292","peripheral nervous system axon ensheathment",0.001,0.751,0.621,"transforming growth factor beta2 production"),
# c("GO:0061326","renal tubule development",0.004,0.761,0.623,"transforming growth factor beta2 production"),
# c("GO:0061008","hepaticobiliary system development",0.007,0.768,0.575,"transforming growth factor beta2 production"),
# c("GO:0060135","maternal process involved in female pregnancy",0.002,0.823,0.424,"transforming growth factor beta2 production"),
# c("GO:0007272","ensheathment of neurons",0.007,0.740,0.668,"transforming growth factor beta2 production"),
# c("GO:0001824","blastocyst development",0.004,0.776,0.692,"transforming growth factor beta2 production"),
# c("GO:0030282","bone mineralization",0.006,0.749,0.665,"transforming growth factor beta2 production"),
# c("GO:0014037","Schwann cell differentiation",0.002,0.739,0.631,"transforming growth factor beta2 production"),
# c("GO:0060674","placenta blood vessel development",0.002,0.770,0.639,"transforming growth factor beta2 production"),
# c("GO:0072132","mesenchyme morphogenesis",0.002,0.768,0.614,"transforming growth factor beta2 production"),
# c("GO:0035272","exocrine system development",0.005,0.771,0.520,"transforming growth factor beta2 production"),
# c("GO:0031017","exocrine pancreas development",0.001,0.776,0.523,"transforming growth factor beta2 production"),
# c("GO:0060281","regulation of oocyte development",0.001,0.666,0.676,"transforming growth factor beta2 production"),
# c("GO:0003170","heart valve development",0.002,0.770,0.676,"transforming growth factor beta2 production"),
# c("GO:0031214","biomineral tissue development",0.007,0.756,0.690,"transforming growth factor beta2 production"),
# c("GO:0003183","mitral valve morphogenesis",0.000,0.783,0.683,"transforming growth factor beta2 production"),
# c("GO:0040019","positive regulation of embryonic development",0.001,0.619,0.586,"transforming growth factor beta2 production"),
# c("GO:0021762","substantia nigra development",0.003,0.759,0.552,"transforming growth factor beta2 production"),
# c("GO:0048702","embryonic neurocranium morphogenesis",0.000,0.784,0.399,"transforming growth factor beta2 production"),
# c("GO:0031100","organ regeneration",0.001,0.777,0.504,"transforming growth factor beta2 production"),
# c("GO:0000959","mitochondrial RNA metabolic process",0.003,0.928,0.027,"mitochondrial RNA metabolism"),
# c("GO:0000291","nuclear-transcribed mRNA catabolic process, exonucleolytic",0.001,0.883,0.537,"mitochondrial RNA metabolism"),
# c("GO:0048024","regulation of mRNA splicing, via spliceosome",0.004,0.762,0.700,"mitochondrial RNA metabolism"),
# c("GO:0006368","transcription elongation from RNA polymerase II promoter",0.005,0.917,0.494,"mitochondrial RNA metabolism"),
# c("GO:0016074","snoRNA metabolic process",0.000,0.921,0.148,"mitochondrial RNA metabolism"),
# c("GO:0016073","snRNA metabolic process",0.003,0.913,0.321,"mitochondrial RNA metabolism"),
# c("GO:0006383","transcription from RNA polymerase III promoter",0.006,0.918,0.170,"mitochondrial RNA metabolism"),
# c("GO:0016078","tRNA catabolic process",0.000,0.886,0.457,"mitochondrial RNA metabolism"),
# c("GO:0043633","polyadenylation-dependent RNA catabolic process",0.000,0.895,0.149,"mitochondrial RNA metabolism"),
# c("GO:0016180","snRNA processing",0.003,0.899,0.299,"mitochondrial RNA metabolism"),
# c("GO:0000469","cleavage involved in rRNA processing",0.001,0.830,0.338,"mitochondrial RNA metabolism"),
# c("GO:0000466","maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)",0.004,0.819,0.446,"mitochondrial RNA metabolism"),
# c("GO:0000460","maturation of 5.8S rRNA",0.004,0.821,0.478,"mitochondrial RNA metabolism"),
# c("GO:0043484","regulation of RNA splicing",0.006,0.768,0.347,"mitochondrial RNA metabolism"),
# c("GO:0071025","RNA surveillance",0.003,0.888,0.510,"mitochondrial RNA metabolism"),
# c("GO:0034243","regulation of transcription elongation from RNA polymerase II promoter",0.003,0.784,0.659,"mitochondrial RNA metabolism"),
# c("GO:0071051","polyadenylation-dependent snoRNA 3'-end processing",0.000,0.913,0.246,"mitochondrial RNA metabolism"),
# c("GO:0031125","rRNA 3'-end processing",0.000,0.832,0.539,"mitochondrial RNA metabolism"),
# c("GO:0048857","neural nucleus development",0.004,0.822,0.028,"neural nucleus development"),
# c("GO:0071695","anatomical structure maturation",0.003,0.821,0.425,"neural nucleus development"),
# c("GO:0007569","cell aging",0.006,0.796,0.429,"neural nucleus development"),
# c("GO:0045598","regulation of fat cell differentiation",0.006,0.673,0.434,"neural nucleus development"),
# c("GO:0060322","head development",0.004,0.822,0.433,"neural nucleus development"),
# c("GO:0031099","regeneration",0.005,0.821,0.437,"neural nucleus development"),
# c("GO:0000086","G2/M transition of mitotic cell cycle",0.007,0.886,0.029,"G2/M transition of mitotic cell cycle"),
# c("GO:0048102","autophagic cell death",0.001,0.854,0.047,"autophagic cell death"),
# c("GO:0071887","leukocyte apoptotic process",0.006,0.829,0.541,"autophagic cell death"),
# c("GO:0036473","cell death in response to oxidative stress",0.003,0.727,0.579,"autophagic cell death"),
# c("GO:1903121","regulation of TRAIL-activated apoptotic signaling pathway",0.000,0.719,0.652,"autophagic cell death"),
# c("GO:1900119","positive regulation of execution phase of apoptosis",0.000,0.682,0.663,"autophagic cell death"),
# c("GO:0008625","extrinsic apoptotic signaling pathway via death domain receptors",0.004,0.659,0.606,"autophagic cell death"),
# c("GO:0010940","positive regulation of necrotic cell death",0.000,0.685,0.474,"autophagic cell death"),
# c("GO:0051541","elastin metabolic process",0.000,0.924,0.077,"elastin metabolism"),
# c("GO:0036508","protein alpha-1,2-demannosylation",0.000,0.918,0.228,"elastin metabolism"),
# c("GO:0097466","glycoprotein ERAD pathway",0.000,0.763,0.470,"elastin metabolism"),
# c("GO:0051438","regulation of ubiquitin-protein transferase activity",0.002,0.745,0.621,"elastin metabolism"),
# c("GO:0002181","cytoplasmic translation",0.003,0.914,0.185,"elastin metabolism"),
# c("GO:0006517","protein deglycosylation",0.000,0.921,0.139,"elastin metabolism"),
# c("GO:0018126","protein hydroxylation",0.003,0.911,0.288,"elastin metabolism"),
# c("GO:0045116","protein neddylation",0.001,0.902,0.319,"elastin metabolism"),
# c("GO:0035269","protein O-linked mannosylation",0.002,0.869,0.596,"elastin metabolism"),
# c("GO:0016925","protein sumoylation",0.002,0.900,0.612,"elastin metabolism"),
# c("GO:0006983","ER overload response",0.001,0.739,0.630,"elastin metabolism"));
# 
# stuff <- data.frame(revigo.data);
# names(stuff) <- revigo.names;
# 
# stuff$abslog10pvalue <- as.numeric( as.character(stuff$abslog10pvalue) );
# stuff$freqInDbPercent <- as.numeric( as.character(stuff$freqInDbPercent) );
# stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) );
# stuff$dispensability <- as.numeric( as.character(stuff$dispensability) );
# 
# # by default, outputs to a PDF file
# pdf( file="revigo_treemap.pdf", width=16, height=9 ) # width and height are in inches
# 
# # check the tmPlot command documentation for all possible parameters - there are a lot more
# treemap(
# 	stuff,
# 	index = c("representative","description"),
# 	vSize = "uniqueness",
# 	type = "categorical",
# 	vColor = "representative",
# 	title = "REVIGO Gene Ontology treemap",
# 	inflate.labels = FALSE,      # set this to TRUE for space-filling group labels - good for posters
# 	lowerbound.cex.labels = 0,   # try to draw as many labels as possible (still, some small squares may not get a label)
# 	bg.labels = "#CCCCCCAA",     # define background color of group labels
# 												       # "#CCCCCC00" is fully transparent, "#CCCCCCAA" is semi-transparent grey, NA is opaque
# 	position.legend = "none"
# )
# 
# dev.off()

img_revigo <- readPNG("results/henske_miR21_revigo.png")
grid.raster(img_revigo)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_GO_genes.csv)

[Download REVIGO Figure](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_revigo.png)

## clusterProfiler
Similar to gprofileR, the tool [clusterProfiler](http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) was used to perform over-representation analysis on GO terms associated with the significant DE genes (padj < 0.05, FC > 1.25). The table displays the list of top 20 GO terms that were significantly enriched among the significant genes. 

**NOTE:** While the GO terms output by clusterProfiler are very similar to those output by gprofileR, the small differences in the GO terms output is due to the different algorithms used by the two different programs.

G Yu, LG Wang, Y Han, QY He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology 2012, 16(5):284-287.

G Yu, LG Wang, GR Yan, QY He. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics 2015, 31(4):608-609.

```{r kegg}
# Create a KEGG dataset
kegg_mouse <- kegg.gsets(species = "mouse", id.type = "kegg")
kegg.gs <- kegg_mouse$kg.sets[kegg_mouse$sigmet.idx]
#head(kegg.gs)


mart <- useDataset("mmusculus_gene_ensembl",
                  useMart('ENSEMBL_MART_ENSEMBL',
                          host =  'useast.ensembl.org'))
attributes <- listAttributes(mart)
entrez <- getBM(attributes= c("ensembl_gene_id", "entrezgene"),
                    mart= mart)
entrez <- getBM(attributes= c("ensembl_gene_id"),
                    mart= mart)
ensembl_background <- entrez

#clusterprofiler analysis
sig_genes <- subset(de_results, padj<0.05 & (abs(log2FoldChange)) > 0.32)
sig_genes$qid <- as.character(sig_genes$qid)
sig_genes <- sig_genes$qid
all_genes <- ensembl_background[, 1, drop=T]
ensembl_foldchanges <- DEG_genenames$log2FoldChange
names(ensembl_foldchanges) <- DEG_genenames$qid
ggo <- groupGO(gene=sig_genes, OrgDb=org.Mm.eg.db, keytype = 'ENSEMBL', ont="BP", level=3, readable=TRUE)
ego <- enrichGO(gene=sig_genes, universe=all_genes, keytype = 'ENSEMBL', OrgDb=org.Mm.eg.db, ont="BP", pAdjustMethod = "BH", qvalueCutoff =0.05, readable=TRUE)

GO_processes <- ego@result
knitr::kable(head(GO_processes, n=20))

cluster_summary <- summary(ego)
#write.csv(GO_processes, "results/henske_miR21_clusterprofiler_GO_processes.csv", quote=F)
#dotplot(ego, showCategory=75)
#plotGOgraph(ego)

#enrichMap(ego, n=75, vertex.label.font=8)
#barplot(ggo, drop=TRUE, showCategory=12)
#barplot(ego, showCategory=20)
ego2 <- ego
ego2@result <- ego@result[c(1,3,5,8,13,14,16,20),]
#cnetplot(ego2, categorySize="pvalue", showCategory = 8, foldChange=ensembl_foldchanges)
```
[Download full list of significant GO processes](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_clusterprofiler_GO_processes.csv)

The dotplot below shows the number of genes associated with the first 75 terms (size) and the p-adjusted values for these terms. 

```{r dotplot}
img_dotplot <- readPNG("results/henske_miR21_dotplot.png")
grid.raster(img_dotplot)
```
[Download clusterProfiler Dotplot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_dotplot.png)

The GOgraph below show the relationship between the top 75 most significantly enriched GO terms, by grouping similar terms together.

```{r clusterprofiler_GOgraph}
img_GOgraph <- readPNG("results/henske_miR21_enrichmap.png")
grid.raster(img_GOgraph)
```
[Download clusterProfiler GOgraph](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_enrichmap.png)

The graph below shows the relationship among eight of the most significant GO terms, and the fold changes of the significant genes associated with these terms. Note that I did not include some of the most significant GO terms due to redundancy with other top terms. For example, I did not include 'ribosome biogenesis' in the plot because it contained similar terms to 'ribonucleoprotein complex biogenesis'. The 'regulation of the apoptotic signaling pathway' and  'intrinisic apoptotic signaling pathway' are among the terms included, and the genes associated with these pathways are shown, in addition to the shared genes among the other top pathways. Of interest, many of the genes shared between the apoptotic pathways and the other pathways are known tumor suppressor genes or oncogenes.

```{r clusterprofiler_cnetgraph}
img_cnet <- readPNG("results/henske_miR21_cnet_selected.png")
grid.raster(img_cnet)
```
[Download clusterProfiler Cnet Graph](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_cnet_selected.png)

## Pathway analysis using SPIA (topology-based method)
The previous analyses did not explore how genes interact with each other (e.g. activation, inhibition, phosphorylation, ubiquitination, etc) to determine the pathway-level statistics. The [SPIA (Signaling Pathway Impact Analysis)](http://bioconductor.org/packages/release/bioc/html/SPIA.html) tool was used to integrate the lists of differentially expressed genes determined by DESeq2 (padj < 0.05), their fold changes, and pathway topology to identify affected pathways.

The table below shows the top 10 significantly dysregulated pathways based on over-representation and signaling perturbations accumulation. The table shows the following information: pSize is the number of genes on the pathway; NDE is the number of DE genes per pathway; tA is the observed total preturbation accumulation in the pathway; pNDE is the probability to observe at least NDE genes on the pathway using a hypergeometric model; pPERT is the probability to observe a total accumulation more extreme than tA only by chance; pG is the p-value obtained by combining pNDE and pPERT; pGFdr and pGFWER are the False Discovery Rate and respectively Bonferroni adjusted global p-values; and the Status gives the direction in which the pathway is perturbed (activated or inhibited). KEGGLINK gives a web link to the KEGG website that displays the pathway image with the differentially expressed genes highlighted in red.

In the plot below, each pathway is a point and the coordinates are the log of pNDE (using a hypergeometric model) and the p-value from perturbations, pPERT. The oblique lines in the plot show the significance regions based on the combined evidence.

The most perturbed pathways include many different cancers, which are all inhibited. While apoptosis and cell death pathways do not show up as being significantly dysregulated in this analysis, there are significant genes identified in the cancer pathways that are associated with apoptosis [(e.g. Small cell lung cancer)](http://www.genome.jp/kegg-bin/show_pathway?mmu05222+12447+12448+12443+16412+16399+16403+12576+20181+20182+22059+11652+17869+19211+18706+18709+320207+13063+12826+12829+12830+16772+16773+16776+16779+16780+94216+11798+19225+16150+18035+12048+22029+22030+22032+13555+242705+12567+12571+19697+27401+12579). 

The full list of significantly dysregulated pathways and the spia plot can be downloaded using the links below the table.

Tarca AL, Kathri P and Draghici S (2013). SPIA: Signaling Pathway Impact Analysis (SPIA) using combined evidence of pathway over-representation and unusual signaling perturbations. [http://bioinformatics.oxfordjournals.org/cgi/reprint/btn577v1](http://bioinformatics.oxfordjournals.org/cgi/reprint/btn577v1).

```{r spia, results="hide",echo=FALSE}
#spia (http://www.gettinggeneticsdone.com/2012/03/pathway-analysis-for-high-throughput.html)

# significant genes is a vector of fold changes where the names
# are ENTREZ gene IDs. The background set is a vector of all the 
# genes represented on the platform.
#convert ensembl to entrez ids

# To create background dataset, I am using all genes in the entrez database matching our Ensembl IDs - SPIA requires Entrez IDs, so lose a lot of genes with conversion. Also, de-duplicating datasets (some Ensembl IDs assoiciated with more than one Entrez IDs) and turning Entrez IDs into characters.

entrez <- getBM(attributes= c("ensembl_gene_id", "entrezgene"),
                    mart= mart)
entrez_results <- subset(entrez, entrezgene != "NA")
all <- as.character(entrez_results[, 2])
dups_all <- duplicated(all)
all <- all[!dups_all]

sig_genes <- subset(de_results, padj<0.05)
names(sig_genes)[1] <- c("ensembl_gene_id")
sig_genes <- merge(sig_genes, entrez_results, by="ensembl_gene_id")


foldchanges <- sig_genes$log2FoldChange
names(foldchanges) <- as.character(sig_genes$entrezgene)
dups_foldchanges <- duplicated(names(foldchanges))
foldchanges <- foldchanges[!dups_foldchanges]
#head(foldchanges)

# run SPIA - in order for spia() to run, the IDs (entrez or ensembl) should be character type, and the background and significant genes lists should be de-duplicated.

spia_result <- spia(de=foldchanges, all=all, organism="mmu", plots=F)

#plotP(spia_result, threshold=0.05)
img_spia <- readPNG("results/henske_miR21_spia.png")
grid.raster(img_spia)

#write.csv(spia_result, "results/henske_miR21_spia_results.csv", quote=F)
```

```{r spia_results, echo=FALSE}
knitr::kable(head(spia_result, n=10))
```
[Download SPIA results](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_spia_results.csv)

[Download SPIA plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_spia.png)

## Gene set enrichment analysis 
Gene set enrichment analysis tools like GAGE and clusterProfiler use ranked lists of genes (here ranked by log2FC) without using a threshold. This allows these gene set enrichment tools to use more information to identify enriched biological processes. The introduction to gene set enrichment analysis goes into more detail about some of the advantages of this approach: [http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896/](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896/). 

### Pathway analysis with clusterProfiler and Pathview
Analysis was performed to identify any pathways that were significantly dysregulated (genes in pathway could be up- or down-regulated simultaneously). The significantly dysregulated pathways (q-value (FDR) < 0.05) are displayed below in the pathway images, which show the degree of dysregulation of the genes (the minus direction (green) is down-regulated, while the positive direction (red) is up-regulated). 

In addition, the Running Enrichment Score plots show the enrichment of higher ranked genes in the significantly dysregulated pathways. 

Since the genes associated with apoptosis did not have very high fold change differences, it is not surprising that these terms were not significant. To note, many of the diseases significantly dysregulated involve the oxidative phosphorylation pathway. 

```{r gsea_mmu04020}
# Attain DEG_background <- data.frame(de_results)
all_results_gsea <- de_results
all_results_gsea <- all_results_gsea[order(abs(all_results_gsea$log2FoldChange), decreasing = T), ]
foldchanges <- all_results_gsea$log2FoldChange
names(foldchanges) <- all_results_gsea$qid
#head(foldchanges)

foldchanges <- sort(foldchanges, T)
ego3 <- gseGO(geneList     = foldchanges,
              OrgDb        = org.Mm.eg.db,
              keytype      = "ENSEMBL",
              ont          = "BP",
              nPerm        = 1000,
              minGSSize    = 100,
              maxGSSize    = 500, 
              pvalueCutoff = 0.05,
              verbose      = FALSE)
gseaGOresult <- ego3@result
#write.csv(gseaGOresult, "results/henske_miR21_gseaGO_results.csv", quote=F)
#gseaplot(ego3, geneSetID = "GO:0006412")
```
[Download GO GSEA results](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_gseaGO_results.csv)

**Huntington's disease**

*Pathway*
```{r gsea_kegg}
# GSEA for the KEGG pathway requires foldchanges vector to be named with Entrez IDs
DEG_background <- data.frame(de_results)

mart <- useDataset("mmusculus_gene_ensembl",
                  useMart('ENSEMBL_MART_ENSEMBL',
                          host =  'useast.ensembl.org'))
attributes <- listAttributes(mart)
entrez <- getBM(filters= "ensembl_gene_id",
                    attributes= c("ensembl_gene_id", "entrezgene"),
                    values= DEG_background$qid,
                    mart= mart)
DEG_background$ensembl_gene_id <- DEG_background$qid
entrez_results <- merge(DEG_background, entrez, by="ensembl_gene_id")
entrez_results <- subset(entrez_results, entrezgene != "NA")

all_results_gsea <- entrez_results
all_results_gsea <- all_results_gsea[order(abs(all_results_gsea$log2FoldChange), decreasing = T), ]
foldchanges <- all_results_gsea$log2FoldChange
names(foldchanges) <- all_results_gsea$entrezgene
foldchanges <- sort(foldchanges, T)

kk2 <- gseKEGG(geneList     = foldchanges,
               organism     = 'mmu',
               nPerm        = 1000,
               minGSSize    = 120,
               pvalueCutoff = 0.05,
               verbose      = FALSE)
kkgsea_results <- kk2@result

mmu05016gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu05016",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu05016 <- readPNG("results/henske_miR21_mmu05016.pathview.png")
grid.raster(img_mmu05016)
```
[Download Huntington Disease GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu05016.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu05016_enrich}
gseaplot(kk2, "mmu05016")
```

**Alzheimer's disease**

*Pathway*
```{r gsea_mmu05010}
mmu05010gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu05010",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu05010 <- readPNG("results/henske_miR21_mmu05010.pathview.png")
grid.raster(img_mmu05010)
```
[Download Alzheimer's disease GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu05010.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu05010_enrich}
gseaplot(kk2, "mmu05010")
```

**Parkinson's disease**

*Pathway*
```{r gsea_mmu05012}
mmu05012gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu05012",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu05012 <- readPNG("results/henske_miR21_mmu05012.pathview.png")
grid.raster(img_mmu05012)
```
[Download Parkinson's disease GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu05012.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu05012_enrich}
gseaplot(kk2, "mmu05012")
```

**Ribosome**

*Pathway*
```{r gsea_mmu03010}
mmu03010gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu03010",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu03010 <- readPNG("results/henske_miR21_mmu03010.pathview.png")
grid.raster(img_mmu03010)
```
[Download Ribosome GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu03010.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu03010_enrich}
gseaplot(kk2, "mmu03010")
```

**Non-alcoholic fatty liver disease**

*Pathway*
```{r gsea_mmu04932}
mmu04932gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu04932",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu04932 <- readPNG("results/henske_miR21_mmu04932.pathview.png")
grid.raster(img_mmu04932)
```
[Download Non-alcoholic fatty liver disease GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu04932.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu04932_enrich}
gseaplot(kk2, "mmu04932")
```

**Spliceosome**

*Pathway*
```{r gsea_mmu03040}
mmu03040gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu03040",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu03040 <- readPNG("results/henske_miR21_mmu03040.pathview.png")
grid.raster(img_mmu03040)
```
[Download Spliceosome GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu03040.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu03040_enrich}
gseaplot(kk2, "mmu03040")
```

**Epstein-Barr virus infection**

*Pathway*
```{r gsea_mmu05169}
mmu05169gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu05169",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu05169 <- readPNG("results/henske_miR21_mmu05169.pathview.png")
grid.raster(img_mmu05169)
```
[Download Epstein-Barr virus infection GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu05169.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu05169_enrich}
gseaplot(kk2, "mmu05169")
```

**cAMP signaling pathway**

*Pathway*
```{r gsea_mmu04024}
mmu04024gsea <- pathview(gene.data  = foldchanges,
                     pathway.id = "mmu04024",
                     species    = "mmu",
                     limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu04024 <- readPNG("results/henske_miR21_mmu04024.pathview.png")
grid.raster(img_mmu04024)
```
[Download cAMP signaling pathway GSEA Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu04024.pathview.png)

*Running Enrichment Scores*
```{r gsea_mmu04024_enrich}
gseaplot(kk2, "mmu04024")
```

### Pathway analysis with GAGE and Pathview 
We also used a different method for gene set enrichment analysis, using the [GAGE (Generally Applicable Gene-set Enrichment for Pathway Analysis)](http://bioconductor.org/packages/release/bioc/html/gage.html) and [Pathview](http://bioconductor.org/packages/release/bioc/html/pathview.html) tools (q-value (FDR) < 0.05). This method also uses the log2 fold changes obtained from the DESeq2 analysis for every gene and identifies pathways that are significantly dysregulated. 

The fold changes between conditions for each gene were used as input to determine whether the pathways were significantly dysregulated. Since the genes associated with apoptosis did not have very high fold change differences, it is not surprising that these terms were not significant. This method did identify significantly dysregulated pathways, including the Ribosome, Oxidative phosphorylation, Pyrimidine metabolism, and Spliceosome, and the images below reflect the degree of dysregulation of the genes (the minus direction (green) is down-regulated, while the positive direction (red) is up-regulated).

Weijun Luo, Michael Friedman, Kerby Shedden, Kurt Hankenson, and Peter Woolf. GAGE: generally applicable
gene set enrichment for pathway analysis. BMC Bioinformatics, 2009. doi:10.1186/1471-2105-10-161.

Weijun Luo and Cory Brouwer. Pathview: an R/Bioconductor package for pathway-based data integration
and visualization. Bioinformatics, 29(14):1830-1831, 2013. doi: 10.1093/bioinformatics/btt285.

**Ribosome Pathway**
```{r gage_single_dir}
# Run gage
# Same direction = T
keggres = gage(foldchanges, gsets=kegg.gs, same.dir=TRUE)

# Look at single up (greater) and down (less), and statatistics.
#lapply(keggres, head, n=20)
sel <- keggres$greater[, "q.val"] < 0.05 & !is.na(keggres$greater[, "q.val"])
path.ids <- rownames(keggres$greater)[sel]

# Get the pathways
keggrespathways = data.frame(id=rownames(keggres$greater), keggres$greater) %>% 
  tbl_df() %>% 
  filter(row_number()<=4) %>% 
  .$id %>% 
  as.character()
#keggrespathways

# Get the IDs.
keggresids = substr(keggrespathways, start=1, stop=8)
#keggresids

#pathview
# Define plotting function for applying later
plot_pathway = function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="mmu", new.signature=FALSE)

# plot multiple pathways (plots saved to disk and returns a throwaway list object)
#tmp = sapply(keggresids, function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="mmu"))

img_pathview <- readPNG("results/henske_miR21_mmu03010.pathview.png")
grid.raster(img_pathview)
```
[Download KEGG Ribosome Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_mmu03010.pathview.png)

**Oxidative phosphorylation**
```{r kegg_gsea_plot2}
img_pathview <- readPNG("results/henske_miR21_gage_mmu00190.pathview.png")
grid.raster(img_pathview)
```
[Download KEGG Oxidative phosphorylation Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_gage_mmu00190.pathview.png)

**Pyrimidine metabolism**
```{r kegg_gsea_plot3}
img_pathview <- readPNG("results/henske_miR21_gage_mmu00240.pathview.png")
grid.raster(img_pathview)
```
[Download KEGG Pyrimidine metabolism Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_gage_mmu00240.pathview.png)

**Spliceosome**
```{r kegg_gsea_plot4}
img_pathview <- readPNG("results/henske_miR21_gage_mmu03040.pathview.png")
grid.raster(img_pathview)
```
[Download KEGG Spliceosome Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske_miR21/henske_miR21_gage_mmu03040.pathview.png)

```{r session_info}
sessionInfo()
```