---
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: true
    theme: simplex
    code_folding: show 
title: "Batch Correction and Differential Expression Analysis"
author: John Hutchinson
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview

RNAseq DE analysis for Shachar Dagan  (shachar_dagan@hms.harvard.edu),  Gu group at HMS. Using new set of LE samples to replace LE2, LE3 and LE4. 

Contact John Hutchinson (jhutchin@hsph.harvard.edu) for additional details.

The most recent update of this html document occurred: `r date()`.

The sections below provide code to reproduce the included results and plots. 

---

# Setup

## Libraries and variables

```{r vars}
library(ggplot2)
#library(gplots)
library(DT)
library(CHBUtils)
library(DESeq2)
#library(Biobase)
library(gProfileR)
library(pheatmap)
library(dplyr)
library(biomaRt)
library(gProfileR)
library(RUVSeq)

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442","#0072B2", "#D55E00", "#CC79A7")

baseDir <- "~/Work/projects/gu_rnaseq/"
dataDir <- file.path(baseDir, "results/final/2016-06-15_project")
resultsDir <- file.path(baseDir, "results/final/2016-06-15_project/report")

pvalcutoff=0.01
lfcutoff=1
numgenes=100
```

---

# Data and metadata import 

```{r import}
project_summary = file.path(dataDir, "project-summary.csv")
counts_file = file.path(dataDir, "combined.counts")

summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE) 
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
summarydata$fastqcode <- NULL
summarydata$phenotype <- NULL

counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality","rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate","Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped","rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.","Genes.Detected", "Unique.Starts.Per.Read","unique_starts_per_read","complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads","Median.insert.size", "Mapped.reads.pct","Total.reads","avg_coverage_per_region", "Mapped.Reads", "Average.insert.size")

metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
metadata$day <- as.character(metadata$day)
metadata$batch <- as.character(metadata$batch)
summarydata$day <- as.character(summarydata$day)
```

---

# Data Manipulations

## Remove rRNA and mitochondrial genes

```{r rRNA removal}
ensemblmart = useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl", host = "jul2015.archive.ensembl.org")
attributes <- listAttributes(ensemblmart)
filters <- listFilters(ensemblmart)

conversions = getBM(attributes = c("ensembl_gene_id", "mgi_symbol", "gene_biotype", "chromosome_name"), mart =ensemblmart)

# id rRNA genes by biotype coding
rrna_biotypes = c("rRNA", "Mt_rRNA", "misc_RNA", "snRNA", "snoRNA", "tRNA", "Mt_tRNA")
rrna_genes <- unique(subset(conversions, gene_biotype %in% rrna_biotypes)$ensembl_gene_id)

#id mitochondrial proteing coding genes by chromosome
mt_genes <- unique(subset(conversions, conversions$chromosome_name=="MT" & conversions$gene_biotype=="protein_coding")$ensembl_gene_id)

# Analysis without mt and rRNA genes
counts <- counts[!rownames(counts) %in% rrna_genes & !rownames(counts) %in% mt_genes, ]
```

## Drop outlier samples

Previous QC analysis showed that LE2, LE3 and LE4 are all outliers with human contamination. Dropped these samples from the analysis.

```{r dropoutliers}
metadata <- metadata[!row.names(metadata) %in% c("LE2_1", "LE3_1", "LE4_1"),]
counts <- counts[,!names(counts) %in% c("LE2_1", "LE3_1", "LE4_1")]
```

## Remove genes with no counts for any of the remaining samples

```{r dropzeros}
counts <- counts[!apply(counts, 1, function(x) all(x==0)),]
```

## Import raw counts into DESeq2

```{r DEobject, results='hide'}
# drop into DESeq2 object so you can take advantage of size factor adjustments
eset <- new("ExpressionSet", exprs = as.matrix(counts))
metadata <- metadata[order(metadata$sampletype, metadata$day),]
pData(eset) <- metadata
validObject(eset)
# not sure the formula matters that much here as you're only using DESeq2 for size adjustments
dds <- DESeqDataSetFromMatrix(countData = exprs(eset), colData = pData(eset), design = ~day+batch+sampletype)
dds <- DESeq(dds)
```



# Data Sanity Check

- does the data make sense, i.e. do the samples cluster the way we expect them to?

Clustering plots can be strongly affected by variance outliers (i.e. genes that show higher than expected amounts of variance). 

For RNA-Seq raw counts, variance increases with the mean. Therefore, we need to use a transformation that will stabilize the variance across the mean for the gene counts. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and the low count genes will dominate the PCA plots. Therfore, we can try transforming to stabilize variance across the mean using vst or rlog. For high counts, rlog gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

## Pick a variance stabilization method {.tabset}

```{r vst,  fig.keep="none"}
library(vsn)
g.raw <- meanSdPlot(as.matrix(counts))
vst = varianceStabilizingTransformation(dds)
g.vst <- meanSdPlot(as.matrix(assay(vst)))
rld = rlog(dds)
g.rld <- meanSdPlot(as.matrix(assay(rld)))
```

### raw counts

```{r vsplots1}
g.raw$gg+ggtitle("raw counts")
```

### vst transformed

```{r vsplots2}
g.vst$gg+ggtitle("vst transformation")
```

### rlog transformed

```{r vsplots3}
g.rld$gg+ggtitle("rlog transformation")
```

- rld stabiliztion looks best (less variance (sd) overall)

## PCA plots on variance stabilized data {.tabset}

### batch and sample type

```{r rawpca1}
plotPCA(rld, intgroup=c("batch", "sampletype"))
```

### batch, day and sample type

```{r rawpca2}
plotPCA(rld, intgroup=c("batch", "sampletype", "day"))
```

# Batch Correction

There is a batch effect, which we will need to eliminate as it is confounded with the sample class. Ideally, all the lung samples would cluster together.

Try using [RUVseq](http://www.nature.com/nbt/journal/v32/n9/full/nbt.2931.html) to remove variation associated wtih batch.

Leverage the two technical replicates we have that span the batches.

```{r ruvseqs}
#RUVs using replicates
## specify replicates, check this!
replicates <- matrix(data=c(c(1,2), c(6,7)), byrow=TRUE, nrow=2)
## run RUVseq on size-adjusted data from DESeq2
# k of 2 completely overlaps replicates on PCA
batch_ruvs_calc <- RUVs(as.matrix(counts, normalize=TRUE), rownames(counts), k=2, replicates, isLog=FALSE)
## variance stabilize RUVseq adjusted results
normed.ruvs.vs <- varianceStabilizingTransformation(batch_ruvs_calc$normalizedCounts)
## PCA plot the variaznce stabilized RUVseqS adjusted results
# have to shove it in this object for plotPCA to work
se.ruvs <- SummarizedExperiment(normed.ruvs.vs, colData=colData(dds))
```

## PCA plots on batch corrected, variance stabilized data {.tabset}

### batch and sample type

```{r ruvpca1}
plotPCA(DESeqTransform(se.ruvs), intgroup=c("sampletype", "batch"))
```

### batch, day and sample type

```{r ruvpca2}
plotPCA(DESeqTransform(se.ruvs), intgroup=c("sampletype", "batch", "day"))
```





# Differential Expression Analysis

RUVseq batch correction is a great improvement over the uncorrected data but it still doesn't completely remove the batch effect (ignoring the one outlier, the lung samples still split themselves by batch). However, this is pretty good, especially with the data we have. 
 
Differential gene expression analysis of count data was performed using the Bioconductor R package, DESeq2. The count data was fit to a negative binomial model and dispersion estimates were generated using the mean values from the maximum likelikhood estimate of log2 fold changes, optimizing the Cox-Reid adjusted profile likelihood. **Before doing this, I also removed the new replicate samples, LE1_2 and LE5_2, as they're superfluous at this point.**

The models used to estimate differential expression accounted for the matched sample design and the RUVseq generated batch correction.
 
As there was a batch effect, we recommend caution in interpreting the results. Significantly differentially expressed genes identified in these analyses will require lab verification. In addition, we recommend looking at the expression heatmaps of the significant DE genes to narrow down genes most likely to be truly differentially expressed between lung and retinal samples.

```{r, DE}
library(DESeq2)
counts <- exprs(eset)
counts <- counts[,!colnames(counts) %in% c("LE1_2", "LE5_2")]

pd <- pData(eset)
pd <- pd[!row.names(pd) %in% c("LE1_2", "LE5_2"),]

ruvs <- batch_ruvs_calc$W
ruvs <- ruvs[!row.names(ruvs) %in% c("LE1_2", "LE5_2"),]

dds.ruvs <- DESeqDataSetFromMatrix(countData=counts, colData=cbind(ruvs,pd), design=~W_1+W_2+day+sampletype)
dds.ruvs <- DESeq(dds.ruvs) 
```

## Statistics for all genes

Adjusted p-value take into account the high number of statistical tests we are performing (multiple testing adjustment). The higher the number of tests we have to adjust for, the more stringent our adjustment, and the fewer statistically significant genes in our result.

To reduce this adjustment, DESeq2 tries to pre-filter out genes for whom statistical tests would have no, or little chance of showing significant evidence, without even looking at their test statistic. DESeq2 does this by filtering out genes with very low average counts overall; these genes are are not likely to show significant differences due to high dispersion. We can also filter out genes that have large outlier values and only test for genes with high fold changes between samples.

Genes which fail these filters have their adjusted p-values set to NA.

**Here I set filtering to maximize the number of genes with an adjusted p-value of `r pvalcutoff` and to only look for genes with log2 fold change of at least `r lfcutoff` (or at least a `r 2^lfcutoff`-fold change).**

  
```{r allstats}
results.dr <- results(dds.ruvs, alpha=pvalcutoff, lfcThreshold = 1)
results.df <- as.data.frame(results.dr)
results.df.annot <- annotate_df(row2colnames(results.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")
  
write.csv(results.df.annot, file.path(resultsDir, "DESeq2_statistics_all_genes.csv"))
```

**[Statistics for ALL genes](../results/final/2016-06-15_project/report/DESeq2_statistics_all_genes.csv)**

## Visualizing differential expression

### Volcano plots {.tabset}
  
The plots represent each gene with a dot. The fold change (log2) is plotted on the x-axis and the adjusted p-value (log10, p-value adjusted/corrected for multiple testing) is on the y-axis. The orange shaded area or the orange colored dots denote genes that are differentially expressed (p-value < `r pvalcutoff`, fold change>`r lfcutoff`). 

As you can see, there are a large number of differentially expressed genes by these criteria.

#### with marginal density plots

```{r volcanoplot1}
results.df.volcstat <- results.df[,c("log2FoldChange", "padj")]
names(results.df.volcstat) <- c("logFC", "Adjusted.PValue")
 
volcano_density_plot(stats=results.df.volcstat, pval.cutoff=pvalcutoff, shade.colour="orange", lfc.cutoff=lfcutoff  )
```

#### basic plot

```{r volcanoplot2}
results.df.volcstat$DE <-  as.logical(results.df.volcstat$Adjusted.PValue<pvalcutoff & abs(results.df.volcstat$logFC)>lfcutoff)
ggplot(results.df.volcstat, aes(y=-log10(Adjusted.PValue), x=logFC, color=DE))+geom_point(alpha=0.5)+scale_color_manual(values=c("grey", "orange"))+theme_bw()
```
  
### MA plot

Similar to the volcano plot, the MA plot is a great way to visualize the comparative expression metrics for a two group comparison. The x–axis is the average/mean expression over all the samples and the y axis is the log2 fold change between WT and KO. The red dots represent the genes that are differentially expressed (adjusted pvalue <`r pvalcutoff`).

```{r MAplot}
DESeq2::plotMA(results.dr, ylim=c(-10,10))
```

## Summary of differential expression

Here we can see a summary of differential expression results with an adjusted pvalue of less than `r pvalcutoff` and a fold change of at least `r 2^lfcutoff` (ignore the LFC >0 and LFC <0, that's a typo from the R package, those values should be `r 2^lfcutoff`). 

```{r results}
summary(results.dr)
```

## Differentially expressed genes

These are the genes with at least a `r 2^lfcutoff`-fold change in expression at an adjusted p-value of `r pvalcutoff`.

```{r outputDE, cache=FALSE}
library(DT)
DEresults.df.annot <- subset(results.df.annot,padj<pvalcutoff & abs(log2FoldChange)>1 )

datatable(DEresults.df.annot, rownames=FALSE)
write.csv(DEresults.df.annot, file.path(resultsDir, "DESeq2_statistics_DE_genes.csv"))
```
  
**[Statisitics for Differentially Expressed genes](../results/final/2016-06-15_project/report/DESeq2_statistics_DE_genes.csv)**

### Heatmaps of top differentially expressed genes {.tabset}
- for top `r numgenes` differentially expressed genes (with adjusted pvalue of less than `r pvalcutoff` and sorted by absolute log2foldchange)
```{r DEheatmap}
# reoder stats by logFC and pull out top  results
top.results.df.annot <- DEresults.df.annot[order(abs(DEresults.df.annot$log2FoldChange)),][1:numgenes,]
# grab the ensemblids of the top reuslts
top.ensemblids <- as.character(top.results.df.annot$ensemblid)
# pull variance stabilizied data into a dataframe
rld.df <- assay(rld)
rld.df <- assay(rld[,which(!colnames((eset)) %in% c("LE1_2", "LE5_2"))])
# subset data to top genes
top.rld.df <- rld.df[row.names(rld.df) %in% top.ensemblids,]
# get gene symbols
top.rld.df.annot <- annotate_df(row2colnames(top.rld.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")
```

#### unscaled

```{r heatmap1, dev="svg", fig.height=12}
pheatmap(top.rld.df, annotation=metadata,labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8)
```

#### scaled and centered by row

```{r heatmap2, dev="svg", fig.height=12}
pheatmap(top.rld.df, annotation=metadata, labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8, scale="row")
```


# Functional Enrichment

## Gene ontology analysis
 - using [g:profiler](http://biit.cs.ut.ee/gprofiler/)  
 - for top 200 differentially expressed genes  
 - categories summarized using [Revigo](http://revigo.irb.hr/)  
 
 
  
```{r goanalysis, cache=FALSE}
library(gProfileR)
top.results.df.annot <- DEresults.df.annot[order(abs(DEresults.df.annot$log2FoldChange)),][1:200,]
top.results.df.annot <- top.results.df.annot[order(abs(top.results.df.annot$log2FoldChange), decreasing=TRUE),]
gprofiler_results  <-   gprofiler(query = as.vector(top.results.df.annot$mgi_symbol),
            organism = "mmusculus",
            ordered_query = T, 
            exclude_iea = F, 
            max_p_value = pvalcutoff, 
            max_set_size = 0,
            correction_method = "gSCS",
            domain_size = "annotated",
            custom_bg = as.vector(results.df.annot$mgi_symbol))

write.table(gprofiler_results, file.path(resultsDir, "GOresults.csv"))
```


**[Gene Ontology results](../results/final/2016-06-15_project/report/GOresults.csv)**

### Gene ontology simplification

Gene ontology enrichment analyses can yield an overwhelming number of enriched categories, many with redundant functionality. We can simplify this output by identifying the most representative subset of the terms, using metrics which measure the semantic similarity of the terms. [Revigo](http://revigo.irb.hr/)  performs such analyses, using an algortithm which forms

>  groups of highly similar GO terms, where the choice of the groups’ representatives is guided by the p-values

The algorithm takes into account the parent-child structure of the gene onotology database

> If the p-values are quite close and one term is a child node of the other, REVIGO will tend to choose the parent term

The algorithm also ytries to find more specific GO terms.

> Very general GO terms, however, are always avoided as cluster representatives ... as they tend to be uninformative

Revigo allows visualization of these representatives and their relations to the terms within their group  as a [treemap](http://en.wikipedia.org/wiki/Treemapping). Here the color depicts a grouping of related terms, the size of a block, it's pvalue from g:profiler and the large text the most representative gene ontology term for the related group. **The gene ontology categories summarized are only the biological process related categories, not the molecular function or cellular component categories.**


```{r revigo}
#write.table(gprofiler_results, "results/KO_control-gprofiler.txt", sep="\t")
allterms <- gprofiler_results$term.id
GOs <- allterms[grep('GO:', allterms)]
pvals <- gprofiler_results$p.value[grep('GO:', allterms)]
GO.pval <- cbind(gprofiler_results$term.id,gprofiler_results$p.value)

#Parameters to change
cutoff <- "0.5" #Allowed values: "0.90" "0.70" "0.50" "0.40"
organism <- "Mus musculus" #Allowed values: See organism.list below
isPValue <- "yes" #Allowed values: "yes"  "no"
whatIsBetter <- "higher" #Allowed values: "higher" "lower" "absolute" "abs_log"
measure <- "SIMREL" #Allowed values: "RESNIK" "LIN" "SIMREL" "JIANG"

#Do not change below
organism.list <- list(
        "whole UniProt"=0,
        "Homo sapiens"=9606,
        "Mus musculus"=10090,
        "Rattus norvegicus"=10116,
        "Bos taurus"=9913,
        "Gallus gallus"=9031,
        "Danio rerio"=7955,
        "Takifugu rubripes"=31033,
        "Xenopus laevis"=8355,
        "Drosophila melanogaster"=7227,
        "Caenorhabditis elegans"=6239,
        "Arabidopsis thaliana"=3702,
        "Oryza sativa"=39947,
        "Zea mays"=4577,
        "Saccharomyces cerevisiae"=4932,
        "Schizosaccharomyces pombe"=4896,
        "Dictyostelium discoideum"=44689,
        "Plasmodium falciparum"=5833,
        "Chlamydomonas reinhardtii"=3055,
        "Escherichia coli"=83333,
        "Bacillus subtilis"=1423,
        "Pseudomonas aeruginosa"=287,
        "Mycobacterium tuberculosis"=1773,
        "Mycoplasma genitalium"=2097,
        "Synechocystis sp."=1148
        )
organism.db <- as.character(organism.list[organism])

mycommand=paste('~/Scripts/Scripts/revigoR/revigo.pl -goterms', paste(GOs,collapse=","), '-gopvals', paste(pvals,collapse=","), '-cutoff', cutoff,  '-organism', organism.db, '-ispvalue', isPValue, '-whatisbetter', whatIsBetter, '-measure', measure, sep=" ")

mytempfile <- tempfile()
system2(command='perl', args=mycommand, stdout=mytempfile)
source(mytempfile)
```


# Counts Output


```{r output}

write.csv(counts, file=file.path(resultsDir, "rawdata.csv"))

write.csv(batch_ruvs_calc$normalizedCounts, file=file.path(resultsDir, "raw.batchcorrected.data.csv"))

write.csv(normed.ruvs.vs, file=file.path(resultsDir, "raw.batchcorrected.variancestabilized.data.csv"))
```




**[Raw data](../results/final/2016-06-15_project/report/rawdata.csv)**

You can use this raw data to reproduce these results using the code below this point.


**[Raw batch corrected data](../results/final/2016-06-15_project/report/raw.batchcorrected.data.csv)**

Use these values for comparing expression values between samples for individual genes.

**[Variance stabilized batch corrected data](../results/final/2016-06-15_project/report/raw.batchcorrected.variancestabilized.data.csv)**

Use these values for any clustering analyses or anything that can be affected by high variance outliers.





