---
output:
  html_document:
    toc: true
    toc_depth: 2
    theme: united
title: "Finding 450K probes around genes and probes of interest"
---

# Load libraries
```{r libraries}
library(biomaRt)
library(knitr)
library(stringr)
library(xlsx)
library(ggplot2)
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

```

# Functions
```{r functions}
bedTools.2in<-function(functionstring="bedtools intersect",bed1,bed2,opt.string="") {
  #create temp files
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) # not to use scientific notation when writing out
 
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
 
  # create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
 
  res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}
bedTools.slop <- function(functionstring="bedtools slop", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  # create the command string and call the command using system()
  command <- paste(functionstring,"-i",a.file,"-g", genomefile, opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

bedTools.flank <- function(functionstring="bedtools flank -s", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out  
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  
  # create the command string and call the command using system()
  command <- paste(functionstring,"-g", genomefile, opt.string, "-i", a.file, ">", out, sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

```

# Set directories

```{r directories}
baseDir <- "~/projects/ab_rrbs_comparison/"
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
```

# Load data
- using mitochondrial data from the [microproteome site](http://www.mitoproteome.org/data/proteome/index.html)

```{r dataload}
annots.450K <- read.csv(file.path(dataDir, "14_April_2014_Updated_humanmethylation450_15017482_v1-2.csv"), skip=7)
annots.refseq <- read.delim(file.path(dataDir, "refseq.genes.bed"), header=F)
points.of.interest <- read.xlsx(file.path(dataDir, "26_Jan_2014_Udpated_450K_Results_for_John.xlsx"), sheetIndex = 1)
points.of.interest <- points.of.interest[!is.na(points.of.interest$Type.of.Locus),]
```

# setup biomart

```{r setupbiomart}
library(biomaRt)
# use GRch37 build
ensemblmart <-  useMart("ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl", host="grch37.ensembl.org")
attributes <- listAttributes(ensemblmart)
filters <- listFilters(ensemblmart)
```

# Data munging of regions of interest

## separate by class

```{r munging}
genes.of.interest <- subset(points.of.interest, Type.of.Locus=="Gene")
CpGs.of.interest <- subset(points.of.interest, Type.of.Locus=="Non-Gene")
```

## extract genes of interest and get corresponding regions from biomart
- first add in new gene of interest "FHAD1"
- used start_position and end_position attributes from biomart, which will include all isoforms

```{r genestobed}
genes.of.interest <- unlist(strsplit(as.character(genes.of.interest$UCSC_RefGene_Name), ";"))
genes.of.interest <- unique(genes.of.interest)
genes.of.interest <- c(genes.of.interest, "FHAD1", "IER3", "ALPPL2")


genes.of.interest.bed <- getBM(genes.of.interest, filters="hgnc_symbol", attributes=c("chromosome_name","start_position", "end_position", "strand", "hgnc_symbol" ), mart=ensemblmart)
genes.of.interest.bed$score=100

names(genes.of.interest.bed) <- c("chr", "start", "end", "strand", "name", "score")
genes.of.interest.bed <- genes.of.interest.bed[,c("chr", "start", "end", "name", "score", "strand")]

genes.of.interest.bed$strand <- ifelse(genes.of.interest.bed$strand=="1", "+", "-")
genes.of.interest.bed$chr <- paste("chr", genes.of.interest.bed$chr, sep="")

genes.of.interest.bed <- genes.of.interest.bed[!grepl("CHR", as.character(genes.of.interest.bed$chr)),]
```

## convert CpGs of interest data into bed format using data supplied by Juan (from Illumina manifest)

```{r cpgstobed}
CpGs.of.interest.bed <- CpGs.of.interest[, c("CHR", "MAPINFO", "Name")]
names(CpGs.of.interest.bed) <- c("chr", "end", "name")
CpGs.of.interest.bed$start <- CpGs.of.interest.bed$end -1
CpGs.of.interest.bed$strand <- "+"
CpGs.of.interest.bed$score <- 100
CpGs.of.interest.bed <- CpGs.of.interest.bed[,c("chr", "start", "end", "name", "score", "strand")]
CpGs.of.interest.bed$chr <- paste("chr", CpGs.of.interest.bed$chr, sep="")
```

# Get coordinates of regions around  genes
5kb upstream of the TSS and 5Kb downstream of transcript end

```{r getpromoters}
genes.pos.flanked <- bedTools.slop(bed1 = genes.of.interest.bed, opt.string = "-b 5000", genomefile = file.path(metaDir, "hg19.genome"))
genes.pos.upstream <- bedTools.flank(bed1 = genes.of.interest.bed, opt.string = "-l 5000 -r 0", genomefile = file.path(metaDir, "hg19.genome"))
genes.pos.downstream <- bedTools.flank(bed1 = genes.of.interest.bed, opt.string = "-l 0 -r 5000", genomefile = file.path(metaDir, "hg19.genome"))
```

# Find 450K probes around genes and CpGs

## Extract coordinates (Human Genome build GRCh37) of 450K probes from annotation file and convert to BED format

```{r 450Kextract}
bed.450K <- annots.450K[,c("CHR", "MAPINFO", "Name")]
# adjust to zero-based coordinate format
bed.450K$start <- bed.450K$MAPINFO-1
names(bed.450K) <- c("chr", "end", "name", "start")
bed.450K$chr <- paste("chr", bed.450K$chr, sep="")
bed.450K <- bed.450K[,c("chr", "start", "end", "name")]
# subset to CpG probes
bed.450K <- bed.450K[grep("cg", bed.450K$name),]
```

## Overlap 450K probe coordinates with flanks of gene coordinates

```{r overlap}
probes450K.in.flanked.genes <- bedTools.2in(bed1=bed.450K,bed2=genes.pos.flanked,opt.string="-wo")
probes450K.in.flanked.genes <- probes450K.in.flanked.genes[,c(1,2,3,4,8)]
names(probes450K.in.flanked.genes) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.flanked.genes <- probes450K.in.flanked.genes[with(probes450K.in.flanked.genes, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.flanked.genes <- probes450K.in.flanked.genes[!duplicated(probes450K.in.flanked.genes),]
probes450K.in.flanked.genes <- merge(probes450K.in.flanked.genes, annots.450K, by.x="probeID", by.y="IlmnID")
probes450K.in.flanked.genes <- probes450K.in.flanked.genes[,c(5:ncol(probes450K.in.flanked.genes))]
names(probes450K.in.flanked.genes) <- sub("^Name$", "IllmnID", names(probes450K.in.flanked.genes))

##########################
##########################
probes450K.in.upstream.genes <- bedTools.2in(bed1=bed.450K,bed2=genes.pos.upstream,opt.string="-wo")
probes450K.in.upstream.genes <- probes450K.in.upstream.genes[,c(1,2,3,4,8)]
names(probes450K.in.upstream.genes) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.upstream.genes <- probes450K.in.upstream.genes[with(probes450K.in.upstream.genes, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.upstream.genes <- probes450K.in.upstream.genes[!duplicated(probes450K.in.upstream.genes),]
probes450K.in.upstream.genes <- merge(probes450K.in.upstream.genes, annots.450K, by.x="probeID", by.y="IlmnID")
probes450K.in.upstream.genes <- probes450K.in.upstream.genes[,c(5:ncol(probes450K.in.upstream.genes))]
names(probes450K.in.upstream.genes) <- sub("^Name$", "IllmnID", names(probes450K.in.upstream.genes))

##########################
##########################

probes450K.in.downstream.genes <- bedTools.2in(bed1=bed.450K,bed2=genes.pos.downstream,opt.string="-wo")
probes450K.in.downstream.genes <- probes450K.in.downstream.genes[,c(1,2,3,4,8)]
names(probes450K.in.downstream.genes) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.downstream.genes <- probes450K.in.downstream.genes[with(probes450K.in.downstream.genes, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.downstream.genes <- probes450K.in.downstream.genes[!duplicated(probes450K.in.downstream.genes),]
probes450K.in.downstream.genes <- merge(probes450K.in.downstream.genes, annots.450K, by.x="probeID", by.y="IlmnID")
probes450K.in.downstream.genes <- probes450K.in.downstream.genes[,c(5:ncol(probes450K.in.downstream.genes))]
names(probes450K.in.downstream.genes) <- sub("^Name$", "IllmnID", names(probes450K.in.downstream.genes))

##########################
##########################

probes450K.in.genes <- bedTools.2in(bed1=bed.450K,bed2=genes.of.interest.bed,opt.string="-wo")
probes450K.in.genes <- probes450K.in.genes[,c(1,2,3,4,8)]
names(probes450K.in.genes) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.genes <- probes450K.in.genes[with(probes450K.in.genes, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.genes <- probes450K.in.genes[!duplicated(probes450K.in.genes),]
probes450K.in.genes <- merge(probes450K.in.genes, annots.450K, by.x="probeID", by.y="IlmnID")
probes450K.in.genes <- probes450K.in.genes[,c(5:ncol(probes450K.in.genes))]
names(probes450K.in.genes) <- sub("^Name$", "IllmnID", names(probes450K.in.genes))

```


## Find other 450K probes within 5kb (either side) of probes of interest

```{r flankprobes}
# get coordinates flanking CpGs of interest
CpGs.pos.flanked <- bedTools.slop(bed1 = CpGs.of.interest.bed, opt.string = "-b 5000", genomefile = file.path(metaDir, "hg19.genome"))
# find other CpGs within flanks
probes450K.around.CpGs <- bedTools.2in(bed1=bed.450K,bed2=CpGs.pos.flanked,opt.string="-wo")
probes450K.around.CpGs <- probes450K.around.CpGs[,c(1,2,3,4,8)]

names(probes450K.around.CpGs) <- c("chr", "start", "end", "probe.450K", "original.probe.450K.of.interest")

# merge with annotations

probes450K.around.CpGs <- merge(probes450K.around.CpGs, annots.450K, by.x="probe.450K", by.y="IlmnID")
probes450K.around.CpGs <- probes450K.around.CpGs[,c(5:ncol(probes450K.around.CpGs))]
```

# Output results

```{r output}
setwd(resultsDir)
write.xlsx2(probes450K.in.flanked.genes, file="probes450K.in.5kb.gene.flanks.xlsx")
write.xlsx2(probes450K.in.upstream.genes, file="probes450K.in.5kb.upstream.of.genes.xlsx")
write.xlsx2(probes450K.in.downstream.genes, file="probes450K.in.5kb.downstream.of.genes.xlsx")
write.xlsx2(probes450K.in.genes, file="probes450K.in.genes.xlsx")
write.xlsx2(probes450K.around.CpGs, file="probes450K.within.5kb.flanking.probes.of.interest.xlsx")
```

[Excel file with probes upstream of genes](../results/probes450K.in.5kb.upstream.of.genes.xlsx)
[Excel file with probes downstream of genes](../results/probes450K.in.5kb.downstream.of.genes.xlsx)
[Excel file with probes in genes](../results/probes450K.in.genes.xlsx)
[Excel file with probes within 5kb of genes (including inside genes)](../results/probes450K.in.5kb.gene.flanks.xlsx)
[Excel file with probes within 5kb of non-gene CpGs](../results/probes450K.within.5kb.flanking.probes.of.interest.xlsx)

