---
bibliography: "references.bib"
csl: "bioinformatics.csl"
title: "Whangbo - Unpaired treatment response RNAseq Analysis"
output:
  html_document:
    code_folding: hide
    theme: yeti
    toc: yes
    toc_float: true
    toc_depth: 4
    number_sections: true
---

```{r setup, echo=FALSE}
library("knitcitations")
cleanbib()
options("citation_format" = "pandoc")

clientname="Jennifer Whangbo"
clientemail="Jennifer.Whangbo@childrens.harvard.edu"
labPI="Ritz"
lablocation="Dana Farber"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

---

RNAseq analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods

Discarded any low quality samples, but kept unmatched pre and post treatment for the same patient. Differential expression was analyzed with DESeq2.

---

# Setup

```{r functions, echo=FALSE}
check_install = function(packages) {
   not_installed = setdiff(packages, rownames(installed.packages()))
   if(length(not_installed) > 0) {
      write(paste("The libraries", not_installed, "are not installed, aborting.",sep=" "), stdout())
      stop()
   }
}

sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}

subset_tximport = function(txi, rows, columns) {
    txi$counts = txi$counts[rows, columns]
    txi$abundance = txi$abundance[rows, columns]
    txi$length = txi$length[rows, columns]
    return(txi)
}

heatmap_fn = function(data, ...) {
  pheatmap(data, annotation=metadata, clustering_method = "ward.D2",
           clustering_distance_cols = "correlation", ...)
}

pca_plot = function(comps, nc1, nc2, colorby, size, title) {
   c1str = paste0("PC", nc1)
   c2str = paste0("PC", nc2)
  ggplot(comps, aes_string(c1str, c2str, color=colorby)) +
    geom_point(size=size) + theme_bw() +
    xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
    ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance")) +
    theme(legend.position="bottom")+
    theme(legend.text=element_text(size=6), legend.title=element_blank())+
    ggtitle(title)
}


annotate_df2 <- function (df, df_ensemblid_header, biomart_ensembl_dataset, biomart_ensemblid_filter, biomart_genesymbol_attribute, biomart_host) {
  require(biomaRt)
  ensembl = useMart("ENSEMBL_MART_ENSEMBL", dataset = biomart_ensembl_dataset, host = biomart_host)
  annot.df = getBM(attributes = c(biomart_ensemblid_filter, biomart_genesymbol_attribute, "description"), filters = c(biomart_ensemblid_filter), values = as.character(df[, df_ensemblid_header]), mart = ensembl)
  m = merge(df, annot.df, by.x = df_ensemblid_header, by.y = biomart_ensemblid_filter,all.x = T)
  return(m)
}

```


```{r load-libraries, cache=FALSE, echo=FALSE, results="hide"}
library(lubripack)
lubripack("ggplot2", "reshape", "gplots", "edgeR", "CHBUtils", "pheatmap","DESeq2", "tximport", "DT", "DEGreport", "dplyr", "rio", "rdrop2", "tximport", "gridExtra")
```

```{r variables}
if (file.exists("/home/jnh7/projects/jw_rnaseq")){
  baseDir <- "/home/jnh7/projects/jw_rnaseq"
} else {
baseDir <- "~/Work/projects/jw_rnaseq"
}
dropboxfiledir <- "/Public/Consults/jw_rnaseq/files"
project_summary = file.path(baseDir, "data/2016-11-02_bcbio/project-summary.csv")
counts_file = file.path(baseDir, "data/2016-11-02_bcbio/combined.counts")
tx2genes_file = file.path(baseDir, "data/2016-11-02_bcbio/tx2gene.csv")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
pvalue.cutoff <- 0.05
lfc.cutoff <- 1

drop_auth()
```

```{r set-random-seed}
# set seed for reproducibility
set.seed(1454944673)
```

## Sample metadata

```{r data_metadata, echo=FALSE}
summarydata = read.table(project_summary, header=TRUE, sep=",")
summarydata = summarydata[,colSums(is.na(summarydata)) < nrow(summarydata)]
# handle newer bcbio-nextgen runs that use description as the key
if("description" %in% colnames(summarydata)) {
  rownames(summarydata) = summarydata$description
  summarydata$Name = rownames(summarydata)
  summarydata$description = NULL
} else {
  rownames(summarydata) = summarydata$Name
  # summarydata$Name = NULL
}
summarydata = summarydata[order(rownames(summarydata)),]
sample_dirs = file.path(baseDir, "data", rownames(summarydata))
salmon_files = file.path(sample_dirs, "salmon", "quant.sf")
sailfish_files = file.path(sample_dirs, "sailfish", "quant.sf")
new_sailfish = file.path(sample_dirs, "sailfish", "quant", "quant.sf")
new_salmon = file.path(sample_dirs, "salmon", "quant", "quant.sf")
if (file.exists(salmon_files[1])) {
  sf_files = salmon_files
} else if (file.exists(sailfish_files[1])) {
  sf_files = sailfish_files
}  else if (file.exists(new_sailfish[1])) {
  sf_files = new_sailfish
}  else if (file.exists(new_salmon[1])) {
  sf_files = new_salmon
}
names(sf_files) = rownames(summarydata)
tx2gene = read.table(tx2genes_file, sep=",", row.names=NULL, header=FALSE)
txi.salmon = tximport(sf_files, type="salmon", tx2gene=tx2gene, reader=readr::read_tsv, countsFromAbundance="lengthScaledTPM")
counts = round(data.frame(txi.salmon$counts, check.names=FALSE))
counts = counts[, order(colnames(counts)), drop=FALSE]
colnames(counts) = gsub(".counts", "", colnames(counts))


# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Average.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads")
summarydata[,"Fragment.Length.Mean"] = summarydata$Average.insert.size
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
metadata = metadata[, colSums(is.na(metadata)) < nrow(metadata), drop=FALSE]
metadata$samplecode <- NULL

exclude.samples <- c("nr_pre_nk_rl", "nr_wk4_tcon_rl")
metadata <- metadata[!(row.names(metadata) %in% exclude.samples),]
summarydata <- summarydata[!(row.names(summarydata) %in% exclude.samples),]
counts <- counts[,!(names(counts) %in% exclude.samples)]

y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)


sanitize_datatable(summarydata, style='bootstrap')
```

# Unsupervised Clustering

## Correlation heatmaps of TMM-normalized counts

```{r newcovars}
metadata$rRNA_rate <- summarydata$rRNA_rate
metadata$Exonic_rate <- summarydata$Exonic.Rate
```

### Tcon cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.tcon, eval=ncol(counts) > 1}
celltype="Tcon"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.tcon, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```

### Treg cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.treg, eval=ncol(counts) > 1}
celltype="Treg"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.treg, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```

### CD8 cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.cd8, eval=ncol(counts) > 1}
celltype="CD8"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.cd8, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```

### NK cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.nk, eval=ncol(counts) > 1}
celltype="NK"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.nk, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```


## PCA plots

### Tcon cells {.tabset}
`r celltype="Tcon"`

```{r pca.tcon, eval=ncol(counts) > 1}
cellindices <- which(metadata$celltype==celltype)
counts.sub <- counts[,cellindices]
summarydata.sub <- summarydata[cellindices,]
dds = DESeqDataSetFromMatrix(countData=counts.sub, colData=summarydata.sub, design=~Name)
vst = varianceStabilizingTransformation(dds)
pca_loadings = function(object, ntop=500) {
  rv <- matrixStats::rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
      length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  names(percentVar) = colnames(pca$x)
  pca$percentVar = percentVar
  return(pca)}
pc = pca_loadings(vst)
comps = data.frame(pc$x)
comps$Name = rownames(comps)
library(dplyr)
comps = comps %>% left_join(summarydata, by=c("Name"="Name"))
```

#### PC1 vs. PC2
```{r PC1-PC2.tcon, fig.height=3}
p1 <- pca_plot(comps, 1, 2, "patientid", size=3, title="PatientID")
p2 <- pca_plot(comps, 1, 2, "status", size=3, title="Status")
p3 <- pca_plot(comps, 1, 2, "timepoint", size=3, title="Timepoint")

grid.arrange(p1, p2, p3, ncol=3, nrow=1)
```

#### PC3 vs. PC4
```{r PC3-PC4.tcon, fig.height=3}
p1 <- pca_plot(comps, 3, 4, "patientid", size=3, title="PatientID")
p2 <- pca_plot(comps, 3, 4, "status", size=3, title="Status")
p3 <- pca_plot(comps, 3, 4, "timepoint", size=3, title="Timepoint")
grid.arrange(p1, p2, p3, ncol=3, nrow=1)

```

#### PC5 vs. PC6
```{r PC5-PC6.tcon, fig.height=3}
p1 <- pca_plot(comps, 5, 6, "patientid", size=3, title="PatientID")
p2 <- pca_plot(comps, 5, 6, "status", size=3, title="Status")
p3 <- pca_plot(comps, 5, 6, "timepoint", size=3, title="Timepoint")
grid.arrange(p1, p2, p3, ncol=3, nrow=1)

```

#### Variance explained by component
```{r component-loadings.tcon, fig.height=3}
ggplot(data.frame(component=reorder(names(pc$percentVar), -pc$percentVar),
                  percent_var=pc$percentVar),
       aes(component, percent_var)) + geom_bar(stat='identity') +
  ylab("percent of total variation") + xlab("") + theme_bw()
```


### Treg cells {.tabset}
`r celltype="Treg"`

```{r pca.treg, ref.label='pca.tcon', eval=ncol(counts) > 1, fig.height=3}
```

#### PC1 vs. PC2
```{r PC1-PC2.treg, ref.label='PC1-PC2.tcon', fig.height=3}
```

#### PC3 vs. PC4
```{r PC3-PC4.treg, ref.label='PC3-PC4.tcon', fig.height=3}
```

#### PC5 vs. PC6
```{r PC5-PC6.treg, ref.label='PC5-PC6.tcon', fig.height=3}
```

#### Variance explained by component
```{r component-loadings.treg, ref.label='component-loadings.tcon', fig.height=3}
```


### CD8 cells {.tabset}
`r celltype="CD8"`

```{r pca.cd8, ref.label='pca.tcon', eval=ncol(counts) > 1, fig.height=3}
```

#### PC1 vs. PC2
```{r PC1-PC2.cd8, ref.label='PC1-PC2.tcon', fig.height=3}
```

#### PC3 vs. PC4
```{r PC3-PC4.cd8, ref.label='PC3-PC4.tcon', fig.height=3}
```

#### PC5 vs. PC6
```{r PC5-PC6.cd8, ref.label='PC5-PC6.tcon', fig.height=3}
```

#### Variance explained by component
```{r component-loadings.cd8, ref.label='component-loadings.tcon', fig.height=3}
```


### NK cells {.tabset}
`r celltype="NK"`

```{r pca.nk, ref.label='pca.tcon', eval=ncol(counts) > 1, fig.height=3}
```

#### PC1 vs. PC2
```{r PC1-PC2.nk, ref.label='PC1-PC2.tcon', fig.height=3}
```

#### PC3 vs. PC4
```{r PC3-PC4.nk, ref.label='PC3-PC4.tcon', fig.height=3}
```

#### PC5 vs. PC6
```{r PC5-PC6.nk, ref.label='PC5-PC6.tcon', fig.height=3}
```

#### Variance explained by component
```{r component-loadings.nk, ref.label='component-loadings.tcon', fig.height=3}
```

---

# Differential expression

- using DESeq2 `r citep("10.1186/s13059-014-0550-8")`

## DESeq2 Setup {.tabset}

```{r }
library(DEGreport)
library(vsn)
```

As with the paired analysis, normalized  and calculated dispersions for all celltypes separately.

### Tcon cells

```{r setcelltype.tcon}
celltype="Tcon"
```

```{r tcondeseqsetup}
counts.sub <- counts[rowSums(counts>0)>1,summarydata$celltype==celltype]
summarydata.sub <- summarydata[summarydata$celltype==celltype,]
txi.salmon.sub = subset_tximport(txi.salmon, rownames(counts.sub), colnames(counts.sub))
dds.sub = DESeqDataSetFromTximport(txi.salmon.sub, colData=summarydata.sub, design=~1)
# model matrix is not full rank, will refactor later
geoMeans.sub = apply(counts.sub, 1, function(row) {
  if (all(row == 0)) 0 
  else 
    exp(mean(log(row[row != 0])))
})
dds.sub = estimateSizeFactors(dds.sub, geoMeans=geoMeans.sub)
dds.sub = estimateDispersions(dds.sub)
```

#### Effect of variance stabilization

```{r tconvstsetup}
notAllZero <- (rowSums(counts(dds.sub))>0)
rld <- rlog(dds.sub)
vsd <- varianceStabilizingTransformation(dds.sub)
rlogMat <- assay(rld)
vstMat <- assay(vsd)
```

```{r tconvstplot, fig.keep="last",fig.height=2.5}
p1 <- meanSdPlot(log2(counts(dds.sub,normalized=TRUE)[notAllZero,] + 1))$gg + ggtitle("log") + theme(legend.position="bottom")
p2 <- meanSdPlot(assay(rld[notAllZero,]))$gg + ggtitle("rlog") + theme(legend.position="bottom")
p3 <- meanSdPlot(assay(vsd[notAllZero,]))$gg + ggtitle("vst") + theme(legend.position="bottom")
grid.arrange(p1,p2,p3, ncol=3)
```

#### Dispersion estimates

```{r tcondispersionestimate}
plotDispEsts(dds.sub)
```

```{r tcondds}
colnames(dds.sub)
dds.Tcon = dds.sub
```

### Treg cells

```{r setcelltype.treg}
celltype="Treg"
```

```{r tregdeseqsetup, ref.label="tcondeseqsetup"}
```

#### Effect of variance stabilization

```{r tregvstsetup, ref.label="tconvstsetup"}
```

```{r tregvstplot, ref.label="tconvstplot",fig.keep="last",fig.height=2.5}
```

#### Dispersion estimates

```{r tregdispersionestimate, ref.label="tcondispersionestimate"}
```

```{r tregdds}
colnames(dds.sub)
dds.Treg = dds.sub
```


### CD8 cells

```{r setcelltype.cd8}
celltype="CD8"
```

```{r cd8deseqsetup, ref.label="tcondeseqsetup"}
```

#### Effect of variance stabilization

```{r cd8vstsetup, ref.label="tconvstsetup"}
```

```{r cd8vstplot, ref.label="tconvstplot", fig.keep="last",fig.height=2.5}
```

#### Dispersion estimates

```{r cd8dispersionestimate, ref.label="tcondispersionestimate"}
```

```{r cd8dds}
colnames(dds.sub)
dds.CD8 = dds.sub
```

### NK cells

```{r setcelltype.nk}
celltype="NK"
```

```{r nkdeseqsetup, ref.label="tcondeseqsetup"}
```

#### Effect of variance stabilization

```{r nkvstsetup, ref.label="tconvstsetup"}
```

```{r nkvstplot, ref.label="tconvstplot", fig.keep="last",fig.height=2.5}
```

#### Dispersion estimates

```{r nkdispersionestimate, ref.label="tcondispersionestimate"}
```

```{r nkdds}
colnames(dds.sub)
dds.NK = dds.sub
```


## Responder versus Non-responder before treatment {.tabset}

```{r responder_vs_nonrespondersetup_before}
design = ~status
```

### Tcon

```{r setcelltype.tcon.responder_vs_nonresponder_before}
celltype="Tcon"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.tcon_before, results='hide'}
counts.sub <- counts[rowSums(counts>0)>1,(summarydata$celltype==celltype & summarydata$timepoint=="pretreatment")]
summarydata.sub <- summarydata[(summarydata$celltype==celltype & summarydata$timepoint=="pretreatment"),]
txi.salmon.sub = subset_tximport(txi.salmon, rownames(counts.sub), colnames(counts.sub))
dds.sub = DESeqDataSetFromTximport(txi.salmon.sub, colData=summarydata.sub, design=design)

geoMeans.sub = apply(counts.sub, 1, function(row) {
  if (all(row == 0)) 0 
  else 
    exp(mean(log(row[row != 0])))
})
dds.sub = estimateSizeFactors(dds.sub, geoMeans=geoMeans.sub)
dds.sub = estimateDispersions(dds.sub)

dds.celltype <- get(paste("dds", celltype, sep="."))

# put in dispersions and size factors from full set
#identical(row.names(counts(dds.celltype)), row.names(counts(dds.sub)))

dispersions(dds.sub) <- dispersions(dds.celltype)
sizeFactors(dds.sub) <- sizeFactors(dds.celltype)[names(sizeFactors(dds.sub))]

dds.sub = DESeq(dds.sub)
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.tcon_before}
stats <- results(dds.sub)[,c("log2FoldChange", "padj")] %>% as.data.frame()
names(stats) <- c("logFC", "Adjusted.Pvalue")
volcano_density_plot(stats, lfc.cutoff=lfc.cutoff, pval.cutoff=pvalue.cutoff)
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.tcon_before, results='asis'}
res =  results(dds.sub)
ymax = max(res$log2FoldChange, na.rm=TRUE)
ymin = min(res$log2FoldChange, na.rm=TRUE)
plotMA(res, ylim=c(ymin, ymax))
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.tcon_before, results='hide'}
out_df = as.data.frame(res)

out_df$id = rownames(out_df)
out_df = out_df[, c("id", colnames(out_df)[colnames(out_df) != "id"])]

out_df <- annotate_df2(df=out_df, df_ensemblid_header = "id", biomart_ensembl_dataset= 'hsapiens_gene_ensembl',biomart_ensemblid_filter = "ensembl_gene_id", biomart_genesymbol_attribute = 'hgnc_symbol', biomart_host="www.ensembl.org" )

norm.counts.celltype <- counts(dds.sub, normalized=TRUE) %>% row2colnames("id") %>% tbl_df()

out_df <-   out_df %>% tbl_df() %>% inner_join(., norm.counts.celltype, by="id") %>% arrange(padj)

tempoutfile <-file.path(tempdir(), paste("pretreatment.responder_vs_nonresponders", celltype, "xlsx", sep="."))
rio::export(out_df, file=tempoutfile)
drop_upload(file=tempoutfile, dest=dropboxfiledir, overwrite=TRUE)
shared_file <-  drop_share(file.path(dropboxfiledir, paste("pretreatment.responder_vs_nonresponders", celltype, "xlsx", sep=".")))$url
```

```{r DESeq-table.responder_vs_nonresponder.tcon_before, results='asis'}
sig_genes = subset(out_df, padj < pvalue.cutoff)
DT::datatable(sig_genes, caption=paste("Significant genes by adjusted pvalue<", pvalue.cutoff, sep=""), rownames=FALSE)
```

```{r setupmetadata}
meta.plotme <- metadata %>% row2colnames("samplecode") %>% tbl_df() 
```


```{r DEseq.plotDE.responder_vs_nonresponder.tcon_before,  fig.height=8}
if(nrow(sig_genes)>10) {
  plotme <- sig_genes[1:10,] } else {
    plotme <- sig_genes
  }

plotme <-   plotme %>%   
    dplyr::select(., -baseMean, -log2FoldChange, -lfcSE, -stat, -padj, -pvalue,-description) %>%
    mutate(., gene = ifelse(hgnc_symbol=="", id, hgnc_symbol)) %>% 
    dplyr::select(., -id, -hgnc_symbol) %>%
    tidyr::gather(., samplecode, value, -gene) %>%
    inner_join(., meta.plotme,by="samplecode" ) %>%
    mutate(., value=as.numeric(value)) 
  ggplot(plotme, aes(x=status, y=value, color=status)) +
    geom_boxplot() +
    facet_wrap(~gene, ncol=2, scales="free_y") + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    theme( legend.position="none")+
    labs(
      title = "Normalized expression levels for selected DE genes",
      subtitle = "(top 10 genes by adjusted pvalue)",
      caption= "Genes without symbols were assigned their Ensembl Id"
    )
```

#### Download Results

**[Results](`r shared_file`)**

### Treg

```{r setcelltype.treg.responder_vs_nonresponder_before}
celltype="Treg"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.treg_before, ref.label='deseq2-expression-analysis.responder_vs_nonresponder.tcon_before', results='hide'}
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.treg_before, ref.label="DESeq.volc.responder_vs_nonresponder.tcon_before"}
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.treg_before, ref.label='DESeq-MA.responder_vs_nonresponder.tcon_before', results='asis'}
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.treg_before, ref.label='DESeq-output.responder_vs_nonresponder.tcon_before', results='hide'}
```

```{r DESeq-table.responder_vs_nonresponder.treg_before, ref.label='DESeq-table.responder_vs_nonresponder.tcon_before', results='asis'}
```

```{r DEseq.plotDE.responder_vs_nonresponder.treg_before, ref.label='DEseq.plotDE.responder_vs_nonresponder.tcon_before', fig.height=8}
```

#### Download Results

**[Results](`r shared_file`)**

### CD8

```{r setcelltype.cd8.responder_vs_nonresponder_before}
celltype="CD8"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.cd8_before, ref.label='deseq2-expression-analysis.responder_vs_nonresponder.tcon_before', results='hide'}
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.cd8_before, ref.label="DESeq.volc.responder_vs_nonresponder.tcon_before"}
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.cd8_before, ref.label='DESeq-MA.responder_vs_nonresponder.tcon_before', results='asis'}
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.cd8_before, ref.label='DESeq-output.responder_vs_nonresponder.tcon_before', results='hide'}
```

```{r DESeq-table.responder_vs_nonresponder.cd8_before, ref.label='DESeq-table.responder_vs_nonresponder.tcon_before', results='asis'}
```

```{r DEseq.plotDE.responder_vs_nonresponder.cd8_before, ref.label='DEseq.plotDE.responder_vs_nonresponder.tcon_before', fig.height=8}
```

#### Download Results

**[Results](`r shared_file`)**

### NK

```{r setcelltype.nk.responder_vs_nonresponder_before}
celltype="NK"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.nk_before, ref.label='deseq2-expression-analysis.responder_vs_nonresponder.tcon_before', results='hide'}
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.nk_before, ref.label="DESeq.volc.responder_vs_nonresponder.tcon_before"}
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.nk_before, ref.label='DESeq-MA.responder_vs_nonresponder.tcon_before', results='asis'}
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.nk_before, ref.label='DESeq-output.responder_vs_nonresponder.tcon_before', results='hide'}
```

```{r DESeq-table.responder_vs_nonresponder.nk_before, ref.label='DESeq-table.responder_vs_nonresponder.tcon_before', results='asis'}
```

```{r DEseq.plotDE.responder_vs_nonresponder.nk_before, ref.label='DEseq.plotDE.responder_vs_nonresponder.tcon_before', fig.height=8}
```

#### Download Results

**[Results](`r shared_file`)**








## Responder versus Non-responder after treatment {.tabset}


```{r responder_vs_nonrespondersetup}
design = ~status
```

### Tcon

```{r setcelltype.tcon.responder_vs_nonresponder}
celltype="Tcon"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.tcon, results='hide'}
counts.sub <- counts[rowSums(counts>0)>1,(summarydata$celltype==celltype & summarydata$timepoint=="week4")]
summarydata.sub <- summarydata[(summarydata$celltype==celltype & summarydata$timepoint=="week4"),]
txi.salmon.sub = subset_tximport(txi.salmon, rownames(counts.sub), colnames(counts.sub))
dds.sub = DESeqDataSetFromTximport(txi.salmon.sub, colData=summarydata.sub, design=design)

geoMeans.sub = apply(counts.sub, 1, function(row) {
  if (all(row == 0)) 0 
  else 
    exp(mean(log(row[row != 0])))
})
dds.sub = estimateSizeFactors(dds.sub, geoMeans=geoMeans.sub)
dds.sub = estimateDispersions(dds.sub)

dds.celltype <- get(paste("dds", celltype, sep="."))

# put in dispersions and size factors from full set
#identical(row.names(counts(dds.celltype)), row.names(counts(dds.sub)))

dispersions(dds.sub) <- dispersions(dds.celltype)
sizeFactors(dds.sub) <- sizeFactors(dds.celltype)[names(sizeFactors(dds.sub))]

dds.sub = DESeq(dds.sub)
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.tcon}
stats <- results(dds.sub)[,c("log2FoldChange", "padj")] %>% as.data.frame()
names(stats) <- c("logFC", "Adjusted.Pvalue")
volcano_density_plot(stats, lfc.cutoff=lfc.cutoff, pval.cutoff=pvalue.cutoff)
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.tcon, results='asis'}
res =  results(dds.sub)
ymax = max(res$log2FoldChange, na.rm=TRUE)
ymin = min(res$log2FoldChange, na.rm=TRUE)
plotMA(res, ylim=c(ymin, ymax))
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.tcon, results='hide'}
out_df = as.data.frame(res)

out_df$id = rownames(out_df)
out_df = out_df[, c("id", colnames(out_df)[colnames(out_df) != "id"])]

out_df <- annotate_df2(df=out_df, df_ensemblid_header = "id", biomart_ensembl_dataset= 'hsapiens_gene_ensembl',biomart_ensemblid_filter = "ensembl_gene_id", biomart_genesymbol_attribute = 'hgnc_symbol', biomart_host="www.ensembl.org" )

norm.counts.celltype <- counts(dds.sub, normalized=TRUE) %>% row2colnames("id") %>% tbl_df()

out_df <-   out_df %>% tbl_df() %>% inner_join(., norm.counts.celltype, by="id") %>% arrange(padj)

tempoutfile <-file.path(tempdir(), paste("week4.responder_vs_nonresponders", celltype, "xlsx", sep="."))
rio::export(out_df, file=tempoutfile)
drop_upload(file=tempoutfile, dest=dropboxfiledir, overwrite=TRUE)
shared_file <-  drop_share(file.path(dropboxfiledir, paste("week4.responder_vs_nonresponders", celltype, "xlsx", sep=".")))$url
```

```{r DESeq-table.responder_vs_nonresponder.tcon, results='asis'}
sig_genes = subset(out_df, padj < pvalue.cutoff)
DT::datatable(sig_genes, caption=paste("Significant genes by adjusted pvalue<", pvalue.cutoff, sep=""), rownames=FALSE)
```

```{r DEseq.plotDE.responder_vs_nonresponder.tcon,  fig.height=8}
if(nrow(sig_genes)>10) {
  plotme <- sig_genes[1:10,] } else {
    plotme <- sig_genes
  }

plotme <- plotme %>%   dplyr::select(., -baseMean, -log2FoldChange, -lfcSE, -stat, -padj, -pvalue,-description) %>%
    mutate(., gene = ifelse(hgnc_symbol=="", id, hgnc_symbol)) %>% 
    dplyr::select(., -id, -hgnc_symbol) %>%
    tidyr::gather(., samplecode, value, -gene) %>%
    inner_join(., meta.plotme,by="samplecode" ) %>%
    mutate(., value=as.numeric(value)) 
ggplot(plotme, aes(x=status, y=value, color=status))+
  geom_boxplot()+
  facet_wrap(~gene, ncol=2, scales="free_y")+ 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  theme( legend.position="none")+
  labs(
    title = "Normalized expression levels for selected DE genes",
    subtitle = "(top 10 genes by adjusted pvalue)",
    caption= "Genes without symbols were assigned their Ensembl Id"
  )
```

#### Download Results

**[Results](`r shared_file`)**

### Treg

```{r setcelltype.treg.responder_vs_nonresponder}
celltype="Treg"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.treg, ref.label='deseq2-expression-analysis.responder_vs_nonresponder.tcon', results='hide'}
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.treg, ref.label="DESeq.volc.responder_vs_nonresponder.tcon"}
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.treg, ref.label='DESeq-MA.responder_vs_nonresponder.tcon', results='asis'}
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.treg, ref.label='DESeq-output.responder_vs_nonresponder.tcon', results='hide'}
```

```{r DESeq-table.responder_vs_nonresponder.treg, ref.label='DESeq-table.responder_vs_nonresponder.tcon', results='asis'}
```

```{r DEseq.plotDE.responder_vs_nonresponder.treg, ref.label='DEseq.plotDE.responder_vs_nonresponder.tcon', fig.height=8}
```

#### Download Results

**[Results](`r shared_file`)**

### CD8

```{r setcelltype.cd8.responder_vs_nonresponder}
celltype="CD8"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.cd8, ref.label='deseq2-expression-analysis.responder_vs_nonresponder.tcon', results='hide'}
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.cd8, ref.label="DESeq.volc.responder_vs_nonresponder.tcon"}
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.cd8, ref.label='DESeq-MA.responder_vs_nonresponder.tcon', results='asis'}
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.cd8, ref.label='DESeq-output.responder_vs_nonresponder.tcon', results='hide'}
```

```{r DESeq-table.responder_vs_nonresponder.cd8, ref.label='DESeq-table.responder_vs_nonresponder.tcon', results='asis'}
```

```{r DEseq.plotDE.responder_vs_nonresponder.cd8, ref.label='DEseq.plotDE.responder_vs_nonresponder.tcon', fig.height=8}
```

#### Download Results

**[Results](`r shared_file`)**

### NK

```{r setcelltype.nk.responder_vs_nonresponder}
celltype="NK"
```

```{r deseq2-expression-analysis.responder_vs_nonresponder.nk, ref.label='deseq2-expression-analysis.responder_vs_nonresponder.tcon', results='hide'}
```

#### Volcano plot

```{r DESeq.volc.responder_vs_nonresponder.nk, ref.label="DESeq.volc.responder_vs_nonresponder.tcon"}
```

Points within the highlighted regions are differentially expressed (i.e. a log2 fold change of at least `r lfc.cutoff` and an adjusted pvalue of less than `r pvalue.cutoff`)

#### MA-plot

```{r DESeq-MA.responder_vs_nonresponder.nk, ref.label='DESeq-MA.responder_vs_nonresponder.tcon', results='asis'}
```

#### Differentially expressed genes

```{r DESeq-output.responder_vs_nonresponder.nk, ref.label='DESeq-output.responder_vs_nonresponder.tcon', results='hide'}
```

```{r DESeq-table.responder_vs_nonresponder.nk, ref.label='DESeq-table.responder_vs_nonresponder.tcon', results='asis'}
```

```{r DEseq.plotDE.responder_vs_nonresponder.nkx, ref.label='DEseq.plotDE.responder_vs_nonresponder.tcon', fig.height=8}
```

#### Download Results

**[Results](`r shared_file`)**



---

# References

```{r writebib, results='hide', echo=FALSE, message=FALSE, cache=FALSE}
write.bibtex(file="references.bib")
```

# SessionInfo

```{r sessioninfo}
sessionInfo()
```