---
bibliography: "references.bib"
csl: "bioinformatics.csl"
title: "Whangbo - Treatment response RNAseq Analysis"
output:
  html_document:
    code_folding: hide
    theme: yeti
    toc: yes
    toc_float: true
    toc_depth: 3
    number_sections: true
---

```{r setup, echo=FALSE}
library("knitcitations")
cleanbib()
options("citation_format" = "pandoc")

clientname="Jennifer Whangbo"
clientemail="Jennifer.Whangbo@childrens.harvard.edus"
labPI="Ritz"
lablocation="Dana Farber"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

---

RNAseq analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods

Discarded any samples which were not matched pre and post treatment for the same patient. Differential expression was analyzed with DESeq2

---

# Setup

```{r check-library-functions, echo=FALSE}
check_install = function(packages) {
   not_installed = setdiff(packages, rownames(installed.packages()))
   if(length(not_installed) > 0) {
      write(paste("The libraries", not_installed, "are not installed, aborting.",sep=" "), stdout())
      stop()
   }
}
```

```{r load-libraries, echo=FALSE}
packages = c("ggplot2", "reshape", "gplots", "edgeR", "CHBUtils", "pheatmap","DESeq2", "tximport", "DT", "DEGreport", "dplyr", "rio", "rdrop2")
check_install(packages)
installed = lapply(packages, library, character.only = TRUE)
```

```{r variables}
if (file.exists("/home/jnh7/projects/jw_rnaseq")){
  baseDir <- "/home/jnh7/projects/jw_rnaseq"
} else {
baseDir <- "~/Work/projects/jw_rnaseq"
}

project_summary = file.path(baseDir, "data/2016-11-02_bcbio/project-summary.csv")
counts_file = file.path(baseDir, "data/2016-11-02_bcbio/combined.counts")
tx2genes_file = file.path(baseDir, "data/2016-11-02_bcbio/tx2gene.csv")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
pvalue.cutoff <- 0.05
```

```{r functions}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}

subset_tximport = function(txi, rows, columns) {
    txi$counts = txi$counts[rows, columns]
    txi$abundance = txi$abundance[rows, columns]
    txi$length = txi$length[rows, columns]
    return(txi)
}

heatmap_fn = function(data, ...) {
  pheatmap(data, annotation=metadata, clustering_method = "ward.D2",
           clustering_distance_cols = "correlation", ...)
}
```

```{r set-random-seed}
# set seed for reproducibility
set.seed(1454944673)
```

## Sample metadata

```{r data_metadata, echo=FALSE, cache=FALSE}
summarydata = read.table(project_summary, header=TRUE, sep=",")
summarydata = summarydata[,colSums(is.na(summarydata)) < nrow(summarydata)]
# handle newer bcbio-nextgen runs that use description as the key
if("description" %in% colnames(summarydata)) {
  rownames(summarydata) = summarydata$description
  summarydata$Name = rownames(summarydata)
  summarydata$description = NULL
} else {
  rownames(summarydata) = summarydata$Name
  # summarydata$Name = NULL
}
summarydata = summarydata[order(rownames(summarydata)),]
if (file.exists(tx2genes_file)) {
  sample_dirs = file.path(baseDir, "data", rownames(summarydata))
  salmon_files = file.path(sample_dirs, "salmon", "quant.sf")
  sailfish_files = file.path(sample_dirs, "sailfish", "quant.sf")
  new_sailfish = file.path(sample_dirs, "sailfish", "quant", "quant.sf")
  new_salmon = file.path(sample_dirs, "salmon", "quant", "quant.sf")
  if (file.exists(salmon_files[1])) {
    sf_files = salmon_files
  } else if (file.exists(sailfish_files[1])) {
    sf_files = sailfish_files
  }  else if (file.exists(new_sailfish[1])) {
    sf_files = new_sailfish
  }  else if (file.exists(new_salmon[1])) {
    sf_files = new_salmon
  }
  names(sf_files) = rownames(summarydata)
  tx2gene = read.table(tx2genes_file, sep=",", row.names=NULL, header=FALSE)
  txi.salmon = tximport(sf_files, type="salmon", tx2gene=tx2gene, reader=readr::read_tsv, countsFromAbundance="lengthScaledTPM")
  counts = round(data.frame(txi.salmon$counts, check.names=FALSE))
} else {
  counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
}
counts = counts[, order(colnames(counts)), drop=FALSE]
colnames(counts) = gsub(".counts", "", colnames(counts))

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Average.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads")
summarydata[,"Fragment.Length.Mean"] = summarydata$Average.insert.size
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
metadata = metadata[, colSums(is.na(metadata)) < nrow(metadata), drop=FALSE]
metadata$samplecode <- NULL

exclude.samples <- c("nr_pre_tcon_bi","nr_pre_treg_bi","nr_pre_cd8_fh","nr_wk4_nk_fh","nr_wk4_cd8_mld","nr_pre_nk_rl","nr_wk4_tcon_rl","nr_wk4_treg_rl","r_pre_treg_ls","r_wk4_treg_ta")
metadata <- metadata[!(row.names(metadata) %in% exclude.samples),]
summarydata <- summarydata[!(row.names(summarydata) %in% exclude.samples),]
counts <- counts[,!(names(counts) %in% exclude.samples)]
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
sanitize_datatable(summarydata, style='bootstrap')
```


## Correlation heatmaps of TMM-normalized counts

```{r newcovars}
metadata$rRNA_rate <- summarydata$rRNA_rate
metadata$Exonic_rate <- summarydata$Exonic.Rate
```

### Tcon cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.tcon, eval=ncol(counts) > 1}
celltype="Tcon"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.tcon, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```

### Treg cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.treg, eval=ncol(counts) > 1}
celltype="Treg"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.treg, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```

### CD8 cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.cd8, eval=ncol(counts) > 1}
celltype="CD8"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.cd8, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```

### NK cells {.tabset}

#### Correlation (Pearson)
```{r pearson-heatmap-normalized.nk, eval=ncol(counts) > 1}
celltype="NK"
normalized_counts.sub <- normalized_counts[,which(metadata$celltype==celltype)]
heatmap_fn(cor(normalized_counts.sub, method="pearson"))
```

#### Correlation (Spearman)
```{r spearman-heatmap-normalized.nk, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts.sub, method="spearman"))
```


## PCA plots

```{r pca-plotter}
pca_plot = function(comps, nc1, nc2, colorby, size) {
   c1str = paste0("PC", nc1)
   c2str = paste0("PC", nc2)
  ggplot(comps, aes_string(c1str, c2str, color=colorby)) +
    geom_point(size=size) + theme_bw() +
    xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
    ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
  }
```

### Tcon cells {.tabset}
`r celltype="Tcon"`

```{r pca.tcon, eval=ncol(counts) > 1}
cellindices <- which(metadata$celltype==celltype)
counts.sub <- counts[,cellindices]
summarydata.sub <- summarydata[cellindices,]
dds = DESeqDataSetFromMatrix(countData=counts.sub, colData=summarydata.sub, design=~Name)
vst = varianceStabilizingTransformation(dds)
pca_loadings = function(object, ntop=500) {
  rv <- matrixStats::rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
      length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  names(percentVar) = colnames(pca$x)
  pca$percentVar = percentVar
  return(pca)}
pc = pca_loadings(vst)
comps = data.frame(pc$x)
comps$Name = rownames(comps)
library(dplyr)
comps = comps %>% left_join(summarydata, by=c("Name"="Name"))
colorby = "timepoint"
```

#### PC1 vs. PC2
```{r PC1-PC2.tcon}
par(mfrow=c(1,3))
pca_plot(comps, 1, 2, "patientid", size=3)
pca_plot(comps, 1, 2, "status", size=3)
pca_plot(comps,1,2, "timepoint", size=3)
```

#### PC3 vs. PC4
```{r PC3-PC4.tcon}
pca_plot(comps, 3, 4, "patientid", size=3)
pca_plot(comps, 3, 4, "status", size=3)
pca_plot(comps, 3, 4, "timepoint", size=3)
```

#### PC5 vs. PC6
```{r PC5-PC6.tcon}
pca_plot(comps, 5, 6, "patientid", size=3)
pca_plot(comps, 5, 6, "status", size=3)
pca_plot(comps, 5, 6, "timepoint", size=3)
```

#### Variance explained by component
```{r component-loadings.tcon}
ggplot(data.frame(component=reorder(names(pc$percentVar), -pc$percentVar),
                  percent_var=pc$percentVar),
       aes(component, percent_var)) + geom_bar(stat='identity') +
  ylab("percent of total variation") + xlab("") + theme_bw()
```


### Treg cells {.tabset}
`r celltype="Treg"`

```{r pca.treg, ref.label='pca.tcon', eval=ncol(counts) > 1}
```

#### PC1 vs. PC2
```{r PC1-PC2.treg, ref.label='PC1-PC2.tcon'}
```

#### PC3 vs. PC4
```{r PC3-PC4.treg, ref.label='PC3-PC4.tcon'}
```

#### PC5 vs. PC6
```{r PC5-PC6.treg, ref.label='PC5-PC6.tcon'}
```

#### Variance explained by component
```{r component-loadings.treg, ref.label='component-loadings.tcon'}
```


### CD8 cells {.tabset}
`r celltype="CD8"`

```{r pca.cd8, ref.label='pca.tcon', eval=ncol(counts) > 1}
```

#### PC1 vs. PC2
```{r PC1-PC2.cd8, ref.label='PC1-PC2.tcon'}
```

#### PC3 vs. PC4
```{r PC3-PC4.cd8, ref.label='PC3-PC4.tcon'}
```

#### PC5 vs. PC6
```{r PC5-PC6.cd8, ref.label='PC5-PC6.tcon'}
```

#### Variance explained by component
```{r component-loadings.cd8, ref.label='component-loadings.tcon'}
```


### NK cells {.tabset}
`r celltype="NK"`

```{r pca.nk, ref.label='pca.tcon', eval=ncol(counts) > 1}
```

#### PC1 vs. PC2
```{r PC1-PC2.nk, ref.label='PC1-PC2.tcon'}
```

#### PC3 vs. PC4
```{r PC3-PC4.nk, ref.label='PC3-PC4.tcon'}
```

#### PC5 vs. PC6
```{r PC5-PC6.nk, ref.label='PC5-PC6.tcon'}
```

#### Variance explained by component
```{r component-loadings.nk, ref.label='component-loadings.tcon'}
```

---

# Differential expression - Difference in response of Responders and Non-repsonders to treatment {.tabset}

- by celltypes

```{r de-setup}
library(DEGreport)
library(vsn)
design = ~status + status:patientid.n + status:timepoint
```

## Tcon

`r celltype="Tcon"`

```{r deseq2-expression-analysis.tcon, results='asis'}
counts.sub <- counts[rowSums(counts>0)>1,summarydata$celltype==celltype]
summarydata.sub <- summarydata[summarydata$celltype==celltype,]
txi.salmon.sub = subset_tximport(txi.salmon, rownames(counts.sub), colnames(counts.sub))
dds = DESeqDataSetFromTximport(txi.salmon.sub, colData=summarydata.sub, design=~1) # model matrix is not full rank, will refactor later

geoMeans = apply(counts.sub, 1, function(row) if (all(row == 0)) 0 else
  exp(mean(log(row[row != 0]))))
dds = estimateSizeFactors(dds, geoMeans=geoMeans)

# refactor design for nested pairing within groups
newcoldata <- colData(dds) %>% tbl_df()
newcoldata <- group_by(temp, status) %>% mutate(., patientid.n=c(1:5)[factor(patientid)])
dds@colData$patientid.n <- as.character(unlist(newcoldata$patientid.n))


newdesign <- ~status + status:patientid.n + status:timepoint
newmodelmatrix <- model.matrix(newdesign, dds@colData)
if(any(apply(newmodelmatrix, 2, function(x) all(x==0)))) {
  removecolumn <- which(apply(newmodelmatrix, 2, function(x) all(x==0)))
  newmodelmatrix <- newmodelmatrix[,-removecolumn]
}
dds = DESeq(dds, full=newmodelmatrix)
```

## Effect of variance stabilization

```{r deseq-diagnostics.tcon, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1))
meanSdPlot(assay(rld[notAllZero,]))
meanSdPlot(assay(vsd[notAllZero,]))
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

## MA-plots

```{r DESeq-output, results='asis'}
res =  results(dds, contrast=list("statusnonresponder.timepointweek4","statusresponder.timepointweek4"))
ymax = max(res$log2FoldChange, na.rm=TRUE)
ymin = min(res$log2FoldChange, na.rm=TRUE)
plotMA(res, ylim=c(ymin, ymax))
```

## Volcano-plots

```{r DESeq-volcano}
stats = as.data.frame(res[,c(2,6)])
p = volcano_density_plot(stats, title="", lfc.cutoff=1.5)
print(p)
```

## Differentially expressed genes

```{r DESeq-tables, results='asis'}
cat(paste("Lowest adjusted p-value hits"))
out_df = as.data.frame(res)
out_df$id = rownames(out_df)
out_df = out_df[, c("id", colnames(out_df)[colnames(out_df) != "id"])]
write.table(out_df, file="treatment.response.interaction.tsv", sep=""), quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
  sig_genes = subset(out_df, padj < pvalue.cutoff)
  DT::datatable(sig_genes)
  cat("\n")
  
  
```

```{r pathway-load-libraries}
orgdb = "org.Hs.eg.db"
biomart_dataset = "hsapiens_gene_ensembl"
keggname = "hsa"
library(dplyr)
library(clusterProfiler)
library(orgdb, character.only=TRUE)
library(biomaRt)
```

```{r biomaRt-entrez}
mart = biomaRt::useMart(biomart = "ensembl", dataset=biomart_dataset)
entrez = biomaRt::getBM(attributes = c("ensembl_gene_id", "entrezgene"), mart=mart)
entrez$entrezgene = as.character(entrez$entrezgene)
```

```{r go-function}
summarize_cp = function(res, comparison) {
  summaries = data.frame()
  for(ont in names(res)) {
     ontsum = summary(res[[ont]])
     ontsum$ont = ont
     summaries = rbind(summaries, ontsum)
  }
  summaries$comparison = comparison
  return(summaries)}

enrich_cp = function(res, comparison) {
  res = res %>%
    left_join(entrez, by=c("rowname"="ensembl_gene_id")) %>%
    filter(!is.na(entrezgene))
  universe = res$entrezgene
  genes = subset(res, padj < 0.05)$entrezgene
  mf = enrichGO(genes, universe=universe,
                OrgDb=orgdb,
                ont="MF",
                pAdjustMethod="BH",
                qvalueCutoff=1,
                pvalueCutoff=1)
  cc = enrichGO(genes, universe=universe,
                OrgDb=orgdb,
                ont="CC",
                pAdjustMethod="BH",
                qvalueCutoff=1,
                pvalueCutoff=1)
  bp = enrichGO(genes, universe=universe,
                OrgDb=orgdb,
                ont="BP",
                pAdjustMethod="BH",
                qvalueCutoff=1,
                pvalueCutoff=1)
  kg = enrichKEGG(gene=genes, universe=universe,
                  organism='mmu',
                  pvalueCutoff=1,
                  qvalueCutoff=1,
                  pAdjustMethod="BH")
  all = list(mf=mf, cc=cc, bp=bp, kg=kg)
  all[["summary"]] = summarize_cp(all, comparison)
  return(all)}
```

```{r gsea-function}
  res = res %>%
    left_join(entrez, by=c("rowname"="ensembl_gene_id")) %>%
    filter(!is.na(entrezgene)) %>%
    filter(!is.na(log2FoldChange)) %>%
    filter(!is.na(lfcSE))
  lfc = data.frame(res)[, "log2FoldChange"]
  lfcse = data.frame(res)[, "lfcSE"]
  genes = lfc/lfcse
  names(genes) = res$entrezgene
  genes = genes[order(genes, decreasing=TRUE)]
  cc = gseGO(genes, ont="CC", OrgDb=orgdb,
             nPerm=500, pvalueCutoff=1, pAdjustMethod="BH",
             verbose=TRUE)
  mf = gseGO(genes, ont="MF", OrgDb=orgdb,
             nPerm=500, pvalueCutoff=1, pAdjustMethod="BH",
             verbose=TRUE)
  bp = gseGO(genes, ont="bp", OrgDb=orgdb,
             nPerm=500, pvalueCutoff=1, pAdjustMethod="BH",
             verbose=TRUE)
  genes = data.frame(res)[, "log2FoldChange"]
  names(genes) = res$entrezgene
  genes = genes[order(genes, decreasing=TRUE)]
  genes = genes[!is.na(genes)]
  kg = gseKEGG(geneList=genes, organism="mmu", nPerm=500,
               pvalueCutoff=1, verbose=TRUE)
  if(orgdb == "org.Hs.eg.db") {
    do = summary(gseDO(geneList=genes, nPerm=500, pvalueCutoff=1,
                       pAdjustMethod="BH", verbose=TRUE))
    do$ont = "DO"
    all = list(mf=mf, cc=cc, bp=bp, kg=kg, do=do)
  }
  else {
    all = list(mf=mf, cc=cc, bp=bp, kg=kg)
  }
  all[["summary"]] = summarize_cp(all, comparison)
  return(all)}
```


