```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, fig.height=8, fig.width=8, figalign="center", echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev='CairoSVG')
options(width=200)

clientname="Rohan Manohar"
clientemail="rohanraoul@gmail.com"
labPI="Wagers"
lablocation="Joslin"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```
---

RNAseq (edgeR) analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

---

## Methods summary

All reads were trimmed to remove adapter sequences and low quality calls.  Trimmed reads were aligned with [Tophat2][1] and gene specific read counts for [Ensembl gene annotations][2] determined with [htseq-count][3]. Read counts were then normalized and differential gene expression calculated with the [edgeR][4] Bioconductor package.

The sections below provide code to reproduce the post-counting results and plots. 

---

## SETUP

### Libraries

Bioconductor and R libraries used to process and visualize the data

```{r libraries}
library(edgeR)
library(plyr)
library(reshape)
library(ggplot2)
library(biomaRt)
library(scales)
library(googleVis)
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
```

### Variables

Working directories, cutoffs and other variables necessary to the analysis.

```{r variables }
if (file.exists("/n/hsphS10/hsphfs1/chb/projects/aw_rnaseq_dev/")) {
  baseDir <- "/n/hsphS10/hsphfs1/chb/projects/aw_rnaseq_dev/"
} else {
  baseDir <- "/Volumes/ody_new/projects/aw_rnaseq_dev/"
}
dataDir <- file.path(baseDir, "htseq-count/")
resultsDir <- file.path(baseDir, "results/deseq")
metaDir <- file.path(baseDir, "meta")
count.file <- file.path(dataDir, "combined.counts")

pvalcutoff=0.01
numsig=10
rowcount.cutoff=10
logfoldcutoff=1

gene_symbol = 'mgi_symbol'
ensembl_gene = 'mmusculus_gene_ensembl'
filter_type = 'ensembl_gene_id'
```

For this analysis, the cutoff for significant differential expression was an FDR (False Discovery Rate) of `r pvalcutoff` and a log2(foldchange) of expression of at least 'r logfoldcutoff`.  

### Functions

Various convenience functions for plotting results and annotating data.

```{r functions}

source("http://dl.dropboxusercontent.com/u/4253254/Resources/functions.r")

annotate_df = function(d) {
  require(biomaRt)
	ensembl = useMart('ensembl', dataset = ensembl_gene)
	a = getBM(attributes=c(filter_type, gene_symbol, "description"),
		filters=c(filter_type), values=d[, 'id'],
		mart=ensembl)
	m = merge(d, a, by.x='id', by.y=filter_type)
	return(m)
}

plotDispEsts = function(cds) {
  estimates = data.frame(means = rowMeans(counts(cds, normalized=TRUE)),
		variance = fitInfo(cds)$perGeneDispEsts)
	xg = 10^seq(-0.5, 5, length.out=300)
	yg = fitInfo(cds)$dispFun(xg)
	fitline = data.frame(xg=xg, yg=yg)
	p = ggplot(estimates, aes(means, variance)) + geom_point(size=1, alpha=0.4) +
		scale_x_log10() + scale_y_log10() +
		geom_line(data=fitline, aes(xg, yg), color="red") +
		labs(title="dispersion estimation while pooling all samples") +
		xlab("mean number of mapped reads per gene") +
		ylab("estimated dispersion")
	p
}

qq = function(pvaldf,  cutoffpval, samples) {
  title=paste("Quantile-quantile plot of p-values", samples, sep=" - ")
  pvaldf <- pvaldf[order(pvaldf$pval, decreasing=F),]
  pvals <- as.vector(unlist(pvaldf$pval))
  padjs <- as.numeric(as.vector(unlist(pvaldf$padj)))
  colors <- as.vector(ifelse(padjs<cutoffpval, "sig", "nonsig"))
  o = -log10(pvals)
  e = -log10( 1:length(o)/length(o) )
  plot=qplot(e,o, color=colors, xlim=c(0,max(e[!is.na(e)])), ylim=c(0,max(o[!is.na(o)]))) + stat_abline(intercept=0,slope=1, col="darkgrey")
  plot=plot+labs(title=title)
  plot=plot+scale_x_continuous(name=expression(Expected~~-log[10](italic(p))))
  plot=plot+scale_y_continuous(name=expression(Observed~~-log[10](italic(p))))
  plot=plot + scale_colour_manual(name="BFH adjusted pvalue", values=c("black", "red"), labels=c(paste("q>", cutoffpval, sep=""),paste("q<", cutoffpval,sep=""))) 
  plot
}


pcaPlot <- function(countdataset, title, palette)  {
  fit <- prcomp(t(countdataset))
  colors <- palette[factor(pData(cds)$condition)]
  legend_values=unique(cbind(colors, as.character(pData(cds)$condition)))
  ##all samples
  plot(fit$x, bg=colors, col="black", cex=2,pch=21, main=title, oma=c(8,5,5,14))
  legend("topright", cex=0.7, col="black", pt.bg=legend_values[,1], pt.cex=1.25, legend=legend_values[,2],  pch=21, bty="n", x.intersp=1)
  }

PCAplot.cds <- function(countdataset=NULL, metadata=NULL, categories=NULL, title=NULL, colorpalette=NULL, alpha=1, numcomponents=6, normalize.counts=TRUE){
  alpha <- sprintf("%x", ceiling(alpha*255))
  colorpalette <- paste(colorpalette, alpha, sep="")
  if (normalize.counts){
    df <- counts(countdataset, normalized=TRUE)
    } else {
      df <- counts(countdataset, normalized=FALSE)
      }
  myPca.core <- prcomp(t(df))
  tmpPCAData.core <- as.data.frame(myPca.core$x[,1:numcomponents])
  colors <- colorpalette[factor(countdataset$condition)]
  legend_values=unique(cbind(colors, as.character(countdataset$condition)))
  pairs(tmpPCAData.core, bg=colors, col="#606060", cex=2, pch=21, main=title, oma=c(8,5,5,14))
  legend("right", cex=0.7, col="#606060", pt.bg=legend_values[,1], pt.cex=1.5, legend=legend_values[,2],  pch=21, bty="n", x.intersp=1)
  }
```

--- 

## DATA IMPORT

The data consists of HTseq counts generated for individual genes (using the Ensembl GTF annotation) from the Tophat aligments. The metadata (or covariates) that describes the experiment consists of the  replicated experiemental classes of "FAPLung", "FAPMuscle" and "MEF"; with 4 (FAPMuscle and MEF) or 5 (FAPLung) replicates per time point. 

These data and metadata were and loaded into a DESeq *CountDataSet*.

```{r dataload_and_reshape}
combined.counts <- read.table(file.path(dataDir, "combined.counts"), header=T, row.names=1)
combined.counts <- combined.counts[, order(names(combined.counts), decreasing=F)]
# use samplenames get covars from the filenames of the counted samples
covars <- as.data.frame(factor(sub("[0-9]", "", names(combined.counts))))
row.names(covars) <- names(combined.counts)
names(covars) <- "group"
covars$replicate <- gsub("[A-Z|a-z]", "", row.names(covars))
````

Drop genes that are not expressed at all.

```{r keep}
keep <- rowSums(combined.counts)>0
combined.counts <- combined.counts[keep,]
```

Convert into DGE (edgeR) object.
Restrict analysis to genes that are expressed in at least two libraries. Keep only genes that have at least 1 count per million reads sequenced in each library. Low numbers of reads will not be able to be reliably tested and having a large amount of genes which cannot be tested will throw off the multiple hypothesis correction later on.

```{r DGEload}
cds <- DGEList(combined.counts, group=covars$group)
keep <- rowSums(cpm(cds) > 1) >= 2
cds <- cds[keep, ]
dim(cds)
```

---
 
## NORMALIZATION

Here is a sample of `r numsig` of the non-normalized counts of reads mapping to
genes in the annotation. These are just chosen at random for display purposes:

```{r rawcounts, results='asis'}
annotated_counts = head(as.data.frame(counts(cds)), numsig)
annotated_counts <- rownames2col(annotated_counts, "id")
annotated_counts = annotate_df(annotated_counts)
annotated_counts.gvis <- gvisTable(as.data.frame(apply(annotated_counts, 2, as.character)), options=list(width=1280))
```

```{r printmetest, results='asis'}
print(annotated_counts.gvis,  "chart")
```

Two types of normalization are performed; the first is scaling the number of reads mapping to each gene by the total number of mappable reads. The second type of normalization is to ensure the RNA composition is similar across the lanes. If a small number of genes are very highly expressed in one sample but not another it can artificially cause more moderately expressed genes in that sample to be under-sampled.

We do not correct for GC content at this time.

```{r norm1}
cds <- calcNormFactors(cds)
# Get the effective library size (uncorrected only)

lib.size = data.frame(lane = colnames(combined.counts), `Library size` = cds$samples$lib.size, `Normalization factor` = cds$samples$norm.factors)

print(gvisTable(as.data.frame(apply(lib.size,2, as.character)), options = list(width = 640)), "chart")
```


### Sample similarity
A quick sanity check using multi-dimensional scaling using the tag counts:

```{r mds1}
numvars <- length(levels(cds$samples$group))
pal <- brewer.pal(numvars, "Dark2")
colours <- pal[cds$samples$group]
plotMDS(cds, main = "MDS plot", labels = cds$samples$group, col = colours)
```

Convert to predictive fold changes and check sample similarity again:

```{r mds2}
y <- predFC(cds, prior.count.total = 2 * ncol(cds))
plotMDS(y, main = "MDS plot", labels = cds$samples$group, col = colours)
```

## Defining the model
The sample that each column represents needs to be described with the experimental condition and any confounding factors. After each column is described, a design matrix is created using a formula based on the variables described for each column. For now we just use the group information without a reference sample (i.e., all sample groups are present as columns). The last term in the model will be the default term that is tested against during differential expression testing.

```{r design}
design = model.matrix(~0 + group, data = cds$samples)
colnames(design) <- levels(cds$samples$group)
print(gvisTable(as.data.frame(apply(rownames2col(design,"ID"), 2, as.character)), options = list(width = 640)), "chart")
```

## Estimating dispersions
The edgeR algorithm fits a negative binomial to the read counts;  for hypothesis testing a measurement of the deviance of the observed counts from the fit called the dispersion must be calculated. There are several different ways to estimate the dispersion. The first estimator is to assume all genes have the same dispersion from the fit; edgeR calls this the common dispersion

Another alternative is to make the assumption that all genes with the same expression level have the same variance; this is the assumption made in DESeq; edgeR calls this a trended dispersion. Tagwise calculates the dispersion for individual tags (the latter is the recommended approach). Difference shown via qq-plots and the coefficient of variation against gene abundance:

```{r disp1}
# Estimate common and trended dispersion
cds <- estimateGLMCommonDisp(cds, design)
cds$common.dispersion
}

de## Variance estimat

ion

With replicates you can get a more accurate repesentation of the biological variability or dispersion. Dispersion describes the fluctuation of the observed counts for a gene around the mean counts for that gene and can be the square of the coefficient of biological variation. For example, if a gene's expression level fluctuates by 20% the calculated dispersion is $0.2^2 = 0.04$.

Below is a plot of the estimated dispersion for genes with given counts; the dispersion increases as the mean count decreases meaning it is more difficult to accurately measure the expression level of
low-count genes.

```{r estimate_sizefactors, results='hide', fig.cap="Empirical and fitted dispersion values plotted against mean expression strength", out.width='50%'}
# sharingMode = maximum, most conservative approach to sharing information across genes to reduce variability of the dispersion estimates
cds <- estimateDispersions(cds, method="pooled", sharingMode="maximum", fitType="parametric")
plotDispEsts(cds)
```

---

## QUALITY CONTROL

### Primary Clustering
- the goal of this analysis is to naiively evaluate the variability within the raw data and determine whether this variability can predict the different sample categories
- Principal Component Analysis  (PCA) is a dimension reduction and visualisation technique that is used to project the multivariate data vector of each sample into a two-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the samples.
- in basic terms, we are checking whether the samples cluster by sample type; its a good sign if they do

The plots show a scatterplot of the samples along all dual combinations of the first four principal components. Each sample group is represented by a separate color. You can use these plots to explore if the samples cluster, and whether this is according to an intended experimental factor or according to unintended causes such as batch effects. 

```{r pca, out.width='50%', fig.show='hold'}
PCAplot.cds(cds, normalize.counts=FALSE, title="Raw counts", colorpalette=cbPalette, alpha=0.8, numcomponents=4)
PCAplot.cds(cds, normalize.counts=TRUE, title="Normalized counts", colorpalette=cbPalette, alpha=0.8, numcomponents=4)
```

---

## DIFFERENTIAL EXPRESSION ANALYSES

### Filtering

Removing genes with no variance in the samples. 

Using this data, I performed independent filtering to eliminate genes that have no, or little chance of showing signicant evidence. Filtering should result in increased detection power, in terms of false discovery rate. Here, genes were filtered  based on the sum of counts from all samples, as below a certain number of counts, it is not possible to get a pvalue below the desired cutoff. I also filtered out any genes without and Ensembl ID annotation.

```{r filter}
## get sum of counts for all samples for each gene
rowcounts <- rowSums(counts(cds, normalized=TRUE))
## filter the data based on the minimal row sum 
use <- (rowcounts > ecdf(rowcounts)(rowcount.cutoff))
cds.filt <- cds[use,]
## annotated?
use <- grepl("ENS", row.names(counts(cds.filt)))
cds.filt <- cds.filt[use,]
```

Here I discarded the genes with less than `r rowcount.cutoff` counts total in all samples, i.e. the bottom `r ecdf(rowcounts)(rowcount.cutoff)*100`% of genes by total counts


---

## R Session Information

```{r save_image}
sessionInfo()
save.image(file.path(resultsDir, "RDATA" ))
```


[1]: http://tophat.cbcb.umd.edu (Tophat2)
[2]: http://tophat.cbcb.umd.edu/igenomes.html (iGenomes)
[3]: http://www-huber.embl.de/users/anders/HTSeq/doc/count.html (htseq-count)
[4]: http://www.bioconductor.org/packages/2.12/bioc/html/edgeR.html (edgeR)
[5]: http://www.bioconductor.org/packages/2.12/bioc/html/OrderedList.html (OrderedList)