---    
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---
  
  ```{r setup0, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', fig.height = 9, fig.width = 12, bootstrap.show.code=FALSE)
```

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Henske RNA-seq - QC report - WT vs. KO"
clientname="Hilaire Lam"
clientemail="hclam@partners.org"
labPI="Elizabeth Henske"
analystname="Mary Piper"
analystemail="piper@hsph.harvard.edu"
```
---
  
# Henske RNA-Seq Quality Control Report - WT vs. KO
RNA-Seq report for the Henske project investigating the cystic lung destruction associated with Tsc2 knockout mice. 

Client: `r clientname`, `r labPI` group.  

Analysts: `r analystname` (`r analystemail`)

The most recent update of this html document occurred: `r date()`

---

# Overview
Experimental design for RNA-Seq: All female mice in wild-type group (WT, n=3), and  female(1)/male(2) mice in the Tsc2-KO group (KO, n=3). WT and KO mice are from the same litter with genotypes, KO: Tbx4-rtTA/TetO-Cre/Tsc2fx/fx, WT: Tsc2fx/fx. All samples mRNA was isolated from the whole lung (perfused prior to harvest) at the same time. The knockout mice have mesenchymal deletion of Tsc2, which should result in hyperactivation of mTORC1.

Genes to focus on: mTOR, PDGF pathway, TGF-beta/BMP pathway, MMP/TIMP, and ECM   STAT3/IL6

```{r qc-setup}
library(CHBUtils)
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(RColorBrewer)
library(knitr)
library(knitrBootstrap)
library(tidyr)
library(reshape)
library(rmarkdown)
library(dplyr)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(Biobase)
library(scales)
library(DESeq2)
library(gProfileR)
library(DEGreport)
library(vsn)
library(ggrepel)
library(sleuth)
library(png)
library(genefilter)
library(gage)
library(pathview)
library(gageData)
library(dplyr)
library(biomaRt)
library(org.Mm.eg.db)
library(clusterProfiler)
library(DOSE)
library(SPIA)
library(R.devices)

#Data files
project_summary = "../2016-08-22_henske_rnaseq/project-summary.csv"
counts_file = "../2016-08-22_henske_rnaseq/combined.counts"
ann_counts_file <- "../2016-08-22_henske_rnaseq/annotated_combined.counts"

metadata <- read.csv("../../metadata/henske_rnaseq.csv")
sex <- factor(c("F", "F", "F", "F", "F", "F", "F", "F", "F", "F", "F", "F", "F", "F", "F", "F", "M", "M", "M", "M", "M", "M", "M", "M"))
metadata <- data.frame(metadata, sex)

summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
summarydata$genotype <- metadata$genotype

counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))

ann_counts <- read.table(ann_counts_file, header=TRUE, row.names="id")

#Analysis variables
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped.reads",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias")

#Combine lanes
#Combine all lanes for each sample
counts_mergedlanes <- data.frame(S1=rowSums(counts[,1:4]), S2=rowSums(counts[,5:8]), S3=rowSums(counts[,9:12]), S4=rowSums(counts[,13:16]), S5=rowSums(counts[,17:20]), S6=rowSums(counts[,21:24]))

genotype_merged <- rep(c("WT", "KO"), each=3)
sex_merged <- c("F", "F", "F", "F", "M", "M")
metadata_merged <- data.frame(genotype = genotype_merged, sex = sex_merged)
row.names(metadata_merged) <- paste0("S", 1:6)
```

# Quality control metrics
## Total reads
The total number of reads for each sample per lane varies between samples (generally between 7 million to 13 million per sample), with S1 (WTrep1) and S2 (WTrep2) libraries containing many more reads than the other libraries. Combining the reads for the same sample across all lanes yields between 25 to 50 million reads per sample.

```{r total-reads}
ggplot(summarydata, aes(x=rownames(summarydata), y=Total.reads/1e6, fill=genotype)) +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        geom_bar(stat = 'identity') +
        ylab("Total reads (million)") + xlab("")
```

## Mapped reads
The number of mapped reads corresponds to the number of total reads (between 6 and 12 million mapped reads per sample/lane). Combining the reads for the same sample across all lanes yields between 25 to 50 million mapped reads per sample, which is a very good depth for aligned reads.

```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped.reads/1e6, fill=genotype)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("Mapped reads (million)") + xlab("")
```

## Genomic mapping rate
The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality or other artifacts.

The percent of input reads mapped is excellent (>95%) for all of the samples. 

```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate, fill=genotype)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Exonic mapping rate
The exonic mapping rates are good, with over 60% of total reads mapping to exons.

```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate, fill=genotype)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
The samples exhibited very low rates of rRNA mapping, which is good.
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA_rate, fill=genotype)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```


## Boxplot of log10 TMM-normalized counts per gene
Generally, we expect similar count spreads for all genes between samples unless the library sizes or total RNA expression are different. The log10 TMM-normalized counts per gene normalization method equates the overall expression levels of genes between samples under the assumption that the majority of them are not DE.Therefore, by normalizing for total RNA expression by sample, we expect the spread of the log10 TMM-normalized counts per gene to be similar for every sample. The samples generally look good with similar log10 TMM-normalized counts per gene distributions. 

Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
Generally, we expect similar count spreads for all genes between samples unless the total expressed RNA per sample is different. The samples generally look good with similar log10 TMM-normalized counts per gene densities. 

```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Exploring similarity using principal component analysis (PCA) and heatmaps
Principal components analysis is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is a way to look at how samples are clustering. 

An intercorrelation heatmap is another way to look at how well samples cluster by plotting the correlation between the expression profiles of the samples.

These techniques (i.e. PCA and intercorrelation heatmaps) can identify outliers better when values have a similar dynamic range. We transformed the data to the log2 scale using rlog transformation (generated using [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)), which is better for datasets in which there are large differences in sequencing depth.

### PCA
In this PCA plot, we are exploring whether the samples look the same across lanes and whether the wildtype and knockout samples cluster together.

The samples cluster very well across lanes, and there is good clustering of wildtype and knockout samples on the PC1 axis (62% variance in the data is accounted for by PC1).

```{r sample_variation_pca_all_groups}

design = ~ genotype

dds = DESeqDataSetFromMatrix(counts,
    colData=metadata, design = design)
dds = DESeq(dds)


#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therefore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
rlogMat <- assay(rld)

## PCA - all groups
data <- plotPCA(rld, intgroup = "genotype", returnData=TRUE)
percentVar <- round(100 * attr(data, "percentVar"))

ggplot(data, aes(PC1, PC2, color=genotype)) + geom_point(size=4) + geom_text_repel(aes(label=name)) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance"))
```

### Heatmap
Similar to the PCA, the heatmap shows the samples to cluster well across lanes, and the expression profiles of the samples cluster by sample group.

```{r heatmap}

row.names(metadata) <- metadata$samplename
metadata_heat <- metadata[3]
files <- row.names(metadata_heat)
file_names <- paste0(files,"_R1")
row.names(metadata_heat) <- file_names
rlog_counts <- assay(rld[notAllZero,])

#Values relate to the row mean subtracted from the normalized count value for each sample value.
pheatmap(cor(rlog_counts), annotation=metadata_heat, method="spearman")
```

## Number of genes detected
Based on the good clustering of samples across lanes, the counts files are combined for samples across lanes. The number of genes detected for each sample is good with over 20,000 genes detected for all samples.

```{r genes-detected-plot}
dd = data.frame(Name=names(counts_mergedlanes), Genes.Detected = colSums(counts_mergedlanes > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation
Gene detection saturation was nearly reached for the majority of the samples, which means that most of the samples were able to detect the majority of expressed genes. While saturation of gene detection was at ~ 50 million reads mapped, reads that had over 30 million reads mapped still detected >22,000 genes.

```{r saturation-plot}
mapped <- data.frame(S1 = sum(summarydata[1:4, 14]), S2 = sum(summarydata[5:8, 14]), S3 = sum(summarydata[9:12, 14]), S4 = sum(summarydata[13:16, 14]), S5 = sum(summarydata[17:20, 14]), S6 = sum(summarydata[21:24, 14])) 
dd = data.frame(Mapped= t(mapped), Genes.Detected = colSums(counts_mergedlanes > 0))
ggplot(dd, aes(x=Mapped/1e6, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped (million)")
```

## Expression of Tsc2

To ensure experimental expectations were met for the knockout of Tsc2, we visualized the mapped reads for this gene. While some exons seem to have many fewer reads for the KO animals, Tsc2 expression is similar to WT for many of the exons. 

```{r tsc2_igv}
tsc2_igv <- readPNG("henske_igv.png")
grid.raster(tsc2_igv)
```

To look in greater depth at the expression levels between samples, the raw counts for each exon in Tsc2 are given below. Since S1 (WT) and S2 (WT) will have greater counts due to having a larger number of total reads, we expect the number of reads aligning to Tsc2 to be higher for these samples. S3 (WT) has a similar number of total reads to the KO samples (S4-6), so it might be more accurate to compare this sample with the KO samples. Overall, there is no knockout of a single exon; however, there does seem to be knockdown for some exons, such as ENSMUSG00000002496:036. 

```{r tsc2_exon}
tsc2_exons <- read.csv(file="henske_tsc2_expression.csv", header=T)
row.names(tsc2_exons) <- tsc2_exons$id
tsc2_exons <- tsc2_exons[2:25]

#Combine lanes
#Combine all lanes for each sample
tsc2_exons_merged <- data.frame(S1=rowSums(tsc2_exons[,1:4]), S2=rowSums(tsc2_exons[,5:8]), S3=rowSums(tsc2_exons[,9:12]), S4=rowSums(tsc2_exons[,13:16]), S5=rowSums(tsc2_exons[,17:20]), S6=rowSums(tsc2_exons[,21:24]))

knitr::kable(tsc2_exons_merged)
```

# Differential Expression analysis - WT vs. KO

Differential gene expression analysis of count data was performed using the Bioconductor R package, [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html). The count data was fit to a negative binomial model and dispersion estimates were generated using the mean values from the maximum likelihood estimate of log2 fold changes, optimizing the Cox-Reid adjusted profile likelihood. 

Several quality metrics were assessed to explore the fit of the count data to the model, and differential expression analysis was performed.

The fit of the count data to the model was good, although we recommend looking at the normalized counts of the significant DE genes to ensure similar expression profiles across all replicates to narrow down genes most likely to be biologically relevant between WT and KO samples. 

**The raw counts files below can be used to perform your own differential expression analysis. The normalized counts file should be used for any sample comparisons or visualizations using scatter or bar plots.**

```{r raw_counts_file}
raw_counts <- counts_mergedlanes
raw_counts$name <- rownames(raw_counts)
ann_counts_names <- ann_counts
ann_counts_names$name <- rownames(ann_counts)
raw_counts <- merge(raw_counts, ann_counts_names, by="name")
rownames(raw_counts) <- raw_counts$name
raw_counts <- raw_counts[,c(2:7,32)]
#write.csv(raw_counts, "henske_raw_counts.csv", quote=F)
```
[Download raw counts file used for all DE Analyses](https://dl.dropboxusercontent.com/u/204381225/henske/henske_raw_counts.csv)

```{r de_normalized counts}
rm(dds)
metadata_merged$genotype <- relevel(metadata_merged$genotype, ref="WT")
design = ~ genotype
dds = DESeqDataSetFromMatrix(counts_mergedlanes,
    colData=metadata_merged, design = design)
dds = DESeq(dds)

# Rlog normalized counts
total_norm_counts <- counts(dds, normalized=T)
total_norm_counts <- data.frame(total_norm_counts)
total_norm_counts$name <- rownames(total_norm_counts) 

ann_counts_names <- ann_counts
ann_counts_names$name <- rownames(ann_counts)
norm_counts <- merge(total_norm_counts, ann_counts_names, by="name")
rownames(norm_counts) <- norm_counts$name
norm_counts <- norm_counts[,c(2:7,32)]
#write.csv(norm_counts, "henske_norm_counts_genenames.csv", quote=F)
```
[Download normalized counts file](https://dl.dropboxusercontent.com/u/204381225/henske/henske_norm_counts_genenames.csv)

### Effect of variance stabilization - WT vs KO
For RNA-Seq gene counts, the variance increases with the mean. To account for this variance, logarithmic transformation (log2) of normalized count values will ensure that these genes won't dominate during visualization. However, due to the noise associated with low count values, the general log2 transformation will worsen this noise, and low count genes will instead dominate. Therefore, we need to use a transformation that will stabilize the variance across the mean for the gene counts. The plots below show the standard deviation of transformed counts using log2, rlog, and vsd transformations by rank(mean) (from top to bottom, respectively). The transformations greatly reduce the standard deviation, with rlog stabilizing the variance the best across the mean. Therefore, we will use the rlog transformed counts for any downstream visualizations of counts.

```{r deseq-diagnostics}

#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therfore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

#Plotting standard deviation by rank(mean)
meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1))
meanSdPlot(assay(rld[notAllZero,]))
meanSdPlot(assay(vsd[notAllZero,]))
```

## Exploring similarity using principal component analysis (PCA) and heatmaps

### PCA
In this PCA plot, the KO samples cluster well, but the WT samples exhibit quite a bit of variation within the group. However, since the WT and KO samples separate on the PC1 axis, which accounts for 61% of the variance, it is likely that we should be able to differentiate expression differences between the sample groups. 


```{r sample_variation_pca}

## PCA 
data <- plotPCA(rld, intgroup = "genotype", returnData=TRUE)
percentVar <- round(100 * attr(data, "percentVar"))

ggplot(data, aes(PC1, PC2, color=genotype)) + geom_point(size=4) + geom_text_repel(aes(label=name)) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance"))
```

### Heatmap
Similar to the PCA, the heatmap shows the expression profiles of the samples to cluster by genotype. Also to note, the samples do not cluster by sex of the animal, which indicates that the differentially expressed genes should not represent sex-related gene expression differences.

```{r sample_variation_heatmap}
#Values relate to the row mean subtracted from the normalized count value for each sample value.
rlog_counts <- assay(rld[notAllZero,])
pheatmap(cor(rlog_counts), annotation=metadata_merged, method="spearman")
```

### Dispersion estimates - WT vs KO
The following plot shows the dispersion by mean of normalized counts. The dispersion looks good, since we expect the dispersion to decrease with increased mean of normalized counts. 

```{r dispersion-estimate}
plotDispEsts(dds)
```

## Differential Expression analysis - WT vs. KO
Using an FDR cut-off of 0.05 (p-values multiple test corrected using the BH method), 2433 genes were identified as significantly differentially expressed. The expression changes listed below for all analyses describe differences in the Tsc2 WT mice relative to the KO mice. 

The table below shows that at an FDR cut-off of 0.05, 889 genes were up-regulated in WT relative to KO, and 1544 genes were down-regulated relative to KO.

```{r de_ko_vs_wt}
res_all <- results(dds)
df <- data.frame(res_all)

# Output all results with associated gene names
df_name <- cbind(df, name = row.names(df))
dfsymbol <- which(row.names(ann_counts_names) %in% row.names(df_name))
dfsymbol <- ann_counts_names[dfsymbol,]

df_genenames <- merge(df_name, dfsymbol, by="name")
row.names(df_genenames) <- df_genenames$name
df_genenames <- df_genenames[,c(2:7, 32)]
df_genenames <- df_genenames[order(df_genenames$padj), ]
df_genenames$padj <- format(df_genenames$padj, scientific = T)
#write.csv(df_genenames, "henske_DE_all_results_wt_ko.csv", quote=F)

summary(res_all, alpha=0.05)
```

### MA-plots - WT vs KO
MA plots explore the mean expression level of the genes with the fold change, allowing for the visualization of the genes that are differentially expressed (red). The MA plot shows that all genes with large fold changes between the two sample groups are found to be significant, in addition to many high expressing genes with smaller fold changes. The red dots represent the significant DE genes (padj < 0.05).

```{r DESeq-output_MA}
#res <- results(dds, alpha=.05)
plotMA(res_all, ylim=c(-5,5))
```

### Significant differentially expressed genes - KO vs WT
The list below of significant DE genes contains the gene expression differences in the KO samples relative to the WT samples. The row names in the table are the Ensembl gene ids, followed by the columns: the mean of the normalized counts for that gene for all samples (`baseMean`), log2 fold change (`log2FoldChange`), standard error (`lfcSE`), Wald statistic (`stat`), Wald test p-value (`pvalue`), BH adjusted p-values, which account for multiple testing (`padj`), and the official gene symbol (`symbol`). Only the top 20 genes (by padj values) are listed in the table below. The entire list of significant genes can be downloaded using the link below the table.

For example, for gene id ENSMUSG00000018339 (official gene symbol = Gpx3), the WT sample group had an expression level log2 fold change of 3.46 relative to the KO sample group, with an average mean expression of 11954.48 (norm. counts), and the gene was up-regulated in KO samples relative WT samples, since the log2FoldChange is positive. 

```{r DE_genes_list}
#Gene list with padj
resSig <- subset(res_all, padj < 0.05)
DEG <- data.frame(baseMean = resSig[[1]], log2FoldChange = resSig[[2]], lfcSE = resSig[[3]], stat = resSig[[4]], pvalue = resSig[[5]], padj = resSig[[6]], row.names = row.names(resSig))
DEG <- cbind(DEG, name = row.names(DEG))
DEGsymbol <- which(row.names(ann_counts_names) %in% row.names(DEG))
DEGsymbol <- ann_counts_names[DEGsymbol,]

DEG_genenames <- merge(DEG, ann_counts_names, by="name")

sig_norm_counts <- cbind(norm_counts, "name" = row.names(norm_counts))
sig_norm_counts <- merge(sig_norm_counts, DEG_genenames, by="name")
row.names(sig_norm_counts) <- sig_norm_counts$name
sig_norm_counts <- sig_norm_counts[, c(2:8)]

row.names(DEG_genenames) <- DEG_genenames$name
DEG_genenames <- DEG_genenames[, c(2:7, 32)]
DEG_genenames <- DEG_genenames[order(abs(DEG_genenames$log2FoldChange), decreasing = T),]
DEG_genenames$padj <- format(DEG_genenames$padj, scientific = T)
knitr::kable(head(DEG_genenames, n=50))
#write.csv(DEG_genenames, "henske_DE_results_0.05padj.csv", quote=F)

#Normalized counts of genes
DEG_genenames <- merge(DEG, DEGsymbol, by="name")
sig_counts <- cbind(counts_mergedlanes, "name" = row.names(counts_mergedlanes))
sig_counts <- merge(sig_counts, DEG_genenames, by="name")
row.names(sig_counts) <- sig_counts$name
sig_counts <- sig_counts[, c(2:7, 38)]

log2_counts <- assay(rld[notAllZero,])
log2_counts <- as.data.frame(log2_counts) 
idx <- rownames(log2_counts) %in% rownames(sig_counts)
log2_sig_counts <- log2_counts[idx,]
log2_sig_counts$name <- rownames(log2_sig_counts)
log2_sig_counts_symbol <- merge(log2_sig_counts, DEG_genenames, by="name")
rownames(log2_sig_counts_symbol) <- log2_sig_counts$name 
log2_sig_counts_symbol <- log2_sig_counts_symbol[, c(2:7,38)]
rownames(log2_sig_counts_symbol) <- log2_sig_counts_symbol$symbol
log2_sig_counts_symbol <- log2_sig_counts_symbol[1:6]
#write.csv(log2_sig_counts_symbol, "henske_DE_rlog_counts.csv", quote=F)
```

[Download Full List Significant DE Genes (can open in Excel)](https://dl.dropboxusercontent.com/u/204381225/henske/henske_DE_results_0.05padj.csv)

[Download All Results (can open in Excel)](https://dl.dropboxusercontent.com/u/204381225/henske/henske_DE_all_results_wt_ko.csv)

#### Heatmap of gene expression for significant genes
This plot shows the expression of the significant differentially expressed genes by sample. The scale values compare the sample expression of a specific gene to the mean expression of the gene across all samples. As expected, these significant genes cluster by sample group, but note that some samples do not show similar expression changes across replicates. The most trustworthy genes are those that show similar expression changes across replicates, so it is useful to check the normalized counts of the genes across replicates to identify those genes most likely to be biologically relevant. 

```{r heatmap_DE_genes_all}

#Values relate to the row mean subtracted from the normalized count value for each sample value.
#pheatmap(log2_sig_counts_symbol, clustering_method = "ward.D2", clustering_distance_cols = "correlation", scale="row", show_rownames = F)
sig_heatmap <- readPNG("henske_sig-genes_heatmap.png")
grid.raster(sig_heatmap)
```
[Download rlog normalized counts used to create heatmap figure](https://dl.dropboxusercontent.com/u/204381225/henske/henske_DE_rlog_counts.csv)

### Expression and results for Tsc2 and expected DE genes
The expression levels of Tsc2 were explored for Tsc2 WT and KO mice. As expected, the levels of Tsc2 expression were higher among WT mice, and the gene was found to be significant DE between the two sample groups.

**Normalized counts of Tsc2:**

```{r tsc2_counts}
tsc2_norm_counts <- subset(norm_counts, symbol=="Tsc2")
knitr::kable(tsc2_norm_counts)
```

**DE results for Tsc2:**

```{r tsc2_results}
tsc2_results <- subset(df_genenames, symbol == "Tsc2")
knitr::kable(tsc2_results)
```


### Functional Analysis - WT vs KO
Using the significant DE genes (padj < 0.05) identified for the KO vs. WT samples, a list of statistically enriched gene ontology (GO) terms for biological process (BP), human phenotype ontology (HP) terms, and KEGG pathways (keg) was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). Note that we did not include in this analysis enriched GO terms for molecular functions or cellular components.

A list including only the significant GO terms was then used as input to [REViGO](http://revigo.irb.hr/), which collapsed redundant and semantically-related terms and output the most significantly enriched functional categories. 

Using these tools, we found the DE genes to exhibit significant enrichment for many different processes and pathways. The top 20 significant GO terms are listed in the table below. Many of the expected affected pathways are represented among the significant GO terms, including 'regulation of TOR signaling' and 'response to BMP stimulus' (within the 'response to fluid shear stress' category in the REVIGO figure below). 

The list of all significant GO/HP/keg terms and associated genes, as well as a high resolution copy of the REViGO figure can be downloaded using the links below the table. 

J. Reimand, T. Arak, P. Adler, L. Kolberg, S. Reisberg, H. Peterson, J. Vilo. g:Profiler -- a web server for functional interpretation of gene lists (2016 update). Nucleic Acids Research 2016; doi: 10.1093/nar/gkw199

Supek F, Bošnjak M, Škunca N, Šmuc T. REVIGO summarizes and visualizes long lists of Gene Ontology terms. PLoS ONE 2011. doi:10.1371/journal.pone.0021800

```{r functional_analysis_all}
gene_list <- DEG_genenames
gene_list <- gene_list[, c(2:7, 32)]
rownames(gene_list) <- rownames(DEG_genenames)
gene_list <- gene_list[order(gene_list$padj),]
gene_list <- gene_list[, c(6,7)]

#gprofileR
library(gProfileR)
gprofiler_results <- gprofiler(query = gene_list, organism = "mmusculus", ordered_query = F, 
                               exclude_iea = F, max_set_size = 0, correction_method = "fdr", 
                               hier_filtering = "none", domain_size = "annotated", custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
#write(GOs, file ="henske_GOterms.csv")
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
#write.csv(GO_genes, file ="henske_GO_genes.csv", quote = FALSE, row.names=T)
knitr::kable(head(gprofiler_results, n=20))

#REVIGO
# # A treemap R script produced by the REVIGO server at http://revigo.irb.hr/
# # If you found REVIGO useful in your work, please cite the following reference:
# # Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# # terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800
# 
# # author: Anton Kratz <anton.kratz@gmail.com>, RIKEN Omics Science Center, Functional Genomics Technology Team, Japan
# # created: Fri, Nov 02, 2012  7:25:52 PM
# # last change: Fri, Nov 09, 2012  3:20:01 PM
# 
# # -----------------------------------------------------------------------------
# # If you don't have the treemap package installed, uncomment the following line:
# # install.packages( "treemap" );
# library(treemap) 								# treemap package by Martijn Tennekes
# 
# # Set the working directory if necessary
# # setwd("C:/Users/username/workingdir");
# 
# # --------------------------------------------------------------------------
# # Here is your data from REVIGO. Scroll down for plot configuration options.
# 
# revigo.names <- c("term_ID","description","freqInDbPercent","uniqueness","dispensability","representative");
# revigo.data <- rbind(c("GO:0007620","copulation",0.001,0.972,0.000,"copulation"),
# c("GO:0060135","maternal process involved in female pregnancy",0.002,0.635,0.647,"copulation"),
# c("GO:0034405","response to fluid shear stress",0.001,0.915,0.000,"response to fluid shear stress"),
# c("GO:0060004","reflex",0.002,0.893,0.199,"response to fluid shear stress"),
# c("GO:1900746","regulation of vascular endothelial growth factor signaling pathway",0.001,0.701,0.695,"response to fluid shear stress"),
# c("GO:0050923","regulation of negative chemotaxis",0.000,0.722,0.471,"response to fluid shear stress"),
# c("GO:0050919","negative chemotaxis",0.001,0.845,0.378,"response to fluid shear stress"),
# c("GO:0048014","Tie signaling pathway",0.000,0.769,0.389,"response to fluid shear stress"),
# c("GO:0035455","response to interferon-alpha",0.001,0.873,0.355,"response to fluid shear stress"),
# c("GO:2000322","regulation of glucocorticoid receptor signaling pathway",0.000,0.747,0.258,"response to fluid shear stress"),
# c("GO:0071873","response to norepinephrine",0.000,0.873,0.626,"response to fluid shear stress"),
# c("GO:0071867","response to monoamine",0.001,0.868,0.564,"response to fluid shear stress"),
# c("GO:0043620","regulation of DNA-templated transcription in response to stress",0.002,0.775,0.277,"response to fluid shear stress"),
# c("GO:0010043","response to zinc ion",0.001,0.888,0.198,"response to fluid shear stress"),
# c("GO:0019934","cGMP-mediated signaling",0.001,0.774,0.246,"response to fluid shear stress"),
# c("GO:0061045","negative regulation of wound healing",0.000,0.723,0.399,"response to fluid shear stress"),
# c("GO:0071398","cellular response to fatty acid",0.001,0.852,0.676,"response to fluid shear stress"),
# c("GO:1902170","cellular response to reactive nitrogen species",0.001,0.864,0.642,"response to fluid shear stress"),
# c("GO:0043567","regulation of insulin-like growth factor receptor signaling pathway",0.001,0.726,0.660,"response to fluid shear stress"),
# c("GO:0071773","cellular response to BMP stimulus",0.001,0.845,0.587,"response to fluid shear stress"),
# c("GO:0032008","positive regulation of TOR signaling",0.002,0.667,0.480,"response to fluid shear stress"),
# c("GO:0097066","response to thyroid hormone",0.001,0.869,0.496,"response to fluid shear stress"),
# c("GO:0098581","detection of external biotic stimulus",0.002,0.885,0.387,"response to fluid shear stress"),
# c("GO:0071548","response to dexamethasone",0.001,0.860,0.642,"response to fluid shear stress"),
# c("GO:0031666","positive regulation of lipopolysaccharide-mediated signaling pathway",0.002,0.615,0.685,"response to fluid shear stress"),
# c("GO:0034198","cellular response to amino acid starvation",0.000,0.847,0.343,"response to fluid shear stress"),
# c("GO:0071526","semaphorin-plexin signaling pathway",0.002,0.761,0.433,"response to fluid shear stress"),
# c("GO:0070542","response to fatty acid",0.001,0.866,0.510,"response to fluid shear stress"),
# c("GO:0038027","apolipoprotein A-I-mediated signaling pathway",0.000,0.783,0.180,"response to fluid shear stress"),
# c("GO:0035633","maintenance of blood-brain barrier",0.000,0.946,0.000,"maintenance of blood-brain barrier"),
# c("GO:0036010","protein localization to endosome",0.001,0.905,0.000,"protein localization to endosome"),
# c("GO:1902305","regulation of sodium ion transmembrane transport",0.001,0.735,0.264,"protein localization to endosome"),
# c("GO:0014909","smooth muscle cell migration",0.002,0.841,0.656,"protein localization to endosome"),
# c("GO:1903305","regulation of regulated secretory pathway",0.001,0.709,0.555,"protein localization to endosome"),
# c("GO:0051937","catecholamine transport",0.002,0.870,0.328,"protein localization to endosome"),
# c("GO:1902946","protein localization to early endosome",0.000,0.908,0.514,"protein localization to endosome"),
# c("GO:0015825","L-serine transport",0.001,0.893,0.652,"protein localization to endosome"),
# c("GO:0042348","NF-kappaB import into nucleus",0.002,0.704,0.562,"protein localization to endosome"),
# c("GO:0015808","L-alanine transport",0.000,0.896,0.143,"protein localization to endosome"),
# c("GO:0010883","regulation of lipid storage",0.002,0.712,0.627,"protein localization to endosome"),
# c("GO:0010874","regulation of cholesterol efflux",0.001,0.722,0.593,"protein localization to endosome"),
# c("GO:0008228","opsonization",0.000,0.897,0.114,"protein localization to endosome"),
# c("GO:0008333","endosome to lysosome transport",0.002,0.875,0.353,"protein localization to endosome"),
# c("GO:0006929","substrate-dependent cell migration",0.001,0.847,0.122,"protein localization to endosome"),
# c("GO:0042982","amyloid precursor protein metabolic process",0.002,0.974,0.000,"amyloid precursor protein metabolism"),
# c("GO:0042983","amyloid precursor protein biosynthetic process",0.000,0.955,0.157,"amyloid precursor protein metabolism"),
# c("GO:0051340","regulation of ligase activity",0.002,0.817,0.003,"regulation of ligase activity"),
# c("GO:0001953","negative regulation of cell-matrix adhesion",0.002,0.780,0.152,"regulation of ligase activity"),
# c("GO:0006883","cellular sodium ion homeostasis",0.000,0.786,0.286,"regulation of ligase activity"),
# c("GO:0035931","mineralocorticoid secretion",0.000,0.512,0.648,"regulation of ligase activity"),
# c("GO:0042135","neurotransmitter catabolic process",0.001,0.802,0.144,"regulation of ligase activity"),
# c("GO:0060586","multicellular organismal iron ion homeostasis",0.001,0.566,0.577,"regulation of ligase activity"),
# c("GO:0032429","regulation of phospholipase A2 activity",0.001,0.816,0.437,"regulation of ligase activity"),
# c("GO:0046885","regulation of hormone biosynthetic process",0.001,0.751,0.307,"regulation of ligase activity"),
# c("GO:1900084","regulation of peptidyl-tyrosine autophosphorylation",0.000,0.817,0.696,"regulation of ligase activity"),
# c("GO:0070373","negative regulation of ERK1 and ERK2 cascade",0.002,0.667,0.586,"regulation of ligase activity"),
# c("GO:0031952","regulation of protein autophosphorylation",0.002,0.802,0.189,"regulation of ligase activity"),
# c("GO:0051443","positive regulation of ubiquitin-protein transferase activity",0.001,0.723,0.561,"regulation of ligase activity"),
# c("GO:0051560","mitochondrial calcium ion homeostasis",0.001,0.776,0.555,"regulation of ligase activity"),
# c("GO:0050999","regulation of nitric-oxide synthase activity",0.001,0.815,0.457,"regulation of ligase activity"),
# c("GO:0010922","positive regulation of phosphatase activity",0.001,0.733,0.497,"regulation of ligase activity"),
# c("GO:0045762","positive regulation of adenylate cyclase activity",0.002,0.709,0.677,"regulation of ligase activity"),
# c("GO:1903019","negative regulation of glycoprotein metabolic process",0.001,0.773,0.410,"regulation of ligase activity"),
# c("GO:0030007","cellular potassium ion homeostasis",0.000,0.785,0.698,"regulation of ligase activity"),
# c("GO:2000463","positive regulation of excitatory postsynaptic membrane potential",0.001,0.769,0.298,"regulation of ligase activity"),
# c("GO:2000479","regulation of cAMP-dependent protein kinase activity",0.001,0.795,0.565,"regulation of ligase activity"),
# c("GO:2000480","negative regulation of cAMP-dependent protein kinase activity",0.000,0.762,0.639,"regulation of ligase activity"),
# c("GO:0071825","protein-lipid complex subunit organization",0.002,0.898,0.011,"protein-lipid complex subunit organization"),
# c("GO:0072319","vesicle uncoating",0.000,0.830,0.487,"protein-lipid complex subunit organization"),
# c("GO:1900227","positive regulation of NLRP3 inflammasome complex assembly",0.000,0.730,0.497,"protein-lipid complex subunit organization"),
# c("GO:0097435","fibril organization",0.001,0.858,0.265,"protein-lipid complex subunit organization"),
# c("GO:0000422","mitochondrion degradation",0.001,0.844,0.394,"protein-lipid complex subunit organization"),
# c("GO:0043113","receptor clustering",0.002,0.792,0.503,"protein-lipid complex subunit organization"),
# c("GO:0044089","positive regulation of cellular component biogenesis",0.002,0.713,0.253,"protein-lipid complex subunit organization"),
# c("GO:1902743","regulation of lamellipodium organization",0.002,0.729,0.275,"protein-lipid complex subunit organization"),
# c("GO:1904375","regulation of protein localization to cell periphery",0.002,0.729,0.686,"protein-lipid complex subunit organization"),
# c("GO:0016241","regulation of macroautophagy",0.002,0.697,0.619,"protein-lipid complex subunit organization"),
# c("GO:0072661","protein targeting to plasma membrane",0.001,0.792,0.545,"protein-lipid complex subunit organization"),
# c("GO:0097090","presynaptic membrane organization",0.001,0.853,0.263,"protein-lipid complex subunit organization"),
# c("GO:1903008","organelle disassembly",0.001,0.864,0.497,"protein-lipid complex subunit organization"),
# c("GO:0030038","contractile actin filament bundle assembly",0.000,0.868,0.360,"protein-lipid complex subunit organization"),
# c("GO:0033687","osteoblast proliferation",0.002,0.943,0.021,"osteoblast proliferation"),
# c("GO:0060011","Sertoli cell proliferation",0.000,0.600,0.678,"osteoblast proliferation"),
# c("GO:0060977","coronary vasculature morphogenesis",0.001,0.596,0.651,"osteoblast proliferation"),
# c("GO:0002043","blood vessel endothelial cell proliferation involved in sprouting angiogenesis",0.000,0.605,0.577,"osteoblast proliferation"),
# c("GO:0048662","negative regulation of smooth muscle cell proliferation",0.002,0.770,0.620,"osteoblast proliferation"),
# c("GO:0003184","pulmonary valve morphogenesis",0.001,0.594,0.666,"osteoblast proliferation"),
# c("GO:0042417","dopamine metabolic process",0.002,0.948,0.023,"dopamine metabolism"),
# c("GO:0097242","beta-amyloid clearance",0.000,0.665,0.024,"beta-amyloid clearance"),
# c("GO:0090596","sensory organ morphogenesis",0.000,0.659,0.340,"beta-amyloid clearance"),
# c("GO:1900221","regulation of beta-amyloid clearance",0.000,0.579,0.390,"beta-amyloid clearance"),
# c("GO:0071827","plasma lipoprotein particle organization",0.002,0.598,0.445,"beta-amyloid clearance"),
# c("GO:0008038","neuron recognition",0.002,0.565,0.666,"beta-amyloid clearance"),
# c("GO:0036075","replacement ossification",0.002,0.642,0.419,"beta-amyloid clearance"),
# c("GO:0021670","lateral ventricle development",0.001,0.588,0.651,"beta-amyloid clearance"),
# c("GO:0016322","neuron remodeling",0.001,0.583,0.577,"beta-amyloid clearance"),
# c("GO:0061036","positive regulation of cartilage development",0.001,0.476,0.699,"beta-amyloid clearance"),
# c("GO:0035987","endodermal cell differentiation",0.002,0.584,0.643,"beta-amyloid clearance"),
# c("GO:0060996","dendritic spine development",0.002,0.527,0.700,"beta-amyloid clearance"),
# c("GO:0030903","notochord development",0.002,0.593,0.555,"beta-amyloid clearance"),
# c("GO:0050702","interleukin-1 beta secretion",0.001,0.587,0.579,"beta-amyloid clearance"),
# c("GO:0097490","sympathetic neuron projection extension",0.000,0.556,0.610,"beta-amyloid clearance"),
# c("GO:0097491","sympathetic neuron projection guidance",0.000,0.494,0.629,"beta-amyloid clearance"),
# c("GO:0030574","collagen catabolic process",0.001,0.627,0.441,"beta-amyloid clearance"),
# c("GO:0045631","regulation of mechanoreceptor differentiation",0.000,0.518,0.592,"beta-amyloid clearance"),
# c("GO:0031077","post-embryonic camera-type eye development",0.001,0.615,0.553,"beta-amyloid clearance"),
# c("GO:0010934","macrophage cytokine production",0.001,0.605,0.434,"beta-amyloid clearance"),
# c("GO:0003300","cardiac muscle hypertrophy",0.002,0.621,0.450,"beta-amyloid clearance"),
# c("GO:0051962","positive regulation of nervous system development",0.002,0.459,0.670,"beta-amyloid clearance"),
# c("GO:0035290","trunk segmentation",0.000,0.651,0.360,"beta-amyloid clearance"),
# c("GO:0071634","regulation of transforming growth factor beta production",0.001,0.558,0.585,"beta-amyloid clearance"),
# c("GO:0097094","craniofacial suture morphogenesis",0.001,0.600,0.520,"beta-amyloid clearance"),
# c("GO:0072171","mesonephric tubule morphogenesis",0.000,0.601,0.681,"beta-amyloid clearance"),
# c("GO:0040019","positive regulation of embryonic development",0.001,0.491,0.623,"beta-amyloid clearance"),
# c("GO:0003071","renal system process involved in regulation of systemic arterial blood pressure",0.001,0.539,0.558,"beta-amyloid clearance"),
# c("GO:0007442","hindgut morphogenesis",0.001,0.619,0.428,"beta-amyloid clearance"),
# c("GO:0060343","trabecula formation",0.002,0.593,0.639,"beta-amyloid clearance"),
# c("GO:0021783","preganglionic parasympathetic fiber development",0.001,0.588,0.484,"beta-amyloid clearance"),
# c("GO:0060033","anatomical structure regression",0.001,0.768,0.025,"anatomical structure regression"),
# c("GO:0050819","negative regulation of coagulation",0.002,0.512,0.677,"anatomical structure regression"),
# c("GO:0060572","morphogenesis of an epithelial bud",0.001,0.744,0.396,"anatomical structure regression"),
# c("GO:2000181","negative regulation of blood vessel morphogenesis",0.000,0.525,0.602,"anatomical structure regression"),
# c("GO:0003323","type B pancreatic cell development",0.001,0.579,0.634,"anatomical structure regression"),
# c("GO:1903375","facioacoustic ganglion development",0.000,0.644,0.477,"anatomical structure regression"),
# c("GO:1901862","negative regulation of muscle tissue development",0.001,0.617,0.590,"anatomical structure regression"),
# c("GO:0051150","regulation of smooth muscle cell differentiation",0.001,0.632,0.372,"anatomical structure regression"),
# c("GO:0010771","negative regulation of cell morphogenesis involved in differentiation",0.002,0.546,0.685,"anatomical structure regression"),
# c("GO:0072537","fibroblast activation",0.000,0.928,0.035,"fibroblast activation"),
# c("GO:0048102","autophagic cell death",0.001,0.919,0.041,"autophagic cell death"),
# c("GO:0034349","glial cell apoptotic process",0.001,0.909,0.497,"autophagic cell death"),
# c("GO:0006651","diacylglycerol biosynthetic process",0.000,0.913,0.041,"diacylglycerol biosynthesis"),
# c("GO:0006704","glucocorticoid biosynthetic process",0.001,0.885,0.305,"diacylglycerol biosynthesis"),
# c("GO:0032799","low-density lipoprotein receptor particle metabolic process",0.001,0.953,0.065,"low-density lipoprotein receptor particle metabolism"),
# c("GO:0032802","low-density lipoprotein particle receptor catabolic process",0.000,0.940,0.666,"low-density lipoprotein receptor particle metabolism"),
# c("GO:2000644","regulation of receptor catabolic process",0.000,0.818,0.665,"low-density lipoprotein receptor particle metabolism"));
# 
# stuff <- data.frame(revigo.data);
# names(stuff) <- revigo.names;
# 
# stuff$abslog10pvalue <- as.numeric( as.character(stuff$abslog10pvalue) );
# stuff$freqInDbPercent <- as.numeric( as.character(stuff$freqInDbPercent) );
# stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) );
# stuff$dispensability <- as.numeric( as.character(stuff$dispensability) );
# 
# # by default, outputs to a PDF file
# pdf( file="revigo_treemap.pdf", width=16, height=9 ) # width and height are in inches
# 
# # check the tmPlot command documentation for all possible parameters - there are a lot more
# treemap(
# 	stuff,
# 	index = c("representative","description"),
# 	vSize = "uniqueness",
# 	type = "categorical",
# 	vColor = "representative",
# 	title = "REVIGO Gene Ontology treemap",
# 	inflate.labels = FALSE,      # set this to TRUE for space-filling group labels - good for posters
# 	lowerbound.cex.labels = 0,   # try to draw as many labels as possible (still, some small squares may not get a label)
# 	bg.labels = "#CCCCCCAA",     # define background color of group labels
# 												       # "#CCCCCC00" is fully transparent, "#CCCCCCAA" is semi-transparent grey, NA is opaque
# 	position.legend = "none"
# )
# 
# dev.off()

img_revigo <- readPNG("henske_revigo.png")
grid.raster(img_revigo)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/henske/henske_GO_genes.csv)

[Download REVIGO Figure](https://dl.dropboxusercontent.com/u/204381225/henske/henske_revigo.png)

## clusterProfiler
Similar to gprofileR, the tool [clusterProfiler](http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) was used to perform over-representation analysis on GO terms associated with the significant DE genes. The table displays the list of top 20 GO terms that were significantly enriched among the significant genes, which is similar to those output by gprofileR. 

**NOTE:** While the GO terms output by clusterProfiler are very similar to those output by gprofileR, the small differences in the GO terms output is due to the different algorithms used by the two different programs.

G Yu, LG Wang, Y Han, QY He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology 2012, 16(5):284-287.

G Yu, LG Wang, GR Yan, QY He. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics 2015, 31(4):608-609.

```{r kegg}
# Create a KEGG dataset
kegg_mouse <- kegg.gsets(species = "mouse", id.type = "kegg")
kegg.gs <- kegg_mouse$kg.sets[kegg_mouse$sigmet.idx]
#head(kegg.gs)

DEG_background <- data.frame(res_all)

mart <- useDataset("mmusculus_gene_ensembl",
                  useMart('ENSEMBL_MART_ENSEMBL',
                          host =  'useast.ensembl.org'))
attributes <- listAttributes(mart)
entrez <- getBM(filters= "ensembl_gene_id",
                    attributes= c("ensembl_gene_id", "entrezgene"),
                    values= row.names(DEG_background),
                    mart= mart)
DEG_background$ensembl_gene_id <- row.names(DEG_background)
entrez_results <- merge(DEG_background, entrez, by="ensembl_gene_id")
entrez_results <- subset(entrez_results, entrezgene != "NA")

#clusterprofiler analysis
sig_genes <- subset(entrez_results, padj<0.05)$entrezgene
sig_genes <- as.character(sig_genes)
all_genes <- entrez_results$entrezgene
all_genes <- as.character(all_genes)
ggo <- groupGO(gene=sig_genes, OrgDb=org.Mm.eg.db, ont="BP", level=3, readable=TRUE)
ego <- enrichGO(gene=sig_genes, universe=all_genes, OrgDb=org.Mm.eg.db, ont="BP", pAdjustMethod = "BH", qvalueCutoff =0.05, readable=TRUE)

GO_processes <- ego@result
knitr::kable(head(GO_processes, n=20))

ego2 <- setReadable(ego, OrgDb = org.Mm.eg.db)
cluster_summary <- (summary(ego2))
#write.csv(GO_processes, "henske_clusterprofiler_GO_processes.csv", quote=F)
#dotplot(ego, showCategory=100)
#plotGOgraph(ego)
#enrichMap(ego, n=100, vertex.label.font=8)
#barplot(ggo, drop=TRUE, showCategory=12)
#barplot(ego, showCategory=8)
#cnetplot(ego, categorySize="pvalue", showCategory = 5, foldChange=foldchanges)
```
[Download full list of significant GO processes](https://dl.dropboxusercontent.com/u/204381225/henske/henske_clusterprofiler_GO_processes.csv)

The dotplot below shows the number of genes associated with the first 100 terms (size) and the p-adjusted values for these terms. 

```{r dotplot}
img_dotplot <- readPNG("henske_GOdotplot.png")
grid.raster(img_dotplot)
```
[Download clusterProfiler Dotplot](https://dl.dropboxusercontent.com/u/204381225/henske/henske_GOdotplot.png)

The GOgraph below show the relationship between the top 100 most significantly enriched GO terms, by grouping similar terms together.

```{r clusterprofiler_GOgraph}
img_GOgraph <- readPNG("henske_GOenrich.png")
grid.raster(img_GOgraph)
```
[Download clusterProfiler GOgraph](https://dl.dropboxusercontent.com/u/204381225/henske/henske_GOenrich.png)

The graph below shows the relationship among the 5 most significant GO terms, and the fold changes of the significant genes associated with these terms. Tsc2 is associated with two of the five terms: 'regulation of anatomical structure size' and 'transmembrane receptor protein tyrosine kinase signaling pathway'.

```{r clusterprofiler_cnetgraph}
img_cnet <- readPNG("henske_cnet.png")
grid.raster(img_cnet)
```
[Download clusterProfiler Cnet Graph](https://dl.dropboxusercontent.com/u/204381225/henske/henske_cnet.png)

## Pathway analysis using SPIA (topology-based method)
The previous analyses did not explore how genes interact with each other (e.g. activation, inhibition, phosphorylation, ubiquitination, etc) to determine the pathway-level statistics. The [SPIA (Signaling Pathway Impact Analysis)](http://bioconductor.org/packages/release/bioc/html/SPIA.html) tool was used to integrate the lists of differentially expressed genes determined by DESeq2, their fold changes, and pathway topology to identify affected pathways.

The table below shows the following significantly dysregulated pathways based on over-representation and signaling perturbations accumulation. The table shows the following information: pSize is the number of genes on the pathway; NDE is the number of DE genes per pathway; tA is the observed total preturbation accumulation in the pathway; pNDE is the probability to observe at least NDE genes on the pathway using a hypergeometric model; pPERT is the probability to observe a total accumulation more extreme than tA only by chance; pG is the p-value obtained by combining pNDE and pPERT; pGFdr and pGFWER are the False Discovery Rate and respectively Bonferroni adjusted global p-values; and the Status gives the direction in which the pathway is perturbed (activated or inhibited). KEGGLINK gives a web link to the KEGG website that displays the pathway image with the differentially expressed genes highlighted in red.

In the plot below, each pathway is a point and the coordinates are the log of pNDE (using a hypergeometric model) and the p-value from perturbations, pPERT. The oblique lines in the plot show the significance regions based on the combined evidence.

The most perturbed pathways include mTOR signaling pathway and ECM-receptor interaction.

Tarca AL, Kathri P and Draghici S (2013). SPIA: Signaling Pathway Impact Analysis (SPIA) using combined evidence of pathway over-representation and unusual signaling perturbations. [http://bioinformatics.oxfordjournals.org/cgi/reprint/btn577v1](http://bioinformatics.oxfordjournals.org/cgi/reprint/btn577v1).

```{r spia, results="hide",echo=FALSE}
#spia (http://www.gettinggeneticsdone.com/2012/03/pathway-analysis-for-high-throughput.html)

# significant genes is a vector of fold changes where the names
# are ENTREZ gene IDs. The background set is a vector of all the 
# genes represented on the platform.
#convert ensembl to entrez ids

sig_genes <- subset(entrez_results, padj<0.05)$log2FoldChange
names(sig_genes) <- subset(entrez_results, padj<0.05)$entrezgene

#Remove NA values
sig_genes <- sig_genes[!is.na(names(sig_genes))] 

# run SPIA.
spia_result <- spia(de=sig_genes, all=entrez_results$entrezgene, organism="mmu", plots=F)

img_spia <- readPNG("henske_spia.png")
grid.raster(img_spia)
#plotP(spia_result, threshold=0.05)
#write.csv(spia_result, "henske_spia_results.csv", quote=F)
```

```{r spia_results, echo=FALSE}
knitr::kable(head(spia_result, n=10))
```
[Download SPIA results](https://dl.dropboxusercontent.com/u/204381225/henske/henske_spia_results.csv)

[Download SPIA plot](https://dl.dropboxusercontent.com/u/204381225/henske/henske_spia.png)

## Gene set enrichment analysis 
Gene set enrichment analysis tools like GAGE and clusterProfiler use ranked lists of genes (here ranked by log2FC) without using a threshold. This allows these gene set enrichment tools to use more information to identify enriched biological processes. The introduction to gene set enrichment analysis goes into more detail about some of the advantages of this approach: [http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896/](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896/). 

### Pathway analysis with clusterProfiler and Pathview
Analysis was performed to identify any pathways that were significantly dysregulated (genes in pathway could be up- or down-regulated simultaneously). The significantly dysregulated pathways (q-value (FDR) < 0.05) are displayed below in the pathway images, which show the degree of dysregulation of the genes (the minus direction (green) is down-regulated, while the positive direction (red) is up-regulated). 

In addition, the Running Enrichment Score plots show the enrichment of higher ranked genes in the significantly dysregulated pathways.

**Calcium signaling pathway**

*Pathway*
```{r gsea_mmu04020}
all_results_gsea <- entrez_results
all_results_gsea <- all_results_gsea[order(abs(all_results_gsea$log2FoldChange), decreasing = T), ]
foldchanges <- all_results_gsea$log2FoldChange
names(foldchanges) <- all_results_gsea$entrezgene
#head(foldchanges)

foldchanges <- sort(foldchanges, T)
ego3 <- gseGO(geneList     = foldchanges,
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              nPerm        = 1000,
              minGSSize    = 100,
              maxGSSize    = 500, 
              pvalueCutoff = 0.05,
              verbose      = FALSE)
gseaGOresult <- ego3@result
#gseaplot(ego3, geneSetID = "GO:0001508")

kk2 <- gseKEGG(geneList     = foldchanges,
               organism     = 'mmu',
               nPerm        = 1000,
               minGSSize    = 120,
               pvalueCutoff = 0.05,
               verbose      = FALSE)
kkgsea_results <- kk2@result

# mmu04020gsea <- pathview(gene.data  = foldchanges,
#                      pathway.id = "mmu04020",
#                      species    = "mmu",
#                      limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu04020 <- readPNG("mmu04020gsea.pathview.png")
grid.raster(img_mmu04020)
```

*Running Enrichment Scores*
```{r gsea_04020_enrich}
gseaplot(kk2, "mmu04020")
```

**Lysosome**

*Pathway*
```{r gsea_mmu04142}

# mmu04142gsea <- pathview(gene.data  = foldchanges,
#                      pathway.id = "mmu04142",
#                      species    = "mmu",
#                      limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu04142 <- readPNG("mmu04142gsea.pathview.png")
grid.raster(img_mmu04142)
```

*Running Enrichment Scores*
```{r gsea_04142_enrich}
gseaplot(kk2, "mmu04142")
```

**Purine metabolism**

*Pathway*
```{r gsea_mmu00230}
# mmu00230gsea <- pathview(gene.data  = foldchanges,
#                      pathway.id = "mmu00230",
#                      species    = "mmu",
#                      limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu00230 <- readPNG("mmu00230gsea.pathview.png")
grid.raster(img_mmu00230)
```

*Running Enrichment Scores*
```{r gsea_00230_enrich}
gseaplot(kk2, "mmu00230")
```

**Parkinson's Disease**

*Pathway*
```{r gsea_mmu05012}
# mmu05012gsea <- pathview(gene.data  = foldchanges,
#                      pathway.id = "mmu05012",
#                      species    = "mmu",
#                      limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu05012 <- readPNG("mmu05012gsea.pathview.png")
grid.raster(img_mmu05012)
```

*Running Enrichment Scores*
```{r gsea_05012_enrich}
gseaplot(kk2, "mmu05012")
```

**Oxidative Phosphorylation**

*Pathway*
```{r gsea_mmu00190}
# #mmu00190gsea <- pathview(gene.data  = foldchanges,
#                      pathway.id = "mmu00190",
#                      species    = "mmu",
#                      limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu00190 <- readPNG("mmu00190gsea.pathview.png")
grid.raster(img_mmu00190)
```

*Running Enrichment Scores*
```{r gsea_00190_enrich}
gseaplot(kk2, "mmu00190")
```

**Alzheimer's Disease**

*Pathway*
```{r gsea_mmu05010}

# #mmu05010gsea <- pathview(gene.data  = foldchanges,
#                      pathway.id = "mmu05010",
#                      species    = "mmu",
#                      limit      = list(gene=max(abs(foldchanges)), cpd=1))
img_mmu05010 <- readPNG("mmu05010gsea.pathview.png")
grid.raster(img_mmu05010)
```

*Running Enrichment Scores*
```{r gsea_05010_enrich}
gseaplot(kk2, "mmu05010")
```

### Pathway analysis with GAGE and Pathview 
We also used a different method for gene set enrichment analysis, using the [GAGE (Generally Applicable Gene-set Enrichment for Pathway Analysis)](http://bioconductor.org/packages/release/bioc/html/gage.html) and [Pathview](http://bioconductor.org/packages/release/bioc/html/pathview.html) tools with a more relaxed significance threshold (q-value (FDR) < 0.1), since this method seems to be more strict. This method also uses the log2 fold changes obtained from the DESeq2 analysis for every gene and identifies pathways that were significantly dysregulated. This method also identified the Lysosome Pathway as being significantly dysregulated, and the image below reflects the degree of dysregulation of the genes (the minus direction (green) is down-regulated, while the positive direction (red) is up-regulated).

Weijun Luo, Michael Friedman, Kerby Shedden, Kurt Hankenson, and Peter Woolf. GAGE: generally applicable
gene set enrichment for pathway analysis. BMC Bioinformatics, 2009. doi:10.1186/1471-2105-10-161.

Weijun Luo and Cory Brouwer. Pathview: an R/Bioconductor package for pathway-based data integration
and visualization. Bioinformatics, 29(14):1830-1831, 2013. doi: 10.1093/bioinformatics/btt285.


```{r gage}
# Run gage
keggres = gage(foldchanges, gsets=kegg.gs, same.dir=FALSE)

# Look at both up (greater), down (less), and statatistics.
#lapply(keggres, head)
sel <- keggres$greater[, "q.val"] < 0.1 & !is.na(keggres$greater[, "q.val"])
path.ids <- rownames(keggres$greater)[sel]

# Get the pathways
keggrespathways = data.frame(id=rownames(keggres$greater), keggres$greater) %>% 
  tbl_df() %>% 
  filter(row_number()<=1) %>% 
  .$id %>% 
  as.character()
#keggrespathways

# Get the IDs.
keggresids = substr(keggrespathways, start=1, stop=8)
#keggresids

#pathview
# Define plotting function for applying later
plot_pathway = function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="mmu", new.signature=FALSE)

# plot multiple pathways (plots saved to disk and returns a throwaway list object)
#tmp = sapply(keggresids, function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="mmu"))

img_pathview <- readPNG("mmu04142.pathview.png")
grid.raster(img_pathview)
```
[Download KEGG Pathview plot](https://dl.dropboxusercontent.com/u/204381225/henske/mmu04142.png)

```{r session_info}
sessionInfo()
```