```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=TRUE,  highlight=TRUE, figalign="center", echo=TRUE, warning=FALSE, error=FALSE, message=FALSE)
```

Array analysis for [Joerg Ermann](mailto:jermann@partners.org) and [Weiguo Zou](zouwg94@gmail.com), Glimcher group at HSPH.  Contact [John Hutchinson](mailto:jhutchin@hsph.harvard.edu) for additional details.

---

# Methods summary
All MoGene 1.0 ST arrays were processed using the 'oligo' BioConductor package[^1], quality-controlled with arrayQualityMetrics[^2] and RMA normalized[^3]. Differentially expressed genes were identified using limma[^4].

---

# Workflow
The sections below provide minimal code to reproduce the included results and plots. 

## Load Bioconductor and R libraies used to process and visualize the data

```{r libraries}
library(oligo)
library(arrayQualityMetrics)
library(limma)
library(pheatmap)
library(RColorBrewer)
library(pathprint)
library(xtable)
library(Biobase)
```

## Setup Data and Results directory variables
```{r variables}
dataDir <- "/Users/johnhutchinson/Work/Consults/Glimcher-AOF2_knockout_analysis_12/data"
resultsDir <- "/Users/johnhutchinson/Work/Consults/Glimcher-AOF2_knockout_analysis_12/results"
```

## Load the data
- load in phenotypes and array names from covars.desc in "data" subdirectory
  - this file contains the names and descriptions of CEL files contained in same directory 
- use array names to load in actual arrays 

```{r dataload, results='hide'}
covars <- read.delim(paste(dataDir, "covars.desc", sep="/"), row.names=1)
celFiles <- list.celfiles(dataDir, full.names=TRUE)
affyRaw <- read.celfiles(celFiles)
pData(affyRaw) <- covars 
````

## Sample metadata

```{r covars, results='asis'}
# Sample information table
pDataTable <- xtable(pData(affyRaw))
print(pDataTable, type='html')
```

## QA/QC - 1 
- using arrayQualityMetrics library
- raw data

```{r rawQC}
arrayQualityMetrics(expressionset=affyRaw, outdir='../results/report_raw', force=TRUE, do.logtransform=TRUE)
```

[raw data QC report](./results/report_raw/index.html)

## Background correct and normalize
- using RMA
- summarize data probe information on the gene (`core`) level

```{r normalize, results='hide'}
affyNorm.core <- rma(affyRaw, target="core", background=TRUE, normalize=TRUE)
# cleanup
rm(affyRaw)
```

## QA/QC - 2 
- using arrayQualityMetrics library
- normalized data

```{r normQC, cache=TRUE}
arrayQualityMetrics(expressionset=affyNorm.core, outdir='../results/report_rma.core', force=TRUE, do.logtransform=FALSE)
```

[normalized data QC report](./results/report_rma.core/index.html)

## QA/QC - 3
### Primary Clustering
- do the samples cluster by sample type?

```{r pca, fig.cap="Primary Component Analysis of samples - all combinations of the 5 first primary components", fig.width=8.5, fig.height=8.5}
eset.core <- exprs(affyNorm.core) 
colnames(eset.core) <- pData(affyNorm.core)$Sample
myPca.core <- prcomp(t(eset.core))
tmpPCAData.core <- as.data.frame(myPca.core$x[,1:4])
colors <- c('darkgrey','darkgrey', 'red','red')
plot(tmpPCAData.core, col=colors, pch=row.names(myPca.core$x))
# cleanup
rm(tmpPCAData.core, myPca.core)
```

## Annotations
- load in feature data, gene symbols in particular

```{r features, results='hide'}
# retrieve NetAffx Biological Annotation
featureData(affyNorm.core) <- getNetAffx(affyNorm.core, "transcript")

# Extract the 'gene assignment' annotation
annot <- pData(featureData(affyNorm.core)[, c('geneassignment')])
categories <- pData(featureData(affyNorm.core)[, c('category')])
# Generate a list of gene symbols from the gene assignment
desc <- annot[, 1]
symbols <- unlist(lapply(desc, function(x) strsplit(x, ' // ')[[1]][2]))
# check to make sure data is correct
identical(length(featureData(affyNorm.core)$probesetid), length(symbols))
head(symbols[!is.na(symbols)])
```

## Analyses
### Limma

#### Create appropriate design matrix 
- make a matrix with arrays as rows, sample groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group, 
- in this case it looks like this, where 1-4 are the 4 microarrays named in the covars.desc file in the same order as in the file

```{r design, results="asis"}
design <- model.matrix(~ -1+factor(pData(affyNorm.core)$Condition))
# make sure the headings match
colnames(design) <- c("KO", "WT")

designTable <- xtable(design)
print(designTable, type='html')
```

#### Fit a linear model
- for each gene based on the given series of arrays

```{r linearmodel}
eset.core <- exprs(affyNorm.core)
fit.core <- lmFit(eset.core, design) 
```

#### Create contrast matrix
- to perform specified pairwise comparisons
- in this table, columns are contrasts/comparisons and rows are sample groups
-  a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts("WT-KO",levels=c("KO", "WT"))
contrastmatrixTable <- xtable(contrast.matrix)
print(contrastmatrixTable, type='html')
```

#### Compute estimated coefficients and standard errors for contrasts
```{r contrastfit}
fit2.core <- contrasts.fit(fit.core, contrast.matrix) 
```

#### Compute moderated t-statistics and log-odds of differential expression 
- by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.core <- eBayes(fit2.core) 
```

## Results
### Volcanoplot to visualize relationship between fold change and signifcance of changes in expression between groups
- genes outside the two vertical blue lines show more than 2-fold changes in expression between groups
- genes which pass an FDR  (false discovery rate) multiple testing correction of 0.2 (20% FDR) are colored blue (none here)
- the larger the pointsize, the more likely there is significant expression change (B = log-odds of differential expression) between the sample groups
- genes with log-odds of differential expression more than 0.5 (~75% chance of being differentially expressed) are labelled by text

```{r volcanoplot, fig.cap="Volcano plot", fig.width=8.5, fig.height=8.5}
stats.core <- topTable(fit2.core, coef=1, sort.by="B",number=length(symbols), genelist=cbind(symbols, fit2.core$genes))
stats.core$Passes.FDR.threshold  <-  as.factor(stats.core$adj.P.Val<0.2)
labeled.stats.core <- stats.core[stats.core$B>=0.5,]
ggplot(data=stats.core, aes(x=logFC, y=-log10(P.Value), color=Passes.FDR.threshold, size=B)) +
  geom_point(alpha=0.5) + 
  geom_text(data=labeled.stats.core, aes(x=logFC,y=-log10(P.Value),label=symbols, size=2), hjust=0.5) +
  geom_vline(xintercept=c(-1,1), color="blue", alpha=0.25) +
  xlab("log2 fold change") + ylab("-log10 p-value")
```

### Obtain top results
- Generates list of top 100 ('number=100') differentially expressed genes as sorted by "B", the log-odds that a gene is differentially expressed
#### The summary table contains the following information: 
- logFC is the log2-fold change
- the AveExpr is the average expression value accross all arrays
- the moderated t-statistic (t) is the logFC to its standard error, the P.Value is the associated p-value
- the adj.P.Value is the p-value adjusted for multiple testing (by FDR) 
- the B-value (B) is the log-odds that a gene is differentially expressed (the-higher-the-better)
- usually one wants to base gene selection on the adjusted P-value rather than the t- or B-value, but there are no significant adjusted P-values
- I avoided selection based on the logFC because the data is quite noisy and I was seeing a very high number of control intronic probes when sorting by logFC
- the last 4 columns contain the log-transformed normalized expression levels for these genes in each sample

```{r toptable}
stats.top.core <- topTable(fit2.core, coef=1, adjust="fdr", sort.by="p", p.value=1, number=100, genelist=cbind(fit2.core$genes,symbols)) 
# expression levels for these genes in all samples
eset.top.core = exprs(affyNorm.core[stats.top.core[, "ID"], ]) 
colnames(eset.top.core) = pData(affyNorm.core)$Sample
eset.stats.top.core <- cbind(stats.top.core, eset.top.core)
```

### Output pvalues and expression levels to files

```{r outputs}
write.table(stats.core, file="../results/all.genes.stats.xls", sep="\t", row.names=F, col.names=T)
write.table(eset.stats.top.core, file='../results/KO_vs_WT.top100.Bvalues.expression_levels.xls', row.names=F, col.names=T, sep='\t')
```
[Link to Excel file with results for all genes](./results/all.genes.stats.xls)

[Link to Excel file with results for genes with top100 log-odds of differential expression](./results/KO_vs_WT.top100.Bvalues.expression_levels.xls)

### Heatmaps of top results
- the row names have the probeset and the gene symbol separated by a dash
- rows with "NA" for the gene symbol are mainly intronic control probes (one of them is an exonic control probe)

```{r heatmaps, fig.cap="KO_vs_WT.top100pvalues.annotated.heatmap", fig.width=8.5, fig.height=15}
## add in gene symbols for each Affy probe
row.names(eset.top.core)=  paste(row.names(eset.top.core), stats.top.core$symbols, sep= "-")
#write out heatmaps with Affy probeset IDs
pheatmap(eset.top.core, main=colnames(fit2.core$coefficients), color=brewer.pal(11, "RdBu"), fontsize_row=8)
```


---
  **How do the samples differ in terms of pathways?**
  
  ## PATHPRINT
  - an unpublished method that compares pathway activation in a set of samples to those of samples within the GEO dataset database that have been analysed on the same microarray platform[^5] 
- in this case, we are comparing the data to ~2500 datasets from the GEO DataSets database that were analysed on the Mouse Gene 1.0 ST array)


```{r pathprintsetup, results='hide'}
mic.data.platform <- "GPL6246"
mic.data.species <- "mouse"
mic.data.names <- pData(affyNorm.core)[,1]
mic.data.fingerprint <- exprs2fingerprint(exprs = eset.core, platform = mic.data.platform, species = mic.data.species, progressBar = TRUE)
dimnames(mic.data.fingerprint)[[2]] <- mic.data.names
```

### Most variable pathways in data 
- allowing disagreement between samples within a group (i.e. KO or WT)
- a positive value indicates higher expression of genes in this pathway than average
- a negative value indicates lower expression of genes in this pathway than average
- a zero value indicates average expression of genes in this pathway

```{r pathprintresults, fig.width=8.5, fig.height=11, fig.cap=""}
pheatmap(mic.data.fingerprint[apply(mic.data.fingerprint, 1, sd) > 0.5, ], fontsize_row=8,col = c("blue", "white", "red"))
```

### Variable pathways that are consistent within individual groups
- samples within a group must agree for call

```{r pathprintresults2,fig.width=8.5, fig.height=11, fig.cap=""}
pheatmap(mic.data.fingerprint[which(mic.data.fingerprint[,1]==mic.data.fingerprint[,2] & mic.data.fingerprint[,3]==mic.data.fingerprint[,4] & mic.data.fingerprint[,1]!=mic.data.fingerprint[,4]),1:4], labCol = mic.data.names[1:4], mar = c(10,25), col = c("blue", "white", "red"))
```

[List of genes within pathways](./data/pathprint.Mm.gs.txt)

[^1]: [BioC](http://www.bioconductor.org)
[^2]: [arrayQualityMetrics](http://www.bioconductor.org/packages/2.10/bioc/html/arrayQualityMetrics.html)
[^3]: [RMA](http://bioinformatics.oxfordjournals.org/content/19/2/185)
[^4]: [limma](http://www.bioconductor.org/packages/release/bioc/html/limma.html)
[^5]: [Geo_DataSets](http://www.ncbi.nlm.nih.gov/gds/)

