---
title: "hg19 RNAseq alignment and QC"
author: John Hutchinson
date: March 23, 2016
output: html_document
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,message=FALSE, prompt=TRUE, comment='', fig.cap='', echo=FALSE)

```

# Overview

Here I aligned all the samples against the hg19 build of the human transcriptome. By comparing these result to the prior mm10 results, we can get more information about what went wrong with the LE2, LE3 and LE4 samples.

```{r check-library-functions, echo=FALSE}
check_install = function(packages) {
   not_installed = setdiff(packages, rownames(installed.packages()))
   if(length(not_installed) > 0) {
      write(paste("The libraries", not_installed, "are not installed, aborting.",sep=" "), stdout())
      stop()
   }
}
```

```{r load-libraries, echo=FALSE}
packages = c("ggplot2", "reshape", "gplots", "edgeR", "CHBUtils", "pheatmap","DESeq2", "tximport", "logging", "DT", "gridExtra")
check_install(packages)
installed = lapply(packages, library, character.only = TRUE)
```

```{r qc-setup}
basicConfig()

if(file.exists("/Volumes/orch/projects/gu_rnaseq/hg19")){
  baseDir <- "/Volumes/orch/projects/gu_rnaseq/hg19"
} else {
  baseDir <- "/Users/johnhutchinson/Work/projects/gu_rnaseq/hg19"
}
project_summary = file.path(baseDir, "results/final/2016-03-21_project_hg19/project-summary.csv")
counts_file = file.path(baseDir, "results/final/2016-03-21_project_hg19/combined.counts")
tx2genes_file = file.path(baseDir, "results/final/2016-03-21_project_hg19/tx2gene.csv")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442","#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]

if (file.exists(tx2genes_file)) {
  loginfo("Using gene counts calculated from the Sailfish transcript counts.")
  sf_files = file.path(baseDir, "results/work/sailfish", rownames(summarydata), "quant.sf")
  names(sf_files) = rownames(summarydata)
  tx2gene = read.table(tx2genes_file, sep=",", row.names=NULL, header=FALSE)
  txi.salmon = tximport(sf_files, type="salmon", tx2gene=tx2gene, reader=readr::read_tsv, countsFromAbundance="lengthScaledTPM")
  counts = round(data.frame(txi.salmon$counts, check.names=FALSE))
} else {
  loginfo("Using gene counts calculated from featureCounts.")
  counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
}
counts = counts[, order(colnames(counts)), drop=FALSE]
colnames(counts) = gsub(".counts", "", colnames(counts))

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Median.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads")
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
```

```{r sanitize-datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}
```

```{r set-random-seed}
# set seed for reproducibility
set.seed(1454944673)
```

## Sample metadata
```{r summary-table, echo=FALSE, cache=FALSE}
sanitize_datatable(summarydata, style='bootstrap')
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

```{r qualimap-run}
qualimap_run = "Mapped" %in% colnames(summarydata)
```


```{r mapped-plot, eval=qualimap_run}
p1 <- ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"), axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("") + 
  ggtitle("Mapped reads")
```


```{r mapping-rate-plot, eval=qualimap_run}
p2 <- ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"), axis.text.x = element_text(angle=90)) + 
  ggtitle("Genomic mapping rate")
```


```{r genes-detected-plot}
dd = data.frame(Name=colnames(counts), Genes.Detected = colSums(counts > 0))
p3 <- ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"), axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("") +
  ggtitle("Number of genes detected")
```

```{r plot1, out.width='100%'}
grid.arrange(p1,p2,p3, nrow=2, ncol=2)
```

###The number of mapped reads, genomic mapping rate and number of genes detected all support samples LE2, LE3 and LE4 being strongly contaminated with human genetic material.

---


```{r exonic-mapping-plot, eval="Exonic.Rate" %in% colnames(summarydata)}
p1 <- ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("") + ggtitle("Exonic mapping rate")
```


```{r intronic-mapping-plot, eval="Intronic.Rate" %in% colnames(summarydata)}
p2 <- ggplot(summarydata, aes(x=Name, y=Intronic.Rate)) +
  geom_bar(stat="identity")+
  theme_bw(base_size=10) +
  theme(panel.grid.major = element_line(size = .5, color = "grey"), axis.text.x = element_text(angle=90)) +
  ylab("intronic mapping rate") + 
  xlab("") + ggtitle("Intronic mapping rate")
```


```{r intergenic-mapping-plot, eval="Intergenic.Rate" %in% colnames(summarydata)}
p3 <- ggplot(summarydata, aes(x=Name, y=Intergenic.Rate)) +
  geom_bar(stat="identity")+
  theme_bw(base_size=10) +
  theme(panel.grid.major = element_line(size = .5, color = "grey"), axis.text.x = element_text(angle=90)) +
  ylab("intergenic mapping rate") + 
  xlab("") + ggtitle("Intergenic mapping rate")
```


```{r plot2, out.width='100%'}
grid.arrange(p1,p2,p3, nrow=2, ncol=2)
```

###The exonic, intronic and intergenic mapping rates all support the hypothesis that samples LE2, LE3 and LE4 were contaminated with human RNA, not DNA. If the contamination were DNA, you'd expect to see an even distribution of reads between exons, introns and intergenic regions instead of the majority mapping to exons.


```{r check-rRNA-eval}
eval_rRNA = "rRNA_rate" %in% colnames(summarydata) & !sum(is.na(summarydata$rRNA_rate)) == nrow(summarydata)
```

```{r rRNA-rate-plot, eval=eval_rRNA, out.width='50%'}
ggplot(summarydata, aes(x=Name, y=rRNA_rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("") + ggtitle("rRNA mapping rate")
```

###The human RNA contamination does not appear to be of rRNA origin
