---
title: "RNA-seq: expression differences in insulin-treated and untreated SH-SY5Y cells"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_depth: 4
    fig_caption: true
    fig_width: 8
    fig_height: 6
author: "Radhika Khetani"
---
```{r setup, echo=FALSE}
# Setup report details
clientname="John Flanagan"
clientemail="flanagan@hms.harvard.edu"
contactname="Melissa Hancock"
contactemail="melissa_hancock@hms.harvard.edu"
analystname="Radhika Khetani"
analystemail="rkhetani@hsph.harvard.edu"
analyst2name="Meeta Mistry"
analyst2email="mmistry@hsph.harvard.edu"
```

RNA-Seq data analysis for `r contactname` (`r contactemail`), from  `r clientname`'s (`r clientemail`) group. 
<br>Contact `r analystname` (`r analystemail`) or `r analyst2name` (`r analyst2email`)  for additional details. 
</br>

Request from client was:

> Goal is to get a list of differentially expressed genes between insulin-treated and untreated cells.

```{r setup1, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
library(rmarkdown)
library(knitrBootstrap)
```

## Workflow
* Load libraries
* Set variables
* Load data and metadata
* Quality Control: PCA and sample-to-sample correlation
* Differential Expression DESeq2
* Functional enrichment analysis: g:Profiler & Revigo

## Setup
### Bioconductor and CRAN libraries used

```{r setup2, echo=TRUE}
library(ggplot2)
library(gplots)
library(RColorBrewer)
library(CHBUtils)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(DESeq2)
library(Biobase)
library(gProfileR)
library(pheatmap)
source("~/Documents/revigo.R")
library(treemap)
```

### Set variables
```{r setup3}
baseDir <- "/Users/rkhetani/Dropbox/HBC\ consults/Flanagan_RNA-Seq/flanagan-bcbio/DESeq2/"
dataDir <- paste(baseDir, "/data", sep="")
resultsDir <- paste(baseDir, "/results", sep="")
metaDir <- paste(baseDir, "/meta", sep="")

heatcolors.1 <- rev(brewer.pal(6, "YlOrRd"))
```

### Load data and metadata
```{r load}
data <- read.delim(file.path(dataDir, 'combined.counts.SHS'), header=T, sep="\t", row.names=1, as.is=T) 
meta <- read.delim(file.path(metaDir, 'exp_design_SHS.txt'), header=T, sep=",", row.names=1)
```

## Quality control

### PCA
PCA (principal components analysis) is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expresson levels for all probes and transforms it in principal component space, reducing each sample into one point (as coordinates within that space). This allows us to separate samples according to expression variation, and identify potential outliers.

After removing **SH-ins-1 and SH-ins-4**, the 2 remaining insulin-treated replicates are clustering together. However, **SH-con-2** looks quite different from the other 2 remaining control replicates. [Note: we also removed **SH-con-1** from the analysis because it looked more like the HepG2 cell line.]

```{r pca, fig.align='center'}
pca_matrix <- prcomp(t(data))$x
df <- cbind(meta, pca_matrix[,c('PC1', 'PC2')])

ggplot(df, aes(PC1, PC2, color = treatment)) + 
  geom_text(aes(PC1, PC2, label = row.names(df)), size = 5, hjust=0.1, vjust=0.1) +
  scale_x_continuous(expand = c(0.3,  0.3))
```

### Sample-to-sample correlation heatmap

The correlation heatmap shows that the 2 insulin-treated replicates are very well correlated, and the 3 control replicates are also fairly well correlated with each other, but the correlation is lower between **SH-con-2** and each of the other 2 replicates. This is similar to the pattern observed in the PCA plot.

```{r heatmaps, fig.align='center'}
annotation <- data.frame(sampletype=meta[,'treatment'], row.names=row.names(meta))
pheatmap(cor(data, method="pearson"), color = heatcolors.1, cluster_rows = T, show_rownames = T,
               annotation=annotation, border_color=NA, cellwidth=50, cellheight=50,
               cluster_cols = F, show_colnames = T, clustering_distance_rows = "euclidean", 
               clustering_distance_cols = "euclidean", 
               fontsize = 10, fontsize_row = 8, height=10)
```

## Differential Expression
Differential expression analysis was performed using the DESeq2 package.

```{r deseq2, fig.align='center'}
eset <- new("ExpressionSet", exprs=as.matrix(data))
meta$treatment <- relevel(meta$treatment, ref="ins")
pData(eset) <- meta

dds <- DESeqDataSetFromMatrix(countData = exprs(eset), colData = pData(eset), design = ~ treatment)
dds <- DESeq(dds)
res <- results(dds)
```

1014 genes are upregulated in the insulin-treated SH cells, and 1009 genes are downregulated, at an adjusted p-value (padj) threshold of 0.01. [padj = p-value after applying the multiple test correction.]
```{r summarize}
summary(res, alpha=0.01)
```

[Note: If we apply a fold change cut off of 1.5-fold (log2fc =0.58), the number of up- and down-regulated genes reduces quite a lot, 418 genes up in insulin-treated, and 316 genes down.]

```{r summarize1, eval=FALSE,echo=FALSE}
p.cutoff <- 0.01
fc.cutoff <- 0.58
res$threshold <-  as.logical(res$padj < p.cutoff & abs(res$log2FoldChange) > fc.cutoff)

sig.genes <- res[which(res$threshold),]
summary(sig.genes)
```

```{r summarize2, eval=FALSE,echo=FALSE}
#If we apply a fold change cut off of 2-fold (log2fc =1), we 
p.cutoff <- 0.01
fc.cutoff <- 1
res$threshold <-  as.logical(res$padj < p.cutoff & abs(res$log2FoldChange) > fc.cutoff)

sig.genes <- res[which(res$threshold),]
summary(sig.genes, alpha=0.01)
```

### Volcano plot

The plot represents each gene with a dot. The fold change (log2) is plotted on the x-axis and the adjusted p-value (log10, p-value adjusted/corrected for multiple testing) is on the y-axis. The purple dots represent the 2023 genes that are differentially expressed.

```{r volcano, fig.align='center'}
p.cutoff <- 0.01
res$threshold <-  as.logical(res$padj < p.cutoff)

df <- data.frame(res)
ggplot(data=df, aes(x=log2FoldChange, y=-log10(padj), colour = threshold)) +
  scale_color_manual(values = c("grey", "purple")) +
  geom_point(alpha=0.75, pch=16) +
  xlim(c(-10,10)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5)),
        axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25))) +
  ggtitle('control vs insulin') +
  xlab("log2 fold change") + ylab("-log10 p-value")
```

### MA plot
Similar to the volcano plot, the MA plot is a great way to visualize the comparative expression metrics for a two-group comparison. The x–axis is the average/mean expression over all the samples and the y-axis is the log2 fold change between insulin-treated and control. The red dots represent the 2023 genes that are differentially expressed.

```{r MAplot, fig.align='center'}
plotMA(res, main="MA plot, alpha=0.01", ylim=c(-2,2))
```

```{r extra, eval=F, echo=FALSE}
### Final summary and results

#write.csv(as.data.frame(res), file="/Users/rkhetani/Dropbox/HBC consults/Flanagan_RNA-Seq/flanagan-bcbio/DESeq2/results/SHS_DE-results.csv")
library(biomaRt)
mart<- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
norm.count <- counts(dds, normalized=T)

gene.names <- getBM(
  filters= "ensembl_gene_id", 
  attributes= c("ensembl_gene_id", "external_gene_name"),
  values= row.names(norm.count),
  mart= mart)

ens.id <- row.names(norm.count)
GeneName <- gene.names[match(ens.id,gene.names$ensembl_gene_id),"external_gene_name"]
new <- data.frame(norm.count,GeneName)
write.table(new, "results//normalized_counts.txt", sep="\t")
```

### g:Profiler
We ran the list of significantly different genes through [g:Profiler](http://biit.cs.ut.ee/gprofiler/), a tool that performs statistical enrichment analysis to find over-representation of information like Gene Ontology terms, biological pathways, regulatory DNA elements, human disease gene annotations, and protein-protein interaction networks.

We took the results from g:Profiler and used the significant (fdr>=0.05) GO terms as input to [REVIGO](http://revigo.irb.hr/) to visualize the most prominent processes and reduce redundancy of terms. Each rectangle is a single cluster representative. The representatives are joined into ‘superclusters’ of loosely related terms, visualized with different colors. Size of the rectangles may be adjusted to reflect either the p-value, or the frequency of the GO term in the underlying GOA database.

```{r functional, fig.align='center'}
sig.genes <- res[which(res$threshold),]
sig.genenames <- rownames(sig.genes)
query <- unique(sig.genenames)
  
gprofiler_results  <-   gprofiler(query = query, 
            organism = "mmusculus",
            ordered_query = F, 
            exclude_iea = F, 
            max_p_value = 0.05, 
            max_set_size = 0,
            correction_method = "fdr",
            hier_filtering = "none", 
            domain_size = "annotated",
            custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep('GO:', allterms)]
pvals <- gprofiler_results$p.value[grep('GO:', allterms)]
GO.pval <- cbind(gprofiler_results$term.id,gprofiler_results$p.value)
runRevigo(GOs, pvals)
```

## Files for download
Below we have listed links to the various files generated in this report which will be useful for any downstream analysis.

### Data Files
* Normalized expression matrix [here](./data/SH-normalized_counts.txt)

### DE results
* Complete [table of results](./results/SH-DESeq2-results.txt)
* Table with results of only the [significant genes](./results/SH-significant.genes.txt)

*Legend for DE results:*
<br>
**baseMean** - mean of normalized counts for all samples
<br>
**log2FoldChange** - log2 fold change: insulin-treated/control
<br>
**lfcSE** -  standard error
<br>
**stat** - Wald statistic
<br>
**pvalue** - Wald test p-value
<br>
**padj** - BH adjusted p-values
</br>

### Functional enrichment results
* Table of [gprofiler results](./results/SH-gprofiler_results.txt)
* Table of [enriched GO terms and associated p-values](./results/SH-GOterms_gprofiler.txt) extracted from the gprofiler results