---
title: "RNA-seq: expression differences in insulin-treated and untreated HepG2 cells"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_depth: 4
    fig_caption: true
    fig_width: 8
    fig_height: 6
author: "Radhika Khetani"
---
```{r setup, echo=FALSE}
# Setup report details
clientname="John Flanagan"
clientemail="flanagan@hms.harvard.edu"
contactname="Melissa Hancock"
contactemail="melissa_hancock@hms.harvard.edu"
analystname="Radhika Khetani"
analystemail="rkhetani@hsph.harvard.edu"
analyst2name="Meeta Mistry"
analyst2email="mmistry@hsph.harvard.edu"
```

RNA-Seq data analysis for `r contactname` (`r contactemail`), from  `r clientname`'s (`r clientemail`) group. 
<br>Contact `r analystname` (`r analystemail`) or `r analyst2name` (`r analyst2email`)  for additional details. 
</br>

Request from client was:

> Goal is to get a list of differentially expressed genes between insulin-treated and untreated cells.

The DE analysis in this report excludes 1 of the 3 insulin replicates that [looked like an outlier](https://dl.dropboxusercontent.com/u/74036176/Flanagan_DESeq2/DESeq2-flanagan-HepG2.html#quality-control), **HepG2-ins-1**.

```{r setup1, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
library(rmarkdown)
library(knitrBootstrap)
```

## Workflow
* Load libraries
* Set variables
* Load data and metadata
* Quality Control: PCA and sample-to-sample correlation
* Differential Expression DESeq2

## Setup
### Bioconductor and CRAN libraries used

```{r setup2, echo=TRUE}
library(ggplot2)
library(gplots)
library(RColorBrewer)
library(CHBUtils)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(DESeq2)
library(Biobase)
library(gProfileR)
library(pheatmap)
source("~/Documents/revigo.R")
library(treemap)
```

### Set variables
```{r setup3}
baseDir <- "/Users/rkhetani/Dropbox/HBC\ consults/Flanagan_RNA-Seq/flanagan-bcbio/DESeq2/"
dataDir <- paste(baseDir, "/data", sep="")
resultsDir <- paste(baseDir, "/results", sep="")
metaDir <- paste(baseDir, "/meta", sep="")

heatcolors.2 <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
```

### Load data and metadata
```{r load}
data <- read.delim(file.path(dataDir, 'combined.counts_ALT.HepG2'), header=T, sep="\t", row.names=1, as.is=T) 
meta <- read.delim(file.path(metaDir, 'exp_design_HepG2_ALT.txt'), header=T, sep=",", row.names=1)
```

## Quality control

### PCA
PCA (principal components analysis) is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expresson levels for all probes and transforms it in principal component space, reducing each sample into one point (as coordinates within that space). This allows us to separate samples according to expression variation, and identify potential outliers.

```{r pca, fig.align='center'}
pca_matrix <- prcomp(t(data))$x
df <- cbind(meta, pca_matrix[,c('PC1', 'PC2')])

ggplot(df, aes(PC1, PC2, color = treatment)) + 
  geom_text(aes(PC1, PC2, label = row.names(df)), size = 5, hjust=0.1, vjust=0.1) +
  scale_x_continuous(expand = c(0.3,  0.3))
```

### Sample-to-sample correlation heatmap

According to this, the samples from the same day (replicates 2 from both, and replicates 3 from both) are better correlated with each other than within the groups (insulin-treated or control).                          
```{r heatmaps, fig.align='center'}
annotation <- data.frame(sampletype=meta[,'treatment'], row.names=row.names(meta))
pheatmap(cor(data, method="pearson"), color = heatcolors.2, cluster_rows = T, show_rownames = T,
               annotation=annotation, border_color=NA, cellwidth=40, cellheight=40,
               cluster_cols = T, show_colnames = T, clustering_distance_rows = "euclidean", 
               clustering_distance_cols = "euclidean", 
               fontsize = 10, fontsize_row = 8, height=10)
```

## Differential Expression
Differential expression analysis was performed using the DESeq2 package.

```{r deseq2, fig.align='center'}
eset <- new("ExpressionSet", exprs=as.matrix(data))
meta$treatment <- relevel(meta$treatment, ref="control")
pData(eset) <- meta

dds <- DESeqDataSetFromMatrix(countData = exprs(eset), colData = pData(eset), design = ~ treatment)
dds <- DESeq(dds)
res <- results(dds)
```

At an adjusted p-value (padj) threshold of 0.1, **there is 1 differentially regulated gene between treated and control samples in this cell type**, LOX (ENSG00000113083).

[Note: The padj threshold applied here is much higher than the one applied for the SH cell type. (padj = p-value after applying the multiple test correction.)]

```{r summarize}
summary(res, alpha=0.1)
```

### Volcano plot
The plot represents each gene with a dot. The fold change (log2) is plotted on the x-axis and the adjusted p-value (log10, p-value adjusted/corrected for multiple testing) is on the y-axis. The purple dot represent the gene that is differentially expressed.

```{r volcano, fig.align='center'}

p.cutoff <- 0.1
res$threshold <- as.logical(res$padj < p.cutoff)

df <- data.frame(res)
ggplot(data=df, aes(x=log2FoldChange, y=-log10(padj), colour = threshold)) +
  scale_color_manual(values = c("grey", "purple")) +
  geom_point(alpha=0.75, pch=16) +
  xlim(c(-10,10)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5)),
        axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25))) +
  ggtitle('control vs insulin') +
  xlab("log2 fold change") + ylab("-log10 p-value")
```

### MA plot
```{r MAplot, fig.align='center'}
plotMA(res, main="MA plot, alpha=0.01", ylim=c(-2,2))
```

***I did not proceed with g:Profiler and Revigo in this case, since the list has only 1 gene.***