---
output:
  html_document:
    toc: true
    toc_depth: 3
    theme: journal
    css: custom.css
title: "Standard Microarray Analysis of U133Plus2.0 Sepsis Studies"
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, cache=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

# STANDARD MICROARRAY ANALYSES OF U133PLUS2.0 STUDIES

---

Analysis of public datasets for differential gene expression signatures between
adults and children for [Les Kobzik](mailto:LKOBZIK@hsph.harvard.edu) as part of
the DARPA 7-day Biodefense Program.

Contact [John Hutchinson](mailto:jhutchin@hsph.harvard.edu) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

----

# GENERAL SETUP

## General R libraries

```{r general_libraries}
library(xtable) # table generation for reports
library(plyr) # library for iteratively working with data structures
library(ggplot2) # library for plotting 
library(RColorBrewer) # library for generating color palettes
library(googleVis) # library for presenting tables
library(CHBUtils)
```

## Data locations
-  directories for data, metadata and results


```{r general_directories}
if (file.exists("/n/hsphS10/hsphfs1/chb/projects/lk_FOY/sepsis")) {
  baseDir <- "/n/hsphS10/hsphfs1/chb/projects/lk_FOY/sepsis"
  }  else if (file.exists("/Volumes/home08/jhutchin/consults/lk_FOY/sepsis")) {
    baseDir <- "/Volumes/home08/jhutchin/consults/lk_FOY/sepsis"
    } else {
      baseDir <- "/Volumes/ody/consults/lk_FOY/sepsis"
      }
dataDir <- file.path(baseDir, "data", "WB")
resultsDir <- file.path(baseDir, "results", "WB", "U133Plus2")
metaDir <- file.path(baseDir, "meta", "WB")
````

----

# Microarray Analysis
- currently only looking at the HG-U133-2.0 Plus arrays, so subsetted the 
unified metadata file to this array (GPL570 for GEO studies, A-AFFY-44 for Array Express studies)
- subsetted to samples for which we have age and raw data
- also remove GSE11375 and GSE36809

```{r subset_data}
refined.metadata <- read.delim(file.path(metaDir,"unified.metadata.refined.tab"))
refined.metadata=refined.metadata[!(refined.metadata$study %in% c("GSE11375", "GSE36809")),]
U133.Plus2.data <- refined.metadata[which(!is.na(refined.metadata$age) & 
  !is.na(refined.metadata$CEL_regex) & 
  grepl("GPL570|A-AFFY-44", refined.metadata$platform)),]

write.table(U133.Plus2.data, 
            file.path(metaDir, "unified.metadata.refined.U133Plus2.0.tab"), 
            quote=F, 
            sep="\t", 
            row.names=F, 
            col.names=T)
```

That metadata file is [here](../meta/WB/unified.metadata.refined.U133Plus2.0.tab)

## Setup

### Microarray libraries
- from [Bioconductor][1]

```{r microarray_analysis_libraries}
# to parse the CEL files and work with intensity values
library(affy) 
# for QC reports
library(arrayQualityMetrics)
# library to do stats 
library(limma) 
# pretty heatmaps
library(pheatmap) 
# annotations for the hgU1332.0Plus array
library(hgu133plus2.db) 
```

### Analysis variables
- color palette for plots
- age ranges and names of the groups  
  - neonate = 0-4 years
  - child = 5-12 years
  - adult = 18-40 years

```{r microarray_analysis_variables}
# colorblind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
# age ranges to compare
child.age.range <- c(5,12)
adult.age.range <- c(18,40)
neonate.age.range <- c(0,4)
# THESE CANNOT OVERLAP!
```

### Metadata
- load and parse the unified, refined and subsetted metadata file 
- to make a covariate table that allows you to load the CEL files using the [affy][2] Bioconductor library
- this table has the CEL file location in the first column, without a header
- the other columns contain the sampleIDs, ages, genders,ethnicities and respective studies

- also limit samples to those wihtin the age ranges described above and label each with their "stage"

```{r covariatedataframe, cache=TRUE}
# U1332.0Plus only for now
covartemplate.file <- "unified.metadata.refined.U133Plus2.0.tab"
covartemplate <- read.table(file.path(metaDir,covartemplate.file ), 
                            header=T, 
                            colClasses="character", 
                            sep="\t")

# convert character ages to numeric to allow numeric subsetting 
covartemplate$age <- as.numeric(covartemplate$age) 

## limit samples to those within the described child, adult and neonate age ranges
children.indices <- which(covartemplate$age<=max(child.age.range) & 
  covartemplate$age>=min(child.age.range))
neonate.indices <- which(covartemplate$age<=max(neonate.age.range))
adult.indices <- which(covartemplate$age<=max(adult.age.range) & 
  covartemplate$age>=min(adult.age.range))
sample.indices <- c(neonate.indices, children.indices, adult.indices)
covartemplate <- covartemplate[sample.indices,]

# convert age back to character value so you can use it in a character matrix
covartemplate$age <- as.character(covartemplate$age) 
covartemplate <- as.matrix(covartemplate)
covars <- aaply(covartemplate, 1, function(n){
  # pull out the info you need piece by piece
  sampleID <- n[grep("sampleID", colnames(covartemplate))] 
  arrayID <- n[grep("arrayID", colnames(covartemplate))]
  age <- n[grep("^age", colnames(covartemplate))]
  gender <- n[grep("gender", colnames(covartemplate))]
  stage <- n[grep("stage", colnames(covartemplate))]
  outcome <- n[grep("outcome", colnames(covartemplate))]
  CELregex <- n[grep("CEL_regex", colnames(covartemplate))]
  CELFTP <- n[grep("CEL_FTP", colnames(covartemplate))]
  study=n[grep("study", colnames(covartemplate))]
  if(!n[grep("study", colnames(covartemplate)) ] %in% c("GSE4607", "BD")){
    # for non GSE4607 studies, get the CEL file name from the FTP location in the metadata file
    CELfileloc <- file.path(dataDir,study, sub(".gz", "", basename(as.character(CELFTP)))) 
  } else {
    # for GSE4607, get the CEL file name directly from the CEL file regex in the metadata file
    CELfileloc <- file.path(dataDir, study, CELregex) 
    }
  if(!file.exists(CELfileloc)){
    # check if the CEL file is actually where its supposed to be, if not notify and move on
    print(paste(CELfileloc, "does not exist", sep=" ")) 
    #print(CELFTP)
    stop()
    } else {
      # if CEL file is where its supposed to be, add location to covariate table
      return(list(CELfileloc=CELfileloc, 
                  sampleID=sampleID, 
                  arrayID=arrayID, 
                  age=age, 
                  gender=gender, 
                  stage=stage, 
                  outcome=outcome, 
                  study=study)
             ) 
      }
  })
covars <- as.data.frame(covars)
covars$age <- as.numeric(covars$age)
# label samples with their stage as determined by age
# don't need to specify adult ranges, as we limited the dataset above to neonates, children and adults
covars$stage <- ifelse(covars$age<=max(neonate.age.range), "neonate", 
                       ifelse(covars$age>=min(child.age.range)
                              & covars$age<=max(child.age.range), "child", 
                              ifelse(covars$age>=min(adult.age.range)
                                     & covars$age<=max(adult.age.range), "adult", NA)))
covars <- covars[order(covars$stage),]
#remove unclassified samples
covars <- covars[!is.na(covars$stage),]
```

### CEL files

```{r load_data, eval=FALSE}
covars <- covars[!(grepl("1088631.cel", covars$CELfileloc)),]
mic.raw <- ReadAffy(filenames=as.character(covars$CELfileloc), phenoData=covars)
save(list="mic.raw", file=file.path(resultsDir, "RDATA.mic.raw.ori"))
```

## Quality Control and Normalization

### Raw QC
- array QC checks on raw data with [ArrayQualityMetrics][3]

```{r rawQC, eval=FALSE}
  arrayQualityMetrics(expressionset=mic.raw, 
                    outdir=file.path(resultsDir, "QCreport_raw"), 
                    force=TRUE, 
                    do.logtransform=TRUE, 
                    intgroup=c("stage", "study"))
````

[Raw Data QC Report](../results/WB/U133Plus2/QCreport_raw/index.html)

### Raw outliers
- arrays that failed array intensity, RLE or NUSE analyses during raw data QC were excluded.

```{r exclude_from_raw, eval=FALSE}
# subset ExpressionSet to studies that are NOT the following
exclude.arrays <-c("M036","GSM1373917","GSM102978","GSM102993","GSM103049","GSM648590","GSM648596","GSM647548","GSM244499","GSM102997","GSM102998","GSM103013","GSM103015","GSM103057","GSM103060","GSM648591","GSM648628","GSM648643","GSM648647","GSM648682","GSM648704","GSM648706","GSM647529","GSM647530","GSM647532","GSM647594","GSM647597")
mic.raw <- mic.raw[,which(!(pData(mic.raw)$arrayID %in% exclude.arrays))]
# change the sample names to more informative
sampleNames(mic.raw) <- as.vector(unlist(pData(mic.raw)$arrayID))
save(list="mic.raw", file=file.path(resultsDir, "RDATA.mic.raw.no.raw.outliers"))
```

### Normalize 
- all arrays using [RMA][4]

```{r normalize_RMA, eval=FALSE}
mic.norm <- rma(mic.raw,normalize=TRUE,background=TRUE)
save(list="mic.norm", file=file.path(resultsDir, "RDATA.mic.norm.no.raw.outliers"))
```


### Normalized QC
- array QC checks on normalized data with ArrayQualityMetrics

```{r normQC, eval=FALSE}
arrayQualityMetrics(expressionset=mic.norm, 
                    outdir=file.path(resultsDir, "QCreport_norm"), 
                    force=TRUE,
                    do.logtransform=FALSE, 
                    intgroup=c("stage", "study"))
```

[Normalized Data QC Report](../results/WB/U133Plus2/QCreport_norm/index.html)

### Normalized outliers

```{r drop_outliers, eval=FALSE}
# subset ExpressionSet to samples that are NOT the following
exclude.arrays <- c("GSM648718","GSM647628","GSM102983","GSM102984","GSM103007","GSM648645","GSM648656","GSM647531","GSM647561","GSM647573","GSM647574","GSM647600")
    
mic.raw <- mic.raw[,which(!(pData(mic.raw)$arrayID) %in% exclude.arrays)]
save(list="mic.raw", file=file.path(resultsDir, "RDATA.mic.raw.no.raw.or.norm.outliers"))
```

### Renormalize without outliers

```{r normalize_RMA2, eval=FALSE}
mic.norm <- rma(mic.raw, normalize=TRUE, background=TRUE)
save(list="mic.norm", file=file.path(resultsDir, "RDATA.mic.norm.no.raw.no.norm.outliers"))
```


```{r load_affybatch, echo=FALSE}
# this is a hack to get around knitr having memory issues with loading and parsing CEL files
# instead of having knitr do the CEL file manipulation, do it interactively and save out an RDATA object that you can load into the final knitr run
load(file.path(resultsDir, "RDATA.mic.norm.no.raw.no.norm.outliers"))
```

### Metadata 
- for remaining U133Plus2.0 studies

```{r print_metadata, results='asis'}
pd <- pData(mic.norm)
pd.gvis <- gvisTable(as.data.frame(apply(pd, 2, as.character)), options=list(width=960))  
print(pd.gvis, "chart")
# rename the samples with the arrayID (to get rid of weird array names)
sampleNames(mic.norm) <- as.vector(unlist(pData(mic.norm)$arrayID))
```

### ALL  Expression Values
- for ALL Genes and Samples

```{r allexprs, cache=TRUE}
exprs.all <-exprs(mic.norm)
probeIDs <- row.names(exprs.all)

symbols <- unlist(mget(rownames(exprs.all), hgu133plus2SYMBOL, ifnotfound=NA))
exprs.all <- cbind(cbind(probeIDs, symbols), exprs.all)
write.table(exprs.all, file.path(resultsDir, "U133_2.0Plus.all.exprs.xls"), sep="\t", row.names=F, col.names=T)
```

[All Expression Values for all Samples](../results/WB/U133Plus2/U133_2.0Plus.all.exprs.xls)  

## Differential Expression Analyses

### Limma

A linear model for microarray data analysis [Limma][5] was performed on the samples to identify differentially expressed genes between all pair-wise comparisons of the sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we construct two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.

#### Design matrix 
- make a matrix with arrays as rows, sample groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group 

```{r design, results="asis"}
design <- model.matrix(~ 0+stage, data=pData(mic.norm))
# make sure the headings match
colnames(design) <- sub("stage", "", colnames(design))
design.gvis <- gvisTable(as.data.frame(apply(row2colnames(design,"ID"), 2, as.character)), options=list(width=640))  
print(design.gvis, "chart")
```

#### Contrast matrix
- to perform specified pairwise comparisons
- in this table, columns are contrasts/comparisons and rows are sample groups
-  a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts(adult-child,adult-neonate,neonate-child, levels=dimnames(design)[[2]])
contrast.gvis <- gvisTable(as.data.frame(apply(row2colnames(contrast.matrix,"contrast"), 2, as.character)), options=list(width=240, height=120))  
print(contrast.gvis, "chart")
```

These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix for each gene based on the given series of arrays
- using the contrast matrix, compute estimated coefficients and standard errors for contrasts
- compute moderated t-statistics and log-odds of differential expression by empirical Bayes shrinkage of the standard errors towards a common value

#### Fit a linear model
- for each gene based on the given series of arrays

```{r linearmodel}
exprs.norm <- exprs(mic.norm)
fit.exprs <- lmFit(exprs.norm, design) 
```

#### Compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2.exprs <- contrasts.fit(fit.exprs, contrast.matrix) 
```

#### Compute moderated t-statistics and log-odds of differential expression 
- by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.exprs <- eBayes(fit2.exprs) 
```

### Calculate statistics for comparisons
- generate statistics describing level of differential expression of genes, as sorted by unadjusted pvalue

```{r calcstats, results='hide', cache=TRUE}
stats.exprs <- lapply(seq(1,3,1), function(n) {
  contrast <- dimnames(fit2.exprs$contrasts)$Contrasts[n]
  pd.contrast <- pData(mic.norm)[apply(design[,colnames(design) %in% names(which(abs(contrast.matrix[,contrast])==1))]==1, 1,any),]
  # calculate statistics for this contrast for ALL genes
  stats.contrast <- topTable(fit2.exprs, coef=n, adjust="fdr", sort.by="p", number=nrow(exprs.norm), genelist=row.names(exprs.norm)) 
  # get expression levels for these genes in the samples involved in this contrast
  exprs.contrast <- exprs.norm[,apply(design[,colnames(design) %in% names(which(abs(contrast.matrix[,contrast])==1))]==1, 1,any)]
  # sanity check
  identical(colnames(exprs.contrast), as.vector(unlist(pd.contrast$arrayID)))
  # rearrange expression values to match order of stats output above
  exprs.contrast <- exprs.contrast[stats.contrast$ID, ]
  # bind columns of stats to columns of sample expressioin values
  stats.exprs.contrast <- cbind(stats.contrast, exprs.contrast)
  # output to file
  write.table(stats.exprs.contrast, file.path(resultsDir, paste("U133_2.0Plus.stats.exprs", gsub(" ", "", contrast), "xls", sep=".")), sep="\t", row.names=F, col.names=T)
  # sanity check
  identical(colnames(exprs.contrast), as.vector(unlist(pd.contrast$arrayID)))
  # save for later use in heatmaps
  return(list(contrast=contrast, pd.contrast=pd.contrast, stats.exprs.contrast=stats.exprs.contrast))
  })
```

#### Excel files with statistics and normalized sample intensities for all genes and samples for each comparison
#### The statistics summary tables contain the following information: 
- logFC is the log2-fold change
- the AveExpr is the average expression value accross all arrays
- the moderated t-statistic (t) is the logFC to its standard error, the P.Value is the associated p-value
- the adj.P.Val is the p-value adjusted for multiple testing (by FDR) 
- the B-value (B) is the log-odds that a gene is differentially expressed (the-higher-the-better)

[Adult vs. Child](../results/WB/U133Plus2/U133_2.0Plus.stats.exprs.adult-child.xls)  
[Adult vs. Neonate](../results/WB/U133Plus2/U133_2.0Plus.stats.exprs.adult-neonate.xls)  
[Neonate vs. Child](../results/WB/U133Plus2/U133_2.0Plus.stats.exprs.neonate-child.xls) 

### Heatmaps
- of top differentially expressed genes
- for each pairwise comparison

Here I first looked for genes with at least a 2 fold change in expression and an adjusted pvalue <0.1. The results were sorted by the unadjusted pvalue and the top 100 taken for the heatmaps.

There is one heatmap for each comparison  with both genders and unknown genders included.

*note that the differentially expressed genes clustered in these heatmaps are based on comparisons that include both genders* 

```{r heatmaps}
for(n in 1:3){
  # get the names of the stages that are being compared in this comparison
  contrast <- stats.exprs[[n]]$contrast
  pd.contrast <- stats.exprs[[n]]$pd.contrast
  stats.exprs.contrast <- stats.exprs[[n]]$stats.exprs.contrast
  top.stats.exprs.contrast <- subset(stats.exprs.contrast, abs(logFC)>=1 & adj.P.Val<=0.1)
  # remove AFFY control probes
  if (any(grepl("AFFX", top.stats.exprs.contrast$ID))){
    top.stats.exprs.contrast <- top.stats.exprs.contrast[-(grep("AFFX", top.stats.exprs.contrast$ID)),]
    }
  # subset to top 100 (if that many present)
  if (nrow(top.stats.exprs.contrast)>=100){
    top.stats.exprs.contrast <- top.stats.exprs.contrast[1:100,]
    }  
  # extract expression values for samples
  top.exprs.contrast <- top.stats.exprs.contrast[,colnames(top.stats.exprs.contrast) %in% pd.contrast$arrayID]
  # setup row names for the heatmap, paste probeset ID to gene symbol
  row.names(top.exprs.contrast) <- paste(as.vector(unlist(mget(top.stats.exprs.contrast$ID, hgu133plus2SYMBOL, ifnotfound=NA))), " (" ,top.stats.exprs.contrast$ID, ")", sep="")

  # heatmap annotations
  heatmap.annots <- pd.contrast[,c("arrayID", "study", "stage", "gender")]
  heatmap.annots <- as.data.frame(apply(heatmap.annots, 2, unlist))
  row.names(heatmap.annots) <- heatmap.annots$arrayID
  heatmap.annots$arrayID <- NULL
  # heatmap annotation colors
  study_colors <- cbPalette
  names(study_colors) <- unique(unlist(pd.contrast$study))
  stage_colors <- c("white", "darkgrey", "black")
  names(stage_colors) <- unique(unlist(pd.contrast$stage))
  gender_colors <- cbPalette[c(3,5,8)]
  names(gender_colors) <- c("MALE", "UNKNOWN", "FEMALE")
  ann_colors = list(gender=gender_colors, study = study_colors, stage = stage_colors)
  ## Heatmaps
  svg(filename = file.path(resultsDir, paste("DE_heatmap", gsub(" ", "", contrast), "svg", sep=".")), width=8, height=11)
  plot.new()
  pheatmap(as.matrix(top.exprs.contrast), 
           annotation=heatmap.annots, 
           color=rev(brewer.pal(11,"RdBu")), 
           cluster_cols = FALSE,
           main=paste(contrast,  "- Unclustered", sep=""), 
           show_colnames=F, 
           fontsize=12, 
           fontsize_row=8,
           annotation_colors=ann_colors, 
           cellwidth=1 ) 
  dev.off()
  }
```

![adult-child heatmap](../results/WB/U133Plus2/DE_heatmap.adult-child.svg)
![adult-neonate heatmap](../results/WB/U133Plus2/DE_heatmap.adult-neonate.svg)
![neonate-child heatmap](../results/WB/U133Plus2/DE_heatmap.neonate-child.svg)

### Heatmaps - ALL studies
- for all top 100 differentially expressed genes from each pairwise comparison in all samples
- wanted to see if patterns were consistent across all the age groups
- combined the top hits from all three comparisons, and looked at their expression patterns in all samples

```{r heatmaps_all_samples}
top.IDs <- unique(unlist(lapply(stats.exprs, function(n) {
  stats.exprs.contrast <- n$stats.exprs.contrast
  top.stats.exprs.contrast <- subset(stats.exprs.contrast, abs(logFC)>=1 & adj.P.Val<=0.1)
  # remove AFFY control probes
  if (any(grepl("AFFX", top.stats.exprs.contrast$ID))){
    top.stats.exprs.contrast <- top.stats.exprs.contrast[-(grep("AFFX", top.stats.exprs.contrast$ID)),]
    }
  # subset to top 100 (if that many present)
  if (nrow(top.stats.exprs.contrast)>=100){
    top.stats.exprs.contrast <- top.stats.exprs.contrast[1:100,]
    } 
  top.stats.exprs.contrast$ID
  })))

# get expression values for these probes
top.exprs.union <- exprs.norm[top.IDs,]

# row labels - add gene symbol to probeset id
row.names(top.exprs.union) <- paste(as.vector(unlist(mget(row.names(top.exprs.union), hgu133plus2SYMBOL, ifnotfound=NA))), " (" ,row.names(top.exprs.union), ")", sep="")
# annotations
pd <- pData(mic.norm)
heatmap.annots <- pd[,c("arrayID", "study", "stage", "gender")]
heatmap.annots <- as.data.frame(apply(heatmap.annots, 2, unlist))
row.names(heatmap.annots) <- heatmap.annots$arrayID
heatmap.annots$arrayID <- NULL
# annotation colors
  study_colors <- cbPalette
  names(study_colors) <- unique(unlist(pd$study))
  stage_colors <- c("white", "darkgrey", "black")
  names(stage_colors) <- unique(unlist(pd$stage))
  gender_colors <- cbPalette[c(3,5,8)]
  names(gender_colors) <- c("MALE", "UNKNOWN", "FEMALE")
  ann_colors = list(gender=gender_colors, study = study_colors, stage = stage_colors)
## Heatmaps
svg(filename = file.path(resultsDir, "allheatmap.svg"), width=8, height=11)
plot.new()

pheatmap(as.matrix(top.exprs.union), 
         annotation=heatmap.annots, 
         color=rev(brewer.pal(11,"RdBu")), 
         cluster_cols = FALSE, 
         main="All Comparisons, All Samples - Unclustered", 
         show_colnames=F, 
         fontsize=12,
         fontsize_row=8,
         annotation_colors=ann_colors,
         cellwidth=1)
dev.off()

```


![all heatmap](../results/WB/U133Plus2/allheatmap.svg)

[1]: http://www.bioconductor.org (BioC)
[2]: http://www.bioconductor.org/packages/release/bioc/html/affy.html (affy)
[3]: http://www.bioconductor.org/packages/2.10/bioc/html/arrayQualityMetrics.html (arrayQualityMetrics)
[4]: http://bioinformatics.oxfordjournals.org/content/19/2/185 (RMA)
[5]: http://www.bioconductor.org/packages/release/bioc/html/limma.html (limma)

