---
title: "DESeq Time Course Analysis"
output:
  html_document:
    toc: true
    theme: united
    fig_width: 12
    fig_height: 12
    fig_caption: false
    highlight: textmate
---


```{r setup, echo=FALSE}
options(width=200)
library(knitr)
opts_chunk$set(tidy=TRUE, cache=FALSE, echo=FALSE, highlight=TRUE, figalign="center", warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120))

clientname="Danielle Gottlieb"
clientemail="danielle.gottlieb@cardio.chboston.org"
lablocation="Children's Hospital"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```
---

RNAseq (DESeq) analysis for `r clientname` (`r clientemail`), at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

---

# METHODS SUMMARY

Raw read counts were then normalized and differential gene expression calculated with the [DESeq][4] Bioconductor package.

The sections below provide code to reproduce the post-counting results and plots. 

---

# SETUP

## Libraries

Bioconductor and R libraries used to process and visualize the data

```{r libraries}
library(DESeq)
library(plyr)
library(reshape2)
library(ggplot2)
library(biomaRt)
library(magrittr)
library(devtools)
install_git("git://github.com/hbc/CHBUtils.git") # misc personal utilities
library(CHBUtils)
library(knitr)
# colorblind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")

library(biomaRt)
ensemblmart <- useMart('ensembl', "hsapiens_gene_ensembl")
attributes <- listAttributes(ensemblmart)
filters <- listFilters(ensemblmart)
```

## Variables

Working directories, cutoffs and other variables necessary to the analysis.

```{r variables}
  baseDir <- "~/projects/dg_cardiac_timeseries"

metaDir <- file.path(baseDir, "meta")
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
rawdata.count.file <- file.path(dataDir, "rawcount_data.csv")
normdata.count.file <-file.path(dataDir, "normalized_data.csv") 

pvalcutoff=0.05
numsig=10
rowcount.cutoff=10
logfoldcutoff=1
```


For this analysis, the cutoff for significant differential expression was an FDR (False Discovery Rate) of `r pvalcutoff` and a log2(foldchange) of expression of at least `r logfoldcutoff`.  

--- 

# DATA IMPORT

The data consists of  counts generated for individual genes from the aligments. The metadata (or covariates) that describes the experiment consists of the  information about sample ID and timepoint.

These data and metadata were and loaded into a DESeq *CountDataSet*.

```{r dataload_and_reshape, results='asis'}
raw.counts <- read.csv(rawdata.count.file, header=T, row.names=1)
names(raw.counts) <- names(raw.counts) %>% sub("_reads_", "", .)
norm.counts <- read.csv(normdata.count.file, header=T, row.names=1)
names(norm.counts) <- names(norm.counts) %>% sub("_norm_", "", .)
# safety check
identical(names(norm.counts), names(raw.counts))

# use ages as covariates
ages <- names(raw.counts) %>% strsplit(., "_") %>% lapply(., function(x) x[2]) %>% unlist(.) %>% sub("wks", "", .) %>% as.numeric(.)
covars <- ages
## load up new count dataset
cds <- newCountDataSet(raw.counts, conditions=covars)
```

---

# DATA PREPROCESSING

## Normalization

To normalize for different depth of sequencing, size factors are calculated for each sample. 

Call $k_{ij}$ the count for gene $i$ in sample $j$. 

For all samples $j$, assume most of the counts for genes will have a constant ratio comparing any two samples in $j$, reflecting the assumption that most genes are not differentially expressed between samples. 

A reference set of counts is created by taking $f_i$, geometric mean of the counts for each gene $i$, and the median of of $\frac{k_i}{f_i}$ for sample $j$ is called the size factor for sample $j$. 

The size factors can be thought of as an attempt to normalize for the total number of reads mapping to transcripts, depending on the assumption that most of the transcripts are not vastly different between samples.

Here are the size factors for each sample in this experiment:

```{r sizefactors, results='asis'}
cds = estimateSizeFactors(cds)
kable(as.data.frame(sizeFactors(cds)))
```

## Variance estimation

With replicates you can get a more accurate repesentation of the biological variability or dispersion. Dispersion describes the fluctuation of the observed counts for a gene around the mean counts for that gene and can be the square of the coefficient of biological variation. For example, if a gene's expression level fluctuates by 20% the calculated dispersion is $0.2^2 = 0.04$.

We can calculate the dispersion from our  replicates. We pool them as we don't have alot of replicates per condition.

i.e. we assume that the estimated variance should not be too affected by the influence of the few differentially expressed genes

After the empirical dispersion values have been calculated for each gene, a dispersion-mean relationship is fitted to share information across genes to reduce the variability of the dispersion estimates. So we end up with two values for each gene: the empirical value (derived only from this gene's data), and the fitted value (i.e., the dispersion value typical for genes with an average expression similar to those of this gene). Because we few replicates, and very few samples, we can only use the fitted value. Other methods or choosing between the empirical and fitted gene dispersion values aredesigned to detect false positives from dispersion outliers (i.e. genes with unusually high variability), but without replicates we can't catch these outliers.

Below is a plot of the estimated dispersion for genes with given counts; the dispersion increases as the mean count decreases meaning it is more difficult to accurately measure the expression level of
low-count genes. The red line represents the fitted value we will use.

```{r estimate_sizefactors, results='hide', out.width='75%'}
# sharingMode = maximum, most conservative approach to sharing information across genes to reduce variability of the dispersion estimates
cds <- estimateDispersions(cds, method="pooled", sharingMode="maximum")
ggplotDispEsts(cds)
```

---

## Quantitative Assessment of Sample Differences by Differential Expression

### General Linearized Model

```{r pairwise_comparisons, cache=TRUE}
fit1 <- fitNbinomGLMs(cds, count ~ pData(cds)$condition) #fits to the formula that the time.course is a covariate
fit0 <- fitNbinomGLMs(cds, count ~ 1 ) #the null hypothesis
stats <- nbinomGLMTest(fit1, fit0 ) #chi squared test for each gene
adj.stats <- p.adjust(stats, method="BH")
```

### Absolute count difference over time course

```{r absdiff}
library(magrittr)
allcounts <- counts(cds, normalize=T)
diffs <- allcounts %>% apply(., 1, function(x) range(na.omit(x))) %>% apply(., 2, diff)
```

### Output results

```{r output}
cds <- cds[,order(ages)]
allcounts <- counts(cds, normalize=T)
genesymbols <- row.names(allcounts)
rowcounts <- rowSums(allcounts)

output <- do.call(cbind, list(rowcounts,stats, adj.stats, allcounts)) %>% as.data.frame(.)
output <- cbind(genesymbols, output, stringsAsFactors=FALSE)

names(output)[1:4] <- c("symbol", "rowcount", "p.value", "adjusted.p.value")

write.table(output, file=file.path(resultsDir, "deseq.stats.xlsx"), sep="\t", quote=F, row.names=T, col.names = NA)
```

