---
output:
  html_document:
    toc: true
    toc_depth: 3
    theme: journal
title: "Preparing data for STEM analysis"
bibliography: "references.bib"
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(knitcitations)
cleanbib()
options("citation_format" = "pandoc")

clientname="Danielle Gottlieb"
clientemail="danielle.gottlieb@cardio.chboston.org"
labPI="Gottlieb"
lablocation="Children's Hospital"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"


library(knitr)
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods Summary  

Pre-normalized RNA-seq data was imported (from Seidman lab) and subset to genes with at least 10 counts per sample. This data subset was then averaged between samples for identical time points prior for import into STEM `r citep("http://www.sb.cs.cmu.edu/stem/")`. STEM analysis was run with default setting with the exceptions of a looser cluster correlation cutoff (0.4), more specific gene ontology terms (level 4) and more stringent pvalue cutoffs for individual gene expression profiles.

---

# Setup

## Libraries

[Bioconductor](http://www.bioconductor.org) and [R](http://cran.r-project.org/) libraries used to process and visualize the data.

```{r libraries}
library(xlsx) # for working with Excel docs
library(reshape2) # reshaping data
library(googleVis) # library for presenting tables
library(pheatmap)# pretty heatmaps
library(ggplot2)
```

## Variables

Working directories, files and other variables necessary to the analysis.

```{r variables}
## Setup Data and Results directory variables
if(file.exists("/n/hsphS10/hsphfs1/chb/projects/dg_cardiac_timeseries/")){
  baseDir="/n/hsphS10/hsphfs1/chb/projects/dg_cardiac_timeseries//"
    } else if (file.exists("/Users/johnhutchinson/projects/dg_cardiac_timeseries//")){
    baseDir="/Users/johnhutchinson/projects/dg_cardiac_timeseries//"
    }

dataDir <- file.path(baseDir, "data")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

pvalue.cutoff=0.01
```

## Functions

Import custom functions to use in analysis

```{r functions}
ggplot.expression.heatmap <- function(mydf, cluster_cols=TRUE, cluster_rows=TRUE, heatmapcolors = c("white", "black"), legend_label="value", x.axis_label="xaxis_label", title="Heatmap", y.axis_label="yaxis_label", text.size=2){
  y <- as.matrix(mydf)
  dd.col <- as.dendrogram(hclust(dist(y)))
  col.ord <- order.dendrogram(dd.col)
  dd.row <- as.dendrogram(hclust(dist(t(y))))
  row.ord <- order.dendrogram(dd.row)
  if (cluster_rows & !cluster_cols) {
    y <- y[col.ord, ]
    } else if (!cluster_rows & cluster_cols) {
      y <-y[,row.ord]
      } else if (cluster_rows & cluster_cols) {
        y <- y[col.ord,row.ord]
        } 
  y_names <- attr(y, "dimnames")
  df <- as.data.frame(y)
  colnames(df) <- y_names[[2]]
  df$gene <- y_names[[1]]
  df$gene <- with(df, factor(gene, levels=gene, ordered=TRUE))
  mdf <- melt(df, id.vars="gene")
  p <- ggplot(mdf, aes(y=gene, x=variable))+geom_tile(aes(fill=value))
  p <- p+scale_fill_continuous(low=heatmapcolors[1],  high=heatmapcolors[2]) 
  p <- p+theme(axis.text.y = element_text(size=text.size))
  p <- p+ggtitle(title)+ylab(y.axis_label)+xlab(x.axis_label)
  p
  }
```

# STEM Preparation

## Data import

I converted filemaker pro data into csv format for import into R 
  - the Excel export [converted some gene names into dates](http://www.biomedcentral.com/1471-2105/5/80)


```{r importdata}
exprdata <- read.csv(file.path(dataDir, "normalized_data_rpkm.csv"))
exprdata <- as.data.frame(exprdata)
```

## Data Transformation
- subset data to genes with at least 10 normalized counts per sample 
  - we can have more confidence in the results of expressed genes
- transpose data
- label samples by time point
- mean aggregate counts for samples by time point

```{r }
# subset data to genes with expression above cutoff
exprdata <- exprdata[apply(as.matrix(exprdata[,2:ncol(exprdata)]),1, function(x) all(x>10)),]
# munge data
exprdata.m <- melt(exprdata)
exprdata.t <- dcast(exprdata.m, variable ~ Gene)
row.names(exprdata.t) <- exprdata.t$variable
exprdata.t$variable <- NULL
row.names(exprdata.t) <- sub("_norm_", "", row.names(exprdata.t))
ages <- unlist(lapply(strsplit(row.names(exprdata.t), "_"), function(x) x[2]))
# aggregate exprs by age by taking mean
aggregate.exprdata <- aggregate(exprdata.t, by=list(ages), function(x) mean(x))
aggregate.exprdata.m <- melt(aggregate.exprdata)
aggregate.exprdata <- dcast(aggregate.exprdata.m, variable ~ Group.1)
names(aggregate.exprdata)[1] <- "Gene_Symbol"
aggregate.exprdata$Gene_Symbol <- toupper(aggregate.exprdata$Gene_Symbol)
```

## Output aggregated data

- in tab-delimited format, with gene names in first column

```{r output}
write.table(aggregate.exprdata, file=file.path(dataDir, "mean.aggregated.rpkm.data.txt"), row.names=F, col.names=T, sep="\t", quote=F)
```

---

#STEM analyses

## Background

> STEM allows researchers to identify significant temporal expression profiles and the genes associated with these profiles and to compare the behavior of these genes across multiple conditions. STEM is fully integrated with the Gene Ontology (GO) database supporting GO category gene enrichment analyses for sets of genes having the same temporal expression pattern.

Sets of genes with the same temporal expression patterns are called an expression "profile". Individual profiles that are similar enough to each other can be grouped together in a "cluster" 

## Methods

I ran STEM with it's default settings except: 
  - I only allowed GO categories below level 4 (default is 3c), this should weed out alot of the non-informative GO categories (eg. "biological process")
  - allowed profile clustering with a correlation of 0.5 and above (default is 0.7), this is a pretty weak an fairly arbitrary correlation cutoff but it saves trudging through a lot of individual profiles. I also found it to best separate the general GO categories I was seeing into individual clusters.
  - profile significance level cutoff was set at 0.01

## General Results

STEM analysis identified 16 expression profiles with a statistically higher number of genes assigned than expected at random. These 16 profiles then cluster based on similarity into 6 clusters (1 cluster with 3 profiles, 5 clusters with 2 profiles each and three singleton profiles).
 
In this diagram, the individual profiles are numbered and the colors represent profile clusters. Black lines within each profile represent the gene expression model profile.

![STEM profiles and clusters](`r file.path(resultsDir, "STEM.all.profile.nogenes.png")`)

HEre we can see the same data, but with the actual expression profiles of individual genes assigned to a model profile.

![STEM profiles and clusters with gene expression patterns shown](`r file.path(resultsDir, "STEM.all.profile.genes.png")`)

Looking for Gene ontology enrichment on gene sets associated with each cluster reveals differing processes enriched in each cluster.

```{r mungeGOprofiles, results='hide'}
# load in GO data
GOfilenames <- list.files(file.path(resultsDir), pattern="GO.txt")
GOfilenames <- file.path(resultsDir, GOfilenames)

#subset to significantly enriched GO categories
GOdata <- lapply(GOfilenames, function(x){
  if(grepl("^profile", basename(x))) {
    input <- read.delim(x,skip=2, sep="\t")
    } else {
      input <- read.delim(x, skip=4, sep="\t")
      }
  input$Corrected.p.value <- ifelse(input$Corrected.p.value=="<0.001", 0.00099, as.numeric(as.character(input$Corrected.p.value)))
  input$significant <- ifelse(as.numeric(as.character(input$Corrected.p.value))<pvalue.cutoff,"YES", "NO")
  input <- subset(input, significant=="YES")
  input$significant <- NULL
  names(input) <- sub("^X.", "", names(input))
  input$Corrected.p.value <- as.character(input$Corrected.p.value)
  input$Corrected.p.value <- ifelse(input$Corrected.p.value=="0.00099", "<0.001", input$Corrected.p.value)
  label <- sub(".txt","", x)
  return(list(label=label, GOdata=input))
  })
```

```{r genedatamunging, results='hide'}
# reorder data columsn for later heatmaps
exprdata <-  exprdata[,c(1,order(ages)+1)]
exprdata$Gene <- toupper(exprdata$Gene)


# load in gene data
genefilenames <- list.files(file.path(resultsDir), pattern="genes.txt")
genefilenames <- file.path(resultsDir, genefilenames)

# get gene names for each cluster and grab expression data from original dataset
genedata <- lapply(genefilenames, function(x){
  label <- sub(".txt","", x)
  
  input <- read.delim(x, skip=4, sep="\t")
  
  stemratios <- input[,grepl("Gene_Symbol|wks", names(input))]
  row.names(stemratios) <- stemratios$Gene_Symbol
  stemratios$Gene_Symbol <- NULL
  
  genes <- input$Gene_Symbol
  
  exprs <- exprdata[match(genes, exprdata$Gene),]
  row.names(exprs) <- exprs$Gene
  exprs$Gene <- NULL
  
  meanexprs <- aggregate.exprdata[match(genes, aggregate.exprdata$Gene_Symbol),]
  row.names(meanexprs) <- meanexprs$Gene_Symbol
  meanexprs$Gene_Symbol <- NULL
    
  return(list(label=label, genes=genes, stemratios=stemratios,exprs=exprs,meanexprs=meanexprs))
  })
```

## Specific Results

### Cluster 0 - profiles 35 and 39
- genes with these expression profiles show enrichment for GO categories associated with the extracellular matrix

![profile 35](`r file.path(resultsDir, "STEM.profile35.png")`)

![profile 39](`r file.path(resultsDir, "STEM.profile39.png")`)

```{r GOcategories0, results='asis'}
index <- intersect(grep("35",GOfilenames), grep("39", GOfilenames))
GOdata.gvis <- gvisTable(as.data.frame(apply(GOdata[[index]]$GOdata, 2, as.character)),, options = list(width = 1280, height=600))  
print(GOdata.gvis, "chart")
```

We can use a heatmap to plot out the log mean expression values per week for genes in this cluster.

```{r heatmap0,out.width='100%',  fig.height=6}
myexprdata <- log(genedata[[index]]$meanexprs)
plottitle <- sub("\\.", " & ", sub("_", " ", sub("_", " - ", sub("_", " ", sub("_GO.txt", "", basename(GOfilenames[index]))))))
ggplot.expression.heatmap(myexprdata, cluster_cols = FALSE, x.axis_label = "Time", y.axis_label = "Gene", title = paste(plottitle, "expression heatmap",sep=" - " ))
```

---

### Cluster 1 - profiles 17 and 38

- genes with these expression profiles show enrichment for GO categories associated with regulation of the immune system

![profile 17](`r file.path(resultsDir, "STEM.profile17.png")`)

![profile 38](`r file.path(resultsDir, "STEM.profile38.png")`)

```{r GOcategories1, results='asis'}
index <- intersect(grep("17",GOfilenames), grep("38", GOfilenames))
GOdata.gvis <- gvisTable(as.data.frame(apply(GOdata[[index]]$GOdata, 2, as.character)),, options = list(width = 1280, height=600))  
print(GOdata.gvis, "chart")
```

We can use a heatmap to plot out the log mean expression values per week for genes in this cluster.

```{r heatmap1,out.width='100%',  fig.height=6}
y <- log(genedata[[index]]$meanexprs)
plottitle <- sub("\\.", " & ", sub("_", " ", sub("_", " - ", sub("_", " ", sub("_GO.txt", "", basename(GOfilenames[index]))))))
ggplot.expression.heatmap(y, cluster_rows = TRUE,cluster_cols = FALSE, x.axis_label = "Time", y.axis_label = "Gene", title = paste(plottitle, "expression heatmap",sep=" - " ))
```

---

### Cluster 2 - profiles 1 and 3

- genes with these expression profiles show enrichment for GO categories associated with muscle contraction

![profile 1](`r file.path(resultsDir, "STEM.profile1.png")`)

![profile 3](`r file.path(resultsDir, "STEM.profile3.png")`)

```{r GOcategories2, results='asis'}
index <- intersect(grep("_1\\.",GOfilenames), grep("\\.3_", GOfilenames))
GOdata.gvis <- gvisTable(as.data.frame(apply(GOdata[[index]]$GOdata, 2, as.character)),, options = list(width = 1280, height=600))  
print(GOdata.gvis, "chart")
```

We can use a heatmap to plot out the log mean expression values per week for genes in this cluster.

```{r heatmap2, out.width='100%',  fig.height=6}
y <- log(genedata[[index]]$meanexprs)
plottitle <- sub("\\.", " & ", sub("_", " ", sub("_", " - ", sub("_", " ", sub("_GO.txt", "", basename(GOfilenames[index]))))))
ggplot.expression.heatmap(y, cluster_rows = TRUE,cluster_cols = FALSE, x.axis_label = "Time", y.axis_label = "Gene", title = paste(plottitle, "expression heatmap",sep=" - " ))
```

---

### Cluster 3 - profiles 10 and 20
 
- genes with these expression profiles show enrichment for GO categories associated with cell division

![profile 10](`r file.path(resultsDir, "STEM.profile10.png")`)

![profile 20](`r file.path(resultsDir, "STEM.profile20.png")`)


```{r GOcategories3, results='asis'}
index <- intersect(grep("10",GOfilenames), grep("20", GOfilenames))
GOdata.gvis <- gvisTable(as.data.frame(apply(GOdata[[index]]$GOdata, 2, as.character)),, options = list(width = 1280, height=600))  
print(GOdata.gvis, "chart")
```

We can use a heatmap to plot out the log mean expression values per week for genes in this cluster.

```{r heatmap3, out.width='100%',  fig.height=6}
y <- log(genedata[[index]]$meanexprs)
plottitle <- sub("\\.", " & ", sub("_", " ", sub("_", " - ", sub("_", " ", sub("_GO.txt", "", basename(GOfilenames[index]))))))
ggplot.expression.heatmap(y, cluster_rows = TRUE,cluster_cols = FALSE, x.axis_label = "Time", y.axis_label = "Gene", title = paste(plottitle, "expression heatmap",sep=" - " ))
```

---

**None of the single profiles show many significantly enriched GO categories**

---

### "Cluster" 4 - profile 27
- no significantly enriched GO categories
- genes with these expression profiles show enrichment for GO categories associated with cell division

![profile 27](`r file.path(resultsDir, "STEM.profile27.png")`)

We can still use a heatmap to plot out the log mean expression values per week for genes in this cluster.

```{r heatmap4, out.width='100%',  fig.height=6}
index <- grep("27",GOfilenames)
y <- log(genedata[[index]]$meanexprs)
plottitle <- sub("\\.", " & ", sub("_", " ", sub("_", " - ", sub("_", " ", sub("_GO.txt", "", basename(GOfilenames[index]))))))
ggplot.expression.heatmap(y, cluster_rows = TRUE,cluster_cols = FALSE, x.axis_label = "Time", y.axis_label = "Gene", title = paste(plottitle, "expression heatmap",sep=" - " ))
```

---

### "Cluster" 5 - profile 40
- genes with these expression profiles show enrichment for a couple GO categories associated with  the immune system

![profile 40](`r file.path(resultsDir, "STEM.profile40.png")`)

```{r GOcategories5, results='asis'}
index <- grep("40",GOfilenames)
GOdata.gvis <- gvisTable(as.data.frame(apply(GOdata[[index]]$GOdata, 2, as.character)),, options = list(width = 1280, height=120))  
print(GOdata.gvis, "chart")
```

We can use a heatmap to plot out the log mean expression values per week for genes in this cluster.

```{r heatmap5, out.width='100%',  fig.height=6}
y <- log(genedata[[index]]$meanexprs)
plottitle <- sub("\\.", " & ", sub("_", " ", sub("_", " - ", sub("_", " ", sub("_GO.txt", "", basename(GOfilenames[index]))))))
ggplot.expression.heatmap(y, cluster_rows = TRUE,cluster_cols = FALSE, x.axis_label = "Time", y.axis_label = "Gene", title = paste(plottitle, "expression heatmap",sep=" - " ))
```

---

### "Cluster" 6 - profile 7
- - genes with these expression profiles show enrichment for a couple GO categories associated with mitochondria and metabolism

![profile 7](`r file.path(resultsDir, "STEM.profile7.png")`)

```{r GOcategories6, results='asis'}
index <- grep("_7_",GOfilenames)
GOdata.gvis <- gvisTable(as.data.frame(apply(GOdata[[index]]$GOdata, 2, as.character)),, options = list(width = 1280, height=120))  
print(GOdata.gvis, "chart")
```

We can use a heatmap to plot out the log mean expression values per week for genes in this cluster.

```{r heatmap6, out.width='100%',  fig.height=6}
y <- log(genedata[[index]]$meanexprs)
plottitle <- sub("\\.", " & ", sub("_", " ", sub("_", " - ", sub("_", " ", sub("_GO.txt", "", basename(GOfilenames[index]))))))
ggplot.expression.heatmap(y, cluster_rows = TRUE,cluster_cols = FALSE, x.axis_label = "Time", y.axis_label = "Gene", title = paste(plottitle, "expression heatmap",sep=" - " ))
```



```{r citations, echo=FALSE, eval=F}
write.bibtex(file="references.bib")
```


