---
output:
  html_document:
    toc: true
    toc_depth: 3
    theme: journal
    css: custom.css
title: "Preparing data for STEM analysis"
bibliography: "references.bib"
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(knitcitations)
cleanbib()
options("citation_format" = "pandoc")

clientname="Danielle Gottlieb"
clientemail="danielle.gottlieb@cardio.chboston.org"
labPI="Gottlieb"
lablocation="Children's Hospital"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"


library(knitr)
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods Summary  

Pre-normalized RNA-seq data was imported (from Seidman lab) and subset to genes with at least 10 counts per sample. This data subset was then averaged between samples for identical time points prior for import into [STEM](http://www.sb.cs.cmu.edu/stem/). STEM analysis was run with default setting with the exceptions of a looser cluster correlation cutoff (0.4), more specific gene ontology terms (level 4) and more stringent pvalue cutoffs for individual gene expression profiles.

---

# Setup

## Libraries

[Bioconductor](http://www.bioconductor.org) and [R](http://cran.r-project.org/) libraries used to process and visualize the data.

```{r libraries}
library(xlsx) # for working with Excel docs
library(reshape2) # reshaping data
library(googleVis) # library for presenting tables
library(pheatmap)# pretty heatmaps
library(ggplot2)# awesome plots
library(magrittr) #chaining commands
library(knitr)
library(igraph)
```

## Variables

Working directories, files and other variables necessary to the analysis.

```{r variables}
## Setup Data and Results directory variables
if(file.exists("/n/hsphS10/hsphfs1/chb/projects/dg_cardiac_timeseries/")){
  baseDir="/n/hsphS10/hsphfs1/chb/projects/dg_cardiac_timeseries//"
    } else if (file.exists("/Users/johnhutchinson/projects/dg_cardiac_timeseries//")){
    baseDir="/Users/johnhutchinson/projects/dg_cardiac_timeseries//"
    }

dataDir <- file.path(baseDir, "data")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

pvalue.cutoff=0.05
listsize=0.9
```

## Functions

Import custom functions to use in analysis

```{r functions}
ggplot.expression.heatmap <- function(mydf, cluster_cols=TRUE, cluster_rows=TRUE, heatmapcolors = c("white", "black"), legend_label="value", x.axis_label="xaxis_label", title="Heatmap", y.axis_label="yaxis_label", text.size=2){
  y <- as.matrix(mydf)
  dd.col <- as.dendrogram(hclust(dist(y)))
  col.ord <- order.dendrogram(dd.col)
  dd.row <- as.dendrogram(hclust(dist(t(y))))
  row.ord <- order.dendrogram(dd.row)
  if (cluster_rows & !cluster_cols) {
    y <- y[col.ord, ]
    } else if (!cluster_rows & cluster_cols) {
      y <-y[,row.ord]
      } else if (cluster_rows & cluster_cols) {
        y <- y[col.ord,row.ord]
        } 
  y_names <- attr(y, "dimnames")
  df <- as.data.frame(y)
  colnames(df) <- y_names[[2]]
  df$gene <- y_names[[1]]
  df$gene <- with(df, factor(gene, levels=gene, ordered=TRUE))
  mdf <- melt(df, id.vars="gene")
  p <- ggplot(mdf, aes(y=gene, x=variable))+geom_tile(aes(fill=value))
  p <- p+scale_fill_continuous(low=heatmapcolors[1],  high=heatmapcolors[2]) 
  p <- p+theme(axis.text.y = element_text(size=text.size))
  p <- p+ggtitle(title)+ylab(y.axis_label)+xlab(x.axis_label)
  p
  }

runrevigo <- function(GOids, GOpvals, listsize){
  mycommand=paste('/Volumes/ody/scripts/revigoR/revigo.pl -goterms', paste(GOids,collapse=","), '-gopvals', paste(GOpvals,collapse=","), '-size', listsize, sep=" ")
  mytempfile <- tempfile()
  system2(command='perl', args=mycommand, stdout=mytempfile)
  source(mytempfile)
}
```

# STEM Preparation

## Data import

I converted filemaker pro data into csv format for import into R 
  - the Excel export [converted some gene names into dates](http://www.biomedcentral.com/1471-2105/5/80)


```{r importdata}
exprdata <- read.csv(file.path(dataDir, "normalized_data.csv"))
exprdata <- as.data.frame(exprdata)
```

## Data Transformation
- subset data to genes with at least 10 normalized counts per sample 
  - we can have more confidence in the results of expressed genes
- transpose data
- label samples by time point
- mean aggregate counts for samples by time point

```{r }
# subset data to genes with expression above cutoff
exprdata <- exprdata[apply(as.matrix(exprdata[,2:ncol(exprdata)]),1, function(x) all(x>10)),]
# munge data
exprdata.m <- melt(exprdata)
exprdata.t <- dcast(exprdata.m, variable ~ Gene)
row.names(exprdata.t) <- exprdata.t$variable
exprdata.t$variable <- NULL
row.names(exprdata.t) <- sub("_norm_", "", row.names(exprdata.t))
ages <- unlist(lapply(strsplit(row.names(exprdata.t), "_"), function(x) x[2]))
# aggregate exprs by age by taking mean
aggregate.exprdata <- aggregate(exprdata.t, by=list(ages), function(x) mean(x))
aggregate.exprdata.m <- melt(aggregate.exprdata)
aggregate.exprdata <- dcast(aggregate.exprdata.m, variable ~ Group.1)
names(aggregate.exprdata)[1] <- "Gene_Symbol"
aggregate.exprdata$Gene_Symbol <- toupper(aggregate.exprdata$Gene_Symbol)
```

## Output aggregated data

- in tab-delimited format, with gene names in first column

```{r output}
write.table(aggregate.exprdata, file=file.path(dataDir, "mean.aggregated.data.txt"), row.names=F, col.names=T, sep="\t", quote=F)
```


## Absolute count differences over time course study

```{r absdiff}
diffs <- aggregate.exprdata[,2:ncol(aggregate.exprdata)] %>% apply(., 1, function(x) range(na.omit(x))) %>% apply(., 2, diff)

expr.diffs <- cbind(aggregate.exprdata, diffs)

write.table(expr.diffs, file=file.path(dataDir, "expr.diffs.xls"), row.names=F, col.names=T, sep="\t", quote=F)
```


---

#STEM analyses

## Background

> STEM allows researchers to identify significant temporal expression profiles and the genes associated with these profiles and to compare the behavior of these genes across multiple conditions. STEM is fully integrated with the Gene Ontology (GO) database supporting GO category gene enrichment analyses for sets of genes having the same temporal expression pattern.

Sets of genes with the same temporal expression patterns are called an expression "profile". Individual profiles that are similar enough to each other can be grouped together in a "cluster" 

## Methods

I ran STEM with it's default settings except: 
  - I only allowed GO categories below level 4 (default is 3c), this should weed out alot of the non-informative GO categories (eg. "biological process")
  - profile significance level cutoff was set at 0.01

## General Results

STEM analysis identified 11 expression profiles with a statistically higher number of genes assigned than expected at random. 
 
In this diagram, the individual profiles are numbered , black lines within each profile represent the gene expression model profile.

![STEM profiles](`r file.path(resultsDir, "STEM.all.profile.nogenes.png")`)

HEre we can see the same data, but with the actual expression profiles of individual genes assigned to a model profile.

![STEM profiles with gene expression patterns shown](`r file.path(resultsDir, "STEM.all.profile.genes.png")`)

Looking for Gene ontology enrichment on gene sets associated with each cluster reveals differing processes enriched in each cluster.

```{r mungeGOprofiles, results='hide'}
# load in GO data
GOfilenames <- list.files(file.path(resultsDir), pattern="^profile.*GO.txt")
GOfilenames <- file.path(resultsDir, GOfilenames)

#subset to significantly enriched GO categories
GOdata <- lapply(GOfilenames, function(x){
  if(grepl("^profile", basename(x))) {
    input <- read.delim(x,skip=2, sep="\t")
    } else {
      input <- read.delim(x, skip=4, sep="\t")
      }
  input$Corrected.p.value <- ifelse(input$Corrected.p.value=="<0.001", 0.00099, as.numeric(as.character(input$Corrected.p.value)))
  input$significant <- ifelse(as.numeric(as.character(input$Corrected.p.value))<pvalue.cutoff,"YES", "NO")
  input <- subset(input, significant=="YES")
  input$significant <- NULL
  names(input) <- sub("^X.", "", names(input))
  input$Corrected.p.value <- as.character(input$Corrected.p.value)
  input$Corrected.p.value <- ifelse(input$Corrected.p.value=="0.00099", "<0.001", input$Corrected.p.value)
  label <- sub(".txt","", x)
  return(list(label=label, GOdata=input))
  })
```

```{r genedatamunging, results='hide'}
# reorder data columsn for later heatmaps
exprdata <-  exprdata[,c(1,order(ages)+1)]
exprdata$Gene <- toupper(exprdata$Gene)


# load in gene data
genefilenames <- list.files(file.path(resultsDir), pattern="^profile.*genes.txt")
genefilenames <- file.path(resultsDir, genefilenames)

# get gene names for each cluster and grab expression data from original dataset
genedata <- lapply(genefilenames, function(x){
  print(x)
  label <- sub(".txt","", x)
  
  input <- read.delim(x, skip=4, sep="\t")
  
  stemratios <- input[,grepl("Gene_Symbol|wks", names(input))]
  row.names(stemratios) <- stemratios$Gene_Symbol
  stemratios$Gene_Symbol <- NULL
  
  genes <- input$Gene_Symbol
  
  exprs <- exprdata[match(genes, exprdata$Gene),]
  row.names(exprs) <- exprs$Gene
  exprs$Gene <- NULL
  
  meanexprs <- aggregate.exprdata[match(genes, aggregate.exprdata$Gene_Symbol),]
  row.names(meanexprs) <- meanexprs$Gene_Symbol
  meanexprs$Gene_Symbol <- NULL
    
  return(list(label=label, genes=genes, stemratios=stemratios,exprs=exprs,meanexprs=meanexprs))
  })
# write out just gene names to files
lapply(genedata, function(x){
  genesymbols <- sort(as.character(x$genes))
  filename <- x$label %>% as.character %>% basename
  filename <- sub("genes", "genesymbols.txt",filename )
  write(x=genesymbols, file=file.path(resultsDir, filename))
})
```

## Comparing Profiles

As each gene is only assigned to a single profile, none of the profiles have any genes in common. However, they can have common GO categories (see Specific Results section below) and similar model expression patterns; examining these can help us figure out which profiles to focus on and how these processes relate to each other.

### GO category relationships

I attempted to visualize these relationships through a network graph, where each node represents a profile, the size of the node reflects how many GO categories are enriched. Profiles with common GO categories are linked by an edge and the width of the edge is determined by how many GO categories they have in common. 

```{r networkvisGO, out.width='100%'}
for (x in 1:length(GOdata)){
  dataset <- GOdata[[x]]
  data <- dataset$GOdata
  data <- data$Category.ID
  label <- sub("_", "_GOids_", sub("_GO", "", basename(dataset$label)))
  assign(label, data)
}

myprofileGOids <- ls()[grep("^profile_GOids_", ls())]

myGOids <- c()
for (x in 1:length(myprofileGOids)){
  GOids <- get(myprofileGOids[x])
  myGOids <- c(myGOids, as.character(GOids))
}
myGOids <- unique(myGOids)

aff.matrix <- c()
for (x in 1:length(myprofileGOids)){
  GOids <- get(myprofileGOids[x])
  status <- ifelse(myGOids %in% GOids, 1,0)
  aff.matrix <- cbind(aff.matrix, status)
  }
dimnames(aff.matrix)[[1]] <- myGOids
dimnames(aff.matrix)[[2]] <- sub("_GOids_", "", myprofileGOids)

aff.matrix <- as.matrix(aff.matrix)
aff.matrix <- t(aff.matrix)
adj.matrix <- aff.matrix%*%t(aff.matrix)
g=graph.adjacency(adj.matrix, mode="undirected", weighted=TRUE, diag=F)
width<-(E(g)$weight/2)*1
number<-diag(adj.matrix)*0.25
plot(g, edge.width=width, vertex.size=number, vertex.color="red", edge.color="darkgrey", vertex.label.family ="sans", vertex.label.color="black")
```

In this graph, profiles 10 and 3 have 145 and 175 enriched GO categories each with only 2 categories in common. In contrast, profile 1 has only  48 enriched GO categories, but shares 38 of those with profile 3. (Profile27 has no enriched GO categories).

### STEM expression patterns

It's difficult to compare the expression profiles when they are all on separate graphs. Here I plotted the model profile expression pattersn for each profile on the same plot. I also produced a simpler plot which includes only those profiles which show at least 10 enriched gene ontology terms.

```{r STEMpatterns}
patterns <- read.csv(file.path(resultsDir, "STEM_patterns.csv"), header=T)

library(reshape2)
patterns.m <- melt(patterns, id.vars="X")

ggplot(patterns.m, aes(x=X, y=value, color=variable, lwd=1.25))+geom_line()+xlab("weeks")+ylab("log2(fold change since week 14)")+ggtitle("Model expression profiles - all profiles")+scale_color_brewer(palette="Paired")

# drop any profiles with less than 10 GO categories
ggplot(subset(patterns.m, variable %in% names(which(rowSums(aff.matrix)>=10))), aes(x=X, y=value, color=variable, lwd=2))+geom_line()+xlab("weeks")+ylab("log2(fold change since week 14)")+ggtitle("Model expression profiles - profiles with >10 enriched GO categories")+scale_color_brewer(palette="Paired")
```

### STEM expression pattern correlations

Previously I used STEM to group individual profiles into "clusters" by their expression patterns. This can simplifiy the picture and allow us to examine fewer profiles, but has the downside of a hard cutoff. A more flexible route is to produce  a heatmap of the inter-profile expression pattern correlations, so we can examine all the relationships. Here  a value of 1 shows a high degress of correlation between profile expression pattern. High levels of correlation (red) imply these genes are changing expression patterns in similar,though not identical ways. 

```{r expressionpattercorrelations}
patterns$X <- NULL 
library(RColorBrewer)
my_palette <- colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(300)
heatmap(cor(patterns, method="spearman"), col=my_palette)
```

Putting all these plots together, we can start piecing together the puzzle.

We might expect to find a few differnt types of patterns based on shared gene ontologies (i.e. share functionality) and shared expression patterns (coregulated genes acting in concert).

For example, profiles with a high amount of shared gene ontologies that have similar expression patterns are likely just subsets of a larger group of genes which affect similar processes in concert (but had enough gene expression pattern differences at one stage to be grouped differently).

You might also expect to see differnt funcitonal units (i.e. two profiles with little shared ontology) with similar expression patterns, indicating these processes occur at the same time.

Looking at the actual data, we can see some level of correlation between profiles 35 (ECM, system development, homophilic cell adhesion, chemotaxis ) and 39 (cell motility, angiogenesis and wound healing) suggesting these genes are acting in concert. The level of common gene ontology categories suggest these genes share some (though not all) functionality. Profiles 35 and 20 (ECM, anatomical structure morphogenesis and cell-matrix adehsion) also share some expression patterns and a high degree of simliarilty at the gene ontology levels, and thus likely similar processes acting in concert. However, profile 39 is anti-correlated with profile 20 for expression patterns, and shares few gene ontology processes. Whatever 39 has in common with 35, it does not have in common with 20. 

We can see even higher levels of correlation between profiles 1 (energy derivation, muscle contraction and development), 7 (energy derivation and small molecule metabolism) and 17 (DNA synthesis, chemotaxis, endotheial cell proliferation). Given the commons expression pattersn, gene ontology connections betwween profiles 1 and 7, and the small size of profile 7, we can likely safely group profiles 7 and 1 together. In contrast, although their expression patterns are simliar profile 1 and profiles 17 share few gene ontology categories and likely represent different processes acting in concert.

Profiles 10 (mitosis) and 39 (angiogeneis, cell motility, wounding response) appear to be anti-correlated, though this can easily be seen from expression levels line plot. 

Profile 20 (ECM, cell adhesion) and profile 40 (defense response) show little correlation in expression. Similarly, the othe r immune related profile (38) also shows little correlation with profile 20.

## Specific Results
- in no particular order
 
### Profile 35

![profile 35](`r file.path(resultsDir, "STEM.profile35.png")`)
`r index <- grep("_35_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_35_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_35_GO.txt)**
    
Gene ontology enrichment analyses can yield an overwhelming number of enriched categories, many with redundant functionality. We can simplify this output by identifying the most representative subset of the terms, using metrics which measure the semantic similarity of the terms. [Revigo](http://revigo.irb.hr/)  performs such analyses, using an algortithm which forms

>  groups of highly similar GO terms, where the choice of the groups’ representatives is guided by the p-values

The algorithm takes into account the parent-child structure of the gene onotology database

> If the p-values are quite close and one term is a child node of the other, REVIGO will tend to choose the parent term

The algorithm also ytries to find more specific GO terms.

> Very general GO terms, however, are always avoided as cluster representatives ... as they tend to be uninformative

Revigo allows visualization of these representatives and their relations to the terms within their group  as a [treemap](http://en.wikipedia.org/wiki/Treemapping). Here the color depicts a grouping of related terms, the size of a block, it's pvalue from g:profiler and the large text the most representative gene ontology term for the related group.

```{r revigo}
GOs <- GOdata[[index]]$GOdata[,"Category.ID"]
pvals <- GOdata[[index]]$GOdata[,"p.value"]
runrevigo(GOids = GOs, GOpvals = pvals, listsize = 0.7)
```

---

### Profile 39

![profile 39](`r file.path(resultsDir, "STEM.profile39.png")`)
`r index <- grep("_39_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_39_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_39_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo39, ref.label='revigo'}
```

---

### Profile 1

![profile 1](`r file.path(resultsDir, "STEM.profile1.png")`)
`r index <- grep("_1_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_1_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_1_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo1, ref.label='revigo'}
```

---

### Profile 17

![profile 17](`r file.path(resultsDir, "STEM.profile17.png")`)
`r index <- grep("_17_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_17_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_17_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo17, ref.label='revigo'}
```

---

### Profile 10

![profile 10](`r file.path(resultsDir, "STEM.profile10.png")`)
`r index <- grep("_10_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_10_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_10_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo10, ref.label='revigo'}
```

---

### Profile 20

![profile 20](`r file.path(resultsDir, "STEM.profile20.png")`)
`r index <- grep("_20_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_20_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_20_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo20, ref.label='revigo'}
```

---

### Profile 3

![profile 3](`r file.path(resultsDir, "STEM.profile3.png")`)
`r index <- grep("_3_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_3_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_3_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo3, ref.label='revigo'}
```

---

### Profile 38

![profile 38](`r file.path(resultsDir, "STEM.profile38.png")`)
`r index <- grep("_38_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_38_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_38_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo38, ref.label='revigo'}
```

---

### Profile 40

![profile 40](`r file.path(resultsDir, "STEM.profile40.png")`)
`r index <- grep("_40_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_40_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_40_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo40, ref.label='revigo'}
```

---

### Profile 7

![profile 7](`r file.path(resultsDir, "STEM.profile7.png")`)
`r index <- grep("_7_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_7_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_7_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo7, ref.label='revigo'}
```

---

### Profile 27

![profile 27](`r file.path(resultsDir, "STEM.profile27.png")`)
`r index <- grep("_27_",GOfilenames)`

The genes which show this expression profile can be accessed via **[this file](../results/profile_27_genesymbols.txt)**

These genes are enriched for the gene ontology categories found in **[this file](../results/profile_27_GO.txt)**

These GO categories can be summarized with Revigo. 

```{r revigo27, ref.label='revigo'}
```


```{r citations, echo=FALSE, eval=F}
write.bibtex(file="references.bib")
```


