```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, figalign="center", 
               echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, 
               tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), 
               dev="CairoSVG",
               fig.path="figure/allsamples-")

options(width=200)

clientname="Christian Bach"
clientemail="cbach@bidmc.harvard.edu)"
labPI="Dan Tenen"
lablocation="BIDMC"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```


Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

## Methods Summary
All MoGene 1.0 ST arrays were processed using the 'oligo' BioConductor package, quality-controlled with arrayQualityMetrics and RMA normalized. Differentially expressed genes were identified using limma.

---

## General Setup

### Working directories, files and other variables necessary to the analysis.

```{r variables}
## Setup Data and Results directory variables
if(file.exists("/n/hsphS10/hsphfs1/chb/projects/dt_leukemias/")){
  baseDir="/n/hsphS10/hsphfs1/chb/projects/dt_leukemias/"
  } else if (file.exists("/Volumes/ody/projects/dt_leukemias/")){
    baseDir="/Volumes/ody/projects/dt_leukemias/"
    }
dataDir <- file.path(baseDir, "data")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # colorblind friendly palette
covarsfilename="covars.desc" # tab delimited file describing samples
probenum <- 15000 # number of probes to assay after ranking by variability
pvalue.cutoff=0.1

```

### Libraries

[Bioconductor](http://www.bioconductor.org) and [R](http://cran.r-project.org/) libraries used to process and visualize the data.

```{r libraries_variables, echo=TRUE}
library(oligo) # array utilities
library(pd.mogene.1.0.st.v1) # array layout annotation
library(mogene10sttranscriptcluster.db) # array probe to gene annotations
library(arrayQualityMetrics) # array quality control reports
library(limma) # array statistical analyses
library(pheatmap) # pretty heatmaps
library(googleVis) # nice charts
library(plyr) # data format utility
library(reshape2) # data format utility
library(devtools) # install libraries from github
install_git("git://github.com/hbc/CHBUtils.git") # misc personal utilities
library(CHBUtils)
library(ggplot2) # pretty graphs
library(ggdendro) # for pretty dendrograms
library(genefilter) # for filtering out low variation genes 
library(RColorBrewer)
```

---

## Import Data and Metadata

### Load Data

- load in phenotypes and array names from metadata file (covars.desc) in "metadata" directory
  - this file contains the names and descriptions of CEL files contained in the data directory 
- use array names to load in arrays 

```{r dataload, results='hide'}
covars <- read.delim(file.path(metaDir, covarsfilename), row.names=1, sep="\t") # simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns
celFiles <- file.path(dataDir, row.names(covars))
affyRaw <- read.celfiles(celFiles)
pData(affyRaw) <- covars 
sampleNames(affyRaw) <- pData(affyRaw)$sampleID
validObject(affyRaw)
rm(covars)
```

### Sample metadata

```{r covars, results='asis'}
# Sample information table
pd.gvis <- gvisTable(as.data.frame(apply(pData(affyRaw), 2, as.character)), options=list(width=640, height=240))  
print(pd.gvis, "chart")
```

---

## Processing 

### Raw Data QC

- using arrayQualityMetrics [library](http://www.bioconductor.org/packages/2.10/bioc/html/arrayQualityMetrics.html)

```{r rawQC, eval=FALSE}
arrayQualityMetrics(expressionset=affyRaw, outdir=file.path(resultsDir, 'report_raw'), force=TRUE, do.logtransform=TRUE, intgroup=c("group"))
```

[Raw Data QC Report](../results/report_raw/index.html)

With the possible exception of MUT2  (which skews lower in its intensity distributions than the other arrays) and to a lesser extent MUT3, the arrays look OK. 
Hierarchical clustering looks promising at this stage. Note that there is more variation between the MUT samples than between the WT samples, which (if I understand the experiment correctly) may reflect the greater variation in sample collection for the MUT samples.  

MUT2 is borderline, but there is no objective reasons to remove it. Rather than excluding arrays at this point, I decided to proceed with the analysis and see the effect on the results.

### Background Correct and Normalize

- using [RMA](http://bioinformatics.oxfordjournals.org/content/19/2/185)

- summarize probesets on the gene ('core') level

```{r normalize, results='hide'}
sampleset <- "All_Samples"
affyNorm.core <- rma(affyRaw, target="core", background=TRUE, normalize=TRUE)
```

### Normalized Data QC
- using arrayQualityMetrics library

```{r normQC, eval=FALSE}
arrayQualityMetrics(expressionset=affyNorm.core, outdir=file.path(resultsDir, paste("report_rma.core", sampleset, sep=".")), force=TRUE, do.logtransform=FALSE, intgroup=c("group"))
```

[Normalized Data QC Report](../results/report_rma.core/index.html)
 
MUT2 now clusters with the WT samples, which is worrisome, but as noted previously, there is no objective reason to exclude it. 

---

## Unsupervised Clustering

### Hierarchical Clustering
The goal of these analyses are to naiively evaluate the variability within the raw data and determine whether this variability can predict the different sample groups

The first method produces a dendrogram by performing  
>  a hierarchical cluster analysis using a set of dissimilarities for the n objects being clustered

Sample IDs are shown and groups highlighted in red and blue. 

```{r cluster1}
plot_dendro(affyNorm.core, title=sampleset, labels.colname="sampleID", colors.colname="group")
```

### Principal Component Analysis (PCA)

This second approach is a dimension reduction and visualisation technique that is used to project the multivariate (i.e.multiple genes) data vector of each array into a lower-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the arrays. The data is typically reduced to a small number of dimensions (or components) which explain most of the sample variability. 

```{r PCAsd, out.width='50%'}
PCAplot.sd.eset(affyNorm.core, title=sampleset)
```

Here, 4 dimensions explain more than 85% of the variation observed in the samples.

As plots with more than 2 dimensions are difficult to visualize, we typically  split up 
the dimensions/components and plot them pairwise against each other; the plots here show scatterplots of the arrays along all dual combinations of the first four principal components. In the first plot, each sample group is represented by a separate color and in the second plot each sample is represented by a different color. 

You can use these plots to explore if the arrays cluster, find outliers, and determine whether this is according to an intended experimental factor or according to unintended causes such as batch effects. 

```{r pca1, fig.cap="Primary Component Analysis of samples - all combinations of the 4 first primary components", out.width='50%'}
PCAplot.eset(affyNorm.core, categories="group", title=sampleset, colorpalette=cbPalette, numcomponents=4)
PCAplot.eset(affyNorm.core, categories="sampleID", title=sampleset, colorpalette=cbPalette, numcomponents=4)
```

The MUT2 sample does not cluster with the other MUT samples. Otherwise, you can see that the MUT2 samples have more variability, and the WT1 sample is actually more differnt from the other WT samples than the MUT2 sample is. 


How about the days from transplantation?

```{r pcagrid, out.width='100%'}
library(gridExtra)
numcomponents=3
eset.core <- exprs(affyNorm.core) 
myPca.core <- prcomp(t(eset.core))
tmpPCAData.core <- as.data.frame(myPca.core$x[,1:numcomponents])
combo.data <- cbind(tmpPCAData.core, pData(affyNorm.core))

combos <- combn(names(tmpPCAData.core), 2)

plots <- apply(combos, 2, function(n) {
  xid <- as.character(n[1])
  xvals <- combo.data[,xid]
  yid <- as.character(n[2])
  yvals <- combo.data[,yid]
  p <-  ggplot(data=cbind(cbind(combo.data,xvals),yvals), aes(x=xvals, y=yvals, color=days, size=days))+
    geom_point()+
    ylab(yid)+
    xlab(xid)
  p+  scale_color_continuous()+scale_size_continuous(range=c(2,10))
})
do.call(grid.arrange,  plots)
```

---

## Annotate

So far we have only been working with the probesets,without reference to the genes they assay. Here we load in metadata about the probesets on the array (feature data), the gene symbols in particular.

```{r features, results='hide'}
featureData(affyNorm.core) <- getNetAffx(affyNorm.core, "transcript") # this will load the Affymetrix annotation, including the probeID, into the fData
# get gene symbols and entrezIDs for all probesets
fData(affyNorm.core)$symbol <- as.character(unlist(mget(featureNames(affyNorm.core), mogene10sttranscriptclusterSYMBOL, ifnotfound=NA))) # curated annotations from Bioconductor 
fData(affyNorm.core)$entrezID <- as.character(unlist(mget(featureNames(affyNorm.core), mogene10sttranscriptclusterENTREZID, ifnotfound=NA))) # curated annotations from Bioconductor 
```

---

## Filter
Reducing the number of genes assayed reduces  the multiple test correction and may allow us to identify more differentially expressed genes.

### By Annotation
- remove the control probes

```{r filter1}
affyNorm.core <- affyNorm.core[which(fData(affyNorm.core)$category=="main"),]
```

### By Cross Hybridization
- some probes are annotated as potentially hybridizing to multiple targets

```{r filter2}
affyNorm.core <- affyNorm.core[which(fData(affyNorm.core)$crosshybtype=="1"),]
```

### By Expression Level
- remove probes with low expression levels in all samples

```{r filter3}
eset.core <- exprs(affyNorm.core)
affyNorm.core <- affyNorm.core[!(apply(eset.core, 1, function(x) all(x<4))),]
```

### By Variability
- remove probes to keep only the `r probenum` probes that show the highest variation among all samples (without regard for group status)

```{r filter4}
eset.core <- exprs(affyNorm.core)
rowmads <- apply(eset.core, 1, mad)

quantile.cutoff <- 1-probenum/nrow(fData(affyNorm.core))
mad.cutoff <- as.numeric(quantile(rowmads, quantile.cutoff))

affyNorm.core <- affyNorm.core[rowmads>mad.cutoff,]
```

---
## Statistical Analyses

### Correlations

```{r correlations}
cors <-  cor(exprs(affyNorm.core))
kable(cors, format="markdown", digits=4)        
write.table(cors, file=file.path(resultsDir, "correlations.tab"), quote=F, sep="\t", row.names=T, col.names=NA)

p <- ggplot(melt(cors), aes(x=Var1, y=Var2, fill=value))+ geom_tile()+theme(axis.title.x = element_blank(), axis.title.y=element_blank())    # Remove x-axis label

p+  scale_fill_gradientn(name="correlation", limits = c(0.75,1), colours=c("#FFFF00", "#EEEE11", "#DDDD22", "#CCCC33", "#AAAA55", "#999966", "#888877", "#777788", "#666699", "#5555AA", "#3333CC", "#2222DD", "#1111EE", "#0000FF"))
```

### Limma

A linear model for microarray data analysis ([Limma][http://www.bioconductor.org/packages/release/bioc/html/limma.html]) was performed on the samples to identify differentially expressed genes between the comparison of the sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we construct two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.

#### Create appropriate design matrix 
- make a matrix with arrays as rows, sample groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group 
- in this case it looks like this, where 1-5 are the 5 non-outlier microarrays

```{r design, results="asis"}
design <- model.matrix(~ 0 + group,  data=pData(affyNorm.core))
# make sure the headings match
colnames(design) <- sub("group", "", colnames(design))

design.gvis <- gvisTable(as.data.frame(apply(rownames2col(design,"ID"), 2, as.character)), options=list(width=640, height=180))  
print(design.gvis, "chart")
```

#### Create contrast matrix
- to perform specified pairwise comparisons
- in this table, columns are contrasts/comparisons and rows are sample groups
-  a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts(WT-MUT, levels=colnames(design))
contrast.gvis <- gvisTable(as.data.frame(apply(rownames2col(contrast.matrix,"contrast"), 2, as.character)), options=list(width=240, height=80))  
print(contrast.gvis, "chart")
```

These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix for each gene based on the given series of arrays
- using the contrast matrix, compute estimated coefficients and standard errors for contrasts
- compute moderated t-statistics and log-odds of differential expression by empirical Bayes shrinkage of the standard errors towards a common value

#### Fit a linear model
- for each gene based on the given series of arrays

```{r linearmodel}
eset.core <- exprs(affyNorm.core)
fit.core <- lmFit(eset.core, design) 
```

#### Compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2.core <- contrasts.fit(fit.core, contrast.matrix) 
```

#### Compute moderated t-statistics and log-odds of differential expression 
- by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.core <- eBayes(fit2.core) 
```

--- 

## Results
### Volcanoplots
- to visualize relationship between fold change and signifcance of changes in expression between groups
- genes outside the two vertical orange lines show more than 2-fold changes in expression between groups
- genes which pass an FDR  (false discovery rate) multiple testing correction of 0.1 (10% FDR) are colored blue
- the larger the pointsize, the more likely there is significant expression change (B = log-odds of differential expression) between the sample groups

```{r volcanoplot, fig.cap="", results='hide', dev="png", fig.height=12, fig.width=12, out.width='100%'}
stats.core <- topTable(fit2.core, coef=1, sort.by="B",number=nrow(fData(affyNorm.core)), genelist=fData(affyNorm.core)[,c("probesetid", "symbol", "entrezID")])
stats.core$Passes.FDR.threshold  <-  as.factor(stats.core$adj.P.Val<pvalue.cutoff)
g <- ggplot(data=stats.core, aes(x=logFC, y=-log10(P.Value), color=Passes.FDR.threshold, size=B)) +
  geom_point(alpha=0.5) +
  geom_vline(xintercept=c(-1,1), color="orange", alpha=0.6) +
  xlab("log2 fold change") + ylab("-log10 p-value") +
  labs(title=paste("WT vs. MUT - ", sampleset, sep="")) +
  scale_x_continuous(limits=c(-4,4))+scale_y_continuous(limits=c(0,7))
g
```

### Top results
Typically, you would want to base gene selection on the adjusted P-value and the logFC, but as you can see from the volcano plot, there are no genes with significant adjusted P-values (FDR<`r pvalue.cutoff`)
Instead, we can select the top 100 genes  with at least 2 fold changes in expression,  as sorted by "B", the log-odds that a gene is differentially expressed and on on the basis of whether the probe was mapped to a gene in the annotation i.e. I dropped all probesets that are not annotated against a gene

```{r toptable}
stats.top.core <- topTable(fit2.core, coef=1, adjust="fdr", sort.by="B", p.value=0.1, lfc=1, number= nrow(fData(affyNorm.core)), genelist=fData(affyNorm.core)[,c("probesetid", "symbol", "entrezID")]) 
if (nrow(stats.top.core)<1) {
  stats.top.core <- topTable(fit2.core, coef=1, adjust="fdr", sort.by="B", p.value=1, lfc=1, number= nrow(fData(affyNorm.core)), genelist=fData(affyNorm.core)[,c("probesetid", "symbol", "entrezID")])
  }
stats.top.core=stats.top.core[which(!is.na(stats.top.core$symbol)),]
if(nrow(stats.top.core)>100) {
  stats.top.core=stats.top.core[1:100,]
  }
# expression levels for these genes in all samples
eset.core <- exprs(affyNorm.core)
if(nrow(stats.top.core)>0) {
  eset.top.core  <-  eset.core[which(row.names(eset.core) %in% stats.top.core$probesetid),] 
  eset.stats.top.core <- cbind(stats.top.core, eset.top.core)
  } else { eset.stats.top.core<-"NA"}
```

#### Output pvalues and expression levels to files
 - note that for all these files, I have not summarized values for genes assayed by multiple probes (i.e. by taking the median value), so you may see multiple instances of the same gene in the results

```{r outputs}
all.exprs=as.data.frame(eset.core)
all.exprs <- rownames2col(all.exprs, "probesetid")
all.exprs<-merge(stats.core, all.exprs, by="probesetid")
write.table(all.exprs, file=file.path(resultsDir, paste("all.genes.stats.exprs", sampleset, "xls", sep=".")),  sep ="\t",, row.names=F, col.names=T)
write.table(eset.stats.top.core, file=file.path(resultsDir, paste("top100.genes.stats.expression.WTvMUT", sampleset, "xls", sep=".")), sep="\t",row.names=F, col.names=T)
```

**The summary tables will contain the following information:**
- logFC is the log2-fold change
- the AveExpr is the average expression value accross all arrays
- the moderated t-statistic (t) is the logFC to its standard error, the P.Value is the associated p-value
- the adj.P.Value is the p-value adjusted for multiple testing (by FDR) 
- the B-value (B) is the log-odds that a gene is differentially expressed (the-higher-the-better)
- the last 5 columns contain the log-transformed normalized expression levels for these genes in each sample

####Link to Excel files with statistics and expression levels of all genes

[ALL genes](../results/all.genes.stats.exprs.`r sampleset`.xls)

####Link to Excel files with statistics and expression levels for genes with top 100 log-odds of differential expression

[For Top 100 differentially expressed genes](../results/top100.genes.stats.expression.WTvMUT.`r sampleset`.xls)

### Heatmap
- of most differentially expressed genes

```{r heatmap, results='hide',out.width='100%', dev="png", fig.height=18, fig.width=18}
row.names(eset.top.core) <- paste(as.vector(unlist(eset.stats.top.core$symbol)), " (" ,as.vector(unlist(eset.stats.top.core$probesetid)),  ")", sep="")
# annotations
pd <- pData(affyNorm.core)
heatmap.annots <- pd[,c("sampleID", "group")]
heatmap.annots <- as.data.frame(apply(heatmap.annots, 2, unlist))
col2rownames(heatmap.annots, "sampleID", removecol=TRUE)
heatmap.annots$arrayID <- NULL
# annotation colors
group_colors <- c(cbPalette[1:2] )
names(group_colors) <- unique(unlist(pd$group))
ann_colors = list(group = group_colors)
## Heatmaps
pheatmap(as.matrix(eset.top.core), 
         annotation=heatmap.annots, 
         color=rev(brewer.pal(11,"RdBu")), 
         cluster_cols = TRUE, 
         main=paste(sampleset, " - Clustered",sep=""), 
         fontsize=12,
         fontsize_row=10,
         annotation_colors=ann_colors,
         show_colnames=TRUE)
```

---
---

# Repeat Analysis without MUT2 sample


Wanted to compare the results above with what we would see if you exclude the MUT2 sample.
Removed this ssample from the raw data and repeated all steps of the analysis.

## Cleanup old variables

```{r setupnomut2}
rm(list=ls()[!grepl("affyRaw|Dir|cbPalette|celFiles|cutoff|probenum", ls())])
sampleset="nomut2"
opts_chunk$set(echo=FALSE, fig.path="figure/nomut2-")
```

---

## Processing

### Remove MUT2 sample and renormalize

```{r outlierfilter, results='hide', cache=TRUE}
affyRaw.filt <- affyRaw[,which(pData(affyRaw)$sampleID!="MUT2")]
affyNorm.core <- rma(affyRaw.filt, target="core", background=TRUE, normalize=TRUE)
```

### Rerun QC

```{r normQC}
```

[Normalized Data without MUT2 QC Report](../results/report_rma.core/index.html)
 
---

## Unsupervised Clustering
### Hierarchical Clustering
 
```{r cluster1, out.width='50%'}
```

### Principal Component Analysis (PCA)

```{r PCAsd, out.width='50%'}
```

Here, 3 dimensions explain more than 85% of the variation observed in the samples.

```{r pca1, fig.cap="Primary Component Analysis of samples - all combinations of the 4 first primary components", out.width='50%'}
```

From this plot we can see that there is good separation between the WT and MUT samples (PC1 vs. PC3), though the MUT samples do not cluster together well for this combination. Consistent with the dendrogram results, WT1 does not cluster with the othe WT samples as well (it is the WT outlier in the PC1 vs. PC3 plot). 

---

## Annotate

```{r features, results='hide'}
```

---

## Filter

### By Annotation

```{r filter1}
```

### By Cross Hybridization

```{r filter2}
```

### By Expression Level

```{r filter3}
```

### By Variability

```{r filter4}
```

---

## Statistical Analyses

### Limma

#### Create appropriate design matrix 

```{r design, results="asis"}
```

#### Create contrast matrix

```{r contrastmatrix, results='asis'}
```

#### Fit a linear model

```{r linearmodel}
```

#### Compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
```

#### Compute moderated t-statistics and log-odds of differential expression 

```{r bayes}
```

--- 

## Results
### Volcanoplots

```{r volcanoplot, fig.cap="", results='hide', dev="png", fig.height=12, fig.width=12, out.width='100%'}
```

### Top results
After excluding MUT2, there are genes with significant differential expression, so we can select genes on this basis as well as those with at least 2 fold changes in expression,  as sorted by "B", the log-odds that a gene is differentially expressed. We continue to select probes  on the basis of whether the probe was mapped to a gene in the annotation 

```{r toptable}
```

### Output pvalues and expression levels to files

```{r outputs}
```

#### Link to Excel files with statistics and expression levels of all genes

[ALL genes](../results/all.genes.stats.exprs.`r sampleset`.xls)

#### Link to Excel files with statistics and expression levels for genes with top log-odds of differential expression

[For Top 100 differentially expressed genes](../results/top100.genes.stats.expression.WTvMUT.`r sampleset`.xls)

### Heatmap

```{r heatmap, results='hide', out.width='100%', dev='png', fig.height=18, fig.width=18}
```


