---
output:
  html_document:
    toc: true
title: "450K probes around nuclear mitochondrial genes"
theme: spacelab
highlight: textmate
---

# Setup

## Load libraries
```{r libraries}
library(biomaRt)
library(knitr)
library(stringr)
library(dplyr)
```

## Functions
```{r functions}
bedTools.2in<-function(functionstring="bedtools intersect",bed1,bed2,opt.string="") {
  #create temp files
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) # not to use scientific notation when writing out
 
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
 
  # create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
 
  res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}
bedTools.slop<-function(functionstring="bedtools slop", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out
  
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  
  # create the command string and call the command using system()
  command <- paste(functionstring,"-i",a.file,"-g", genomefile, opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

```

## Set directories  

```{r directories}
baseDir <- "~/Work/projects/ab_nucmito/"
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
```

## Load data
- using mitochondrial data from the [microproteome site](http://www.mitoproteome.org/data/proteome/index.html)

```{r dataload}
annots.450k <- read.csv(file.path(dataDir, "14_April_2014_Updated_humanmethylation450_15017482_v1-2.csv"), skip=7)
annots.mito <- read.delim(file.path(dataDir, "mitoproteome.tsv"))
```

---

# Analysis

## Find genomic locations of mitochondrial genes using biomaRt  

## setup biomart

```{r setupbiomart}
ensemblmart <-  useMart("ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl", host="grch37.ensembl.org")
attributes <- listAttributes(ensemblmart)
filters <- listFilters(ensemblmart)
```

## pull down chromosomal positions using uniprot ids from mito annotations
- current biomart build is GRCh37
- transcript start and ends
- some ids have multiple transcripts, keep all

```{r getpos}
mito.pos <- getBM(annots.mito$ACCESSION,filters="uniprot_swissprot_accession", attributes=c("uniprot_swissprot_accession","chromosome_name","transcript_start","transcript_end", "strand", "hgnc_symbol"), mart=ensemblmart)
```

## Subset genes to those on nuclear chromosomes    
- dicard all mitochondrial, patches and polymorphic regions
- discard genes without annotated gene symbol
- convert to [BED format](http://genome.ucsc.edu/FAQ/FAQformat.html#format1)

```{r subset}
chrs <- c(seq(1,22), "X", "Y")
mito.pos <- mito.pos[mito.pos$chromosome_name %in% chrs,]

# rearrange columns into bed format
mito.pos$score <- 0
mito.pos <- mito.pos[,c(2,3,4,6,7,5)]
mito.pos$chromosome_name <- paste("chr", mito.pos$chromosome_name, sep="")
# drop genes without known symbol
mito.pos <- mito.pos[which(str_length(mito.pos$hgnc_symbol)>0),]
```

## Mitochondrial genes on nuclear chromosomes  

```{r printmitogenes, results='asis'}
#kable(mito.pos)
write.table(mito.pos, file=file.path(resultsDir, "nuclear.mitochondrial.genes.xls"), quote=F, sep="\t", row.names=F, col.names=T)
```

### [Excel file with mitochondrial genes on nuclear chromosome](../results/nuclear.mitochondrial.genes.xls)

## Get coordinates of regions around nuclear mitochondrial genes  
3kb upstream of the TSS and 3Kb downstream of transcript end

```{r getpromoters}
mito.pos.flanked <- bedTools.slop(bed1 = mito.pos, opt.string = "-b 3000", genomefile = "~/Work/projects/ab_nucmito/meta/hg19.genome")
```

## Find 450K probes within flanks and genes of nuclear mitochondrial genes  

## Extract coordinates (Human Genome build GRCh37) of 450K probes from annotation file and convert to BED format
```{r 450kextract}
bed.450K <- annots.450k[,c("CHR", "MAPINFO", "Name")]
# adjust to zero-based coordinate format
bed.450K$start <- bed.450K$MAPINFO-1
names(bed.450K) <- c("chr", "end", "name", "start")
bed.450K$chr <- paste("chr", bed.450K$chr, sep="")
bed.450K <- bed.450K[,c("chr", "start", "end", "name")]
```

## Overlap 450K probe coordinates with flanked nuclear mitochondrial gene coordinates

```{r overlap}
probes450K.in.flanked.nucmitos <- bedTools.2in(bed1=bed.450K,bed2=mito.pos.flanked,opt.string="-wo")
probes450K.in.flanked.nucmitos <- probes450K.in.flanked.nucmitos[,c(1,2,3,4,8)]
names(probes450K.in.flanked.nucmitos) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.flanked.nucmitos <- probes450K.in.flanked.nucmitos[with(probes450K.in.flanked.nucmitos, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.flanked.nucmitos <- probes450K.in.flanked.nucmitos[!duplicated(probes450K.in.flanked.nucmitos),]

probes450K.in.flanked.nucmitos <- probes450K.in.flanked.nucmitos %>% tbl_df() %>% inner_join(., annots.450k, by=c("probeID"="IlmnID"))
```

---

# Results

```{r output}
#kable(probes450K.in.flanked.nucmitos, row.names=F)
write.table(probes450K.in.flanked.nucmitos, file=file.path(resultsDir, "probes450K.in.flanked.nucmitos.xls"), quote=F, sep="\t", row.names=F, col.names=T)
```


### [Excel file with results](../results/probes450K.in.flanked.nucmitos.xls)





